{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "88196fc8-cfc7-46f1-9061-8c1edb73114e",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">8.1 Exploring peak data and motif discovery</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1db538a4-6106-4543-8695-d14fd3a3e963",
   "metadata": {},
   "source": [
    "Last week, we used <code>bowtie2</code> to align our raw reads to the hg19 reference genome. We then took those aligned reads and visualized the aligned reads using IGV. We then used <code>macs2</code> to identify significant and enriched peaks in our taz_1 and taz_2 datasets, and we were able to see how the reads pile up on both sides of the footprint of a DNA binding protein.\n",
    "\n",
    "Today, first start off with some housekeeping as we're now generating a lot more data and also a lot more large data files. Then, we'll explore our peak data some more, making use of a program called <a href=\"http://homer.ucsd.edu/homer/\" rel=\"noopener noreferrer\" target=\"_blank\"><u>HOMER (Hypergeometric Optimization of Motif EnRichment)</u></a>. First, we'll build HOMER into our Biology Hub server, then we will use it to take a look at the distribution of our reads around our summits. <strong>While you can run HOMER in your notebook, it'll be easier to run the installation and your HOMER commands in Terminal.</strong>\n",
    "\n",
    "We'll also take a closer look at our peak data by looking at the intersection of our peaks, and taking a look at the concordance of the two TAZ replicates. Recall from lecture that Dr. Ingolia talked about analyzing the reproducibility of our data using the Irreproducibile Discovery Rate (IDR). We'll take a qualitative look at the IDR of our two TAZ replicates when we play around with the intersection of their called peaks.\n",
    "\n",
    "Once we've identified our set of reproducible peaks, we can perform motif discovery on the overlapping peaks to see what are the binding sites recognized by the DNA binding protein(s) that TAZ is interacting with in order to regulate gene expression.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "\n",
    "<ul>\n",
    "    <li>General housekeeping</li>\n",
    "    <li>Learn to install and use HOMER</li>\n",
    "    <li>Learn to plot the peak profile</li>\n",
    "    <li>Identify overlapping peaks</li>\n",
    "    <li>Take a basic look at the concordance of the two replicates</li>\n",
    "    <li>Pull data from peaks ranking highly in both replicates</li>\n",
    "    <li>Identify binding motifs for TEAD+TAZ</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba131691-0eae-4ed6-9fe1-c1fa01150ef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d767cc9-90b0-4703-8ed1-99731baa266b",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Housekeeping and backing up files</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86f3ff4a-c33a-41f8-bcd9-4bb2645fa8df",
   "metadata": {},
   "source": [
    "Since we're generating large data files, some of which are transitional files, we can do some general housekeeping before proceeding with today's lesson to help the Biology Hub maintainers keep storage needs and costs down. We'll play around in Terminal to poke around our server and see if we can reduce some of our storage footprint. We'll do this portion of the lab notebook in Terminal since it'll be easier to move around."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eef29cbf-c418-449f-846e-062578ded488",
   "metadata": {},
   "source": [
    "<h2>Checking directory sizes through Terminal</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bcd1c7d-a020-4db6-8822-fc4ef0150fc9",
   "metadata": {},
   "source": [
    "While we can hover over files in our File Browser to take a look at the file size, when we do the same thing with directories, we don't get any file size shown. So we'll make use of Terminal to take a look at how large our directories are overall and to identify potential areas where we can tidy up our server."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cf49100-ad2a-4f09-a568-d1feb5e893db",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfe07690-db60-415c-bb9c-0dd4e66a2f2a",
   "metadata": {},
   "source": [
    "<h2>Back up <code>MCB201A_Su2025</code></h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffd0c7a8-eedf-492f-ba19-7923ac319a55",
   "metadata": {},
   "source": [
    "For those of you who want to keep your MCB201A files, go ahead and download your <code>MCB201A_Su2025</code> directory, so that you have it saved as a compressed folder on your laptop."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6968eb05-c79e-4dde-b098-c51750b647e6",
   "metadata": {},
   "source": [
    "<h2>Delete your server's copy of <code>MCB201A_Su2025</code></h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c01a0831-6b5a-4a07-90cc-eb8f31a34b69",
   "metadata": {},
   "source": [
    "With <code>MCB201A_Su2025</code> backed up, let's go ahead and delete our server's copy of it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66f79b0e-26f9-4420-b15c-8f3677d7b0a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e39a7e78-bae3-4d90-89e0-919a3b1e83c3",
   "metadata": {},
   "source": [
    "<h2>Finding files known to be large</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62ebb00e-412a-4dd2-9954-cdaf088426b9",
   "metadata": {},
   "source": [
    "We know that we have particular output files that are known to be large, so we can use the <code>find</code> command to then identify those large files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec1379e1-ed6c-44b0-b618-46041c3ec556",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3759f742-b102-4cf1-b28a-943dac42d375",
   "metadata": {},
   "source": [
    "Then we can feed the output from the <code>find</code> command to the <code>du</code> command to get an output of the size of those files. We'll also modify how both commands are running in order to get a specific output from <code>find</code> that can be understood by <code>du</code> and to let <code>du</code> know that it will read in the file names from the standard input."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5ae11c1-efb0-4b1c-a92c-f371c7d5fd8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ae77594-d496-480f-8fb0-745df613af3f",
   "metadata": {},
   "source": [
    "We can then also use the <code>find</code> command to then delete those files as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dca9503-64b9-4ddf-aef2-58e500fd4f5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e939db5e-d1c8-4998-9ecb-8083506e4fe3",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Installing HOMER</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c05c7993-d506-4fdb-9f03-d2efc5f24d0d",
   "metadata": {},
   "source": [
    "HOMER is a multifunctional program that can not only perform motif discovery, which is what we'll be primarily using it for, but it also has a lot of next-gen sequencing analysis functionality. For instance, it can also be used to call peaks, much like macs2.\n",
    "\n",
    "HOMER was primarily developed by Chris Benner at UCSD.\n",
    "\n",
    "I've placed a <code>configureHomer.pl</code> file in our shared class ChIP-seq directory. This file will be used to install the files needed for HOMER to run on your Biology Hub server.\n",
    "\n",
    "First, you'll want to make a new directory in your home directory that will hold a <u>copy</u> of the <code>configureHomer.pl</code> file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e1bee9b-e8dd-428c-a75a-7e7cc9ccb9bc",
   "metadata": {},
   "source": [
    "<h2>Change to home directory</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adf21d0b-e901-4b74-919d-d9f56eef8960",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36fc62f1-acd4-4a37-9ee9-0cf1388ae2e1",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c85618ae-9d11-4180-8849-2ee057f43b00",
   "metadata": {},
   "source": [
    "<h2>Make a new directory called <code>homer</code> in your home directory</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f7112c3-15d7-42b4-a493-91aee43469d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b664cc2e-10c1-4767-b945-60dcf9342d58",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e143dc6e-623a-4fac-a437-ef545020cc3f",
   "metadata": {},
   "source": [
    "<h2>Enter <code>homer</code> directory and copy over <code>configureHomer.pl</code> to it</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08905daf-2e7a-4671-b447-7411f1f4501c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74748d70-e5f5-4a2d-b2f3-e116855064cd",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6fefb88-1959-409e-bc99-ee32969a922e",
   "metadata": {},
   "source": [
    "<h2>Run <code>configureHomer.pl</code></h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8269aa35-b07a-4701-9b38-11949e8189b0",
   "metadata": {},
   "source": [
    "Once you've copied over <code>configureHomer.pl</code>, you'll now need to run it in order to install the necessary files. Since this file uses Perl, another coding language, we will need to run it directly in command line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d401a0e-4e3e-40a6-bb13-17627ab413aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e52599-22a7-4be3-80fb-c1b69f22f026",
   "metadata": {},
   "source": [
    "There might be a few warnings, but there shouldn't be any errors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c4fd934-80c8-4a26-8798-7a972ab3a2a1",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2990be8d-ff77-496c-87bb-4cec4274c84a",
   "metadata": {},
   "source": [
    "<h2>Update PATH to be able to run HOMER</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79286a3c-cca7-42bf-8078-549c62c8b333",
   "metadata": {},
   "source": [
    "We'll need to tell Terminal where to go to find the different files that HOMER uses. This temporarily adds the directories that HOMER uses to your <code>PATH</code> variable, but since the server reinitializes things, you'll need to go to your <code>homer</code> directory and re-run this line to update the <code>PATH</code> variable each time you start up your server."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "084b096d-843c-4a9c-9f55-fbf2963e1a7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07462558-2c4e-46b5-bf63-714177e81261",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce1d9bda-8b3b-4dfa-a5ec-f53acca04d10",
   "metadata": {},
   "source": [
    "<h2>Install hg19 genome</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "926d0043-3260-4b6d-a3ba-d7d974a547dc",
   "metadata": {},
   "source": [
    "Since our alignments were to the hg19 genome, we'll need to install this genome to use with HOMER and then prepare it for use later on when we perform motif discovery. First, let's look at what we currently have installed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "042cf0e0-ea7f-48c3-b989-307a8baa76e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4106e139-f3fb-45d2-9ed7-c90080479d7e",
   "metadata": {},
   "source": [
    "You should see that we don't have the hg19 genome installed, as indicated by the <code>-</code> by the name of the genome. So let's go ahead and install it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "909efa9b-b144-400e-b827-66e90b2d95e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47942041-8c0c-483b-99cd-9a4102015a96",
   "metadata": {},
   "source": [
    "Once you've installed the hg19 genome, you'll see a lot of new files under <code>~/homer/data/genomes/h19</code>, and if an analysis requires you to provide an input genome, you can simply specifiy <code>hg19</code>. Since it is a set of saved files and directories on your server, you don't have to reinstall it each time you start up your Biology Hub server."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f32cee7d-3d32-44d7-b552-9159265fa016",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b173c26-4e19-4339-80e6-3eb61eca9eff",
   "metadata": {},
   "source": [
    "<h2>Preparse hg19 genome</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c00f6f8c-6b5d-4fe6-9c08-06cc179455f5",
   "metadata": {},
   "source": [
    "Before we can use the hg19 genome though, we'll have to prepare it for use in motif discovery by preparsing the genome. This parses the genome into 50bp fragments that we can later use for motif discovery."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf4e8cc1-c558-4ef7-a61b-ff50a87cb686",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc2c1714-8fd5-4961-8d36-a933f2afe0b4",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a38b9f6-9923-49f6-8606-3dff9a7ea6bf",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Plot peak profile</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b0c227-5587-4f10-8cf7-541f05d5d845",
   "metadata": {},
   "source": [
    "To get an idea of how all the reads are distributed around their summits, we can plot the peak profile for our two replicates. This allows us to visualize the average pileup around our peak summits to perform QC and to get an understanding of the binding footprint of our protein of interest."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ede7805f-ba57-4bac-8d1c-f7831e29c60c",
   "metadata": {},
   "source": [
    "<h2>Change directories to prepare to plot peak profiles</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f452e7fc-e0c9-4af9-8381-cef80930f53f",
   "metadata": {},
   "source": [
    "Change your current working directory to this week's directory and make a folder named <code>plot-profile</code>. Set your current working directory to the new folder you created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08905d41-5c99-4716-8af7-35d7c8c58f62",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c91d6c1f-65a9-48e2-bf6b-75e64262ea61",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca47bab0-0d77-4c98-82be-227b241f8c7f",
   "metadata": {},
   "source": [
    "<h2>Use <code>annotatePeaks.pl</code> to generate histograms</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1dbe3ba-c276-4bc2-9de7-1865dc931737",
   "metadata": {},
   "source": [
    "HOMER has a program for peak annotation called <code>annotatePeaks.pl</code>, which like the other HOMER functions, you can run directly in Terminal. <a href=\"http://homer.ucsd.edu/homer/ngs/annotation.html\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Documentation for <code>annotatePeaks.pl</code> can be found here.</u></a>\n",
    "\n",
    "<code>annotatePeaks.pl</code> is the workhorse of HOMER. It is multifunctional and can be used for a variety of purposes. It's primary purpose is, as it's name implies, to annotate the peaks identified through your peak calling program by identifying nearby genes, distance to the nearest transcription start sites (TSSs), and annotate your peaks with genomic features. It can also be used to generate histograms and heatmaps, which is one of the functions we'll be using it for here.\n",
    "\n",
    "The syntax to use this program in command line is fairly straightforward; however, there are a number of different options that you can invoke that will change how the program runs, resulting in pretty different outputs.\n",
    "\n",
    "If we take a look at the documentation to see how we can use this program:\n",
    "\n",
    "<pre style=\"width: 450px; margin-top: 15px; margin-bottom: 15px; color: #000000; background-color: #EEEEEE; border: 1px solid; border-color: #AAAAAA; padding: 10px; border-radius: 15px; font-size: 12px;\">annotatePeaks.pl &lt;peak file &vert; tss&gt; &lt;genome version&gt; &lbrack;options&rbrack;</pre>\n",
    "\n",
    "We can either supply it with a peak file, such as our <code>.narrowPeak</code> or <code>*_summits.bed</code> output from <code>macs2</code>.\n",
    "\n",
    "Alternatively, we can specify it to carry out a transcription start site (TSS)-centric analysis, in which case a peak file is not needed. Another way that we can alter how <code>annotatePeaks.pl</code> runs is by supplying it with additional options such as <code>-hist</code>, which will generate a histogram output if we supply it with coverage data in the form of a bedGraph file using the option <code>-bedGraph</code>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b28b3048-f106-4f23-8ddb-1d936eef26e7",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "032102db-9ff8-4593-8a23-9978391bc67a",
   "metadata": {},
   "source": [
    "<h2>Obtain distribution of coverage/read counts</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e8a605-a0b7-431b-ac06-e64168dafe0b",
   "metadata": {},
   "source": [
    "To obtain the distribution of coverage or reads around the summit for taz_1, you'll need your <code>10M_taz_1_summits.bed</code> file, as well as the bedgraph file that macs2 output, which contains the distribution of reads across the whole genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dff61dae-6c4e-4b0d-9045-68bc762bb8da",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e1f9ce-e79a-4c0a-9c3c-4523860093b6",
   "metadata": {},
   "source": [
    "<h2>Let's breakdown the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8a9735f-3d37-43a2-a0b3-f28c85dc6680",
   "metadata": {},
   "source": [
    "<code>annotatePeaks.pl</code>\n",
    "\n",
    "This calls up the annotatePeaks program within HOMER.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>~/MCB201B_F2025/Week_7/macs2/10M_taz_1_summits.bed</code>\n",
    "\n",
    "Here, we are providing the annotatePeaks program with a bed file containing the coordinates for our peak summits because we're interested in taking a look at the coverage/distribution of read counts around our summits. \n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>hg19</code>\n",
    "\n",
    "We specify the reference genome that we want the program to use. We can just provide it with the name of the reference genome <code>hg19</code>, and HOMER will know the directory containing the files for our reference genome that we installed earlier.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-bedGraph ~/MCB201B_F2025/Week_7/macs2/10M_taz_1_treat_pileup.bdg</code>\n",
    "\n",
    "While we've provided annotatePeaks with the genomic coordinates to each of our summits, we need to also provide it with information on the read coverage throughout the genome so that it can determine the distribution of reads around our summits. So here, we supply it with the bedGraph file that was generated by <code>macs2</code>, since this file contains information on read coverage for each location in the genome.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-size 2000</code>\n",
    "\n",
    "The <code>-size</code> option specifies the size of our region of interest, centered around what we're interested in, which in our case are the peak summits. By specifying a size of 2000, we will be generating a histogram of a region 1kb upstream and downstream of the summit.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-hist 10</code>\n",
    "\n",
    "The <code>-hist</code> option changes how annotatePeaks runs. This changes it essentially to \"histogram\" mode, where it will output (into Terminal) the distribution of read coverage around in our specified region of interest. The argument following the <code>-hist</code> option specifies the bin size for the histogram, so with <code>10</code>, we will be looking at the read distribution in intervals of 10.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>&gt; 10M_taz_1_summit_profile.txt</code>\n",
    "\n",
    "We want annotatePeaks to save the output into a .txt file rather than dumping it into Terminal, so we redirect the output into <code>taz_1_summit_profile.txt</code>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fdc232e-e9d8-4529-9986-285cee548d22",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise set 1</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44b20649-bdd0-44c2-b87c-4c9f862f1f5c",
   "metadata": {},
   "source": [
    "<h2>Exercise #1: Obtain coverage data around each summit for 10M_taz_2</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "245fa54c-f0bf-4635-b291-fb5390c8f12b",
   "metadata": {},
   "source": [
    "Now see if you can run the <code>annotatePeaks.pl</code> program to generate the histogram for 10M_taz_2 using the same setup that we used for 10M_taz_1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437b4c10-27d1-4285-baef-21eb7c56c961",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd60357f-00c9-4e1b-9dc8-66450317af62",
   "metadata": {},
   "source": [
    "<h2>Exercise #2: Import HOMER outputs for both replicates</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28517db0-7f21-4cbc-9acc-1158a8821a94",
   "metadata": {},
   "source": [
    "Now that we have both outputs, we'll make use of Python to take the output from HOMER to take a look at how the reads are distributed around the summits for both of our TAZ replicates.\n",
    "\n",
    "First, you'll need to import both files into their own pandas DataFrames using the <code>pd.read_csv()</code> function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e19858c8-116b-4863-827b-78a922a49662",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5ea6d2ca-2a5b-490f-9eb9-c8766b8334e8",
   "metadata": {},
   "source": [
    "<h2>Exercise #3: Simplify column headers</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff1f5960-ac0f-4b9f-a27b-3ce63ecec21e",
   "metadata": {},
   "source": [
    "Take a look at the column headers. You'll see that they are a bit long with extra informaiton on our command line input that we don't need for plotting purposes. So replace your column header with the following:\n",
    "\n",
    "<ol>\n",
    "    <li><code>'distance_from_center'</code></li>\n",
    "    <li><code>'coverage'</code></li>\n",
    "    <li><code>'plus_tags'</code></li>\n",
    "    <li><code>'minus_tags'</code></li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "316558c7-adda-4876-919e-81368a732866",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d08456a8-b9b7-4b6b-befa-569057c4ff7d",
   "metadata": {},
   "source": [
    "<h2>Exercise #4: Plot peak profiles</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27ce0b03-aeb5-4131-82d3-95545ccf8669",
   "metadata": {},
   "source": [
    "Now plot the peak profiles for both replicates as a line plot overlaid on the same plot with the X-axis as the distance from the center of the summit and the Y-axis as the coverage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f727a05-49f1-4d0d-a17b-efaa7f097100",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "399c1f13-50d5-4413-9128-b9891eaac6aa",
   "metadata": {},
   "source": [
    "You can see in your plot that while the coverage pattern is similar between the two replicates, they vary slightly in their read depth, with taz_2 having overall higher coverage than taz_1. This is normally expected, since each replicate will have slightly different read counts. The important consideration is that the plot profiles look similar in their shape.\n",
    "\n",
    "Looking at the shape of both of our peak profiles, we can see that it is a sharp, narrow peak profile which is characteristic of the peak profiles for transcription factors as they recognize and bind to specific sequences in the genome based on what their DNA binding domains recognize. By plotting the peak profile, we can now see that TEAD+TAZ recognize and bind to a specific sequence as opposed to just broadly recognizing DNA. We'll later on pull together our peaks to identify the motif(s) recognized by TEAD when its bound to TAZ. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d3412ed-b228-41d0-9731-092187ee8e34",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Generate heatmap of coverage/read density</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d39e3f30-a4e7-41f4-b455-84327e9ecd59",
   "metadata": {},
   "source": [
    "Among the documentation for <code>annotatePeaks.pl</code>, we can see that there is an additional option <code>-ghist</code> that we can specify, which will change the output generated by <code>annotatePeaks.pl</code>. In this case, rather than outputting the average coverage along all peaks, it will output a data matrix, where we are then able to see the coverage for each peak individually."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ccc8855-b5f3-415a-b267-785558be90e5",
   "metadata": {},
   "source": [
    "<h2>Create <code>heatmap</code> directory in <code>Week_8</code> and enter it</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f42a66cd-b3f3-4cd0-ad99-d58a3598efbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea5e269b-3fac-4d72-8b03-6ab99eaa8555",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3763c95-cdb2-4dd5-9259-53c8e568d2d9",
   "metadata": {},
   "source": [
    "<h2>Obtain data matrix for both TAZ replicates</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f479152c-8bf0-465f-8a43-3088942682c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40f4da20-8dfa-491e-a3e0-08c2e28c7eeb",
   "metadata": {},
   "source": [
    "So you can see that we run the code with the same input, except now we add in an additional option <code>-ghist</code>, which changes how it outputs the coverage data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e90f1323-d1c4-42f4-8039-5ee10452e511",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2368ce0a-a716-4322-8c00-17134820623c",
   "metadata": {},
   "source": [
    "<h2>Import our data matrices</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd7afd79-62cd-472a-a1e4-fd6868914588",
   "metadata": {},
   "source": [
    "Let's go ahead and import our data matrices, and then to simplify plotting, we can drop the <code>'Gene'</code> column since we don't need that for our heatmaps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8cbc496-f2c7-4a80-9d6b-d80bedc12ef1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "66d2ac5a-88cd-47e3-b919-e6c4bd7860a1",
   "metadata": {},
   "source": [
    "Let's do a quick sanity check."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e8140cd-e090-4c98-bc37-e9c329b43492",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ffdf2445-657a-48f4-bba6-5b49c3527cd5",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ae08223-0620-430e-828b-2afbe60ed058",
   "metadata": {},
   "source": [
    "<h2>Generate heatmaps for both replicates</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe03a2f2-9b61-4f42-8274-32ab0e5932ba",
   "metadata": {},
   "source": [
    "With everything imported, now we can make use of either <code>matplotlib</code> or <code>seaborn</code> to generate our heatmaps. For this example, we'll make use of <code>sns.heatmap()</code>. <a href=\"https://seaborn.pydata.org/generated/seaborn.heatmap.html\" rel=\"noopener noreferrer\">Documentation for <code>sns.heatmap()</code> can be found here.</u></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0310bbd4-c5c8-4e53-8222-c6239325a401",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d036b5d1-97ae-4f4e-94b0-c81222c6b39c",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Plot TSS-centric read distribution</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "037285ca-1e99-4efe-93aa-09beb916d8b1",
   "metadata": {},
   "source": [
    "The multifunctionality of <code>annotatePeaks.pl</code> means that we can also perform a transcription start site (TSS)-centric analysis. In this case, we can supply the program with the <code>tss</code> argument along with our bedGraph files, so it can provide us with the distribution of reads around TSSs. This can help give us an idea into if our protein of interest is regulating its target genes via promoter binding or through more distant interactions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5f506ef-017d-493b-a644-cee64663bacc",
   "metadata": {},
   "source": [
    "<h2>Obtain read distributions around TSS</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b176507-d2a9-4d0a-9610-f5cbf93d458b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d5814a1-2c4f-47c9-bd04-41a88c358868",
   "metadata": {},
   "source": [
    "Here, we make minor alterations to how we ran the annotatePeaks program to instruct it to focus on TSSs and to provide a larger region of interest. So we provide the <code>tss</code> argument and increase the <code>-size</code> to <code>8000</code>. That way, we'll get an output containing the read distribution centered around TSSs, which we can plot like we did with the summit profile."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71d3026c-a18e-4550-9967-1715be12f3f5",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise set 2</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a99d855b-d693-41d9-bcfd-5ebad2ba35f7",
   "metadata": {},
   "source": [
    "<h2>Exercise #1: Obtain read distributions around TSS for 10M_taz_2</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f544ea33-ac3c-43ab-bc35-04b51a48f898",
   "metadata": {},
   "source": [
    "Now give it a try to update your input to instead analyze the distribution around TSSs for 10M_taz_2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d6617e5-be2a-4655-8559-e4b47e97c28a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d52aa92-43ab-472f-a345-760142f0843f",
   "metadata": {},
   "source": [
    "<h2>Exercise #2: Import your TSS analysis outputs</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6078896-6427-4774-8e37-1c99d5ee00f1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5da6c93e-572b-41ae-b641-0d42dd0ea54c",
   "metadata": {},
   "source": [
    "Like with our plot profiles earlier, update the column headers for convenience.\n",
    "\n",
    "<ol>\n",
    "    <li><code>'distance_from_center'</code></li>\n",
    "    <li><code>'coverage'</code></li>\n",
    "    <li><code>'plus_tags'</code></li>\n",
    "    <li><code>'minus_tags'</code></li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf154d4-1075-4632-923c-0444b734c26c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7babc623-6e49-4a25-9363-a590a338ff6f",
   "metadata": {},
   "source": [
    "<h2>Exercise #3: Plot the coverage around TSS for both replicates</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4daa9e20-4179-4841-8ba3-39ea0d5e87b4",
   "metadata": {},
   "source": [
    "Once you've imported your datasets, go ahead and plot the coverage around TSSs, using the same approach that you used for plotting the peak profile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af110756-9105-4e3a-b155-52dc68741a32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "522ab86c-e7fc-4c62-b7fc-e24ecb01f718",
   "metadata": {},
   "source": [
    "What does the distribution tell us about the way that TEAD+TAZ is potentially regulating gene transcription?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf1e35c0-1e15-41b1-9647-4ccd5297ef75",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Determine distance from TSS</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b05a47c-3eb6-4b68-af55-379aa63cb474",
   "metadata": {},
   "source": [
    "The results that we get when we look at the average coverage around TSSs seem to suggest that our peaks may be more distant than &pm;4kbp. What we can then do is take a look at the distribution of distances that our peaks are located from their nearest TSS. That can help give us a little bit more insight into our above result.\n",
    "\n",
    "We'll once again make use of <code>annotatePeaks.pl</code> with slight modifications. If we take a look at <a href=\"http://homer.ucsd.edu/homer/ngs/annotation.html\" rel=\"noopener noreferrer\"><u>the documentation for <code>annotatePeaks.pl</code></u></a>, among one of the outputs for its basic annotate function is a column containing information on the distance to the nearest TSS, which is what we're interested in.\n",
    "\n",
    "So we can use the basic annotate function of <code>annotatePeaks.pl</code> and focus on TSS annotations."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08feab34-e053-4d4a-8a89-1b5b434bac98",
   "metadata": {},
   "source": [
    "<h2>Create new directory <code>annotation</code> in this week's directory</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cca6ec2e-95ac-4f98-b466-1189ad4c5042",
   "metadata": {},
   "source": [
    "Let's make a new directory <code>annotation</code>, and then enter it too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94f96c93-b275-4ead-9fc5-f17e85c976a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ac68fce-b4e2-44b2-9adf-640abfdfc0c0",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7f3f5e6-cb6a-42d1-8f9a-36ae1bed5cd7",
   "metadata": {},
   "source": [
    "<h2>Annotate to find distance to nearest TSS for 10M_taz_1</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de51f9a5-f4ac-4707-b20c-a95bf520697d",
   "metadata": {},
   "source": [
    "Now lets modify how <code>annotatePeaks.pl</code> runs, and this time have it return to us the distances to TSSs for each peak summit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b36f990b-6d4c-408a-bbe8-0b025416138b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7562014f-689e-439f-90ee-76b5ff2befc4",
   "metadata": {},
   "source": [
    "<h2>Let's breakdown the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf468d5c-90b9-4fd2-a30a-8215e675070e",
   "metadata": {},
   "source": [
    "<code>annotatePeaks.pl</code>\n",
    "\n",
    "This calls up the annotatePeaks program within HOMER.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>~/MCB201B_F2025/Week_7/macs2/10M_taz_1_summits.bed</code>\n",
    "\n",
    "Here, we are providing the annotatePeaks program with a bed file containing the coordinates for our peak summits. This time we're interested in identifying the distances of our summits to the nearest TSS.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>hg19</code>\n",
    "\n",
    "We indicate that we want to use the hg19 genome, which we have installed.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-noann</code>\n",
    "\n",
    "This option alters how the annotatePeaks program runs. By default the program will mark peak locations based on the genomic feature in which its located (intron, exon, TSS, CDS, etc), and the program will also identify the nearest gene and TSS. By providing the <code>-noann</code>, we are instructing the annotatePeaks program not to annotate peaks for genomic features, and instead, just focus on identifying nearby genes and TSSs. \n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>&gt; 10M_taz_1_annotation.txt</code>\n",
    "\n",
    "We redirect the output data into a .txt file that we can then load into Python in order to pull relevant information."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57310813-eb18-4c64-a52b-54b5afdb31a9",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise Set 3</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71a0c1f2-c706-4f1b-a53e-1600962910cb",
   "metadata": {},
   "source": [
    "<h2>Exercise #1: Obtain distance to nearest TSS for 10M_taz_2</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f72934cb-4985-40b8-af14-7a58072ad118",
   "metadata": {},
   "source": [
    "Run <code>annotatePeaks.pl</code> now for 10M_taz_2 to identify the distance to the nearest TSS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3171ae2-569d-42b7-842e-f107afba0ee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f45c802-65eb-46ee-b166-0efa7fa7e5b4",
   "metadata": {},
   "source": [
    "<h2>Exercise #2: Import your outputs</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a522e24e-e602-45b1-b32c-18ce96a7b0ed",
   "metadata": {},
   "source": [
    "Now go ahead and import your output files containing the distances to the nearest TSS for each replicate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d111e2fc-c859-440c-aca9-e0197400d75a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e33141aa-a0dd-45fd-9258-e9cc70ff323c",
   "metadata": {},
   "source": [
    "Update the first column's header real quick to <code>'Peak ID'</code> for convenience."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14ef9d7b-5cc3-425e-b2d5-e12932d06917",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "16b4955d-c89f-41bd-afdd-5f2cd3331086",
   "metadata": {},
   "source": [
    "<h2>Exercise #2: Calculate the absolute value of distances for each peak</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11611744-350e-4290-ad15-bcfd185b36ca",
   "metadata": {},
   "source": [
    "Since our dataset contains both positive and negative distances for <code>'Distance to TSS'</code>, which just indicate relative position to the TSS either upstream or downstream, calculate the absolute value of the distances, since we're only really interested in how far away is TEAD+TAZ binding for this exercise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ac5e69c-da91-432b-aea8-c59e37c15d74",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "736a7279-9e5b-4ab4-a5b5-cdb2dea4839e",
   "metadata": {},
   "source": [
    "Let's convert to kilobases for convenience."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce6afe98-af3b-449d-8629-ef92aaa27323",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dbf98df7-9d4e-412b-84f0-3aafa8b30ff0",
   "metadata": {},
   "source": [
    "<h2>Exercise #3: Plot a histogram of the distribution of distances</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82e29b4f-62c2-4ef7-a573-ba9cea9a2e2c",
   "metadata": {},
   "source": [
    "Now see if you can plot a histogram of the distribution of distances according to the following bins:\n",
    "\n",
    "<ul>\n",
    "    <li>From 0kb to 1kb</li>\n",
    "    <li>From 1kb to 10kb</li>\n",
    "    <li>From 10kb to 100kb</li>\n",
    "    <li>>100kb</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8146737b-fe22-4dcd-a620-2a466dd9b5bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "30954281-5bb4-41c1-a3ed-1fd91f21e69a",
   "metadata": {},
   "source": [
    "What do the distribution of peak distances to the nearest TSS suggest to us about how TAZ is regulating gene expression?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b2298b6-1fcc-4787-9213-e4da06fc5ecf",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Find concordant peaks between two replicates</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b293a5cd-d994-4648-8fec-618d8d205b2a",
   "metadata": {},
   "source": [
    "Currently, we have two replicates with thousands of identified peaks that were found to be significant compared to the background levels (using our pooled IgG sequencing data). However, not all those peaks will be the same between replicates, due to biological and experimental variability, so we'll want to focus on just those peaks that are present in both our replicates for a more focused downstream analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0cefc57-48e4-4d52-85b4-0a451567d6e2",
   "metadata": {},
   "source": [
    "<h2>HOMER's mergePeaks program</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d9edb76-117b-48eb-a401-25073fd5011f",
   "metadata": {},
   "source": [
    "To find peaks that are shared between our two replicates, we will make use of the <code>mergePeaks</code> command, and we'll provide it with the .narrowPeak files for both our replicates for analysis.\n",
    "\n",
    "<a href=\"http://homer.ucsd.edu/homer/ngs/mergePeaks.html\" rel=\"noopener noreferrer\"><u>Documentation for the <code>mergePeaks</code> utility can be found here.</u></a>\n",
    "\n",
    "If we take a look at the command line structure to use this program to pull overlapping peaks:\n",
    "\n",
    "<pre style=\"width: 750px; margin-top: 15px; margin-bottom: 15px; color: #000000; background-color: #EEEEEE; border: 1px solid; border-color: #AAAAAA; padding: 10px; border-radius: 15px; font-size: 12px;\">mergePeaks -d &lt;max distance to merge&gt; &lt;peak file 1&gt; &lt;peak file 2&gt; &lbrack;peak file N&rbrack; &lbrack;options&rbrack; -prefix &lt;base name&gt;</pre>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bf12f5f-137f-44f7-bf05-79c290a4aff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57cb058e-983b-4217-8b41-180636121cb6",
   "metadata": {},
   "source": [
    "<h2>Let's breakdown the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "873735af-3572-4fa8-ba1c-57f1ccd50a86",
   "metadata": {},
   "source": [
    "<code>mergePeaks</code>\n",
    "\n",
    "Here, we are calling up the mergePeaks program.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-d 100</code>\n",
    "\n",
    "Here, we provide the program with the maximum distance that we would like it to merge peaks based on the distance between the peak centers. So peaks whose centers are greater than 100bp apart will not be considered overlapping.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>~/MCB201B_F2025/Week_7/macs2/10M_taz_1_peaks.narrowPeak</code>\n",
    "\n",
    "The .narrowPeak file corresponding to our first replicate that we want it to pull peak information from.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>~/MCB201B_F2025/Week_7/macs2/10M_taz_1_peaks.narrowPeak</code>\n",
    "\n",
    "The .narrowPeak file corresponding to our second replicate. mergePeaks will identify the overlapping peaks from the two files we supply it. We can also supply it with additional files, and it will analyze all of them.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-prefix merged</code>\n",
    "\n",
    "Here, we supply it with the basename that we want it to add to all the files that it outputs. Based on how we set up the command, we will have three output files, two that correspond to unique peaks present in their respective replicates and one that contains the information on overlapping peaks, which is the file that we're most interested in."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88691d9d-b37b-4f6a-a93c-e38bb14eb9d7",
   "metadata": {},
   "source": [
    "<h2>Simplify output file names</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8dc62f0-5924-4796-9669-36455c9b26e3",
   "metadata": {},
   "source": [
    "If you take a look in your File Browser, you can see the three outputs, and they have rather long, unwieldy names, especially the file containing our overlapping peaks. So let's just directly in the File Browser rename the files for simplicity's sake.\n",
    "\n",
    "Rename the file containing the overlapping peaks to: <code>merged_taz_peaks.narrowPeak</code>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04b5c82d-1faf-4c14-a567-929bbcf2ecc9",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 1px solid; border-color: #BBBBBB; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3be81cca-4453-48a3-82f1-857b45308039",
   "metadata": {},
   "source": [
    "<h2>Import <code>merged_taz_peaks.narrowPeak</code> as a DataFrame</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55abbecc-789e-4d68-b2d1-76ac79a87cba",
   "metadata": {},
   "source": [
    "Now let's take a look at the information contained within the file by importing into our Python notebook.\n",
    "\n",
    "First, import the .narrowPeak file containing the merged peaks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a5be475-0c62-44dc-a2a6-95220d92c131",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0ae49ec7-bc74-4e75-9deb-ebed941d5a77",
   "metadata": {},
   "source": [
    "Now let's take a look to see if it imported okay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00857388-90aa-416f-bef1-0cfef8080e56",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b2e87a07-0fff-4e32-a7e3-3493fe8e25c1",
   "metadata": {},
   "source": [
    "So we can see that the column headers are a bit long, so we can shorten things for our sake later on by using the <code>pd.DataFrame.rename()</code> function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2674602-5d82-4aef-a298-63732a0de24e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "08ce7e90-00c8-46cc-b0d1-7731f62b9408",
   "metadata": {},
   "source": [
    "And let's check to see if it updated properly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3251d5d-a059-42d0-8df5-fb70e29d58c3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b52ac5aa-3186-43c3-9e9f-34c091c51617",
   "metadata": {},
   "source": [
    "Now that we have our overlapping peaks, it's not enough to just know which peaks are shared between our two replicates. We also need to know how reproducible these peaks are between our two replicates."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4590bd43-1349-470b-b4c2-7652d938b9d9",
   "metadata": {},
   "source": [
    "<h2>Assessing reproducibility</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6e7a958-c565-4701-82f6-90de1c15befe",
   "metadata": {},
   "source": [
    "Dr. Ingolia mentioned in lecture that between replicates, there is variability in which peaks are called as well as their significance, which can be due to normal biological or experimental variability. <a href=\"https://doi.org/10.1101/gr.136184.111\" rel=\"noopener noreferrer\"><u>The standard way (set by the ENCODE and modeENCODE consortia)</u></a> of determining if the peaks that we identify through our peak calling program is to calculate the Irreproducible Discovery Rate (IDR), developed by <a href=\"https://doi.org/10.1214/11-AOAS466\" rel=\"noopener noreferrer\"><u>Li et al 2011</u></a>. \n",
    "\n",
    "The IDR framework allows for us to determine the reproducibility of our peaks (or in general the findings of high-throughput experiments) by comparing replicates to one another. IDR provides us with a quantitative assessment of when two replicates are no longer consistent with one another, which could point to findings that are less reproducible. In our case, that would suggest that those peaks may not necessarily be reproducible in repeated experiments. It's in a way conceptually similar to the FDR, where a lower IDR indicates a lower chance that the peak is irreproducible.\n",
    "\n",
    "We'll just take a qualitative look at the concordance of our data, which will give us an idea of which peaks are probably more consistent between our replicates. And then we'll use those peaks for motif discovery.\n",
    "\n",
    "To do this, we'll first need our original <code>.narrowPeak</code> files, so that we can rank the peaks according to their significance <code>'-log10(qvalue)'</code>. Then we can use our <code>merged_taz_peaks.narrowPeak</code> file containing our overlapping peaks to pull all the information only for the overlapping peaks. And then we can plot the two replicates against one another to see how correlated the ranks are for each of the overlapping peaks. We can also do the same for the significance of each overlapping peak."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a97a4596-0a29-4ffd-98ef-22049e212397",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise Set 3</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de0ff97e-81b2-4fb0-bd20-405cd6e82d86",
   "metadata": {},
   "source": [
    "<h2>Exercise #1: Import <code>.xls</code> files for both replicates</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a0d6a83-981e-4e83-839a-09a09ed2a1cd",
   "metadata": {},
   "source": [
    "Let's import our original <code>.xls</code> files that was output by <code>macs2</code>, so we have all the information for each peak."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "737de54e-36b4-477f-8dc8-2b8c349e52a1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "efc23ae6-c7ec-48f2-aa1d-9a53f44ee2f0",
   "metadata": {},
   "source": [
    "Let's check to see how the file imported."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08d0727b-55ed-4240-ae3c-dcb9c6f2ad22",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d8e33d11-59b2-4205-b4aa-bbf8a8077b4c",
   "metadata": {},
   "source": [
    "<h2>Exercise #2: Rank data according to the <code>'-log10(qvalue)'</code></h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f974947e-e656-494a-b1fb-202588055663",
   "metadata": {},
   "source": [
    "Now, we'll want to rank all the peaks according to their significance, with the greatest <code>'-log10(qvalue)'</code> having the assigned rank of 1. \n",
    "\n",
    "So we'll need to sort our data based on the <code>'-log10(qvalue)'</code>, and then we can assign ranks to each peak."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae4bdad7-bf90-454d-a24c-beebd8f9005d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "29ee83b3-0cac-4dbf-b980-3eb9c2a7713c",
   "metadata": {},
   "source": [
    "<h2>Guided Exercise: Pull together taz_1 peaks and taz_2 peaks based on the overlapping peaks</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50347292-b269-48a2-8b61-2817e659220f",
   "metadata": {},
   "source": [
    "Now that we've ranked our individual replicates, we'll go ahead and then use our imported merged peaks to pull together the DataFrames for both replicates. To do this, we'll make use of the <code>pd.DataFrame.merge()</code> function, which will allow us to merge together two DataFrames based on a key or index. \n",
    "\n",
    "<a href=\"https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html\" rel=\"noopener noreferrer\"><u>Documentation for <code>pd.DataFrame.merge()</code> can be found here.</u></a>\n",
    "\n",
    "Digging into the documentation, we can see that the function is set up such that the DataFrame preceding the <code>.merge()</code> is the \"left\" DataFrame to be merged, while the DataFrame passed as an argument to the <code>.merge()</code> is the \"right\" DataFrame to be merged.\n",
    "\n",
    "We can also provide it with the column containing the values that will act as the key that the merge will be based on with the <code>left_on</code> and <code>right_on</code> parameters. And if we want to distinguish the source of the columns after the merge, we can add suffixes to the column names (if the column name is shared between both DataFrames)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75e3fd2f-b7ad-4b24-b6bb-5cb74f855deb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "da744392-3c4d-4569-b768-8a6cca292a32",
   "metadata": {},
   "source": [
    "<h2>Exercise #3: Plot each replicates' ranks against each other</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3cbd948-ff76-4fa1-a7db-777d1ddc3f0d",
   "metadata": {},
   "source": [
    "We can then plot the peaks' ranks against one another to visualize the reproducibility of our dataset. For this exercise, plot the ranks for taz_1 peaks on the x-axis and the ranks for the corresponding taz_2 peaks on the y-axis as a scatter plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c59c8775-b91f-4a3b-a741-7cba6d7d4e4b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ac521e53-604b-4e2f-be80-8fe5f288a097",
   "metadata": {},
   "source": [
    "What you should be able to see is a plot that begins to fan out as the value for the rank increases (so overall less significant peak), suggesting that those peaks identified in either set that do not have a high -log10(qvalue) may not be very reproducible. So in our more focused analysis, we will want to avoid those peaks and just look into those that consistently rank high (rank values closer to 1)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06ba8d67-9089-426c-8c90-8babb159904a",
   "metadata": {},
   "source": [
    "<h2>Exercise #4: Plot significance of two replicates against one another</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26bb34ce-a5c7-4646-88ff-dae87928f346",
   "metadata": {},
   "source": [
    "We can also do the same for the <code>'-log10(qvalue)'</code> of our overlapping peaks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88912f34-0f3e-4b91-9a61-d8d92a6665ce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "053a34af-dc0a-41e9-ba63-6f57352434ab",
   "metadata": {},
   "source": [
    "To visually increase the spread of the lower significance samples, we can log transform the <code>'-log10(qvalue)'</code> using <code>np.log10()</code> when we plot each data point.\n",
    "\n",
    "Alternatively, you can change the x- and y-axes to log scale."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c051173-a207-4d48-9ccf-8674ffee6661",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e3e051ae-f953-439b-a57c-58df498c9e31",
   "metadata": {},
   "source": [
    "Again, we can see that peaks that are highly significant in one replicate are likely to also be similarly significant in the second replicate, which suggest that they are more consistent between replicates. However, not all the peaks are necessarily highly consistent between our two replicates."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a215319-5493-4ff6-a12a-0bfdd2263b24",
   "metadata": {},
   "source": [
    "<h2>Exercise #5: Narrow down peaks to consistently high ranked peaks</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "531b84f2-7467-417f-a948-971002d27e6e",
   "metadata": {},
   "source": [
    "For our motif discovery, we'll want to have a focused set of peaks to work with, so we can just focus on those peaks that are highly consistent between our two replicates. Normally, we would quantify the Irreproducible Discovery Rate (IDR), but here we'll just set a cut off of those peaks that are consistently in the top 500 in both replicates. See if you can pull just those highly reproducible peaks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5df1e86c-e95e-442e-939d-7bd81bd8d435",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "840232fd-5604-4270-987c-482e76404a97",
   "metadata": {},
   "source": [
    "<h2>Exercise #6: Highlight the consistently high ranked peaks</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee843f32-8e2c-4577-aeb2-f451e0d4385b",
   "metadata": {},
   "source": [
    "Similar to what we did in notebook 7-2, let's highlight just the consistently high ranked peaks to visually distinguish them from the rest of the peaks in our scatter plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9409898-dad6-4b39-8c04-0645d7d6ff07",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f5423977-9704-4797-a2be-59c45006165d",
   "metadata": {},
   "source": [
    "<h2>Exercise #7: Export your concordant peaks to a <code>.narrowPeak</code> file</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b7c5e06-d0b2-486b-9b37-840519f87d11",
   "metadata": {},
   "source": [
    "Let's now export our top concordant peaks into a <code>.narrowPeak</code> file. Recall that the files that we've been working with are essentially plain text files with tabs as a delimiter to separate values that just have a custom file extension indicating that the file has a particular format/organization to its data. \n",
    "\n",
    "We can make use of <code>pd.DataFrame.read_csv()</code> to export our DataFrame containing our top concordant peaks. We don't need to export our <code>header</code> nor our <code>index</code>, so you'll have to modify how <code>.read_csv()</code> is running.\n",
    "\n",
    "You can go ahead and export it to your <code>Week_8</code> directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4344b122-6935-4e5a-9c84-506276235b94",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a3b5796c-2d25-4feb-9ead-49dbb8e52401",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Motif discovery</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7d99383-f9d6-41f3-897e-4d2c95568177",
   "metadata": {},
   "source": [
    "Let's return to Terminal and HOMER. Here, we'll make use of the core function of HOMER, which is to identify motifs. To do this, we'll need to go back to Terminal to run the <code>findMotifsGenome.pl</code> program. \n",
    "\n",
    "<a href=\"http://homer.ucsd.edu/homer/ngs/peakMotifs.html\" rel=\"noopener noreferrer\"><u>Documentation for <code>findMotifsGenome.pl</code> can be found here.</u></a>\n",
    "\n",
    "If we look into the documentation, we can see that the basic command line structure to use this code is:\n",
    "\n",
    "<pre style=\"width: 600px; margin-top: 15px; margin-bottom: 15px; color: #000000; background-color: #EEEEEE; border: 1px solid; border-color: #AAAAAA; padding: 10px; border-radius: 15px; font-size: 12px;\">findMotifsGenome.pl &lt;peak file&gt; &lt;genome version&gt; &lt;output directory&gt; -size &#35; &lbrack;options&rbrack;</pre>\n",
    "\n",
    "So to run the <code>findMotifsGenome.pl</code> program, we will supply it with our <code>.narrowPeak</code> file containing our focused list of overlapping and reproducible peaks, the genome version, an output directory, a fragment size, and any additional options."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "011c848a-8aea-49f0-b056-c67298ec7855",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38316ede-1ace-4f8d-9290-5dfb8756d811",
   "metadata": {},
   "source": [
    "<h2>Let's break down the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66a6f464-b2f4-4f4d-88c6-11a4585d097a",
   "metadata": {},
   "source": [
    "<code>findMotifsGenome.pl</code>\n",
    "\n",
    "We call up the findMotifsGenome program.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>~/MCB201B_F2025/Week_8/top_concordant_peaks.narrowPeak</code>\n",
    "\n",
    "We provide it with our <code>.narrowPeak</code> file containing just our focused set of overlapping peaks.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>hg19</code>\n",
    "\n",
    "We indicate the genome version that it will use to pull sequences and find motifs. Recall that when we first set up HOMER, we preparsed the genome for 50bp fragments, and now we will be using that preparsed genome for motif discovery.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>./</code>\n",
    "\n",
    "We indicate the output directory as our current working directory.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-size 50</code>\n",
    "\n",
    "We indicate the fragment size around the peak center where we want it to look for motifs. Here, it will look for motifs in the region &pm;50bp from the peak center.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-S 5</code>\n",
    "\n",
    "This option specifies how many <i>de novo</i> motifs we want it to look for. The default option is 25, which is a lot. To keep the runtime down, we'll have it look for just 5 <i>de novo</i> motifs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15d4d324-c0be-4a3f-a7cf-6e4f00a310f7",
   "metadata": {},
   "source": [
    "<h2>Take a look at the identified motifs</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fdc6140-c51b-4ff1-beda-90e2428c57bc",
   "metadata": {},
   "source": [
    "Let's go ahead and take a look at the HTML output files to see what motifs we have identified within our peaks."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5091f52f-d3e6-43db-a82e-592086817294",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Visualize motif distributions</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e621bfdd-5ea3-4fc8-9af5-b95560466e08",
   "metadata": {},
   "source": [
    "Now let's take a look at all the distribution of all our various TEAD motifs around our peak center just to get an idea of where TEAD might be binding relative to the center of our peaks, which we know should theoretically map to where TEAD+TAZ binds to DNA.\n",
    "\n",
    "To do this, we'll make use of annotatePeaks' histogram mode, but we'll provide it with our TEAD motifs that we've found through our analysis. What annotatePeaks will do is go through our focused set of peaks and determine the distribution of our TEAD motifs around the center of all the peaks."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "852664f5-f648-4366-a2fd-5a9fd981150d",
   "metadata": {},
   "source": [
    "<h2>Obtain motif distributions around peaks</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b950e6e0-55fc-4ee7-bbce-2085f7259a34",
   "metadata": {},
   "source": [
    "To make things a little more convenient, we'll stay in our <code>motif</code> directory but specify our output into our <code>annotation</code> directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b3829be-eb12-47d1-bd90-56dd22c45ddb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fdff7d8-59c9-4a42-a391-90b13f21f897",
   "metadata": {},
   "source": [
    "<h2>Let's break down the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2c7e37a-9fb1-4ed5-840a-0070dfbc3a31",
   "metadata": {},
   "source": [
    "<code>annotatePeaks.pl</code>\n",
    "\n",
    "This calls up the annotatePeaks program.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>~/MCB201B_F2025/Week_8/top_concordant_peaks.narrowPeak</code>\n",
    "\n",
    "This provides HOMER with our focused set of overlapping peaks.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>hg19</code>\n",
    "\n",
    "We specify the genome version.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-m</code>\n",
    "\n",
    "This tells the program that we want it to do an analysis of a specific motif or set of motifs. In our case, we are providing it with motifs that we want it to use to determine their distributions within our focused set of peaks.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>&lt;.motif files&gt;</code>\n",
    "\n",
    "We provide the motif file(s) that we want the annotatePeaks program to use. We can supply it with a single motif file or more than one. In either case, the output will be a single file that contains information for each motif we provide it. Each motif will have 3 corresponding columns of data in the output: one for total counts, one for (+) counts, and one for (-) counts.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-size 2000</code>\n",
    "\n",
    "We specify the size of the region of interest that we're interested in looking at motif distributions. With a size set to 2000, we will be looking at a region &pm;1000bp centered on our peaks.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-hist 20</code>\n",
    "\n",
    "We invoke the histogram mode of annotatePeaks and provide it with the bin size that it will use to determine the distribution of motifs in our specified region.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>&gt; motif_distribution.txt</code>\n",
    "\n",
    "We redirect the output to a file called <code>motif_distribution.txt</code> that we can then import into this Python notebook to plot."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d54823a8-2b12-4155-b3f1-5f3789a1185d",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise Set 4</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84265cc4-501e-4c89-97f3-2d704713daab",
   "metadata": {},
   "source": [
    "This exercise set will have you plot the motif distributions of your TEAD motifs that you analyzed with HOMER. This way, we can get an idea of how the motifs are distributed around our peaks, which can tell us more about how TEAD+TAZ is recognizing and binding to DNA."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fa16132-e55d-45ac-9250-6bb6cbf70498",
   "metadata": {},
   "source": [
    "<h2>Exercise #1: Import motif distribution file</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c0ddb13-09b5-4b1c-b3f1-9fd81b779199",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2c6422d1-a537-4763-81f9-624ae66ec011",
   "metadata": {},
   "source": [
    "<h2>Exercise #2: Plot motif distributions as a line plot</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b57f75df-c959-4a66-bee9-3342da995109",
   "metadata": {},
   "source": [
    "For this exercise, see if you can plot the motif distributions for TEAD1-4 and TEAD (family). While there are still unwieldy column headers, we only really care about the ones for total sites, so we can pull those out to plot. We can just update the first column header and pull the others as needed when plotting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c005687f-d8ff-4a9d-b84c-2d12fb809e3c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ac3e063c-b88c-4ec2-bd8b-852b28ab787d",
   "metadata": {},
   "source": [
    "What does this motif distribution tell us about how TEAD+TAZ binds to DNA?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c2bc1de-e313-40d5-abf5-f7f9137cb32b",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">References</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cf61a94-9865-409a-8acc-daef73261ab7",
   "metadata": {},
   "source": [
    "<a href=\"https://www.nature.com/articles/ncb3216\" rel=\"noopener noreferrer\"><u>Zanconato et al 2015 Nat Cell Biol:</u></a> Genome-wide association between YAP/TAZ/TEAD and AP-1 at enhancers drives oncogenic growth\n",
    "\n",
    "<a href=\"https://genome.cshlp.org/content/22/9/1813\" rel=\"noopener noreferrer\"><u>Landt et al 2012 Genome Res:</u></a> ChIP-seq guidelines and practices of the ENCODE and modENCODE consortia\n",
    "\n",
    "<a href=\"https://projecteuclid.org/journals/annals-of-applied-statistics/volume-5/issue-3/Measuring-reproducibility-of-high-throughput-experiments/10.1214/11-AOAS466.full\" rel=\"noopener noreferrer\"><u>Li et al 2011 Ann Appl Stat:</u></a> Measuring reproducibility of high-throughput experiments"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
