{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1519adb7-4ab9-4fe1-9678-6c9cdf0ddc6f",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">8.2 Exploring chromatin domains</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0910c41f-4eab-4662-8d3a-13d3cb985019",
   "metadata": {},
   "source": [
    "If we have time, we'll take a look at chromatin markers that signify an active enhancer region, specifically around our focused set of overlapping peaks. This will give us an idea of which of our peaks overlap with enhancers that have been marked as active based on the region's epigenetic signature.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "\n",
    "<ul>\n",
    "    <li>Continue to practice using HOMER to analyze CHIP-seq data</li>\n",
    "    <li>Plot chromatin ChIP-seq data</li>\n",
    "    <li>Infer transcriptional activity based on epigenetic signatures</li>\n",
    "    <li>Plot TF read coverage with histone read coverage</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9990533b-13f5-45e6-9237-215247b75cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93a8132c-824f-4128-93c2-2a3d333e2762",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Analyze epigenetic signatures</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "951d9697-57b7-4eef-a365-f1f933a1a8c2",
   "metadata": {},
   "source": [
    "I have a <code>histone</code> directory located within our shared <code>chip</code> directory, containing the bedGraph files from <a href=\"https://doi.org/10.1186/1471-2164-15-331\" rel=\"noopener noreferrer\"><u>Rhie et al 2014</u></a> (<a href=\"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE49651\" rel=\"noopener noreferrer\" target=\"_blank\"><u>GSE49651</u></a>), which Zanconato et al 2015 also used in their analyses. We'll be using their read coverage data to get an idea of the epigenetic signature around our peaks. \n",
    "\n",
    "This notebook will be on the independent side, where you'll make use of the commands we used in notebook 8-1 but applied now to chromatin coverage data. As you're analyzing the data, you'll want to think about what the epigenetic signatures are telling us about the activity of the enhancers that is associated with these set of peaks and what that means for TEAD+TAZ regulation of gene transcription."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f78f5c3-ee6d-4499-a791-2922d8e7bf55",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise set</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b4efd87-73d4-40c6-8801-067295526ba5",
   "metadata": {},
   "source": [
    "<h2>Exercise #1: Obtain average coverage profile around our concordant peaks</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30576049-3b50-4bd8-a7a5-ac443cdd0c03",
   "metadata": {},
   "source": [
    "Using the focused set of peaks that we identified in notebook 8-1, use HOMER to analyze the epigenetic signatures in a &plusmn;1kb region. You'll need to separately analyze each <code>.bedGraph</code> file in the <code>histone</code> subdirectory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47d2e9b6-7c54-4569-bad6-d63d74a1fdef",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96cf3641-6ed0-4f52-953f-5a114c10d23e",
   "metadata": {},
   "source": [
    "<h2>Exercise #2: Plot the profile for all three histone marks overlaid on a single plot</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8201731-77fd-49ec-b245-5bbe924a4054",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "467bebbf-5188-44f0-862f-a55a7abbef6f",
   "metadata": {},
   "source": [
    "What does the average epigenetic signature tell us about how TEAD+TAZ is potentially regulating transcription of their target genes for these top peaks we're focusing on?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a2ce6e6-af63-496a-831e-fc7386f55989",
   "metadata": {},
   "source": [
    "<h2>Exercise #3: Overlay our two TAZ replicate's plot profiles</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b7c3557-2167-4701-85d9-b1c5ac204e2a",
   "metadata": {},
   "source": [
    "Now let's get the plot profile for both TAZ replicates but just for the focused set of peaks and then overlay that on top of the histone coverage plot that we generated. So we can rerun our HOMER analysis, but using now our concordant peaks file instead of the full set of peaks for each replicate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cec9262-c9b3-4df1-a856-4773b09763fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe43b2a1-137c-45d1-a47c-bdf5e1eaf9e5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b1e8f951-5809-4eb1-9712-3d162d49443b",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">References</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "359729bf-6f95-4617-a15a-d927bbbd0572",
   "metadata": {},
   "source": [
    "<a href=\"https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-15-331\" rel=\"noopener noreferrer\"><u>Rhie et al 2014 BMC Genomics:</u></a> Nucleosome positioning and histone modifications define relationships between regulatory elements and nearby gene expression in breast epithelial cells"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
