{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c0f082e-15c6-4b1c-8ffe-88f9ceaedd87",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">7.1 ChIP-seq alignment</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77c2444a-90bb-4a58-8c25-197495aef49f",
   "metadata": {},
   "source": [
    "Today, we'll align our datasets to the genome using <code>bowtie2</code>. After we perform our alignments, we'll take a look at the resulting files generated by the aligner, and then we'll load the files into the Integrative Genome Viewer (IGV) to take a look at where reads have aligned to the genome. We'll explore how the reads can be visualized, and how we can begin to see reads piling up on either side of our protein of interest.\n",
    "\n",
    "While you can follow along in this notebook running command line, there will be parts where we will run things simultaneously, so you'll want to be able to run it in the background in Terminal.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "<ul>\n",
    "    <li>Learn to use bowtie2 to perform alignments</li>\n",
    "    <li>Explore bowtie2 outputs</li>\n",
    "    <li>Learn to use samtools to work with alignment outputs</li>\n",
    "    <li>Learn to use IGV to visualize aligned reads</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4b39e0f-83f8-426e-ada3-f1b47573e8f3",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">ChIP-seq alignment with <code>bowtie2</code></h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "383049f2-5dd7-435d-a34b-ca3c6012a4a5",
   "metadata": {},
   "source": [
    "Today, we'll be aligning our four truncated datasets to our reference genome in class, and your homework this week will be to align the full datasets. Please note that the alignment can take quite a long time (~3 hours or so).\n",
    "\n",
    "Since Zanconato et al 2015 aligned their ChIP-seq data to the human genome build hg19, we'll be doing the same. We'll be making use of an aligner that Dr. Ingolia talked about in lecture, <code>bowtie2</code>. <a href=\"https://bowtie-bio.sourceforge.net/bowtie2/index.shtml\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Documentation for <code>bowtie2</code> can be found here.</u></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58c13d65-e24b-46d6-9547-9ba3b176bdf4",
   "metadata": {},
   "source": [
    "<h2>Alignment</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83b7a605-f1d0-40f0-bdbf-83aea8e5bdac",
   "metadata": {},
   "source": [
    "We'll be doing things a little bit in reverse today to help save time, since even with the truncated dataset, the alignment still takes some time to run. In this case, it should take ~20-30 minutes. So what we'll do is we'll first set up an alignment run, and then we'll go into the technical details about running the alignment as it runs in Terminal."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fed4fa8a-a516-4d65-bb28-3aafd5bce845",
   "metadata": {},
   "source": [
    "<h3>Prepare directory to receive alignment outputs</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd82f8b8-3b01-4f85-b898-c52edc692b75",
   "metadata": {},
   "source": [
    "Let's go ahead and prepare a new directory <code>alignment</code> that can receive our output, which will be our alignment files, and once we've made <code>alignment</code>, let's change into it. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9b9d51e-9f50-4e48-9ad4-9e62899def9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6278257-5b46-4c35-b4b9-f47b9dbe68fd",
   "metadata": {},
   "source": [
    "<h3>Align sequence data to h19 reference genome</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70a8e7cc-a6bd-4d75-9370-f6c697a0d88e",
   "metadata": {},
   "source": [
    "Let's align our first sequence file using <code>bowtie2</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7227eed1-0995-4b94-8986-ff4ff1992c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b668433f-4036-4b65-98f3-41311365213e",
   "metadata": {},
   "source": [
    "<h2>Bowtie2</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16a65e8d-bdcd-4992-80ce-643318d294b3",
   "metadata": {},
   "source": [
    "While that's running, let's review alignments, specifically looking at <code>bowtie2</code>.\n",
    "\n",
    "Aligning sequences to the genome can be quite tricky and computationally intensive. Recall from lecture that the sequences may not always be a perfect match, and in some cases, there may be short gaps in the sequences to get a good alignment (<strong>Fig 1</strong>). Moreover, the sequence reads that we are trying to align are usually around 50-150bp with millions of read for each sequence file, and each of those reads needs to be aligned to the reference genome, which for humans is approximately 3 billion basepairs.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 1</strong></h4>\n",
    "<img src=\"./images/7_1_fig_1.png\" style=\"height: 250px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Image from: Harvard Chan Bioinformatics Core</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ad719c5-dc94-4657-bbc6-dd04315943ba",
   "metadata": {},
   "source": [
    "<h3>Reference genome and index</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79bc5bbf-fe98-4884-9733-3c362c105556",
   "metadata": {},
   "source": [
    "As Dr. Ingolia mentioned in lecture, alignment software use a genome index. While theoretically, you can align using the reference genome, converting the reference genome into an organized index enables faster searches through the genome to find alignments. As a result, <code>bowtie2</code> is able to efficiently and relatively quickly align sequence reads. To keep it's memory usage on the lower end, <code>bowtie2</code> uses an FM Index (which uses the Burrows-Wheeler Transform mentioned in lecture).\n",
    "\n",
    "So normally before you perform your alignments, you will need to build an index from your reference genome, which can take some time, especially for larger genome sizes. Fortunately for us, bowtie2's documentation also provides a link to a pre-built index for hg19. The index files can be found in the shared folder for this course, and we've called them up when we ran our first alignment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2d4857e-7d48-4d44-8722-208c22770856",
   "metadata": {},
   "source": [
    "<h3>Alignment</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be76128a-ec26-4827-86eb-456dc8c02771",
   "metadata": {},
   "source": [
    "To get an idea of how to use bowtie2 to align our sequences, let's take a look at how its documentation notes to set it up:\n",
    "\n",
    "<pre style=\"width: 700px; margin-top: 15px; margin-bottom: 15px; color: #000000;background-color: #EEEEEE; border: 1px solid; border-color: #AAAAAA; padding: 10px; border-radius: 15px; font-size: 12px;\">&dollar;bowtie2 &lbrack;options&rbrack;* -x &lt;bt2-idx&gt; {-1 &lt;m1&gt; -2 &lt;m2&gt; &vert; -U &lt;r&gt; &vert; --interleaved &lt;i&gt; &vert; -b &lt;bam&gt;} &lbrack;-S &lt;sam&gt;&rbrack;</pre>\n",
    "\n",
    "So to use the bowtie2 aligner, you first specify the <code>bowtie2</code> command followed by options and arguments."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc78b723-f1d1-4b80-bd35-43583c04ba05",
   "metadata": {},
   "source": [
    "<h2>Let's breakdown the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0565fc8-7353-4ae3-b644-9e5d9890e6c4",
   "metadata": {},
   "source": [
    "Now let's go back and take a look at the code that we ran to align our <code>10M_ctrl_1.fastq</code> sequence file and break down the input with respect to the information we pulled from the manual.\n",
    "\n",
    "<code>bowtie2</code> \n",
    "\n",
    "This is the command to call up the aligner.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-x ~/shared/2025-fall/courses/1547808/chip-seq/chip-index/hg19</code> \n",
    "\n",
    "This argument specifies the file path to where the index files are located along with the basename of the index files.\n",
    "\n",
    "So for our input, we provided the file path to where the index files were located <code>~/shared/2025-fall/courses/1547808/chip-seq/chip-index/</code> along with the basename of our index files <code>hg19</code>.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-U ~/shared/2025-fall/courses/1547808/chip-seq/truncated/10M_ctrl_1.fastq</code> \n",
    "\n",
    "This argument specifies the file that you want to align. <code>bowtie2</code> is able to align paired-end reads indicated as <code>-1 &lt;m1&gt; -2 &lt;m2&gt;</code>, or unpaired reads indicated as<code>-U &lt;r&gt;</code>, interleaved <code>.fastq</code> files indicated as <code>--interleaved &lt;i&gt;</code>, and unaligned BAM files indicated as<code>-b &lt;bam&gt;</code>.\n",
    "\n",
    "So in our case, the ChIP-seq dataset that we're working with is unpaired, so we can use <code>-U</code> to specify that we're providing unpaired sequence reads followed by the file name for our sequence file.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-S 10M_ctrl_1.sam</code>\n",
    "\n",
    "Here, we specify our output file name, and if we want to save the SAM file in a directory that is not our current working directory, we can include the file path as well. But since we'll be saving the output to our current directory, we can just specify the file name."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a25e7381-3fc2-43e7-bca1-ad33e62c1bef",
   "metadata": {},
   "source": [
    "<h2>Perform alignment for 10M_taz_1</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daedf049-a69e-461d-af8f-3d98015aa4a3",
   "metadata": {},
   "source": [
    "Now that we've done an example for a control file, let's run an alignment using the 10M_taz_1 dataset <u>in the background</u> by adding an <code>&amp;</code> at the end of our input. Once everyone's alignment is running, we'll continue with the next section, while <code>bowtie2</code> runs in the background.\n",
    "\n",
    "If time permits, we can also run alignments for the other <code>.fastq</code> files, but for now, we'll just run this one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f99d0f7-c6c3-447c-9fbd-c0788ffa8041",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ea7b8ab-2d12-4e3b-979d-4d96307b05af",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exploring <code>bowtie2</code> output</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "061a796a-c0c3-4f12-a65f-324d768c1c70",
   "metadata": {},
   "source": [
    "<h2>Reading the SAM file</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38136fc8-f204-4203-8c48-499a30184e67",
   "metadata": {},
   "source": [
    "Recall from lecture that SAM files have a specific structure (<strong>Fig 2</strong>), where an aligned sequence read retains its original identifying information but now this information is also associated with a specific genomic location, and things are just slightly rearranged a bit.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 2</strong></h4>\n",
    "<img src=\"./images/7_1_fig_2.png\" style=\"height: 300px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Image from Dr. Nick Ingolia</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deab3821-875b-4e14-a5f0-1b8df2a04e69",
   "metadata": {},
   "source": [
    "Let's take a look at our <code>10M_ctrl_1.sam</code> output file generated by <code>bowtie2</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e4f89e5-1c25-49ba-8b5e-ffa8186b95b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c12eea3-2548-4d91-8852-fc5a47a1aaef",
   "metadata": {},
   "source": [
    "You should see a lot of header lines indicated by an &commat;. The alignment information are the lines after those header lines.\n",
    "\n",
    "There are a fair number of columns to the SAM file, so you probably see that a single line will wrap to the next line, making it more difficult to read. You might have noticed that the values within the SAM file are separated by tabs, so try to export just the first 100 lines of our SAM file into a tab-separated values <code>.tsv</code> file. That way, we can open it up just the first 100 lines in excel to see each column more easily.\n",
    "\n",
    "Once we've exported the first 100 lines into a <code>.tsv</code> file, let's open it up in Excel to view its contents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ee7159d-19b6-4356-acf3-90da9841baff",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52f6ea4f-011f-4dae-95e2-1ece5fb51f6f",
   "metadata": {},
   "source": [
    "Unfortunately, the columns do not come with headers, so <a href=\"https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#sam-output\" rel=\"noopener noreferrer\"><u>we'll have to look into the <code>bowtie2</code> documentation to figure out what each column indicates.</u></a>\n",
    "\n",
    "To figure out what the FLAG column is telling us, we can look back to the bowtie2 documentation, or <a href=\"https://broadinstitute.github.io/picard/explain-flags.html\" rel=\"noopener noreferrer\"><u>we can also use the Broad Institute's sam FLAG  decoder.</u></a>\n",
    "\n",
    "For the optional tags, their definitions can be found also in the <code>bowtie2</code> documentation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bc52a7d-6617-46c1-8ac5-96ab22285c82",
   "metadata": {},
   "source": [
    "<h2>Sort and index alignments</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aee210d5-9845-4a14-a2f4-6e3dbd1fe721",
   "metadata": {},
   "source": [
    "Recall from lecture that genome viewers usually assume that the alignment data that you give it is sorted by its mapped location. The genome viewer we will be using (IGV) requires us to provide it with a sorted and indexed BAM alignment file in order for it to display the aligned reads."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1471c8fb-b590-4e67-8428-4c3ee9eb0c4c",
   "metadata": {},
   "source": [
    "<h3>Convert SAM to BAM</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a72edf1-800a-431f-b63e-138d86c0e2ca",
   "metadata": {},
   "source": [
    "The first thing we'll need to do is to convert our SAM file to a BAM file, which is a compressed, binary file. To do this, we'll use the <code>samtools</code> command. <a href=\"https://www.htslib.org/doc/samtools.html\" rel=\"noopener noreferrer\"><u>Documentation for <code>samtools</code> can be found here.</u></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "803e1075-75db-430c-8f84-d8f9bad112db",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24f76f9e-a822-4df0-b813-4a000ead97b9",
   "metadata": {},
   "source": [
    "<h2>Let's breakdown the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2b01fde-f3f8-452c-b630-cfe51ca6eb5d",
   "metadata": {},
   "source": [
    "<code>samtools</code>\n",
    "\n",
    "This is the command to call up <code>samtools</code>.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>view</code> \n",
    "\n",
    "This indicates that we want to make use of the sub-command <code>view</code>, which is used to convert SAM to BAM and vice versa.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-b</code> \n",
    "\n",
    "This tells the view sub-command that we want it to output a BAM file, so essentially convert the SAM file to a BAM file.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-o 10M_ctrl_1.bam</code> \n",
    "\n",
    "This specifies the output file information. Here is where we provide the sub-command with the file name.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>10M_ctrl_1.sam</code> \n",
    "\n",
    "This is the name of the file that we are providing to the sub-command to convert to a BAM file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44ab12f9-5f62-4dfb-8c07-c6f18d7025af",
   "metadata": {},
   "source": [
    "<h3>Sort the BAM file</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e2201b3-a1be-4dbd-820d-3fb1d6173372",
   "metadata": {},
   "source": [
    "To sort your BAM file, you can use the <code>sort</code> sub-command within the <code>samtools</code> command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "651c7337-e154-4db9-b56e-9e045e5df9df",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efa2aea2-7f6a-4587-aa62-b89a990ae024",
   "metadata": {},
   "source": [
    "<h2>Let's breakdown the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7186d538-9732-4bad-a0d3-800720f4cf98",
   "metadata": {},
   "source": [
    "<code>samtools</code> \n",
    "\n",
    "This is the command to call up <code>samtools</code> just like before.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>sort</code> \n",
    "\n",
    "This calls up the <code>sort</code> sub-command, which gives us the ability to sort our aligned reads by their chromosomal position.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-o 10M_ctrl_1-sorted.bam</code> \n",
    "\n",
    "This specifies the output file information.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>10M_ctrl_1.bam</code> \n",
    "\n",
    "This is the file that contains the alignments that we want to sort based on their chromosomal position."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47ed0c3b-34f1-4b46-a897-58a92b6bd4b4",
   "metadata": {},
   "source": [
    "<h3>Index the sorted BAM file</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ef82562-a112-4dcb-a717-fbb0dc9a86d6",
   "metadata": {},
   "source": [
    "Like with how we needed to create a genome index to allow <code>bowtie2</code> to quickly access the sequence information, we need to index our sorted BAM file to allow for quick access to our alignment data. To index our sorted BAM files, we will use the <code>index</code> sub-command and provide it with the BAM file we want it to index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96f16177-b458-43a6-8c58-41f4564dbcd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67033341-e106-4211-aa39-6f2c8aeda775",
   "metadata": {},
   "source": [
    "You should now have three outputs from converting, sorting, and indexing:\n",
    "\n",
    "<ul>\n",
    "    <li><code>10M_ctrl_1.bam</code></li>\n",
    "    <li><code>10M_ctrl_1-sorted.bam</code></li>\n",
    "    <li><code>10M_ctrl_1-sorted.bai</code></li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8c4059b-ee9c-4afe-a931-9dbe0357c59d",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Visualize aligned reads</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "582acf3f-1e92-47e0-90df-e1a7fc582e54",
   "metadata": {},
   "source": [
    "To prepare to visualize your aligned reads in IGV, download the following files:\n",
    "\n",
    "<ol>\n",
    "    <li>Your sorted BAM file: <code>10M_ctrl_1-sorted.bam</code></li>\n",
    "    <li>The sorted and indexed BAM file: <code>10M_ctrl_1-sorted.bai</code></li>\n",
    "</ol>\n",
    "\n",
    "Now, let's open up IGV and visualize our alignments."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b0a82f4-948b-4d63-ab6b-a09b77acf28c",
   "metadata": {},
   "source": [
    "If there's time, we can process our <code>10M_taz_1.sam</code> file for visualization as well."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "799d47de-5b13-4390-8079-6f2909d3c162",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">References</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c5aa707-7c39-4632-8e84-258ac82a12e4",
   "metadata": {},
   "source": [
    "<p style=\"padding-left: 20px;\"><a href=\"https://pubmed.ncbi.nlm.nih.gov/19549630/\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Whiteford et al 2009 Bioinformatics:</u></a> Swift: primary data analysis for the Illumina Solexa sequencing platform</p>\n",
    "\n",
    "<p style=\"padding-left: 20px;\"><a href=\"https://bionumbers.hms.harvard.edu/bionumber.aspx?id=100679&ver=5&trm=gc+content+human+genome&org=\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Bionumbers for human genome GC content</u></a></p>\n",
    "\n",
    "<p style=\"padding-left: 20px;\"><a href=\"https://www.nature.com/articles/ncb3216\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Zanconato et al 2015 Nat Cell Biol:</u></a> ChIPâ€“seq: advantages and challenges of a maturing technology</p>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
