{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1579e369-02b8-476a-a638-b2fbe98d6598",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">6.1 Intro to ChIP-seq pipeline</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6ef9b79-3e8a-40df-a62c-f1504006ec6e",
   "metadata": {},
   "source": [
    "We're now moving on to the ChIP-seq analysis module of this course. The goal of this module is to use big data analysis to identify the genes that may be regulated by TAZ by looking at the regions of DNA to which TAZ (and its associated TEAD DNA-binding protein) actually bind. In class, we'll practice using a truncated dataset to allow us to perform the analyses on a small subset of the larger dataset, and outside of class (aka for homework assignments), you will run the analysis using the full dataset. If there's time in class, you can also start running analyses on the complete dataset to save some time. The full dataset is rather large, so the analysis will take some time to run. You will want to keep this in mind, since your assignments will require you to turn in the results of your analysis.\n",
    "    \n",
    "To start, we'll go over the basic concepts that underlie ChIP-seq, and then we'll have a small command line refresher and also learn more about command line syntax to prepare ourselves to work with our sequencing files. We'll then run a QC to understand the quality of our sequencing data.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "<ul>\n",
    "    <li>Understand concepts behind ChIP-seq</li>\n",
    "    <li>Review command line/Terminal</li>\n",
    "    <li>Learn more about how to work in Terminal</li>\n",
    "    <li>Learn to make sense of your environment</li>\n",
    "    <li>Learn to perform QC on sequencing data</li>\n",
    "    <li>Understand how to interpret QC results</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51f3876e-b019-48cf-907a-038ef0584793",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">ChIP-seq overview</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b89c2101-4e71-4755-aa92-e5c81e25a338",
   "metadata": {},
   "source": [
    "<h2>Experimental design</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e128ee3-b32d-4968-b329-2bd0db447a4c",
   "metadata": {},
   "source": [
    "One of the main concepts that underlies ChIP-seq is a technique called chromatin immunoprecipitation (ChIP). Like with the more general technique of immunoprecipitation, ChIP takes advantage of the specificity of antibodies in order to select for specific proteins, namely proteins that interact with DNA either directly (such as histones) or indirectly through interactions with a DNA-binding protein (Park 2009).\n",
    "\n",
    "Prior to immunoprecipitation, samples are crosslinked with formaldehyde in order to stabilize molecular interactions, such as protein-protein or protein-DNA interactions (<strong>Fig 1</strong>). Then the DNA is fragmented via sonication, producing smaller fragments of DNA ranging from 200 to 600 base pairs, with any associated proteins still bound to it. DNA fragmentation is followed by immunoprecipitation using antibodies specific to the target protein or interest. As the target protein is pulled down, any bound DNA is pulled down along with it, allowing us to enrich for the regions in which the target protein is closely associated. To release the DNA from the protein, the crosslinking reaction is reversed, and the DNA can be purified. Afterwards, a sequencing library is prepared from the released DNA fragments, and next-generation sequencing (NGS) is carried out.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 1</strong></h4>\n",
    "<img src=\"./images/6_1_fig_1.png\" style=\"height: 450px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Image from: Park 2009 Nat Rev Genet</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe4597ed-b3dd-496a-ae52-09e2e240ad94",
   "metadata": {},
   "source": [
    "<h2>ChIP-seq analysis pipeline overview</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f138468a-2c48-4d15-98c3-7ea94cd508ae",
   "metadata": {},
   "source": [
    "Below is a brief overview of the main steps taken to analyze ChIP-seq data, and in later lectures, you'll learn about them in more detail."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dbe5e56-b4d5-4b09-b6c3-fd4ebbd40ca5",
   "metadata": {},
   "source": [
    "<h3>Raw sequence file</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c929f47-9c36-4590-b9b8-5d0fbac573d3",
   "metadata": {},
   "source": [
    "As you'll recall from MCB201A, NGS allows for massively parallel sequencing of the DNA fragments in a library prep. The resulting sequence output contains all the sequence reads and the associated quality information for each base. The type of file it usually comes as is a compressed fastq (<code>.fastq.gz</code>) file. A <code>.fastq</code> file has a specific structure, where each sequencing read corresponds to 4 lines within the file, so a fastq file containing the sequence information for 10 reads will be 40 lines long.\n",
    "\n",
    "Each line for a sequence read contains a specific set of information:\n",
    "\n",
    "<table style=\"margin:auto;\">\n",
    "    <tr style=\"border: 1px solid; border-color: #000000;\">\n",
    "        <th>Line</th>\n",
    "        <th>Details</th>\n",
    "    </tr>\n",
    "    <tr style=\"border: 1px solid; border-color: #000000;\">\n",
    "        <td>1</td>\n",
    "        <td>A sequence identifier that contains information on the run</td>\n",
    "    </tr>\n",
    "    <tr style=\"border: 1px solid; border-color: #000000;\">\n",
    "        <td>2</td>\n",
    "        <td>The sequence information</td>\n",
    "    </tr>\n",
    "    <tr style=\"border: 1px solid; border-color: #000000;\">\n",
    "        <td>3</td>\n",
    "        <td>A separater line. Begins with <strong>&plus;</strong> and sometimes followed by the sequence identifier</td>\n",
    "    </tr>\n",
    "    <tr style=\"border: 1px solid; border-color: #000000;\">\n",
    "        <td>4</td>\n",
    "        <td>Quality score information for each nucleotide encoded using ASCII to represent the numerical quality value</td>\n",
    "    </tr>\n",
    "</table>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "762c4900-69e3-4517-ac81-0a4471961cb0",
   "metadata": {},
   "source": [
    "How this looks like if you pull out the information for a single sequence read from a fastq file is something like:\n",
    "\n",
    "```\n",
    "@LH00328:659:235GCFLT3:5:1101:37786:1048 1:N:0:ATTCTGGC+TCTGGACA\n",
    "CNTCCGGATGGCTTTGTTGATCCGCTGCACAGCGTGGAGCATGGCTTGTTCCTGATCACCCTTTGTGTTGGCTGCAGCCACACCAGCCTGGACTTCCTCCTTTTCCAGAAGCTCCACCAGGCCCAGCTCCTGTGCCTTCTGCTCTCTGTC\n",
    "+\n",
    "I#IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII9IIIIIIIII9IIIIIIIIIIIIIIIIIIIIIIIII\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bfe0c36-02ab-4b45-8726-a17d30ccdcd0",
   "metadata": {},
   "source": [
    "The quality score contained within the fourth line tells us the probability that the called base is incorrect. In other words, it tells us the likelihood that the sequence read is incorrect for that specific nucleotide. Using the example above, we can see that the first nucleotide <code>C</code> has an associated quality score encoded as <code>I</code>, which corresponds to a Phred-33 quality score of 40 (<a href=\"https://people.duke.edu/~ccc14/duke-hts-2018/bioinformatics/quality_scores.html#Phred-Quality-Scores\" rel=\"noopener noreferrer\" target=\"_blank\"><u>More information on Phred quality scores can be found here</u></a>). This tells us that the probability of the first base being incorrectly sequenced is 0.00008.\n",
    "\n",
    "<code>.fastq</code> files are often incredibly large, so they are usually compressed into <code>.fastq.gz</code> file. Most programs that analyze sequencing data are able to handle both <code>.fastq</code> and <code>.fastq.gz</code> files, so you don't need to unzip the .fastq.gz file to use them for analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c0332fc-ad24-4528-9c8c-f7e1bb6c1746",
   "metadata": {},
   "source": [
    "Let's jump into Terminal and pull out some sequence information from one of our truncated <code>.fastq</code> files.\n",
    "\n",
    "For this, you have the option of following along either in your notebook, where you will run bash commands in a code cell or by following along in Terminal. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "039f5161-70fa-4c5e-81fe-b5d253d352f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR1810912.1 HWI-ST1210:136:C1RBDACXX:5:1101:1472:2121 length=50\n",
      "CACTGTAATTTCCCTAAAACAGGAGGTCACGGTAAGTCCTATGGATAGAT\n",
      "+SRR1810912.1 HWI-ST1210:136:C1RBDACXX:5:1101:1472:2121 length=50\n",
      "??????????????????????????????????????????????????\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#Each cell where we want to run command line in this notebook will start with %%bash\n",
    "#This opens up a temporary shell for us to run command line\n",
    "#You'll want to keep in mind that you need to do this for each code cell,\n",
    "#and each instance is separate, so if you change directories in one\n",
    "#It won't be reflected in another\n",
    "\n",
    "#We're pulling out the first four rows\n",
    "#Each read in a fastq file has 4 rows of information\n",
    "#So here, we're only pulling out a single read\n",
    "head -4 ~/shared/2025-fall/courses/1547808/chip-seq/truncated/10M_ctrl_1.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e97857f0-5b42-47af-8ce9-56e97bc89544",
   "metadata": {},
   "source": [
    "<h3>Quality assessment</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d07a8be3-8d47-498c-b82d-f5c79a2a0a0d",
   "metadata": {},
   "source": [
    "Prior to performing your analysis, you'll want to check the quality of your sequence reads in order to determine if there might be issues with your samples that you will want to keep in mind when performing your analyses. These issues could have arised from sample prep or the sequencing run itself, and they can point to biases or other confounding factors that can impact the interpretation of your sequencing data. This is the case for essentially anytime you're analyzing your NGS data, and we'll be doing this for our ChIP-seq dataset and later on for our RNA-seq dataset.\n",
    "    \n",
    "A common method to performing quality checks on raw sequence data is called <a href=\"https://www.bioinformatics.babraham.ac.uk/projects/fastqc/\" rel=\"noopener noreferrer\" target=\"_blank\"><u>FastQC</u></a>, developed by Babraham Bioinformatics. FastQC takes in <code>.fastq</code> (and the compressed form <code>.fastq.gz</code>) files and analyzes the raw sequence data with a series of modules that can highlight potential issues in a dataset. <a href=\"https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Information on the different modules and an explanation on how to interpret their results can be found in the FastQC documentation.</u></a>\n",
    "\n",
    "After we run through the overview of ChIP-seq and our command line refresher/practice, we'll run a QC on our ChIP-seq data files using FastQC."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16115ceb-4a23-4a8b-b336-e4b4c006db92",
   "metadata": {},
   "source": [
    "<h3>Genome alignment</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b111ce92-b42b-47c2-849b-8eeeb5f71920",
   "metadata": {},
   "source": [
    "To determine where our sequence reads came from, we need to align them to a reference genome (<strong>Fig 2</strong>), and the resulting alignments allow us to determine the distribution of sequence reads along the genome. In ChIP-seq, the resulting distribution of aligned sequences gives us an idea of the downstream transcriptional targets of a protein of interest. In our case, it will shed light on where TAZ (and TEAD) are binding and provide hints to what genes TAZ may be regulating, since we should be recovering a greater number of fragments corresponding to regions in the genome to which TAZ (and TEAD) are binding.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 2</strong></h4>\n",
    "<img src=\"./images/6_1_fig_2.png\" style=\"height: 250px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Image from: Harvard Chan Bioinformatics Core</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26a075c3-bc85-4539-a6ea-53427a9e7f13",
   "metadata": {},
   "source": [
    "<h3>Peak calling and visualization</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1d0723c-194c-4faa-b658-908793f41f9f",
   "metadata": {},
   "source": [
    "Following genome alignment, regions of high alignment can be determined by peak calling (<strong>Fig 3</strong>). Tools such as MACS2 use the aligned reads to calculate the density of read counts throughout the genome in order to identify regions that have a large number of aligned reads. These peaks in read density tell us that the identified region is enriched within a dataset, suggesting a potential binding site for a protein of interest.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 3</strong></h4>\n",
    "<img src=\"./images/6_1_fig_3.png\" style=\"width: 400px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Image from: <a href=\"#wilbanksfacciotti2010\"><u>Wilbanks and Facciotti 2010 PLoS ONE</u></a></p>\n",
    "\n",
    "After we call our peaks, we'll visualize them using the Integrative Genomics Viewer (IGV), which will allow us to take a look at how our reads pile up and contribute to the determination of peaks by the MACS2 algorithm."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6b04087-116b-40bc-ab4d-1cd111648f77",
   "metadata": {},
   "source": [
    "<h3>Motif discovery</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a6cd886-eaad-4797-9090-9c80bdd9dece",
   "metadata": {},
   "source": [
    "While peak calling can provide us with information on potential binding sites for a protein of interest, we can also perform further analyses to obtain information on specific binding motifs for a protein of interest. This process, called motif discovery, is used to identify similar patterns in peak sequences (<strong>Fig 4</strong>). These patterns can suggest that the enriched sequences are recognized by your protein of interest and can indicate a potential transcription factor binding motif.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 4</strong></h4>\n",
    "<img src=\"./images/6_1_fig_4.png\" style=\"width: 500px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Image from: <a href=\"https://meme-suite.org/meme/doc/examples/meme_example_output_files/meme.html\"><u>MEME Suite</u></a></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18e7e532-d7c3-4854-8cc7-ae840f84cd69",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Data files for this module</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cbd3602-1380-4a3c-8f23-0725e64038c2",
   "metadata": {},
   "source": [
    "Since we didn't perform ChIP-seq on our MDA-MB-231 cells, we'll be making use of a publicly available dataset from <a href=\"#zanconatoetal2015\"><u>Zonconato et al 2015 Nat Cell Biol</u></a>, where they performed ChIP-seq analyses on MDA-MB-231 cells to identify how the YAP/TAZ transcriptional response is mediated.\n",
    "\n",
    "The data files for ChIP-seq can be found in the <code>shared</code> directory. This is a directory that is available to everyone on the Biology Hub, but cannot be modified. Inside the <code>shared</code> directory, you'll find a number of subdirectories. The directory path that we're specifically interested in is <code>~/shared/2025-fall/courses/1547808</code>. There, you'll find our <code>chip-seq</code> and <code>rna-seq</code> subdirectories that will contain the files you will need for ChIP-seq (and our RNA-seq files, too).\n",
    "\n",
    "During class, we'll be doing a guided analysis using truncated sequence files that contain just 10 million reads (vs the almost 100 million reads of the full sequence file). These files have a prefix of <code>10M_</code> to indicate that this is the truncated sequence file. You will take what you've learned in class to perform the analysis on the full dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7533f209-7608-4bec-933c-f4ddb89c3611",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Command line refresher</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcae3d0b-353d-438a-acb8-fbe52d733add",
   "metadata": {},
   "source": [
    "Since a number of steps both for analyzing our ChIP-seq dataset and our RNA-seq dataset involve using shell commands, we'll refresh ourselves with some more Terminal (our command line interface). "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a967bd73-2bb0-4a61-9ffc-78af8de1aa47",
   "metadata": {},
   "source": [
    "<h2>The command line</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fe60e06-ab66-4231-94cf-ad1b90447cb8",
   "metadata": {},
   "source": [
    "Recall from MCB201A 16-1, that when you start up Terminal, you'll be greeted with an empty prompt that looks something like:\n",
    "\n",
    "<pre style=\"width: 350px; margin-top: 15px; margin-bottom: 15px; color: #000000; background-color: #EEEEEE; border: 1px solid; border-color: #AAAAAA; padding: 10px; border-radius: 15px; font-size: 12px;\">jovyan@jupyter-CalNetID:~$</pre>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee3c10b1-19aa-48eb-9207-f2cc6e4de255",
   "metadata": {},
   "source": [
    "<h2>Some command line interface functionality</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de4a6bfd-e9e7-489c-816b-f4ccbd758425",
   "metadata": {},
   "source": [
    "Small note that if you're following along running bash commands in this notebook, this functionality isn't available in the same way in your notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2282b8c-d8ed-46a7-9c7e-8f56e2956c92",
   "metadata": {},
   "source": [
    "<h3>Pull up a previous/subsequent line</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02aade88-533d-4c1b-ac99-0c79b36f5e30",
   "metadata": {},
   "source": [
    "If you want to pull up a previously executed command, you can press the <kbd>&uarr;</kbd> arrow key to bring up a previous command, and you can use the <kbd>&darr;</kbd> to go to the subsequent command. This can come in handy if you want to either re-run a previously executed command or if you want to bring up a previous line and make modificaitons to it, so you don't have to retype everything."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bcefa2a-6bfc-4915-9bdd-0993722595c5",
   "metadata": {},
   "source": [
    "<h3>Terminate an actively running process</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9342c9f1-c61f-4147-96de-e87f67ffe08e",
   "metadata": {},
   "source": [
    "If you're running something in Terminal and you want to forcefully terminate the process, you can do so by pressing <kbd>Ctrl</kbd>+<kbd>C</kbd>, which will essentially abort the process that is actively running in Terminal."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5eadd17-68b8-4ee2-8d8f-f7dd77e0bb9e",
   "metadata": {},
   "source": [
    "<h3>Suspend an actively running process</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84cbbcf0-0891-4b38-8278-40c5ca178c0a",
   "metadata": {},
   "source": [
    "In a more \"gentler\" way, you can also suspend an actively running process, which will not terminate it/abort its operation. You can do this by instead pressing <kbd>Ctrl</kbd>+<kbd>Z</kbd>. This will temporarily pause a program without shutting it down.\n",
    "\n",
    "You can then restart the program back into the foreground by inputting <code>fg</code> into the command line and pressing <kbd>Enter</kbd> to execute the command."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a14b448-7564-422a-a35e-4af5963ecbf5",
   "metadata": {},
   "source": [
    "<h3>Run a process in the background</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba03ef4a-d593-4124-858f-e77d0ffe689a",
   "metadata": {},
   "source": [
    "To run a process in the background, you can either end the command line with an <code>&amp;</code>, which will indicate that you want the process to run in the background when the command is executed. This can be helpful for when you want to have the ability to use Terminal while some process runs in the background.\n",
    "\n",
    "If you forget to input a <code>&amp;</code> at the end of the line, you can always press <kbd>Ctrl</kbd>+<kbd>Z</kbd>, which will pause the process. Then if you input <code>bg</code> and execute the command, the paused process will resume running in the background."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb984514-f86c-43cd-b4b0-43b783840bc9",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Command line syntax</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93b0a07b-22fa-4278-9ba0-df0b19ef4a8e",
   "metadata": {},
   "source": [
    "Since now we'll start using the command line to do more than just move around our Biology Hub, you'll need to understand the syntax for command line so that Terminal can understand and execute your input correctly and as you intended.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 5</strong></h4>\n",
    "<img src=\"./images/6_1_fig_5.png\" style=\"width: 300px; margin: auto;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7ce56c-e0eb-4355-a417-8ac5176000ce",
   "metadata": {},
   "source": [
    "In a loose way, the organization of command line syntax isn't too different from Python. The <strong>prompt</strong>, while not present in Python, indicates the start of a command line. The <strong>command</strong> is like a function in Python, where it executes some pre-defined set of operations. The <strong>option</strong> and the <strong>argument</strong> can both be thought of as parameters for Python functions. Options alter how the command behaves, while arguments tell the command what it should operate on. OPtions are preceded by either a <code>&dash;</code> or <code>&dash;&dash;</code>, which indicate a short form (abbreviated) or long form option, respectively. Most commands have options that can be written in a short form or an equivalent long form, and you can find the different options for a command through its documentation.\n",
    "\n",
    "The fact that there are sometimes two forms to an option, such as <code>&dash;h</code> and <code>&dash;&dash;help</code>, is kind of an artifact from when options used to have to be a single character."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b4758169-893f-444e-9331-4b002806dae9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: ls [OPTION]... [FILE]...\n",
      "List information about the FILEs (the current directory by default).\n",
      "Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.\n",
      "\n",
      "Mandatory arguments to long options are mandatory for short options too.\n",
      "  -a, --all                  do not ignore entries starting with .\n",
      "  -A, --almost-all           do not list implied . and ..\n",
      "      --author               with -l, print the author of each file\n",
      "  -b, --escape               print C-style escapes for nongraphic characters\n",
      "      --block-size=SIZE      with -l, scale sizes by SIZE when printing them;\n",
      "                               e.g., '--block-size=M'; see SIZE format below\n",
      "  -B, --ignore-backups       do not list implied entries ending with ~\n",
      "  -c                         with -lt: sort by, and show, ctime (time of last\n",
      "                               modification of file status information);\n",
      "                               with -l: show ctime and sort by name;\n",
      "                               otherwise: sort by ctime, newest first\n",
      "  -C                         list entries by columns\n",
      "      --color[=WHEN]         colorize the output; WHEN can be 'always' (default\n",
      "                               if omitted), 'auto', or 'never'; more info below\n",
      "  -d, --directory            list directories themselves, not their contents\n",
      "  -D, --dired                generate output designed for Emacs' dired mode\n",
      "  -f                         do not sort, enable -aU, disable -ls --color\n",
      "  -F, --classify             append indicator (one of */=>@|) to entries\n",
      "      --file-type            likewise, except do not append '*'\n",
      "      --format=WORD          across -x, commas -m, horizontal -x, long -l,\n",
      "                               single-column -1, verbose -l, vertical -C\n",
      "      --full-time            like -l --time-style=full-iso\n",
      "  -g                         like -l, but do not list owner\n",
      "      --group-directories-first\n",
      "                             group directories before files;\n",
      "                               can be augmented with a --sort option, but any\n",
      "                               use of --sort=none (-U) disables grouping\n",
      "  -G, --no-group             in a long listing, don't print group names\n",
      "  -h, --human-readable       with -l and -s, print sizes like 1K 234M 2G etc.\n",
      "      --si                   likewise, but use powers of 1000 not 1024\n",
      "  -H, --dereference-command-line\n",
      "                             follow symbolic links listed on the command line\n",
      "      --dereference-command-line-symlink-to-dir\n",
      "                             follow each command line symbolic link\n",
      "                               that points to a directory\n",
      "      --hide=PATTERN         do not list implied entries matching shell PATTERN\n",
      "                               (overridden by -a or -A)\n",
      "      --hyperlink[=WHEN]     hyperlink file names; WHEN can be 'always'\n",
      "                               (default if omitted), 'auto', or 'never'\n",
      "      --indicator-style=WORD  append indicator with style WORD to entry names:\n",
      "                               none (default), slash (-p),\n",
      "                               file-type (--file-type), classify (-F)\n",
      "  -i, --inode                print the index number of each file\n",
      "  -I, --ignore=PATTERN       do not list implied entries matching shell PATTERN\n",
      "  -k, --kibibytes            default to 1024-byte blocks for disk usage;\n",
      "                               used only with -s and per directory totals\n",
      "  -l                         use a long listing format\n",
      "  -L, --dereference          when showing file information for a symbolic\n",
      "                               link, show information for the file the link\n",
      "                               references rather than for the link itself\n",
      "  -m                         fill width with a comma separated list of entries\n",
      "  -n, --numeric-uid-gid      like -l, but list numeric user and group IDs\n",
      "  -N, --literal              print entry names without quoting\n",
      "  -o                         like -l, but do not list group information\n",
      "  -p, --indicator-style=slash\n",
      "                             append / indicator to directories\n",
      "  -q, --hide-control-chars   print ? instead of nongraphic characters\n",
      "      --show-control-chars   show nongraphic characters as-is (the default,\n",
      "                               unless program is 'ls' and output is a terminal)\n",
      "  -Q, --quote-name           enclose entry names in double quotes\n",
      "      --quoting-style=WORD   use quoting style WORD for entry names:\n",
      "                               literal, locale, shell, shell-always,\n",
      "                               shell-escape, shell-escape-always, c, escape\n",
      "                               (overrides QUOTING_STYLE environment variable)\n",
      "  -r, --reverse              reverse order while sorting\n",
      "  -R, --recursive            list subdirectories recursively\n",
      "  -s, --size                 print the allocated size of each file, in blocks\n",
      "  -S                         sort by file size, largest first\n",
      "      --sort=WORD            sort by WORD instead of name: none (-U), size (-S),\n",
      "                               time (-t), version (-v), extension (-X)\n",
      "      --time=WORD            change the default of using modification times;\n",
      "                               access time (-u): atime, access, use;\n",
      "                               change time (-c): ctime, status;\n",
      "                               birth time: birth, creation;\n",
      "                             with -l, WORD determines which time to show;\n",
      "                             with --sort=time, sort by WORD (newest first)\n",
      "      --time-style=TIME_STYLE  time/date format with -l; see TIME_STYLE below\n",
      "  -t                         sort by time, newest first; see --time\n",
      "  -T, --tabsize=COLS         assume tab stops at each COLS instead of 8\n",
      "  -u                         with -lt: sort by, and show, access time;\n",
      "                               with -l: show access time and sort by name;\n",
      "                               otherwise: sort by access time, newest first\n",
      "  -U                         do not sort; list entries in directory order\n",
      "  -v                         natural sort of (version) numbers within text\n",
      "  -w, --width=COLS           set output width to COLS.  0 means no limit\n",
      "  -x                         list entries by lines instead of by columns\n",
      "  -X                         sort alphabetically by entry extension\n",
      "  -Z, --context              print any security context of each file\n",
      "  -1                         list one file per line.  Avoid '\\n' with -q or -b\n",
      "      --help     display this help and exit\n",
      "      --version  output version information and exit\n",
      "\n",
      "The SIZE argument is an integer and optional unit (example: 10K is 10*1024).\n",
      "Units are K,M,G,T,P,E,Z,Y (powers of 1024) or KB,MB,... (powers of 1000).\n",
      "Binary prefixes can be used, too: KiB=K, MiB=M, and so on.\n",
      "\n",
      "The TIME_STYLE argument can be full-iso, long-iso, iso, locale, or +FORMAT.\n",
      "FORMAT is interpreted like in date(1).  If FORMAT is FORMAT1<newline>FORMAT2,\n",
      "then FORMAT1 applies to non-recent files and FORMAT2 to recent files.\n",
      "TIME_STYLE prefixed with 'posix-' takes effect only outside the POSIX locale.\n",
      "Also the TIME_STYLE environment variable sets the default style to use.\n",
      "\n",
      "Using color to distinguish file types is disabled both by default and\n",
      "with --color=never.  With --color=auto, ls emits color codes only when\n",
      "standard output is connected to a terminal.  The LS_COLORS environment\n",
      "variable can change the settings.  Use the dircolors command to set it.\n",
      "\n",
      "Exit status:\n",
      " 0  if OK,\n",
      " 1  if minor problems (e.g., cannot access subdirectory),\n",
      " 2  if serious trouble (e.g., cannot access command-line argument).\n",
      "\n",
      "GNU coreutils online help: <https://www.gnu.org/software/coreutils/>\n",
      "Full documentation <https://www.gnu.org/software/coreutils/ls>\n",
      "or available locally via: info '(coreutils) ls invocation'\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#Example of basic command\n",
    "ls --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa487c48-99e4-4029-bf7c-85f8cffff817",
   "metadata": {},
   "source": [
    "<h2>Spaces</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e650728-5e34-40d5-a2dc-fd5093d4ec3b",
   "metadata": {},
   "source": [
    "Spaces in command line have meaning and are used to separate the different parts of an input. If you refer back to <strong>Fig 5</strong>, you can see that each part of an input is separated by a space, so the command an dindividual options and arguments are all separated from each other with a space. This means that you will want to be careful not to include spaces in your file name, otherwise it may appear as though you are providing additional arguments or options. If your file name does have spaces, you'll need to wrap it in single or double quotes, such as <code>\"file name\"</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6dbc419-427b-4ef8-85b0-ba8ce993cd2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#Example of handling spaces in command line\n",
    "#Need to flank with single quote or double quotes\n",
    "mkdir -p \"my cat\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5223a63-51f7-420c-946b-30d470c75d8e",
   "metadata": {},
   "source": [
    "<h2>Line breaks</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c633ce2c-e6fa-480a-8ffd-5fddf0c1b420",
   "metadata": {},
   "source": [
    "There will be times where we'll encounter long command line inputs, so we'll need to use line breaks to separate out the physical lines but still have them be interpreted as a single logical line by our command line interpreter. To introduce physical line breaks, you can input a backslash <code>&bsol;</code> and then pressing <kbd>Enter</kbd>. This will create a physical line break and allow you to continue inputting the next physical line without executing your command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "92004507-a698-48cb-8024-e6e0f5cc613d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir \\\n",
    "-p \\\n",
    "liebchen\n",
    "\n",
    "#Can separate lines in command line with line breaks indicated by a backslash\n",
    "#You don't need to type in the preceding > that shows up in Terminal for each new line\n",
    "#Can just type as shown here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5676ab2-0c62-4104-9cda-428e5ddd658a",
   "metadata": {},
   "source": [
    "In Terminal, you'll see that when you initiate a line break, there will be a <code>&gt;</code> at the beginning of each new physical line. This just indicates that the empty line continues from the preceding one."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aa05d2d-37d6-4971-88ed-c486ffaac52d",
   "metadata": {},
   "source": [
    "<h2>Finding documentation</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "643688e5-05e1-4970-af15-cf94ff0e6f4e",
   "metadata": {},
   "source": [
    "You can output the documentation for different commands by inputting <code>&dash;&dash;help</code> as an option following the command. The documentation usually provides details on how to use that command as well as possible options that you can input to modify how the command executes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "df7b0c3b-9c33-4d3a-91ed-d1fc36e3a36a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: ls [OPTION]... [FILE]...\n",
      "List information about the FILEs (the current directory by default).\n",
      "Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.\n",
      "\n",
      "Mandatory arguments to long options are mandatory for short options too.\n",
      "  -a, --all                  do not ignore entries starting with .\n",
      "  -A, --almost-all           do not list implied . and ..\n",
      "      --author               with -l, print the author of each file\n",
      "  -b, --escape               print C-style escapes for nongraphic characters\n",
      "      --block-size=SIZE      with -l, scale sizes by SIZE when printing them;\n",
      "                               e.g., '--block-size=M'; see SIZE format below\n",
      "  -B, --ignore-backups       do not list implied entries ending with ~\n",
      "  -c                         with -lt: sort by, and show, ctime (time of last\n",
      "                               modification of file status information);\n",
      "                               with -l: show ctime and sort by name;\n",
      "                               otherwise: sort by ctime, newest first\n",
      "  -C                         list entries by columns\n",
      "      --color[=WHEN]         colorize the output; WHEN can be 'always' (default\n",
      "                               if omitted), 'auto', or 'never'; more info below\n",
      "  -d, --directory            list directories themselves, not their contents\n",
      "  -D, --dired                generate output designed for Emacs' dired mode\n",
      "  -f                         do not sort, enable -aU, disable -ls --color\n",
      "  -F, --classify             append indicator (one of */=>@|) to entries\n",
      "      --file-type            likewise, except do not append '*'\n",
      "      --format=WORD          across -x, commas -m, horizontal -x, long -l,\n",
      "                               single-column -1, verbose -l, vertical -C\n",
      "      --full-time            like -l --time-style=full-iso\n",
      "  -g                         like -l, but do not list owner\n",
      "      --group-directories-first\n",
      "                             group directories before files;\n",
      "                               can be augmented with a --sort option, but any\n",
      "                               use of --sort=none (-U) disables grouping\n",
      "  -G, --no-group             in a long listing, don't print group names\n",
      "  -h, --human-readable       with -l and -s, print sizes like 1K 234M 2G etc.\n",
      "      --si                   likewise, but use powers of 1000 not 1024\n",
      "  -H, --dereference-command-line\n",
      "                             follow symbolic links listed on the command line\n",
      "      --dereference-command-line-symlink-to-dir\n",
      "                             follow each command line symbolic link\n",
      "                               that points to a directory\n",
      "      --hide=PATTERN         do not list implied entries matching shell PATTERN\n",
      "                               (overridden by -a or -A)\n",
      "      --hyperlink[=WHEN]     hyperlink file names; WHEN can be 'always'\n",
      "                               (default if omitted), 'auto', or 'never'\n",
      "      --indicator-style=WORD  append indicator with style WORD to entry names:\n",
      "                               none (default), slash (-p),\n",
      "                               file-type (--file-type), classify (-F)\n",
      "  -i, --inode                print the index number of each file\n",
      "  -I, --ignore=PATTERN       do not list implied entries matching shell PATTERN\n",
      "  -k, --kibibytes            default to 1024-byte blocks for disk usage;\n",
      "                               used only with -s and per directory totals\n",
      "  -l                         use a long listing format\n",
      "  -L, --dereference          when showing file information for a symbolic\n",
      "                               link, show information for the file the link\n",
      "                               references rather than for the link itself\n",
      "  -m                         fill width with a comma separated list of entries\n",
      "  -n, --numeric-uid-gid      like -l, but list numeric user and group IDs\n",
      "  -N, --literal              print entry names without quoting\n",
      "  -o                         like -l, but do not list group information\n",
      "  -p, --indicator-style=slash\n",
      "                             append / indicator to directories\n",
      "  -q, --hide-control-chars   print ? instead of nongraphic characters\n",
      "      --show-control-chars   show nongraphic characters as-is (the default,\n",
      "                               unless program is 'ls' and output is a terminal)\n",
      "  -Q, --quote-name           enclose entry names in double quotes\n",
      "      --quoting-style=WORD   use quoting style WORD for entry names:\n",
      "                               literal, locale, shell, shell-always,\n",
      "                               shell-escape, shell-escape-always, c, escape\n",
      "                               (overrides QUOTING_STYLE environment variable)\n",
      "  -r, --reverse              reverse order while sorting\n",
      "  -R, --recursive            list subdirectories recursively\n",
      "  -s, --size                 print the allocated size of each file, in blocks\n",
      "  -S                         sort by file size, largest first\n",
      "      --sort=WORD            sort by WORD instead of name: none (-U), size (-S),\n",
      "                               time (-t), version (-v), extension (-X)\n",
      "      --time=WORD            change the default of using modification times;\n",
      "                               access time (-u): atime, access, use;\n",
      "                               change time (-c): ctime, status;\n",
      "                               birth time: birth, creation;\n",
      "                             with -l, WORD determines which time to show;\n",
      "                             with --sort=time, sort by WORD (newest first)\n",
      "      --time-style=TIME_STYLE  time/date format with -l; see TIME_STYLE below\n",
      "  -t                         sort by time, newest first; see --time\n",
      "  -T, --tabsize=COLS         assume tab stops at each COLS instead of 8\n",
      "  -u                         with -lt: sort by, and show, access time;\n",
      "                               with -l: show access time and sort by name;\n",
      "                               otherwise: sort by access time, newest first\n",
      "  -U                         do not sort; list entries in directory order\n",
      "  -v                         natural sort of (version) numbers within text\n",
      "  -w, --width=COLS           set output width to COLS.  0 means no limit\n",
      "  -x                         list entries by lines instead of by columns\n",
      "  -X                         sort alphabetically by entry extension\n",
      "  -Z, --context              print any security context of each file\n",
      "  -1                         list one file per line.  Avoid '\\n' with -q or -b\n",
      "      --help     display this help and exit\n",
      "      --version  output version information and exit\n",
      "\n",
      "The SIZE argument is an integer and optional unit (example: 10K is 10*1024).\n",
      "Units are K,M,G,T,P,E,Z,Y (powers of 1024) or KB,MB,... (powers of 1000).\n",
      "Binary prefixes can be used, too: KiB=K, MiB=M, and so on.\n",
      "\n",
      "The TIME_STYLE argument can be full-iso, long-iso, iso, locale, or +FORMAT.\n",
      "FORMAT is interpreted like in date(1).  If FORMAT is FORMAT1<newline>FORMAT2,\n",
      "then FORMAT1 applies to non-recent files and FORMAT2 to recent files.\n",
      "TIME_STYLE prefixed with 'posix-' takes effect only outside the POSIX locale.\n",
      "Also the TIME_STYLE environment variable sets the default style to use.\n",
      "\n",
      "Using color to distinguish file types is disabled both by default and\n",
      "with --color=never.  With --color=auto, ls emits color codes only when\n",
      "standard output is connected to a terminal.  The LS_COLORS environment\n",
      "variable can change the settings.  Use the dircolors command to set it.\n",
      "\n",
      "Exit status:\n",
      " 0  if OK,\n",
      " 1  if minor problems (e.g., cannot access subdirectory),\n",
      " 2  if serious trouble (e.g., cannot access command-line argument).\n",
      "\n",
      "GNU coreutils online help: <https://www.gnu.org/software/coreutils/>\n",
      "Full documentation <https://www.gnu.org/software/coreutils/ls>\n",
      "or available locally via: info '(coreutils) ls invocation'\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#To pull up documentation, you can use -h or --help options depending on the command\n",
    "\n",
    "ls --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be5b68b-b18e-4a3a-b708-8253284f1753",
   "metadata": {},
   "source": [
    "<h2>Navigating in Terminal refresher</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41881aee-a74c-40c8-8b9d-24dd743a51de",
   "metadata": {},
   "source": [
    "Below is just a quick refresher to navigate through directories and subdirectories using Terminal. For those of you following in your notebooks, you'll want to keep in mind that once your code cell finishes executing your bash command, it'll exit its instance of Terminal. So if you change directories, your current working directory will be reset back to what it was before when you run in a new code cell, as if you started up a new Terminal session."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "abccf6b2-33d1-41fd-b1e1-f99ee5667a4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "cat\n",
      "fastqc-results\n",
      "images\n",
      "liebchen\n",
      "liebchen_1.txt\n",
      "liebchen_head.txt\n",
      "liebchen meows\n",
      "liebchen.txt\n",
      "ls_help.txt\n",
      "my\n",
      "my cat\n",
      "POSTCLASS_6-1_Intro_ChIP_seq.ipynb\n",
      "week_6_homework\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#Navigating through directories\n",
    "\n",
    "#list contents\n",
    "ls\n",
    "\n",
    "#change directory to the specified one\n",
    "cd cat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6adb2ca-e871-4a7a-b21c-2a8e15dfb003",
   "metadata": {},
   "source": [
    "<ul>\n",
    "    <li><code>ls</code> - this command lists the contents of your current working directory (or if you provide a path, the contents of the specified directory</li>\n",
    "    <li><code>cd</code> - this command changes the current working directory to the specified directory</li>\n",
    "    <li><code>mkdir</code> - this command lets you make directories in the current directory or if you provide a path, then in the specified path</li>\n",
    "    <li><code>mv</code> - this command lets you move a file (or folder) to the specified path</li>\n",
    "    <li><code>cp</code> - this command lets you copy a file or directory (using the <code>&dash;r</code> option) into the specified path</li>\n",
    "    <li><code>rm</code> - this command lets you delete a file or directory (using the <code>&dash;r</code> option)</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f6061cff-df6f-44c7-98cd-e59d135c727c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "cat\n",
      "fastqc-results\n",
      "images\n",
      "liebchen\n",
      "liebchen_1.txt\n",
      "liebchen_head.txt\n",
      "liebchen meows\n",
      "liebchen.txt\n",
      "ls_help.txt\n",
      "my\n",
      "my cat\n",
      "POSTCLASS_6-1_Intro_ChIP_seq.ipynb\n",
      "week_6_homework\n",
      "changing directory\n",
      "below this is cat directory contents\n",
      "liebchen.txt\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#Run ls command again\n",
    "ls\n",
    "\n",
    "#Can see that the working directory is changed even though earlier cell changed to cat\n",
    "#So each code cell is operating separately from others\n",
    "echo \"changing directory\"\n",
    "cd cat\n",
    "echo \"below this is cat directory contents\"\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "616b6760-5573-4102-a8e6-b7629427896f",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Standard input/output streams</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15187dfc-ff0c-493b-a4c9-7b10071ffeab",
   "metadata": {},
   "source": [
    "In the command line, there are standard streams for handling input and output(s), and in the command line/Terminal, the input fed into a command is referred to as the standard input <code>stdin</code>. Outputs from the command can go through two different standard streams, either the <code>stdout</code> or the <code>stderr</code>. The <code>stdout</code> is the standard output which handles the normal outputs from a command, and the <code>stderr</code> is the standard error, which as its name suggests is usually used to handle error outputs and diagnostic outputs from a command.\n",
    "\n",
    "For example, the <code>echo</code> command will output into the <code>stdout</code>, which we can see in Terminal. This command functions more or less the same as the <code>print()</code> command that we've been using in Python, but for Terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7af1fc85-d866-4217-9398-0144cd0e72ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MEOW!\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#echo command acts basically like the print() function in Python\n",
    "echo \"MEOW!\"\n",
    "#echo will output to the stdout stream, which is what we see normally"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fce56798-7dff-48df-8655-d4cc942ac290",
   "metadata": {},
   "source": [
    "We can also see the same using the <code>ls</code> command that we use to output file names. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "da2a7fed-2be8-4491-8919-0d63dc347ec2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "cat\n",
      "fastqc-results\n",
      "images\n",
      "liebchen\n",
      "liebchen_1.txt\n",
      "liebchen_head.txt\n",
      "liebchen meows\n",
      "liebchen.txt\n",
      "ls_help.txt\n",
      "my\n",
      "my cat\n",
      "POSTCLASS_6-1_Intro_ChIP_seq.ipynb\n",
      "week_6_homework\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#ls command also outputs to stdout\n",
    "ls\n",
    "\n",
    "#In general, there are two output streams that handle outputs\n",
    "#Both will show up in Terminal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36996412-4626-4bc7-a3ec-052381cab424",
   "metadata": {},
   "source": [
    "<h2>Redirecting an output</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "121032d4-16c1-406d-b42b-63010c6a1392",
   "metadata": {},
   "source": [
    "We can redirect the standard output streams by making use of the <code>&gt;</code> redirect symbol. If we follow this with an output file name, we can redirect to that specific file to store the output. In a way, this type of redirection can be thought of as using the <code>open()</code> + <code>file.write()</code> commands that we've used in Python with the <code>'w'</code> write mode for the <code>open()</code> function. So the result of redirecting an output with <code>&gt;</code> is that the output file is generated (if it doesn't exist), and the <code>stdout</code> is written to that file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e01c8d8d-6c0c-46ab-8705-f854c79c995f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#You can control the output though via redirecting it\n",
    "#The > indicates that you are redirecting the stdout of echo to the following file\n",
    "echo \"MEOW!\" > liebchen_1.txt\n",
    "#You should see now there's no output in Terminal\n",
    "#It is instead redirected to the newly created file\n",
    "#Note that > will overwrite anything if the file already exists"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3feb40e7-0789-4037-8c79-4d782252f63e",
   "metadata": {},
   "source": [
    "We also have the option to append outputs to an existing file, like how <code>open()</code> also has an <code>'a'</code> append mode. To do this in command line, we can use <code>&gt;&gt;</code> to redirect the <code>stdout</code> to append to the end of a file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5f2facb5-65b2-4534-80b1-a29271fa258b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#You can redirect to append to a file's existing contents as well\n",
    "echo \"MEOW! MEOW!\" >> liebchen_1.txt\n",
    "#Again, there will be no output in Terminal\n",
    "#because it is being redirected to be appended to the indicated file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da1a7a50-d3ff-4c99-b8ec-7e4c54f1e296",
   "metadata": {},
   "source": [
    "<h2>Piping</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9969ea51-333f-4130-a6f3-de90bccbcac3",
   "metadata": {},
   "source": [
    "Piping is kind of a subset of redirecting an output, where you pass (or \"pipe\") the output of one command as an input for the next command. To do this, you'll use the <code>&vert;</code> pipe symbol, where the command to the left of the pipe is used as the input for the command to the right of the pipe.\n",
    "\n",
    "For example, we can generate a list of unsorted files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "dd30fcd9-27ab-4d28-bd9c-7330bb7eaed0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "liebchen meows\n",
      "my\n",
      "ls_help.txt\n",
      ".\n",
      "POSTCLASS_6-1_Intro_ChIP_seq.ipynb\n",
      "week_6_homework\n",
      "my cat\n",
      "cat\n",
      "liebchen\n",
      "liebchen_head.txt\n",
      ".gitkeep\n",
      "fastqc-results\n",
      "liebchen.txt\n",
      "images\n",
      "liebchen_1.txt\n",
      ".ipynb_checkpoints\n",
      " \n",
      "..\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#Use -f option for the ls command to generate a list of unsorted files\n",
    "ls -f"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b4e08e4-57a1-40e9-8bc5-e71393d468fd",
   "metadata": {},
   "source": [
    "And then we can pipe that into a command to sort the output from the <code>ls</code> command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cd2726bb-6cd6-49ed-b621-67ea3b4b1560",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      ".\n",
      "..\n",
      "cat\n",
      "fastqc-results\n",
      ".gitkeep\n",
      "images\n",
      ".ipynb_checkpoints\n",
      "liebchen\n",
      "liebchen_1.txt\n",
      "liebchen_head.txt\n",
      "liebchen meows\n",
      "liebchen.txt\n",
      "ls_help.txt\n",
      "my\n",
      "my cat\n",
      "POSTCLASS_6-1_Intro_ChIP_seq.ipynb\n",
      "week_6_homework\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#Then we can pipe that into other things\n",
    "\n",
    "ls -f | sort"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3fc5e49-9b7f-4d83-ab30-b6c107e21249",
   "metadata": {},
   "source": [
    "We can also combine this with the <code>&gt;</code> redirect symbol to then redirect the output to a new <code>.txt</code> file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "370ae9b1-f604-4f64-b731-b10434d58eec",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "ls -f | sort > sorted_files.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b310f8-2d15-427c-8796-c23cda0a9872",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Opening files in command line</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a8181e4-780b-4816-8817-c23807140554",
   "metadata": {},
   "source": [
    "There are a few different ways of opening up files in command line to view the contents, and a common way to open up plain text files is by using the <code>cat</code> command, which simply concatenates the file into the <code>stdout</code> stream."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "44468226-694d-47d8-aece-ed543f2feb6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MEOW!\n",
      "MEOW! MEOW!\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "cat liebchen_1.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6ab85f4-bca8-4591-a8ef-b6190ac1dddf",
   "metadata": {},
   "source": [
    "We can also use this to concatenate files together if we combined it with the <code>&gt;</code> redirect symbol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9a05b3f1-0ed3-4323-bd69-238b72681614",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cat liebchen_1.txt sorted_files.txt > meow.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e0e6829-de54-4275-9fac-9f19d2748461",
   "metadata": {},
   "source": [
    "Similar to Python, you can also view just the first set of lines <code>head</code> or the last set of lines <code>tail</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d8adf5d5-45ba-4f53-8b09-f543a5e63441",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MEOW!\n",
      "MEOW! MEOW!\n",
      " \n",
      ".\n",
      "..\n",
      "cat\n",
      "fastqc-results\n",
      ".gitkeep\n",
      "images\n",
      ".ipynb_checkpoints\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head meow.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5d99f454-6ef6-47bb-b13b-542b8141377a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "liebchen_1.txt\n",
      "liebchen_head.txt\n",
      "liebchen meows\n",
      "liebchen.txt\n",
      "ls_help.txt\n",
      "my\n",
      "my cat\n",
      "POSTCLASS_6-1_Intro_ChIP_seq.ipynb\n",
      "sorted_files.txt\n",
      "week_6_homework\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "tail meow.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96576cfd-5e0e-41c4-a042-dce7de23746d",
   "metadata": {},
   "source": [
    "By default, these commands will output 10 lines (from top or from bottom). But if we wanted to specify the number of lines to output, we can provide that to the command to modify how many lines we want it to output into the <code>stdout</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "92026cfc-bdc3-4acb-abfd-681107706128",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: ls [OPTION]... [FILE]...\n",
      "List information about the FILEs (the current directory by default).\n",
      "Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head -4 ls_help.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33a24578-99e4-44ba-8d3c-cae484c79abb",
   "metadata": {},
   "source": [
    "For those of you following along in Terminal, you can also open up your file to a temporary page viewer by using the <code>less</code> command. This command will pull up the contents of the indicated file, but it will not be part of your <code>stdout</code>. Instead, it'll be in a temporary viewer, which you can scroll through and also exit by pressing <kbd>Q</kbd>, and you'll return to an empty prompt. Note that this won't work in this notebook, and it'll be something you do in Terminal.\n",
    "\n",
    "You can also pipe an output to a temporary viewer as well by piping into the <code>less</code> command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "37dd73ca-8009-4a8c-b90c-3c8f18c8af27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: ls [OPTION]... [FILE]...\n",
      "List information about the FILEs (the current directory by default).\n",
      "Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.\n",
      "\n",
      "Mandatory arguments to long options are mandatory for short options too.\n",
      "  -a, --all                  do not ignore entries starting with .\n",
      "  -A, --almost-all           do not list implied . and ..\n",
      "      --author               with -l, print the author of each file\n",
      "  -b, --escape               print C-style escapes for nongraphic characters\n",
      "      --block-size=SIZE      with -l, scale sizes by SIZE when printing them;\n",
      "                               e.g., '--block-size=M'; see SIZE format below\n",
      "  -B, --ignore-backups       do not list implied entries ending with ~\n",
      "  -c                         with -lt: sort by, and show, ctime (time of last\n",
      "                               modification of file status information);\n",
      "                               with -l: show ctime and sort by name;\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#Note that this should be run in Terminal since it doesn't work the same for notebooks\n",
    "head -15 ls_help.txt | less"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4947ebac-858e-43a2-9147-afac1e5c8f09",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise set</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef13bcf5-e0a3-495a-ba79-06742dcf6425",
   "metadata": {},
   "source": [
    "This is just a set of exercises for you to refresh yourself in Terminal and practice some new commands that you learned today just to refamiliarize yourself on working with command line.\n",
    "\n",
    "<ol>\n",
    "    <li>Return to your home directory</li>\n",
    "    <li>Navigate to this week's directory</li>\n",
    "    <li>Create a new folder named <code>fastqc-results</code> in this week's directory</li>\n",
    "    <li>Navigate to <code>fastqc-results</code> and create a subdirectory named <code>seq-view</code></li>\n",
    "    <li>Change directories to <code>seq-view</code></li>\n",
    "    <li>Output the first 10 <u>reads</u> of <code>10M_ctrl_1.fastq</code> into <code>stdout</code></li>\n",
    "    <li>Redirect the output to a new <code>.fastq</code> file in your <code>seq-view</code> folder</li>\n",
    "    <li>Repeat #6-8 for <code>10M_taz_1.fastq</code></li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c5445972-296a-4dea-b39c-9288ae03d818",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR1810912.1 HWI-ST1210:136:C1RBDACXX:5:1101:1472:2121 length=50\n",
      "CACTGTAATTTCCCTAAAACAGGAGGTCACGGTAAGTCCTATGGATAGAT\n",
      "+SRR1810912.1 HWI-ST1210:136:C1RBDACXX:5:1101:1472:2121 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810912.2 HWI-ST1210:136:C1RBDACXX:5:1101:1374:2127 length=50\n",
      "CCCCGTGCTCCAGCCCAGCCAGGCCGCGCTGTCAGAAGGGATCTATCANN\n",
      "+SRR1810912.2 HWI-ST1210:136:C1RBDACXX:5:1101:1374:2127 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810912.3 HWI-ST1210:136:C1RBDACXX:5:1101:1373:2149 length=50\n",
      "ACAAAAAAACACAATACAATTTAAAAAGTATGAGTTATCTGAATATACAT\n",
      "+SRR1810912.3 HWI-ST1210:136:C1RBDACXX:5:1101:1373:2149 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810912.4 HWI-ST1210:136:C1RBDACXX:5:1101:1298:2192 length=50\n",
      "ACGGGATTATAAATCCCCATGTGCACTAAAGGGAAGGGTCTAGAAGGTCN\n",
      "+SRR1810912.4 HWI-ST1210:136:C1RBDACXX:5:1101:1298:2192 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810912.5 HWI-ST1210:136:C1RBDACXX:5:1101:1327:2194 length=50\n",
      "CTAAACAGTTGTGGTCTATGACCTACAATGTAAATTTAAAAAAAAAAANN\n",
      "+SRR1810912.5 HWI-ST1210:136:C1RBDACXX:5:1101:1327:2194 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810912.6 HWI-ST1210:136:C1RBDACXX:5:1101:1499:2206 length=50\n",
      "TTTAACCTGTTTGAGGCTAAGGCTTATGCTATTTAGATTAATTGTTACAT\n",
      "+SRR1810912.6 HWI-ST1210:136:C1RBDACXX:5:1101:1499:2206 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810912.7 HWI-ST1210:136:C1RBDACXX:5:1101:1358:2223 length=50\n",
      "ATGAATTTGTGTGTCATCCTTTTGTAGGGGCCCTGCTAATCTTCTCTGTN\n",
      "+SRR1810912.7 HWI-ST1210:136:C1RBDACXX:5:1101:1358:2223 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810912.8 HWI-ST1210:136:C1RBDACXX:5:1101:1332:2225 length=50\n",
      "TAATGCAATAAATATCTGATATAGAGGTCCTATTCAAAATTTGCCAACTN\n",
      "+SRR1810912.8 HWI-ST1210:136:C1RBDACXX:5:1101:1332:2225 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810912.9 HWI-ST1210:136:C1RBDACXX:5:1101:1253:2247 length=50\n",
      "AGGCTGTGTGACTGTGGGCAGATCACACTACCCTCTCTGAACCTCAGTCN\n",
      "+SRR1810912.9 HWI-ST1210:136:C1RBDACXX:5:1101:1253:2247 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810912.10 HWI-ST1210:136:C1RBDACXX:5:1101:1652:2148 length=50\n",
      "CTTTCTCCTTCTTTTCCTTCTCTCCCCCCTGTATGCACACCCTGTCCCTC\n",
      "+SRR1810912.10 HWI-ST1210:136:C1RBDACXX:5:1101:1652:2148 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.1 HWI-ST1210:136:C1RBDACXX:5:1101:1269:2131 length=50\n",
      "GCCCTGCTGGAAGACTGAGTTGCGGCGGGGCTTGCAGATGATGAGGGCAG\n",
      "+SRR1810906.1 HWI-ST1210:136:C1RBDACXX:5:1101:1269:2131 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.2 HWI-ST1210:136:C1RBDACXX:5:1101:1317:2146 length=50\n",
      "AGAATGTTAGCACTTCAAACTTAGCAATACATATTATATTGAAAAAGNNN\n",
      "+SRR1810906.2 HWI-ST1210:136:C1RBDACXX:5:1101:1317:2146 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.3 HWI-ST1210:136:C1RBDACXX:5:1101:1360:2159 length=50\n",
      "CCAGGTCAAGGCTATTCTACCAATCAGCACATAATTATTTCCTATTGTAC\n",
      "+SRR1810906.3 HWI-ST1210:136:C1RBDACXX:5:1101:1360:2159 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.4 HWI-ST1210:136:C1RBDACXX:5:1101:1383:2180 length=50\n",
      "CTCAGAACTCTTTCTTTTTCCTTCTTTCTTGCTCATGGATTTTCAGCTGT\n",
      "+SRR1810906.4 HWI-ST1210:136:C1RBDACXX:5:1101:1383:2180 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.5 HWI-ST1210:136:C1RBDACXX:5:1101:1268:2193 length=50\n",
      "TTTCAAATCTCTCTAGGAGATCCGTGCACAGACCCTTCCATCCTCGTACN\n",
      "+SRR1810906.5 HWI-ST1210:136:C1RBDACXX:5:1101:1268:2193 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.6 HWI-ST1210:136:C1RBDACXX:5:1101:1433:2243 length=50\n",
      "GGCAGGCCTCCTTGAGCTGTGGTGGGCTCCACCCAGTTCGAGCTTCCCGG\n",
      "+SRR1810906.6 HWI-ST1210:136:C1RBDACXX:5:1101:1433:2243 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.7 HWI-ST1210:136:C1RBDACXX:5:1101:1460:2249 length=50\n",
      "AGATTTTGTATTCTTTTATCAACATTTCTTCAACCTTTGCTTCCCCCCAC\n",
      "+SRR1810906.7 HWI-ST1210:136:C1RBDACXX:5:1101:1460:2249 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.8 HWI-ST1210:136:C1RBDACXX:5:1101:1585:2185 length=50\n",
      "TTTGTACTAACCCCCTTTCCCTGGTCTCTTCCAGCTTGTCTTCTTCTCTC\n",
      "+SRR1810906.8 HWI-ST1210:136:C1RBDACXX:5:1101:1585:2185 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.9 HWI-ST1210:136:C1RBDACXX:5:1101:1519:2191 length=50\n",
      "TGATGAACATGAATGCAAAATCCTCAACAAAATGTCAGCAAACTGAATTC\n",
      "+SRR1810906.9 HWI-ST1210:136:C1RBDACXX:5:1101:1519:2191 length=50\n",
      "??????????????????????????????????????????????????\n",
      "@SRR1810906.10 HWI-ST1210:136:C1RBDACXX:5:1101:1733:2227 length=50\n",
      "TGTTTTGAGACTGAGTCTCACTTTGTTGCCCAGGCTGGAGTCTGGTGGTG\n",
      "+SRR1810906.10 HWI-ST1210:136:C1RBDACXX:5:1101:1733:2227 length=50\n",
      "??????????????????????????????????????????????????\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#Change to home directory\n",
    "cd ~\n",
    "#Can also just be cd\n",
    "\n",
    "#Navigate to this week's directory\n",
    "cd ~/MCB201B_F2025/Week_6\n",
    "\n",
    "#Create a new folder\n",
    "mkdir -p fastqc-results\n",
    "#You can add the -p option to avoid the error if file already exists\n",
    "\n",
    "#Navigate into fastqc-results\n",
    "cd fastqc-results\n",
    "\n",
    "#Create a new folder\n",
    "mkdir -p seq-view\n",
    "\n",
    "#Navigate into seq-view\n",
    "cd seq-view\n",
    "\n",
    "#Output first 10 reads into stdout\n",
    "head -40 ~/shared/2025-fall/courses/1547808/chip-seq/truncated/10M_ctrl_1.fastq\n",
    "\n",
    "#Redirect instead to a file\n",
    "head -40 ~/shared/2025-fall/courses/1547808/chip-seq/truncated/10M_ctrl_1.fastq > 10_reads_ctrl_1.fastq\n",
    "\n",
    "#Repeat for taz_1\n",
    "head -40 ~/shared/2025-fall/courses/1547808/chip-seq/truncated/10M_taz_1.fastq\n",
    "head -40 ~/shared/2025-fall/courses/1547808/chip-seq/truncated/10M_taz_1.fastq > 10_reads_taz_1.fastq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b57944f-9ceb-4d26-823e-31d609a66d51",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">FastQC</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25b6814f-b7ed-4826-a96d-2e6a6fe19452",
   "metadata": {},
   "source": [
    "<a href=\"https://www.bioinformatics.babraham.ac.uk/projects/fastqc/\" rel=\"noopener noreferrer\" target=\"_blank\"><u>FastQC</u></a> is developed by Simon Andrews at the Babraham Institute in the UK. FastQC is a quick way to take a look at the quality of your raw sequence reads in order to get a general idea of how it looks and if there may be a bias in your data that could affect your analysis and/or interpretation of your data. There is a range of analyses that FastQC performs to assess the quality of your raw sequence reads, and they help you to get an idea of whether an issue may have occurred during sample prep, library prep, or sequencing.\n",
    "\n",
    "There are a number of analysis modules with more detailed information and examples that can be found on <a href=\"https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/\" rel=\"noopener noreferrer\" target=\"_blank\"><u>the FastQC documentation pages.</u></a>\n",
    "\n",
    "While it is helpful to run FastQC to get an idea of the quality of your sequence data, you should also be aware that each library prep will be different and just because it shows up as a <strong>Fail</strong> in FastQC, it may be fine in the context of what you're studying. The results of the FastQC file may point you to where you may want to look if it the outcome is not what you expect."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67db9403-e67b-4efb-b998-9b4adae6ad0c",
   "metadata": {},
   "source": [
    "<h2>Running FastQC in Terminal</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db1a3e2e-ef2d-43a5-9574-a8d5f29708e5",
   "metadata": {},
   "source": [
    "We'll do a guided practice to perform a FastQC analysis on truncated sequence files:\n",
    "\n",
    "<ul>\n",
    "    <li><code>10M_ctrl_1.fastq</code></li>\n",
    "    <li><code>10M_taz_1.fastq</code></li>\n",
    "</ul>\n",
    "\n",
    "I recommend following along in Terminal, but you can also do this in the notebook as well, but you'll then need to have all your command lines in a single code cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "19744d1e-6bd6-42a7-90b3-d7bc5a6bf062",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of 10M_ctrl_1.fastq\n",
      "Approx 5% complete for 10M_ctrl_1.fastq\n",
      "Approx 10% complete for 10M_ctrl_1.fastq\n",
      "Approx 15% complete for 10M_ctrl_1.fastq\n",
      "Approx 20% complete for 10M_ctrl_1.fastq\n",
      "Approx 25% complete for 10M_ctrl_1.fastq\n",
      "Approx 30% complete for 10M_ctrl_1.fastq\n",
      "Approx 35% complete for 10M_ctrl_1.fastq\n",
      "Approx 40% complete for 10M_ctrl_1.fastq\n",
      "Approx 45% complete for 10M_ctrl_1.fastq\n",
      "Approx 50% complete for 10M_ctrl_1.fastq\n",
      "Approx 55% complete for 10M_ctrl_1.fastq\n",
      "Approx 60% complete for 10M_ctrl_1.fastq\n",
      "Approx 65% complete for 10M_ctrl_1.fastq\n",
      "Approx 70% complete for 10M_ctrl_1.fastq\n",
      "Approx 75% complete for 10M_ctrl_1.fastq\n",
      "Approx 80% complete for 10M_ctrl_1.fastq\n",
      "Approx 85% complete for 10M_ctrl_1.fastq\n",
      "Approx 90% complete for 10M_ctrl_1.fastq\n",
      "Approx 95% complete for 10M_ctrl_1.fastq\n",
      "Approx 100% complete for 10M_ctrl_1.fastq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for 10M_ctrl_1.fastq\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#Change to shared directory containing truncated files\n",
    "cd ~/shared/2025-fall/courses/1547808/chip-seq/truncated\n",
    "\n",
    "#Running fastqc\n",
    "fastqc -o ~/MCB201B_F2025/Week_6/fastqc-results 10M_ctrl_1.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e655ac3-5601-4c3d-89d9-ef42cbb39d0d",
   "metadata": {},
   "source": [
    "<h2>Let's breakdown the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac4f0c50-5b27-46e0-bdb8-df9328674fa6",
   "metadata": {},
   "source": [
    "<code>fastqc</code>\n",
    "\n",
    "This is our command. We are instructing Terminal to run the fastqc program.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-o</code>\n",
    "\n",
    "This is an option that we input to modify how the command behaves. In this case, we are instructing it to output the resulting files to a specific path that we provide after this option.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>~/MCB201B_F2025/Week_6/fastqc-results</code>\n",
    "\n",
    "This is the path to the directory that we want the fastqc results to be saved. In other words, this is the output directory.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>10M_ctrl_1.fastq</code>\n",
    "\n",
    "This is the <code>.fastq</code> file that we want the fastqc program to analyze. If our current working directory did not have this file name, then we would also have to include the path to the <code>.fastq</code> file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80ec1a7f-8f3b-494c-9e7f-e592b017f5d5",
   "metadata": {},
   "source": [
    "If we want to very crudely rewrite/'translate' this to the Python syntax that we're familiar with, we can kind of see what we're doing in command line isnt too different than calling up a function in Python and providing it with specific arguments so that it knows what to do:\n",
    "\n",
    "```\n",
    "fastqc(o='~/MCB201B_F2025/Week_6/fastqc-results',\n",
    "       input='10M_ctrl_1.fastq'\n",
    "       )\n",
    "```\n",
    "Note that this isn't actually something you can call up in Python, it's just for example."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee4e832a-0ce1-4aea-bace-35dfe451ed99",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">FastQC exercise set</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a9528ee-ba4b-4850-a1d4-87ac6fd85c53",
   "metadata": {},
   "source": [
    "Now give it a try for the other ChIP-seq <code>.fastq</code> files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f574cc1-cca5-4baf-9604-6c99697da04a",
   "metadata": {},
   "source": [
    "<ol>\n",
    "    <li>Run <code>fastqc</code> for the other truncated ChIP-seq files</li>\n",
    "    <li>Once you're done with the analyses, use the File Browser to open up the HTML output files</li>\n",
    "    <li>Take a look at the analysis results</li>\n",
    "    <li>How does the quality of the sequencing data look like for these samples?</li>\n",
    "    <li>If you have extra time in class, you can feel free to get started on running the fastqc analysis on the full datasets (this will be part of this week's assignment)</li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e4d3c7e1-c78e-44a2-9d3e-039a3ef96e4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of 10M_ctrl_2.fastq\n",
      "Approx 5% complete for 10M_ctrl_2.fastq\n",
      "Approx 10% complete for 10M_ctrl_2.fastq\n",
      "Approx 15% complete for 10M_ctrl_2.fastq\n",
      "Approx 20% complete for 10M_ctrl_2.fastq\n",
      "Approx 25% complete for 10M_ctrl_2.fastq\n",
      "Approx 30% complete for 10M_ctrl_2.fastq\n",
      "Approx 35% complete for 10M_ctrl_2.fastq\n",
      "Approx 40% complete for 10M_ctrl_2.fastq\n",
      "Approx 45% complete for 10M_ctrl_2.fastq\n",
      "Approx 50% complete for 10M_ctrl_2.fastq\n",
      "Approx 55% complete for 10M_ctrl_2.fastq\n",
      "Approx 60% complete for 10M_ctrl_2.fastq\n",
      "Approx 65% complete for 10M_ctrl_2.fastq\n",
      "Approx 70% complete for 10M_ctrl_2.fastq\n",
      "Approx 75% complete for 10M_ctrl_2.fastq\n",
      "Approx 80% complete for 10M_ctrl_2.fastq\n",
      "Approx 85% complete for 10M_ctrl_2.fastq\n",
      "Approx 90% complete for 10M_ctrl_2.fastq\n",
      "Approx 95% complete for 10M_ctrl_2.fastq\n",
      "Approx 100% complete for 10M_ctrl_2.fastq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for 10M_ctrl_2.fastq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of 10M_ctrl_3.fastq\n",
      "Approx 5% complete for 10M_ctrl_3.fastq\n",
      "Approx 10% complete for 10M_ctrl_3.fastq\n",
      "Approx 15% complete for 10M_ctrl_3.fastq\n",
      "Approx 20% complete for 10M_ctrl_3.fastq\n",
      "Approx 25% complete for 10M_ctrl_3.fastq\n",
      "Approx 30% complete for 10M_ctrl_3.fastq\n",
      "Approx 35% complete for 10M_ctrl_3.fastq\n",
      "Approx 40% complete for 10M_ctrl_3.fastq\n",
      "Approx 45% complete for 10M_ctrl_3.fastq\n",
      "Approx 50% complete for 10M_ctrl_3.fastq\n",
      "Approx 55% complete for 10M_ctrl_3.fastq\n",
      "Approx 60% complete for 10M_ctrl_3.fastq\n",
      "Approx 65% complete for 10M_ctrl_3.fastq\n",
      "Approx 70% complete for 10M_ctrl_3.fastq\n",
      "Approx 75% complete for 10M_ctrl_3.fastq\n",
      "Approx 80% complete for 10M_ctrl_3.fastq\n",
      "Approx 85% complete for 10M_ctrl_3.fastq\n",
      "Approx 90% complete for 10M_ctrl_3.fastq\n",
      "Approx 95% complete for 10M_ctrl_3.fastq\n",
      "Approx 100% complete for 10M_ctrl_3.fastq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for 10M_ctrl_3.fastq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of 10M_taz_1.fastq\n",
      "Approx 5% complete for 10M_taz_1.fastq\n",
      "Approx 10% complete for 10M_taz_1.fastq\n",
      "Approx 15% complete for 10M_taz_1.fastq\n",
      "Approx 20% complete for 10M_taz_1.fastq\n",
      "Approx 25% complete for 10M_taz_1.fastq\n",
      "Approx 30% complete for 10M_taz_1.fastq\n",
      "Approx 35% complete for 10M_taz_1.fastq\n",
      "Approx 40% complete for 10M_taz_1.fastq\n",
      "Approx 45% complete for 10M_taz_1.fastq\n",
      "Approx 50% complete for 10M_taz_1.fastq\n",
      "Approx 55% complete for 10M_taz_1.fastq\n",
      "Approx 60% complete for 10M_taz_1.fastq\n",
      "Approx 65% complete for 10M_taz_1.fastq\n",
      "Approx 70% complete for 10M_taz_1.fastq\n",
      "Approx 75% complete for 10M_taz_1.fastq\n",
      "Approx 80% complete for 10M_taz_1.fastq\n",
      "Approx 85% complete for 10M_taz_1.fastq\n",
      "Approx 90% complete for 10M_taz_1.fastq\n",
      "Approx 95% complete for 10M_taz_1.fastq\n",
      "Approx 100% complete for 10M_taz_1.fastq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for 10M_taz_1.fastq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of 10M_taz_2.fastq\n",
      "Approx 5% complete for 10M_taz_2.fastq\n",
      "Approx 10% complete for 10M_taz_2.fastq\n",
      "Approx 15% complete for 10M_taz_2.fastq\n",
      "Approx 20% complete for 10M_taz_2.fastq\n",
      "Approx 25% complete for 10M_taz_2.fastq\n",
      "Approx 30% complete for 10M_taz_2.fastq\n",
      "Approx 35% complete for 10M_taz_2.fastq\n",
      "Approx 40% complete for 10M_taz_2.fastq\n",
      "Approx 45% complete for 10M_taz_2.fastq\n",
      "Approx 50% complete for 10M_taz_2.fastq\n",
      "Approx 55% complete for 10M_taz_2.fastq\n",
      "Approx 60% complete for 10M_taz_2.fastq\n",
      "Approx 65% complete for 10M_taz_2.fastq\n",
      "Approx 70% complete for 10M_taz_2.fastq\n",
      "Approx 75% complete for 10M_taz_2.fastq\n",
      "Approx 80% complete for 10M_taz_2.fastq\n",
      "Approx 85% complete for 10M_taz_2.fastq\n",
      "Approx 90% complete for 10M_taz_2.fastq\n",
      "Approx 95% complete for 10M_taz_2.fastq\n",
      "Approx 100% complete for 10M_taz_2.fastq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for 10M_taz_2.fastq\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#Change to shared directory containing truncated files\n",
    "cd ~/shared/2025-fall/courses/1547808/chip-seq/truncated\n",
    "\n",
    "#Running fastqc\n",
    "fastqc -o ~/MCB201B_F2025/Week_6/fastqc-results 10M_ctrl_2.fastq\n",
    "fastqc -o ~/MCB201B_F2025/Week_6/fastqc-results 10M_ctrl_3.fastq\n",
    "fastqc -o ~/MCB201B_F2025/Week_6/fastqc-results 10M_taz_1.fastq\n",
    "fastqc -o ~/MCB201B_F2025/Week_6/fastqc-results 10M_taz_2.fastq\n",
    "\n",
    "#Later on we'll learn to script in bash, but for now, we'll just get used to working in command line\n",
    "#Running one command at a time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ed02aed-852f-48d2-9235-4dc5f0e8acdc",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Interpreting FastQC results</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21203fb4-b4e4-40b5-86e8-85f0e3298ba8",
   "metadata": {},
   "source": [
    "After running <code>fastqc</code>, you'll have two output files, one of which is an HTML file and the other which is a .zip file containing the individual graphs as separate files, as well as the data for each of those graphs.\n",
    "\n",
    "<a href=\"https://www.bioinformatics.babraham.ac.uk/projects/fastqc/\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Documentation for FastQC can be found here.</u></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3735e57d-4b4a-4ef8-ba7e-97ff0288d9c4",
   "metadata": {},
   "source": [
    "<h2>Basic statistics</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2eacac1-0203-4aac-8178-0065ec3163e5",
   "metadata": {},
   "source": [
    "The first thing you'll see is the summary statistics and general file information for your sequence file (<strong>Fig 6</strong>). This provides you general information about your sequence file.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 6</strong></h4>\n",
    "<img src=\"./images/6_1_fig_6.png\" style=\"height: 250px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Summary statistics for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2114f1cd-8775-4ffc-924c-26d4325d69f2",
   "metadata": {},
   "source": [
    "<h2>Per base sequence quality</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e0bee16-d937-4b09-bc3f-5b500ab2d01e",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 7</strong></h4>\n",
    "<img src=\"./images/6_1_fig_7.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Per base sequence quality for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43ee5d5c-5585-4e87-b576-2cc69d484fbb",
   "metadata": {},
   "source": [
    "Next, you'll see the per base sequence quality, which displays the stastitical information on your sequence read quality (<strong>Fig 2</strong>). The X-axis is the position along the read, while the Y-axis is the average quality score at that specific position for all reads. Recall from lecture and lesson 6.1 that .fastq files also contain the read quality information for each base of a sequence read, and the information is usually conveyed through a Phred score:\n",
    "\n",
    "```\n",
    "@SRR1810912.1 HWI-ST1210:136:C1RBDACXX:5:1101:1472:2121 length=50\n",
    "CACTGTAATTTCCCTAAAACAGGAGGTCACGGTAAGTCCTATGGATAGAT\n",
    "+SRR1810912.1 HWI-ST1210:136:C1RBDACXX:5:1101:1472:2121 length=50\n",
    "??????????????????????????????????????????????????\n",
    "```\n",
    "\n",
    "This module takes the average score at each position for all sequences and visualizes the data in the form of a box-and-whisker plot, with the inner quartile range covered by the yellow box and the whiskers depicting the lower and upper 10th percentiles. The blue line represents the calculated average Phred score at that postion, and in the case of the ChIP-seq data we're using, all the bases appear to have a Phred-33 score of 30, represented by the ASCII <code>&quest;</code>.\n",
    "\n",
    "The results of this analysis module can be quite important in that it lets you know if there may have been something that went wrong during sequencing, especially if most of the bases are of poor or highly variable quality. Usually, towards the end of a sequence read, the quality should dip a bit because of the nature of sequencing. Somtimes the elongation fails to continue in a greater and greater proportion of the strands or the clonal strands desync, so then the base can't be accurately called from the mixed fluorescent signal."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc44a598-8fdb-41f8-9f6d-c535203613cc",
   "metadata": {},
   "source": [
    "<h2>Per tile sequence quality</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01bdf515-f111-45a0-bccc-e6f127233d23",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 8</strong></h4>\n",
    "<img src=\"./images/6_1_fig_8.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Per tile sequence quality for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5946259-8d1a-4141-8902-d04d4bb0018a",
   "metadata": {},
   "source": [
    "For Illumina sequencing platforms, each sequence read retains its original sequence identifier, so the read can be traced back to its specific flow cell tile (<strong>Fig 9</strong>).\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 9</strong></h4>\n",
    "<img src=\"./images/6_1_fig_9.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Image from Whiteford et al 2009</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5c5a337-33d1-4027-aea4-f20bc7b32cce",
   "metadata": {},
   "source": [
    "The Illumina sequencing platform makes use of flow cells, which contain nanoscale wells that allow for clonal library amplification and subsequent sequencing by synthesis. If you recall from MCB201A and Dr. Ingolia's lectures, Illumina sequencing by synthesis technology uses fluorescently labeled nucleotides whose signal can be captured via imaging. Repeated images of the same section (a tile) of a flowcell allow for the sequence of the clonally amplified fragments to be read out as the progression of fluorescence signals when the clonal population of DNA fragments is extended. A tile is then considered a defined section of the flow cell that is repeatedly imaged, and a single tile can capture the sequences of a large number of clonal fragments (or clusters). Within a single flow cell, there are many, many tiles.\n",
    "\n",
    "So the information presented by this module can be thought of as a 3D matrix, where the 1st axis (or X-axis) contains the position of that base within the sequence reads, the 2nd axis (or Y-axis) contains essentially the tile ID, and the 3rd axis (the heatmap color) is the average read quality at that position (Phred score) for that tile.\n",
    "\n",
    "Issues that are identified through this analysis module generally point to some issue with the sequencing run, such as an issue calling bases or something obscuring the view of the imager, which would lead to an inability to call the base with high confidence. Since the data we use for ChIP-seq has a Phred-33 score of <code>&quest;</code> for all bases, we see a homogeneous graph for the per tile sequence quality."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff8ac9f0-1ee2-470f-9067-a6f8d040f37a",
   "metadata": {},
   "source": [
    "If we take a quick look at one sequence file from our RNA-seq data in <strong>Fig 10</strong>:\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 10</strong></h4>\n",
    "<img src=\"./images/6_1_fig_10.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Per tile sequence quality for <code>2024 RNA-seq data</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae758d78-a865-45d2-95ea-3d66c3e7cf73",
   "metadata": {},
   "source": [
    "We can see that most of the tiles look good, but for two of the tiles, there is an issue with the read quality in the later positions of the sequence reads. This type of issue generally points to a potential obstruction of the flow cell that occurred during imaging, which occurred partway through sequencing, so the later sequence read quality is impacted. This could be due to something like dust or some other particulate falling onto the flow cell or potentially something getting stuck in the flow cell, both of which can obscure those tiles and their fluorescence signals."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6917e762-0925-4d76-b202-1e4738c067aa",
   "metadata": {},
   "source": [
    "<h2>Per sequence quality score</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bc7a749-5ff7-4304-90bd-1b028a29ea51",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 11</strong></h4>\n",
    "<img src=\"./images/6_1_fig_11.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Per sequence quality score for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf84e93f-c59c-4366-982e-4b14293dc76a",
   "metadata": {},
   "source": [
    "The next analysis module is the per sequence quality score (<strong>Fig 11</strong>), which plots the distribution of the different Phred scores among all your sequence reads. So it's essentially like a histogram, where the X-axis is the average Phred score for a sequence read (not position), and the Y-axis is the number of reads that have that mean Phred score. A good sequencing run will have the peak of your distribution towards the higher end of the mean Phred score, indicating that your reads are mostly all good quality.\n",
    "\n",
    "Again, since the reads for the ChIP-seq dataset is all <code>&quest;</code>, we can get a better feel for this analysis module by looking at our RNA-seq dataset in <strong>Fig 12</strong>:\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 12</strong></h4>\n",
    "<img src=\"./images/6_1_fig_12.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Per sequence quality score for <code>2024 RNA-seq data</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "336f1fa0-500c-41df-b99f-793a18064595",
   "metadata": {},
   "source": [
    "We can see that there is a nice peak towards the higher Phred scores, indicating that the vast majority of our reads are of good quality."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e3a5dd4-82a1-4f4c-b284-540af5bcd6b0",
   "metadata": {},
   "source": [
    "<h2>Per base sequence content</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5f2a0c0-c5f2-4f74-b359-8f9647b94e1e",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 13</strong></h4>\n",
    "<img src=\"./images/6_1_fig_13.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Per base sequence content for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "554a9b59-61df-4f68-b40f-4c634f81c884",
   "metadata": {},
   "source": [
    "This analysis module takes a look at the proportions of each nucleotide for your sequence reads (<strong>Fig 13</strong>). Since our data comes from amplified fragments of the human genome, we would expect that the proportions of nucleotides should be close to the human genome. Since the GC content of the human genome is 41% (Bionumbers), we would expect both the G and C bases to comprise ~20% of the genome each, and both A and T to make up ~30% of the genome each. Over the millions of reads, the proportions to be similar to that of the human genome at each position of a read, so the four lines should be roughly parallel to one another.\n",
    "\n",
    "Deviations greater than 10% trigger a warning, while those greater than 20% will trigger a fail. This may point to some bias in your library leading to an over- or under-representation of certain bases. In the case of this ChIP-seq dataset, the complete drop in the proportion of A nucleotides at the last position suggest that there may have been trimming of this dataset to remove Illumina adapters from the sequence read since the Illumina universal adapter starts with an A. The homogeneous read quality also suggest that it may have been trimmed.\n",
    "\n",
    "You'll see later on with our RNA-seq data that it is usually expected to fail this analysis module."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2ee0df5-d61b-4aba-a232-4082d4c5b708",
   "metadata": {},
   "source": [
    "<h2>Per sequence GC content</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce5444f0-9c6c-41cb-826e-fd6f6337f3ff",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 14</strong></h4>\n",
    "<img src=\"./images/6_1_fig_14.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Per sequence GC content for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c30fef1d-035e-4326-ad77-c7b1195e5a26",
   "metadata": {},
   "source": [
    "While the per base sequence content looks at the proportions of each base at each position in a read, the per sequence GC content takes a look at the average GC content for all your sequence reads (<strong>Fig 14</strong>). In this analysis module, you're provided with the distribution of %GC content for your sequence file, and it gives you an idea if there may be an issue with your library prep. Usually, you should expect that without any experimental biases, the peak of your %GC content distribution should be match up roughly to the %GC content of your model organism's genome. The theoretical curve is calculated based on the sequence data you provide to FastQC, where it assumes a normal distribution around the modal GC content. If you see an unexpected distribution of %GC content, then that may point to an issue with your library, potentially an overexpressed gene or some contamination of your prep."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49257433-d88b-4960-96bd-f9857cff0f47",
   "metadata": {},
   "source": [
    "<h2>Per base N content</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566a9013-f5e5-4efe-8857-c849fc8967b8",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 15</strong></h4>\n",
    "<img src=\"./images/6_1_fig_15.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Per base N content for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6269f351-1639-4f9d-b776-c42d62548a22",
   "metadata": {},
   "source": [
    "This analysis module will let you know if there may have been issues calling bases along a specific position among all your sequence reads (<strong>Fig 15</strong>). In the ChIP-seq dataset that we're using, you can see that towards the end of the read, there is a fairly large increase in the proportion of Ns, causing this analysis module to be flagged as Fail. This usually points to an issue with the sequencing run.\n",
    "\n",
    "Taken into consideration with the per base sequence content, there may have been an issue towards the end of the sequencing run."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ab0ea9c-91cc-4ade-87e7-aa1f61f30169",
   "metadata": {},
   "source": [
    "<h2>Sequence length distribution</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17d3ae5e-63c9-4f97-bbf0-008f52b846e9",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 16</strong></h4>\n",
    "<img src=\"./images/6_1_fig_16.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Sequence length distribution for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78348415-4b7a-4c81-a622-901f2fcea352",
   "metadata": {},
   "source": [
    "This plots the distribution of the lengths of your sequence reads (<strong>Fig 16</strong>). Since Illumina sequencing platforms create reads of all the same length, you should expect a single length for all your sequence reads. Here, we can see that all the sequences are 50bp in length.\n",
    "\n",
    "Other sequencing platforms may have a range of sequence read lengths, and you would be able to see there distribution through this analysis module."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78f8976b-b68e-48c2-a175-0e4f44231d38",
   "metadata": {},
   "source": [
    "<h2>Sequence duplication levels</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27b1bbfc-2b37-46b3-abf6-7cd687145fab",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 17</strong></h4>\n",
    "<img src=\"./images/6_1_fig_17.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Sequence duplication levels for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1b2880b-db9b-415e-9c6b-bd2dd566d5a8",
   "metadata": {},
   "source": [
    "The sequence duplication levels module plots the distribution of the number of duplicate sequences within your sequence file (<strong>Fig 17</strong>). This analysis module assumes that an ideal library will have low levels of a diverse range of different sequence reads, indicating broad coverage. So a high level of duplication could suggest that there may be an issue with your library prep, resulting in some bias and an overrepresentation of certain fragments/sequences. However, your experimental design can also influence the composition of your library, and sometimes, you may expect there to be high levels of duplication of certain sequences, such as for very highly expressed genes. Another potential reason why this module may get flagged is if you have saturated your sequencing run, so subsequent reads are no longer as unique."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6c9827f-6f43-4123-bcb2-f6e784809df3",
   "metadata": {},
   "source": [
    "<h2>Overrepresented sequences</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52605726-221f-4d14-b2c6-a757871f2955",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 18</strong></h4>\n",
    "<img src=\"./images/6_1_fig_18.png\" style=\"height: 50px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Overrepresented sequences for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "246f4b99-7b7e-4f69-8e3c-b41a1c1885f9",
   "metadata": {},
   "source": [
    "This module takes a look at the sequnces that appear to be overrepresented in your sample (<strong>Fig 18</strong>), and it can point to potential adapter contamination or some other bias in your dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f523cc3-fa06-4e5c-acc2-0e51835ec1fe",
   "metadata": {},
   "source": [
    "<h2>Adapter content</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e344e872-ce4c-4a64-bc7b-534f93f64a70",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 19</strong></h4>\n",
    "<img src=\"./images/6_1_fig_19.png\" style=\"height: 350px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Adapter content for <code>10M_ctrl_1.fastq</code></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3706b7d7-8831-42f5-819b-7666c3a968a4",
   "metadata": {},
   "source": [
    "Illumina sequencing requires adapters to be ligated to both ends of DNA fragments to be sequenced in order for the DNA fragment to be anchored to the flow cell and for it to be clonally amplified. These adapters also are needed for the sequencing primers to bind to the single strand DNA template for sequencing by synthesis to occur. This analysis module takes a look at your sequence reads to determine if there is potential adapter sequences within them (<strong>Fig 19</strong>). An unusually high amount of adapter sequences that were captured in your sequencing run would be flagged by this  module and would indicate that you may want to perform some adapter trimming prior to aligning your reads.\n",
    "\n",
    "In the case of the ChIP-seq dataset that we're working with, it looks like there's no adapter contamination."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6177f15d-3576-4549-bc28-1f1a443f8fe2",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">References</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6257add-4732-4130-a593-85009327e4d4",
   "metadata": {},
   "source": [
    "<p style=\"padding-left: 20px;\"><a href=\"https://www.nature.com/articles/nrg2641\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Park 2009 Nat Rev Genet:</u></a> ChIPseq: advantages and challenges of a maturing technology</p>\n",
    "\n",
    "<p style=\"padding-left: 20px;\"><a href=\"https://www.nature.com/articles/ncb3216\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Zanconato et al 2015 Nat Cell Biol:</u></a> ChIPseq: advantages and challenges of a maturing technology</p>\n",
    "\n",
    "<p style=\"padding-left: 20px;\"><a href=\"https://www.nature.com/articles/ncb3216\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Wilbanks and Facciotti 2010 PLoS ONE:</u></a> Evaluation of Algorithm Performance in ChIP-Seq Peak Detection</p>\n",
    "\n",
    "<p style=\"padding-left: 20px;\"><a href=\"https://pubmed.ncbi.nlm.nih.gov/19549630/\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Whiteford et al 2009 Bioinformatics:</u></a> Swift: primary data analysis for the Illumina Solexa sequencing platform</p>\n",
    "\n",
    "<p style=\"padding-left: 20px;\"><a href=\"https://bionumbers.hms.harvard.edu/bionumber.aspx?id=100679&ver=5&trm=gc+content+human+genome&org=\" rel=\"noopener noreferrer\" target=\"_blank\"><u>Bionumbers for human genome GC content</u></a></p>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
