{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8ec5764-f92f-4a59-8564-c46295ccb69d",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Week 11 Assignment</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae0dff8-efec-409f-aa3e-71cd00f1a460",
   "metadata": {},
   "source": [
    "<strong><u>Due: Saturday, November 15 at 11:59PM</u></strong>\n",
    "\n",
    "<strong><u>Total points: 5 pts</u></strong>\n",
    "\n",
    "<strong><u>To turn in your assignment:</u></strong>\n",
    "<ol>\n",
    "    <li>Complete assignment and make sure you have all the outputs requested for the notebook</li>\n",
    "    <li>Go to your server's file directory and locate the <code>week_11_homework</code> directory containing all your outputs from this assignment including this notebook</li>\n",
    "    <li>Right click and select <strong>Download as an Archive</strong> to download the <code>week_11_homework</code> directory as a ZIP folder. Do not mix up with <strong>Download Current Folder as an Archive</strong></li>\n",
    "    <li>Upload your ZIP folder to your individual feedback Google Drive folder under <code>MCB201B</code></li>\n",
    "</ol>\n",
    "\n",
    "<strong><u>How this assignment will be graded:</u></strong>\n",
    "\n",
    "You will be graded on whether or not you have <u>all requested outputs and files</u> and if your output is correct. You will also be graded on if your code is expected to be able to run without errors. For those of you running code in Terminal or on your own computer, you will need to write out your input(s) into the empty code cell for this notebook. You don't need to run them here. If you used line breaks, make sure not to transfer over the <code>&gt;</code> symbol at the start of each new line. There can be more than one correct way to answer a question. If you find that you prefer to separate your answer for a single question into multiple code cells, feel free to do so.\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "066427bd-fc76-4e74-957c-57340af567c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22f0a9bc-68de-4438-ba39-4cd70b861500",
   "metadata": {},
   "source": [
    "<h2>Question 1</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecb8e3de-a9d3-42b3-98ee-3d1963122ff5",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a50ff93a-12f6-4b29-a3e2-5b7ffa753c29",
   "metadata": {},
   "source": [
    "Import your group's read counts <code>.txt</code> output from <code>htseq-count</code> into this notebook. Pull out the read counts statistics for each sample condition for the following into its own DataFrame.\n",
    "<ul>\n",
    "    <li>total reads counted</li>\n",
    "    <li>of the reads not counted:</li>\n",
    "    <ul>\n",
    "        <li>alignment not unique</li>\n",
    "        <li>not aligned</li>\n",
    "        <li>too low alignment quality</li>\n",
    "        <li>ambiguous</li>\n",
    "        <li>did not align to an exon</li>\n",
    "    </ul>\n",
    "</ul>\n",
    "\n",
    "Output the DataFrame as a stylized table into this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f404eb42-3f34-41cd-a0e5-29a0d1925597",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7ee1ef82-7704-42ef-8241-bd6fa7fe4830",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e475884c-0715-45d4-9d3b-c9d9f4bd54f7",
   "metadata": {},
   "source": [
    "<h2>Question 2</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54fc3958-a475-423e-80cf-1e2615185a0f",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1609f652-0d8a-40c9-9489-18a2993bf629",
   "metadata": {},
   "source": [
    "Filter your original DataFrame to obtain just the genes along with their associated counts. Export this DataFrame as a <code>.csv</code> file to be included with your assignment submission."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4364ed82-c5a2-402d-a401-eddf01d3856f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "08ecdcb5-5f38-4afe-a70c-ab3e6321c009",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7616cee9-84b1-4c05-8c98-999f702d0f5c",
   "metadata": {},
   "source": [
    "<h2>Question 3</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2ad5b7a-e1ec-4963-a1ad-3b6eaa266fb6",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a309c27-3cb6-4cee-a27f-ceb4cec7d979",
   "metadata": {},
   "source": [
    "Create a stacked bar plot visualizing both sample conditions' count statistics as percents of their respective total using the data from question 1. Add axis labels and legend as appropriate, and output your stacked bar plot as a PDF to be included with your assignment submission."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cc7bc3f-e5f6-4618-91f3-5a910107b691",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5b0bb14f-8cd3-4a74-bf04-606ff2e96d28",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88ed4b78-1c34-421d-814a-e2f7d98828ff",
   "metadata": {},
   "source": [
    "<h2>Question 4</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6e1c5ea-de3a-48f5-b277-9edc0bc6bc75",
   "metadata": {},
   "source": [
    "2pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "787be2ce-7ddc-4300-9fc3-03521088e6cd",
   "metadata": {},
   "source": [
    "Visualize each gene's read counts as a scatter plot with the x-axis as each gene's read count for the control condition and the y-axis as each gene's read count for the tazko condition. Highlight using different colors the points for TAZ (WWTR1) and YAP1, and add text annotations to label TAZ and YAP1 in your visualization. Add axis labels as appropriate, and output your scatterplot as a PDF to be included with your assignment submission."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54477175-27cb-470f-82e0-cd9b529caa17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2e654b78-10ce-40a3-ad5d-e58bbe71f044",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
