{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b281f0cb-f072-490e-a133-712cd815b040",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">9.1 RNA-seq analysis pipeline and QC</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0e748ce-873f-4cd7-8db3-f8c317759407",
   "metadata": {},
   "source": [
    "Now that you've gotten more familiar with ChIP-seq analysis and particularly with Terminal command line, we'll now move on to transcriptomics analysis. Specifically, we'll be analyzing our RNA-seq data that we generated during the summer. Much like with our ChIP-seq analysis module, the goal of our RNA-seq analysis is to continue to provide you with additional practice analyzing large datasets. ChIP-seq allows us to infer what genes may be regulated by TAZ based on where TAZ (and TEAD) is binding in the genome. RNA-seq complements that by providing us with information on the transcriptomic changes that occur following TAZ KO, which provides a differnt angle for us to infer downstream transcriptional targets by looking at what genes are differentially expressed when we KO TAZ. The end goal is to bring together our analyses from ChIP-seq and RNA-seq to then derive information about what genes are direct vs indirect transcriptional targets of TAZ and get a better sense of how TAZ acts as an oncogene or tumor suppressor gene.\n",
    "\n",
    "A lot of the broad concepts that we learned for ChIP-seq will be applicable to RNA-seq analysis as well even though the specifics may differ. Our sequence files are in the standard fastq format, and we'll still align our reads to a reference genome. And we can visualize the alignments again using IGV. After that, instead of looking at peaks, we'll be quantifying transcript abundance based on our alignment data, and then using that data to determine which genes are differentially expressed when we KO TAZ.\n",
    "\n",
    "The setup of this module will be more fun because we'll be working with all the data that we've generated in the summer. In class, we'll work with a truncated dataset like we did with our ChIP-seq analysis. Each group will perform the initial QC, alignment, and quantification on their own replicates. Then, the fun part is that we'll as a class bring together all our replicates to play with the data and perform differential expression analysis.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "\n",
    "<ul>\n",
    "    <li>Continue practicing command line and bash</li>\n",
    "    <li>Learn to set up shell scripts</li>\n",
    "    <li>Understand RNA-seq pipeline set up and QC</li>\n",
    "    <ul>\n",
    "        <li>Output some reads to look at later</li>\n",
    "        <li>Perform QC on truncated dataset</li>\n",
    "    </ul>\n",
    "    <li>Review fastq sequence files and FastQC analysis</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a37fc210-ad0a-4d61-9bdc-85cc70d835e6",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Shell scripts</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb7b3dc6-7308-4c6e-ac76-e912b6d433f6",
   "metadata": {},
   "source": [
    "Now that you're now more familiar with using Terminal and command line, we'll begin learning how to set up shell scripts, so that we can run a series of commands and programs by executing the script. This way, you don't need to sit and wait for Terminal to finish what it's doing to then tell it what to do next. \n",
    "\n",
    "We'll first learn to set up and run a basic shell script, then we'll begin adding additional layers of complexity to set ourselves up to build a shell script to analyze our RNA-seq dataset. Along the way, we'll continue to build upon what we know about command line syntax and control flow. We'll be updating our practice example script while also taking a look to see how it changes along the way by running it in Terminal as we make different modifications."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "320f7d1d-bfae-49e7-a3fc-45104c7eb7a2",
   "metadata": {},
   "source": [
    "<h2>What is a shell script?</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbc30e13-ebb0-4863-a28c-63954d62811a",
   "metadata": {},
   "source": [
    "Shell scripts (or Bash scripts for <strong><u>B</u></strong>ourne <strong><u>a</u></strong>gain <strong><u>sh</u></strong>ell in our case) allow you to automate a series of commands/tasks, so you don't have to input each step separately after the previous one has completed. This allows you to set up a wrapper that brings together the programs in your full analysis set up/pipeline. \n",
    "\n",
    "Like Python scripts, shell scripts are just plain text files with a particular extension, and in the case of shell scripts, the file extension is <code>.sh</code>. The script contains all the commands that you want to run as well as any associated comments."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76eecf3b-4a36-4478-be2b-c168a765e886",
   "metadata": {},
   "source": [
    "<h2>Structure of a shell script</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68902dfb-9124-4a3b-bc33-fc91d11051d0",
   "metadata": {},
   "source": [
    "The shell script structure is fairly flexible although you'll want to keep in mind conventions as well as proper syntax and control flow. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1f02d40-e497-45b2-9c28-6fc37ca22d04",
   "metadata": {},
   "source": [
    "<h3>Shebang</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51dcb53f-320a-483f-8704-6fb7e574bcee",
   "metadata": {},
   "source": [
    "The first line of shell scripts must begin with a hashbang, often referred to as a shebang, which consists of two characters <code>#!</code>. The shebang specifies the interpreter that will be used to execute the script.\n",
    "\n",
    "<strong>Find path to interpreter</code>\n",
    "\n",
    "To find the path to the interpreter that we've been using when we run commands in Terminal, we can call up the variable <code>0</code> and output the interpreter to the standard output using <code>echo</code>.\n",
    "\n",
    "To retrieve a value assigned to a variable, you can use the special character <code>$</code>. For this example, we can use it to retrieve the value assigned to the variable <code>0</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce22c981-477d-40c9-bff4-fcccc1bd9ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#!/bin/bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1e403ac-7356-4d16-b02c-116409de4001",
   "metadata": {},
   "source": [
    "Now that we know the path to our interpreter, we can include that alongside our shebang to specify what interpreter we want to use to execute our shell script.\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4884d93-ddd3-4ca1-aae7-48f1978ffb26",
   "metadata": {},
   "source": [
    "<h2>Comments</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96cf18a1-5910-4421-b8be-545ee5eef53a",
   "metadata": {},
   "source": [
    "Like Python, we can add comments to our shell script to help others (and you) to understand the code. At the beginning of the shell script, after the shebang, you can include comments that provide a brief overview of the script's contents. Other information that is sometimes included in these top-level comments are copyright, author information, and version history.\n",
    "\n",
    "Comments are specified by a hash <code>#</code>, and that everything in that line after the <code>#</code> is considered a comment.\n",
    "\n",
    "```bash\n",
    "# Liebchen was here.\n",
    "```\n",
    "\n",
    "Comments can also be added throughout the script to help explain things that might not be super obvious from the code itself or if you want to explain the logic/reasoning of how you set up the code."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a3957ad-c9cc-416a-83af-dad888401980",
   "metadata": {},
   "source": [
    "<h2>Functions and commands</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3b59b1c-5c7a-4ad7-8631-c5fb85399829",
   "metadata": {},
   "source": [
    "After the top-level comments is usually where functions are defined, and after that, follows the main code defining variables, running commands, and generating outputs. Essentially, all the different commands that we've run in Terminal, we can write out as lines of code in our shell script in the same way."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "460e2431-7c35-45d2-9c73-4ca34a20b0f2",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Setting up a shell script</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ded3ffa7-517e-45b1-b3f8-ef2a6c08a9b2",
   "metadata": {},
   "source": [
    "Let's now set up a simple shell script that we can run in Terminal. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11b060e3-7e8f-4646-97f3-819d1dc51a3c",
   "metadata": {},
   "source": [
    "<h2>Set up <code>.sh</code> file</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "056e4d84-bda5-4b34-8a2f-04d869c6eaa0",
   "metadata": {},
   "source": [
    "First, in the File Browser, navigate to this week's directory. Then open up your Launcher and start up a Text file, which is located under the section \"Other\". \n",
    "\n",
    "Then input the shebang along with the path to our bash interpreter:\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "```\n",
    "\n",
    "Let's then save the file as <code>week-9-script.sh</code>. Make sure that you replace the <code>.txt</code> extension with the <code>.sh</code> extension. When you finish saving the file, you'll notice that the color of the shebang has changed to be like the above example. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a83478ee-8fe7-4ce1-8497-19adf008f063",
   "metadata": {},
   "source": [
    "<h2>Make your shell script executable</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a79106d0-2c6c-4147-8bbf-17b1c3fe29de",
   "metadata": {},
   "source": [
    "In order to run our shell script in Terminal, we'll need to make it an executable file. This is kind of like turning a kitchen appliance on, so that you can use it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a24236b-7ba2-4340-82ff-a76d412b8924",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "chmod +x ./week-9-script.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec9354ab-9eac-430f-a21e-228b4fd36cf9",
   "metadata": {},
   "source": [
    "To confirm that our shell script is executable, we can then make use of <code>ls</code> but with an additional option <code>-l</code> to also include file permission information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "631583a8-92ea-4075-89f6-39fe7be55b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "ls -l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dba04f3e-80e3-44cf-9871-25859eed29ac",
   "metadata": {},
   "source": [
    "An executable file will have a <code>x</code> bit set within the file permissions information, indicating that it is executable with a command. So now to run our shell script, we can just call it up like we would with a command using Terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8b93468-3b96-4912-b824-a42a28f0f4aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "./week-9-script.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9c762d4-cf85-4bbd-b97e-7e216b4af022",
   "metadata": {},
   "source": [
    "There won't be any output since we haven't specified any commands and just added the shebang, but if your script is able to run, you should not get an error.\n",
    "\n",
    "Now as we make modifications and save our shell script, we can just run our script again in Terminal to test things out and see how it runs. Since this is a bit set for the file, the nice thing is that it stays permanently executable even after our server is shut down and started back up."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "309c34e0-39e9-483f-9c91-92ec998c5fdc",
   "metadata": {},
   "source": [
    "<h2>Add a top-level comment</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe52997c-59e0-4dba-9872-740cb02b80e9",
   "metadata": {},
   "source": [
    "We'll then go ahead and add a top-level comment that describes what our script will do. In this case, we can just note that this is our test script to play around in.\n",
    "\n",
    "```bash\n",
    "# This is a script to play around in and learn.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d308ce4a-eb8d-4677-a1bb-040d92f2ab0a",
   "metadata": {},
   "source": [
    "Now let's run our script. We shouldn't see any output still because our script only just contains the shebang and comments, so there shouldn't be any outputs, but there also shouldn't be any errors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf01e7cd-7153-457a-acc3-86a118597ac5",
   "metadata": {},
   "source": [
    "<h2>Set up a function</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf924d4a-de6f-4323-8c93-c549a1c5589e",
   "metadata": {},
   "source": [
    "Conventionally, functions are defined after the top-level comments before all the main body of code/commands. For practice purposes, we can set up a simple function to output some text to the standard output stream. Although the syntax for defining a function is different than what we know from Python, the overall concept and purpose is the same, where we define a reusable block of code that we can then call up later on when we need it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc9f1338-9366-42ed-bd2c-133ab4a52a87",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#####################\n",
    "# Meow at me for help\n",
    "#####################\n",
    "help_me_pls() {\n",
    "    echo \"MEOWWWWW!\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65e1ffda-35c1-4732-a6fc-713e81e413af",
   "metadata": {},
   "source": [
    "Now let's run our script, and we should still see that there's no output because all we've done is define our function."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6748b39-4721-46cd-bb77-6de8b4e96373",
   "metadata": {},
   "source": [
    "<h2>Input a basic command</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "339d6e4e-82b2-4ffe-8a75-23119be5072e",
   "metadata": {},
   "source": [
    "Now let's set up the main body of the shell script containing the commands that we want to automate. First, let's test out the function that we defined earlier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ce01ea2-7cdd-4e68-acd2-94c327bdbf17",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "help_me_pls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09bdf95e-25dc-42b1-8c47-89fa84188aa2",
   "metadata": {},
   "source": [
    "Let's now run our script in Terminal to see if our function and script works. Don't forget to save your script prior to running it in Terminal."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b73ddfb4-2952-40aa-a790-5d175fca84a4",
   "metadata": {},
   "source": [
    "<h3>Automate a series of commands</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c3cecab-a135-45a2-b31a-db7402c01883",
   "metadata": {},
   "source": [
    "So now that we know everything works well. Let's set up a quick script to automate a simple task. For this example script, we can pull out the first read from one of our sequence files. This example shown here is for group 1, but you can feel free to work with your group's data.\n",
    "\n",
    "Recall that we can use <code>cat</code> to open up files and display them in our <code>stdout</code> stream in Terminal. Since our sequence files are compressed (to help save storage usage), we'll make use of the option <code>zcat</code>, which does the same thing but for our compressed files. And since we're interested in just the first read, we can pipe <code>&vert;</code> the output from <code>zcat</code> into <code>head</code> to pull out just the first read, which we can redirect <code>&gt;</code> to a new file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1f4e211-3381-47a2-84db-27a9f693a213",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "zcat ~/shared/2025-fall/courses/1547808/rna-seq/truncated/g1/1M_g1_ctrl_r1.fastq.gz \\\n",
    "    | head -4 \\\n",
    "    > g1_ctrl_r1_read.txt\n",
    "\n",
    "zcat ~/shared/2025-fall/courses/1547808/rna-seq/truncated/g1/1M_g1_ctrl_r2.fastq.gz \\\n",
    "    | head -4 \\\n",
    "    > g1_ctrl_r2_read.txt\n",
    "\n",
    "zcat ~/shared/2025-fall/courses/1547808/rna-seq/truncated/g1/1M_g1_tazko_r1.fastq.gz \\\n",
    "    | head -4 \\\n",
    "    > g1_tazko_r1_read.txt\n",
    "\n",
    "zcat ~/shared/2025-fall/courses/1547808/rna-seq/truncated/g1/1M_g1_tazko_r2.fastq.gz \\\n",
    "    | head -4 \\\n",
    "    > g1_tazko_r2_read.txt\n",
    "\n",
    "#You can imagine that this can be set up as a for loop.\n",
    "#We'll get to that in the subsequent parts of this notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4036ce73-bbe0-4bb7-aaf0-4ffd1b5ee44c",
   "metadata": {},
   "source": [
    "Let's save and then run our script in Terminal."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37a289c9-d0e6-477d-8726-c6530b7a7331",
   "metadata": {},
   "source": [
    "<h2>Variables</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "109c77b5-2c56-4f04-a188-ade3333a0ba7",
   "metadata": {},
   "source": [
    "Like with Python, we can save things to variables, so we can call them up later to use. This is both something that we can do in Terminal and also in our shell scripts and is helpful to simplify our code, particularly if we're calling up things repeatedly. However, unlike Python, variables are untyped in bash, meaning that they are essentially interpreted as strings. There are some instances, like for example, integers, where the type can be inferred based on the context.\n",
    "\n",
    "We can go ahead and set up a couple of variables, and see how they are understood when we execute our script in Terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a36bcd8d-3d3a-4df7-af9e-4f814b7b331c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#Like with Python, we can assign things to variables\n",
    "meow=hungry\n",
    "cats=1\n",
    "\n",
    "#There are different ways of calling up a variable\n",
    "\n",
    "#Note how meow alone is not enough to indicate a variable\n",
    "echo meow\n",
    "\n",
    "#Requires the $ parameter expansion symbol \n",
    "echo $meow\n",
    "\n",
    "#Can add clarity with the curly brackets\n",
    "echo ${meow}\n",
    "\n",
    "#And then even more clarity with double quotes \n",
    "echo \"${meow}\"\n",
    "\n",
    "#Variables are untyped, and type can be inferred\n",
    "#For example here, arithmetic expansion\n",
    "#The 1 is then inferred to be an integer type\n",
    "echo $((cats + cats))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80593574-ec75-4e21-a0f1-eea62c40129b",
   "metadata": {},
   "source": [
    "Let's run our script again. Don't forget to save before running."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "674abd14-d33d-421c-a07b-152fdfb66cce",
   "metadata": {},
   "source": [
    "<h3>Positional parameters</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beaf660b-6e36-44ab-abd1-335921bf8a9d",
   "metadata": {},
   "source": [
    "When we call up commands, we can also provide positional parameters which will automatically be saved to particular variables based on their position in the command line input with the first position following the command saved to the variable <code>1</code> with subsequent positions assigned to each subsequent variable. Recall that each position is separated by a space, so that's how the bash interpreter is able to identify the command and each positional parameter.\n",
    "\n",
    "There are also ways to set up named parameters, which we're more familiar with as options and arguments that we provide when inputting commands, but for this example, let's add a positional parameter/variable to our shell script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6fdd508-623b-4a9e-ae42-0f3556694eea",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#Can call up positional variables\n",
    "echo \"${1}\"\n",
    "\n",
    "echo \"${2}\"\n",
    "\n",
    "#We stopped here."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d47e0eb-0c71-45e8-82f8-fc5cbf2654dc",
   "metadata": {},
   "source": [
    "Let's save the file, and now instead of running the script on its own, we can provide an argument for the first position when we run it in Terminal. You should see now that there is an additional output that comes from our first positional parameter. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76312e29-aca1-4b32-ad16-17a22f6289a4",
   "metadata": {},
   "source": [
    "<h3>Save file paths to variables</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fa7475b-6d07-42f2-8e20-631c26d38969",
   "metadata": {},
   "source": [
    "Since variables are generally untyped and handled essentially as strings, we can store our file paths and file names to variables in the same way to simplify our code and make it easier to read."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c2041bf-fd3d-4565-acba-033ba4af5ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb084f05-340b-4705-9e61-ef2e28c2fa54",
   "metadata": {},
   "source": [
    "Recall that variables are generally handled as strings, so that means we can append additional information onto our file paths, such as additional subdirectories, file names, and file extensions. You have some experience with this already when you added additional paths to your <code>PATH</code> variable so that Terminal is able to find the directory containing your HOMER executables.\n",
    "\n",
    "```bash\n",
    "PATH=$PATH:/home/jovyan/homer/bin\n",
    "```\n",
    "\n",
    "You can see that the additional string <code>:/home/jovyan/homer/bin</code> is appended to the <code>PATH</code> variable, and we can do the same thing to more conveniently specify subdirectories and files with additional modifications for clarity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd837927-acc2-40ac-82fe-d37b11a890ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c6c8645-5b54-49db-83a8-4247e4e015df",
   "metadata": {},
   "source": [
    "Let's now take a look at our script output."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47abd98a-0608-4464-b039-809c3f531b04",
   "metadata": {},
   "source": [
    "<h3>Playing around with variables</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e15c103-7ee9-4638-9b45-f40ac8a106ec",
   "metadata": {},
   "source": [
    "We can make use of a set of operations in bash called parameter expansion and pattern matching, and since our variables are generally handled as strings, this allows us to play around with the content of our variables. We won't go over all the different operations we can do, but we'll make use of one of them to allow us to remove substrings.\n",
    "\n",
    "The <code>$</code> special character initiates the expansion, which allows us to then make use of pattern matching to play with our variables. We can then make use of the <code>%</code> symbol to pattern match from the end of the string and remove that matching pattern. This can be handy for us to quickly pull out a base name much like we did in our Python scripts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c296dc4b-fb69-456c-9199-f0b99bc82b92",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c804ae37-6aac-4125-8a60-f9568f67dd28",
   "metadata": {},
   "source": [
    "Let's take a look to see how the output looks like and how it changes when we match to different patterns within a file name."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13dbd11a-fee5-408b-89d8-9cb71ae8b7fb",
   "metadata": {},
   "source": [
    "<h2>Create and assign an array</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce921084-877c-49d3-a3fa-31c93055bdb5",
   "metadata": {},
   "source": [
    "Like with Python, we can also work with compound data types in bash, specifically we can create and work with arrays. For our shell script example today, let's create a simple array and assign it to a variable. You'll want to keep in mind the syntax for an array, which is different in bash compared to what we're familiar with from Python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38bc329a-64e6-475d-8aa9-4b974a09999c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d88cfa15-fd5b-4162-b3e3-6370ecf4bdbc",
   "metadata": {},
   "source": [
    "Now let's take a look at how our script runs. Don't forget to save your script."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3ef43d2-8595-42ca-b3a6-9d05a3cf82fc",
   "metadata": {},
   "source": [
    "<h3>Positions in arrays</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56de1718-4069-4924-a097-baa67a9db6aa",
   "metadata": {},
   "source": [
    "Like with the compound data types that we're familiar with from Python, arrays in bash are also indexed based on their position within their array, and the way to call up a particular element in an array is also similar. One thing to note is that bash arrays are zero-indexed (so also not too different from what we know from Python)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2febf9f2-405d-40c5-98c1-aa6d21134b02",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e4c6cdf-9719-465e-8369-545722e91e38",
   "metadata": {},
   "source": [
    "Let's save and then run our script to take a look at how this is being interpreted by Terminal."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dc8ce34-f3ef-41c9-8570-15f109611887",
   "metadata": {},
   "source": [
    "<h3>Expand an array</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1cc84e3-b48d-47ae-9cdf-0ae494a1da48",
   "metadata": {},
   "source": [
    "As you might have seen with <code>echo \"${cats}\"</code>, we only saw the first element outputted to <code>stdout</code>. We can make use of some particular bash syntax to expand our array, so that we are able to see/output all elements within it at once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91241e24-7ffb-4735-a14c-0eb545e5ccad",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c528010-c233-4eeb-869f-4044c9a2a882",
   "metadata": {},
   "source": [
    "What you can see is that with <code>&lbrack;@&rbrack;</code>, we can specify that we want to expand our array, essentially pulling out all the elements of the array. We can add <code>!</code> to retrieve the index values of all our array elements, which will show us that the array is by default zero-indexed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4263170f-e567-4124-a637-dc6665934e75",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e96a8f0-30e9-4d94-8b09-1c8bac7aefac",
   "metadata": {},
   "source": [
    "<h3>Length of an array</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fa5023d-d586-4bc6-950d-530aeb350c1e",
   "metadata": {},
   "source": [
    "We can make use of array expansion to also pull the length of an array by simultaneously expanding the array and quantifying the number of elements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6af6dbe-9b4d-4ec7-91b7-aa3adc5d6d15",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a4d1238-158f-4f46-8095-5f9c40fce85f",
   "metadata": {},
   "source": [
    "<h3>Pull file names into arrays</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6ea8dc0-e311-4e2f-9b09-6932e2663b26",
   "metadata": {},
   "source": [
    "Like how we can prepare a list of file names in Python, we can do the same thing in bash as well, which will allow us to set up a way to then work through all our files that we want to analyze. For example, you'll be setting up a script today to pull out reads from your group's sequence files and run a FastQC analysis on each of your sequence files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef03d3f1-54db-4699-bd8b-4f4c1c2f2075",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "471b5b2e-dc44-4c12-914e-0a664bbfab17",
   "metadata": {},
   "source": [
    "Let's run this and see how our output looks.\n",
    "\n",
    "In an easier way than in Python, we can make use of pattern matching to pull particular files that we're interested based on whether or not they have a specific file extension or contain a specific pattern in their file name. Here, we make use of pattern matching but instead use the wildcard symbol <code>*</code>. This symbol lets us pull in any files matching the specified pattern around the wildcard symbol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e335f904-99c5-42ab-b657-cedadb64309d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ff17755-f152-44b2-86ce-7f10278ebb99",
   "metadata": {},
   "source": [
    "Let's take a look at the files that we pull in this way."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00f00f25-bfac-4ad3-bacf-15155c749453",
   "metadata": {},
   "source": [
    "<h2>Control flow</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afa1bc91-f23a-424b-aece-55c7386596be",
   "metadata": {},
   "source": [
    "Our scripts, much like with our Python code, reads from the first line down to the last line of our script, and there are particular keywords that we can use to control the flow of code execution. We'll make use of two whose logic you're already familiar with, specifically <code>if</code>-<code>elif</code>-<code>else</code> conditional statements and for loops."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e59a7f70-4f26-4cae-9f73-7034d21c1466",
   "metadata": {},
   "source": [
    "<h3>Conditional statements</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd354655-c406-4918-a08d-369cd5bb1fb8",
   "metadata": {},
   "source": [
    "The logic for conditional statements is the same in bash, but you'll want to keep in mind that the syntax is slightly different than what you're used to from Python.\n",
    "\n",
    "In bash, conditional statements are set up with <code>if</code> followed by a condition specified within double brackets <code>&lbrack;&lbrack; ... &rbrack;&rbrack;</code>. In the same line, the conditional is initialized by a <code>; then</code> with the subsequent block of code containing the code block to be executed if the condition is satisfied/true.\n",
    "\n",
    "The set up after for <code>elif</code> follows a similar syntax as the set up for <code>if</code> (kind of like Python), and the final <code>else</code> handles all remaining conditions.\n",
    "\n",
    "Conditional statements are then ended/closed with a <code>fi</code> keyword, which is apparently just <code>if</code> backwards.\n",
    "\n",
    "```bash\n",
    "if [[ your condition here ]]; then\n",
    "    command to be executed\n",
    "    \n",
    "elif [[ another condition here ]]; then\n",
    "    other command to be executed\n",
    "    \n",
    "else\n",
    "    handle remaining conditions\n",
    "    \n",
    "fi\n",
    "```\n",
    "\n",
    "Something to keep in mind also is that <code>true</code> is represented as <code>0</code> in bash, also indicating a success command, whereas any non-zero integer <code>1</code> or greater indicates false or an error/failure.\n",
    "\n",
    "For our example script, we can set up a conditional statement to handle the first positional parameter, which can set your script up to take in a file path to your group's RNA-seq data and to exit when a help option or no positional argument is specified."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60f9200c-ee93-44f4-9a5a-72d80eceebc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b692931f-83b1-4469-9da3-375bf86873b3",
   "metadata": {},
   "source": [
    "Let's take a look at our script to see how it runs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b75e75e-50bf-4306-8df0-e092191ce55f",
   "metadata": {},
   "source": [
    "<h3>For loops</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22c8c55d-ca32-4ea4-bd9b-066c0f1a8134",
   "metadata": {},
   "source": [
    "The logic of for loops remain the same, but you'll just want to keep in mind the slight differences in syntax, like how conditional statements have slightly different syntax in bash as well.\n",
    "\n",
    "You can set up a for loop with the <code>for</code> keyword followed by a variable and <code>in</code> and the array you want to loop through, and in the same line that you initialize your loop starts with a <code>; do</code>. The subsequent code block then contains the operations/commands to be repeatedly done until there are no more elements to loop through. The end of the for loop is specified by a <code>done</code> keyword.\n",
    "\n",
    "```bash\n",
    "for element in \"${array[@]}\"; do\n",
    "    command to be repeated\n",
    "done\n",
    "```\n",
    "\n",
    "We can make use of our <code>cats</code> array from earlier to try out a for loop in our script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca5cfde9-c4f5-44e0-9e70-bef5b4a974b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "257c8781-cc97-4d86-b637-06f57e698ed1",
   "metadata": {},
   "source": [
    "Let's take a look at our script and see how the for loop runs.\n",
    "\n",
    "We can also then pull index values or calculate our own range of values (like how we specified <code>np.arange()</code> and <code>range()</code> in our Python for loops) to then iterate through multiple arrays. Let's create a second array and work through both our <code>cats</code> array and the <code>fave_snack</code> array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aa7ce20-94a0-4b74-9f18-4401dc2d91eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98adfcd2-6d60-4396-9e6b-a22b8c04eb1e",
   "metadata": {},
   "source": [
    "Now let's see how our for loop handles both arrays."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc37ba74-ea90-48ee-a0ff-34687f06aedb",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise set</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95e6b638-429f-4886-acdf-49f2e1d12eb0",
   "metadata": {},
   "source": [
    "For this exercise set, you'll be applying what you've learned in this notebook to set up a new shell script called <code>rna-seq-qc.sh</code> that can perform the following tasks:\n",
    "\n",
    "<ol>\n",
    "    <li>Make new directory to hold some reads in this week's directory</li>\n",
    "    <li>Save the file path to your group's RNA-seq truncated data set to a variable</li>\n",
    "    <li>Pull in your group's file names into an array</li>\n",
    "    <li>Pull out the first 10 reads from each of your sequence files</li>\n",
    "    <li>Make a directory to hold FastQC analysis outputs</li>\n",
    "    <li>Run a FastQC analysis on your group's RNA-seq truncated data files (you can refer back to notebook 6-1 on how to run <code>fastqc</code></li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f34dd625-d2f5-4817-ae56-b7a1816d2395",
   "metadata": {},
   "source": [
    "We'll reconvene once everyone has their shell scripts to take a look to see how people set it up.\n",
    "\n",
    "Once we all have a working shell script, we can then proceed with running them, and while it runs, we'll review the RNA-seq pipeline in preparation for the rest of this module of our course."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6f3ece0-023a-4c2d-b750-85fb5447db5e",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">RNA-seq overview</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccccce8c-8614-4ca9-9558-aef1663d51a7",
   "metadata": {},
   "source": [
    "RNA-seq comprises a series of experiments and analysis programs much like ChIP-seq. First, an appropriate sample needs to be obtained and prepared prior to sequencing, followed by next-gen sequencing, then alignment. Where RNA-seq differs from ChIP-seq is that rather than taking the alignments and using them to call peaks, RNA-seq takes the alignment information and quantifies the reads associated with genomic features, specifically genes. The read counts between different conditions are normalized and used for differential expression analysis to determine how the expression profile changes with different experimental manipulations. The resulting dataset can then provide insights into interesting candidates for follow up or provide a broad overview of how the transcriptome is changing between different conditions or treatments."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "758f7bff-3448-4061-8e62-08803bcfac22",
   "metadata": {},
   "source": [
    "<h2>RNA collection</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "184ac838-62e8-4b2d-89b1-7179d825246e",
   "metadata": {},
   "source": [
    "Recall from MCB201A that we collected RNA from our MDA-MB-231 cells to take a closer look at what genes are being regulated by TAZ to gain insights into how TAZ is regulating tumorigenesis. So as Dr. Ingolia mentioned in lecture, one of the important steps to be able to effectively sequence mRNA is to enrich for the mRNA in our total RNA sample. This is because the total RNA within a cell is almost entirely made up of ribosomal RNA (rRNA), which you can see from our Bioanalyzer traces (<strong>Fig 1</strong>), where the very prominent rRNA peaks allow us to assess the overall integrity of our RNA sample.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 1</strong></h4>\n",
    "<img src=\"./images/9_1_fig_1.png\" style=\"height: 200px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Example Bioanalyzer trace from Group 3</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "304903fe-c08d-4dbb-87c9-d2b11fa428ba",
   "metadata": {},
   "source": [
    "<h2>Library prep</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8e4fd1b-55df-498e-8760-34e9c90547f7",
   "metadata": {},
   "source": [
    "In order to get enough material to sequence, a cDNA library needs to be prepared from the enriched mRNA. Our unprepped RNA was sent to the FGL for QC and library prep, where they used the KAPA RNA HyperPrep Kit in order to prepare a cDNA library."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba31cff0-f7f4-4e7b-8ef5-5b81dade8547",
   "metadata": {},
   "source": [
    "<h3>1. mRNA enrichment</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f32c5fb-1ce7-418b-9746-aef0f94015d9",
   "metadata": {},
   "source": [
    "After polyA selection, which enriches for mRNA by hybridizing their polyA tails to oligo-dTs. The oligo-dTs are attached to magnetic beads, which allow your mRNA to be pelleted and any unbound RNAs (your rRNAs, ncRNAs, etc) to be washed away. So this is like any standard pull-down experiment. The mRNA is then dissociated from the beads, leaving you with a sample enriched for mRNA."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a34893d-697d-4e4f-b98f-fdcab8691d2d",
   "metadata": {},
   "source": [
    "<h3>2. mRNA fragmentation</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6f307f0-f566-44f8-b4f1-dbeaec98b302",
   "metadata": {},
   "source": [
    "Following mRNA enrichment, the mRNAs are fragmented using a combination of heat and magnesium, breaking the mRNAs into smaller fragments. This allows short reads to cover the span of the mRNA rather than being localized to just the ends of the mRNA."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8859b2b9-d381-4f2d-b642-e7a7c558d77f",
   "metadata": {},
   "source": [
    "<h3>3. Library preparation</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb5f1706-ff04-49d0-8b1a-66eb2218537c",
   "metadata": {},
   "source": [
    "A cDNA is then synthesized from the mRNA fragments and adapters are added onto the ends of the cDNA. Recall that Dr. Ingolia talked about how mRNA is transcribed from one strand of the DNA acting as a template, so mRNA has a specific strandedness derived from which strand of DNA acted as the template for mRNA transcription. This information is lost during cDNA synthesis if steps aren't taken to preserve or retain strandedness.\n",
    "\n",
    "Since genes can be transcribed from either the plus strand or the minus strand of DNA, and in some instances, you may have genes that overlap with each other in terms of their chromosomal location but are transcribed from opposite strands. So in order to reduce ambiguity when quantifying reads and assigning them to genes, cDNA preparation now often includes a slight modification that allows for this information to be preserved/retained."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d25604e6-0028-44fe-99a4-6085134f2ce7",
   "metadata": {},
   "source": [
    "<h4>RNA-seq strandedness</h4>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3142da57-f647-45ad-8d0b-e5c469497cba",
   "metadata": {},
   "source": [
    "The basic idea is that when you create your cDNA library, you can selectively degrade one of your cDNA strands, and the one that is degraded depends on the specific protocol you are using. So when you amplify your cDNA library, you only have sequencing primers that are oriented a specific way, where each primer will only be associated with one strand of the cDNA rather than both (<strong>Fig 2</strong>).\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 2</strong></h4>\n",
    "<img src=\"./images/9_1_fig_2.png\" style=\"height: 500px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Modified image from Azenta Life Sciences</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "177d9282-8bc8-467a-89d1-a74fac45efcf",
   "metadata": {},
   "source": [
    "In our case, our library was prepped with dUTPs incorporated into the second cDNA strand. Thus, the second cDNA strand cannot be amplified, and the first cDNA strand is sequenced. Therefore, our sequence data is considered <strong><u>directional on the first strand.</u></strong> (UC Berkeley QB3 Functional Genomics Lab).\n",
    "\n",
    "As Dr. Ingolia talked about during lecture, the strandedness of your library has an impact on downstream alignment and read counting because some RNAs will be coded on the plus strand and others on the minus strand. We'll encounter this later on when we do our analyses, and we'll play around with an alignment where we provide the incorrect strandedness to see what impact that has on our read counts."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40902fbf-01a4-4911-8b82-fbf9810a9ca0",
   "metadata": {},
   "source": [
    "<h2>Next-gen sequencing</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7174a915-4492-47ee-82c5-8df8df375737",
   "metadata": {},
   "source": [
    "Sequencing of cDNA libraries for our RNA-seq experiments is conceptually the same as for ChIP-seq. One key difference in our case is that our samples were sequenced using paired-end sequencing, which means that each fragment was sequenced from both ends.\n",
    "\n",
    "This means that our sequencing data for each replicate has <strong><u>two</u></strong> <code>.fastq.gz</code> files associated with it, where one corresponds to the reads from the first mate <code>&ast;_r1.fastq.gz</code> and the other from the second mate <code>&ast;_r2.fastq.gz</code>. Other than that, the type of results from sequencing are pretty similar to what we've seen with our ChIP-seq data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5692478f-78ee-4e96-a5e7-85719d537d34",
   "metadata": {},
   "source": [
    "<h2>FastQC analysis of RNA-seq data</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aed078d3-c593-4193-b4e3-7e7709a6a233",
   "metadata": {},
   "source": [
    "As you've seen with your shell script, we make use again of FastQC to take a look at whether or not there were issues with our sequencing runs. There are some notable differences in how the FastQC outputs look between ChIP-seq data and RNA-seq data. Like with all our data, you'll want to keep in mind what the biology is that gave rise to the results that we're seeing."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "399dc7a9-afa2-4799-af90-6f3fa9b6229b",
   "metadata": {},
   "source": [
    "<h3>Per base sequence content</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c57d26fb-9945-46cb-8ee4-8c57c966f37a",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 3</strong></h4>\n",
    "<img src=\"./images/9_1_fig_3.png\" style=\"height: 350px; margin: auto;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "416f153f-9a36-4bc2-bb85-eab718e59a6d",
   "metadata": {},
   "source": [
    "Following mRNA fragmentation, the mRNA needs a primer to bind for cDNA synthesis to occur. Because all the mRNA fragments will have different \"starting\" sequences, priming needs to be random. In reality though, random priming is not completely random, and there is often a bias with enrichment for certain bases. So RNA-seq data will often fail this analysis module (<strong>Fig 3</strong>)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f42913d-5a1f-48f2-b99c-6f78d66c7940",
   "metadata": {},
   "source": [
    "<h3>Sequence duplication  levels</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2af5587-a615-4034-9b42-ce0f75177fe7",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 4</strong></h4>\n",
    "<img src=\"./images/9_1_fig_4.png\" style=\"height: 350px; margin: auto;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0193808-a609-4d36-8b59-9712f1a88517",
   "metadata": {},
   "source": [
    "Another difference that will show up in RNA-seq FastQC analyses compared to what we saw with ChIP-seq is the output of the analysis module quantifying the sequence duplication levels. Recall that FastQC assumes that an ideal library is a diverse one, which would mean that there shouldn't be overrepresented sequences in your sequencing data. However, for RNA-seq experiments, highly overexpressed genes will often be overrepresented in your dataset. So a warning or failure for this module isn't usually too big of an issue unless you weren't expecting it based on your experimental setup."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7eb62de-c338-415d-be13-15c77d54a656",
   "metadata": {},
   "source": [
    "<h2>Alignment of RNA-seq data</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "929beba6-3b35-4486-8c18-4c2a5047456b",
   "metadata": {},
   "source": [
    "Like with ChIP-seq, we need to align the reads to a reference. This allows us to later use the read alignments for downstream quantification. In ChIP-seq, the read alignments were usd to build coverage maps, and in the case of RNA-seq, we use the read alignments to determine the levels of expression for each gene in order to identify differentially expressed genes between our two conditions (<strong>Fig 5</strong>). There are special considerations that we need to keep in mind when aligning data from RNA-seq, which Dr. Ingolia touched upon, and we'll see when we perform the alignments next week.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 5</strong></h4>\n",
    "<img src=\"./images/9_1_fig_5.png\" style=\"height: 400px; margin: auto;\"/>\n",
    "<p style=\"text-align:center;\">Image from Kukurba and Montgomery 2015</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10eaec44-484d-49b5-8f8c-e75aeb99e237",
   "metadata": {},
   "source": [
    "<h2>Quantification of aligned reads (read counting)</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27b1e902-636a-4b9a-8af2-302e976e6dff",
   "metadata": {},
   "source": [
    "The next step after aligning reads is to quantify the number of reads mapped to each gene to give us an idea of the level of expression of each gene within our cells (<strong>Fig 5</strong>). A matrix can then be created from all the raw counts, which is then fed into a program to determine which genes are differentially expressed between conditions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df329883-028e-4e78-8ce6-d0db43d6a0a6",
   "metadata": {},
   "source": [
    "<h2>Differential expression analysis</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf176601-83f6-4d8d-a8be-7cebc6a79c55",
   "metadata": {},
   "source": [
    "For differential expression analysis, we'll make use of both Python (and R) to obtain a dataset of genes and their change in expression relative to controls along with any significance values and log2 fold change in expression. This data can also then be used to generate visualizations to help us get an overview of the results of our analysis (<strong>Fig 6</strong>).\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 6</strong></h4>\n",
    "<img src=\"./images/9_1_fig_6.png\" style=\"height: 350px; margin: auto;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d9c5732-16cb-4911-84dd-d28856bfeef0",
   "metadata": {},
   "source": [
    "<h2>Clustering</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ef95d9e-12b7-4219-bfa6-595d271e7a93",
   "metadata": {},
   "source": [
    "From there, we can play around with our results to identify potential genes of interest. We can also cluster our data in various ways to see how consistent the replicates are with one another (<strong>Fig 7</strong>). This can also help you identify potential experimental variability, such as batch effects, that may impact your interpretation of the data.\n",
    "\n",
    "<h4 style=\"text-align: center;\"><strong>Fig 7</strong></h4>\n",
    "<img src=\"./images/9_1_fig_7.png\" style=\"height: 350px; margin: auto;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb261708-027d-4517-9b12-dbdb3c091c56",
   "metadata": {},
   "source": [
    "<h2>Functional enrichment and annotation</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d37c8b40-b5d7-4a43-8f42-6e2121f92388",
   "metadata": {},
   "source": [
    "We can take our differential expression dataset and take a look at what biological processes, pathways, or functions are enriched in our dataset. This can give us a broad picture of what changes are occurring within the cell based on the changes we find when we look at different broad groupings of our dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e74ade-ad10-4242-9a5a-9df2298057db",
   "metadata": {},
   "source": [
    "<h2>Multi-omics</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e856b14-3a95-483e-8fd9-60c9eb673255",
   "metadata": {},
   "source": [
    "Towards the end of the course, we'll also be pulling together the results of our ChIP-seq analysis and our RNA-seq analysis to take a multi-omics view of our data by integrating our two analyses and gain additional insights into how TAZ is regulating tumorigenesis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "892eb7d0-7df5-4aca-a829-3c3e1a3070b3",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">References</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d43ab115-b692-4424-8801-a123aa4df5f6",
   "metadata": {},
   "source": [
    "<p style=\"padding-left: 20px;\"><a href=\"https://doi.org/10.1101/pdb.top084970\" rel=\"noopener noreferrer\"><u>Kukurba and Montgomery 2015 CSH Protocols:</u></a> RNA Sequencing and Analysis</p>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
