{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8ec5764-f92f-4a59-8564-c46295ccb69d",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Week 10 Assignment</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae0dff8-efec-409f-aa3e-71cd00f1a460",
   "metadata": {},
   "source": [
    "<strong><u>Due: Saturday, November 8 at 11:59PM</u></strong>\n",
    "\n",
    "<strong><u>Total points: 5 pts</u></strong>\n",
    "\n",
    "<strong><u>To turn in your assignment:</u></strong>\n",
    "<ol>\n",
    "    <li>Complete assignment and make sure you have all the outputs requested for the notebook</li>\n",
    "    <li>Go to your server's file directory and locate the <code>week_10_homework</code> directory containing all your outputs from this assignment including this notebook</li>\n",
    "    <li>Right click and select <strong>Download as an Archive</strong> to download the <code>week_10_homework</code> directory as a ZIP folder. Do not mix up with <strong>Download Current Folder as an Archive</strong></li>\n",
    "    <li>Upload your ZIP folder to your individual feedback Google Drive folder under <code>MCB201B</code></li>\n",
    "</ol>\n",
    "\n",
    "<strong><u>How this assignment will be graded:</u></strong>\n",
    "\n",
    "You will be graded on whether or not you have <u>all requested outputs and files</u> and if your output is correct. You will also be graded on if your code is expected to be able to run without errors. For those of you running code in Terminal or on your own computer, you will need to write out your input(s) into the empty code cell for this notebook. You don't need to run them here. If you used line breaks, make sure not to transfer over the <code>&gt;</code> symbol at the start of each new line. There can be more than one correct way to answer a question. If you find that you prefer to separate your answer for a single question into multiple code cells, feel free to do so.\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "066427bd-fc76-4e74-957c-57340af567c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22f0a9bc-68de-4438-ba39-4cd70b861500",
   "metadata": {},
   "source": [
    "<h2>Question 1</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecb8e3de-a9d3-42b3-98ee-3d1963122ff5",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a50ff93a-12f6-4b29-a3e2-5b7ffa753c29",
   "metadata": {},
   "source": [
    "From your previous <strong>full analysis</strong> of our ChIP-seq datasets for <code>taz_1</code> and <code>taz_2</code>, assign ranks to each peak that was called by <code>macs2</code> separately for each TAZ replicate using the <code>'-log10(qvalue)'</code> from <code>*_peaks.xls</code> output files. Output into this notebook the top 10 peaks for each replicate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f404eb42-3f34-41cd-a0e5-29a0d1925597",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7ee1ef82-7704-42ef-8241-bd6fa7fe4830",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7616cee9-84b1-4c05-8c98-999f702d0f5c",
   "metadata": {},
   "source": [
    "<h2>Question 2</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2ad5b7a-e1ec-4963-a1ad-3b6eaa266fb6",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a309c27-3cb6-4cee-a27f-ceb4cec7d979",
   "metadata": {},
   "source": [
    "Use HOMER to identify peaks present in both <code>taz_1</code> and <code>taz_2</code>. Then create a scatter plot that visualizes the ranks of each replicate's peaks with their rank in <code>taz_1</code> along the x-axis and their rank in <code>taz_2</code> along the y-axis. Label axes accordingly. Output the scatter plot into this notebook, and you do not need to separately submit a PDF output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cc7bc3f-e5f6-4618-91f3-5a910107b691",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#Copy over your Terminal code here, but you do not need to run it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a836e2e9-c988-495e-960b-897e9bc2e6a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This code cell and additional code cells you make\n",
    "#can be used for the rest of this question."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b0bb14f-8cd3-4a74-bf04-606ff2e96d28",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88ed4b78-1c34-421d-814a-e2f7d98828ff",
   "metadata": {},
   "source": [
    "<h2>Question 3</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6e1c5ea-de3a-48f5-b277-9edc0bc6bc75",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "787be2ce-7ddc-4300-9fc3-03521088e6cd",
   "metadata": {},
   "source": [
    "Identify peaks present in both <code>taz_1</code> and <code>taz_2</code>, and are ranked in the top 500 in both replicates. Use these peaks for motif analysis with HOMER. Plot the motif distribution around your top set of peaks for your top 3 motifs. Label axes and provide legend as appropriate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54477175-27cb-470f-82e0-cd9b529caa17",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#Copy over your Terminal code here, but you do not need to run it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c648a18c-60f5-4bf8-9dcb-6fcaed6bbbf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This code cell and additional code cells you make\n",
    "#can be used for the rest of this question."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e654b78-10ce-40a3-ad5d-e58bbe71f044",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3243bffc-1424-4fc7-ab13-b91b6ac54aa6",
   "metadata": {},
   "source": [
    "<h2>Question 4</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f36026a8-213e-49fe-8f2a-8bb36c0be54b",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9fc4ac9-a287-4796-8eae-3a9e81250dc7",
   "metadata": {},
   "source": [
    "Use either your bash script from class or Jack's bash script (making adjustments as needed to analyze your group's dataset) to analyze your <strong>full</strong> RNA-seq dataset for your group. You will need to:\n",
    "\n",
    "<ol>\n",
    "    <li>Pull out 10 reads from each of your group's sequence files. <strong>Submit the outputs as part of your homework submission.</strong></li>\n",
    "    <li>Run FastQC on all four of your group's sequence files. <strong>Submit the outputs as part of your homework submission.</strong></li>\n",
    "    <li>Align your paired-end reads for both sample conditions to the hg19 reference genome. You do not need to submit any alignment output files.</li>\n",
    "    <li>Process your alignment outputs as we did in class for visualization in IGV and for read counting with <code>htseq-count</code>. You do not need to submit any output files from samtools, but make sure to delete your SAMs.</li>\n",
    "    <li>Quantify read abundance for each gene using <code>htseq-count</code> as we did in class. <strong>Submit your output <code>.txt</code> file from <code>htseq-count</code>.</li>\n",
    "</ol>\n",
    "\n",
    "No need to copy over any code or script for this question."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe264baa-46e4-4ea7-95cb-c4b004a05e43",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1872c40-89bc-4e6f-8169-9516e938eb6f",
   "metadata": {},
   "source": [
    "<h2>Question 5</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99e748cb-5ffb-4b62-b14d-485088654fb4",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "678301d3-2b51-4862-99b2-f56c8f0be741",
   "metadata": {},
   "source": [
    "Load in your control sample's position-sorted <code>.bam</code> and <code>.bam.bai</code> and your TAZ KO sample's position-sorted <code>.bam</code> and <code>.bam.bai</code> files into IGV. Navigate to the following genomic range for TAZ (WWTR1): \n",
    "\n",
    "```\n",
    "chr3:149,232,978-149,325,800\n",
    "```\n",
    "\n",
    "Confirm that you can visualize the reads for TAZ, and save as a PNG to include with your homework submission. \n",
    "<strong>Once you've completed your homework assignment, go ahead and delete all your <code>.bam</code> and <code>.bam.bai</code> files (both name-sorted and position-sorted) to reduce storage usage on the Biology Hub.</strong> Jack will provide backups if you need, so feel free to delete your BAMs, BAM.BAIs, and SAMs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8daadb15-dc4b-487b-a455-e6324cc5ba2d",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
