{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "deaf89fe-afa9-4b97-ac58-21fb2e62fa4c",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">10.1 RNA-seq alignment</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c938df20-c85d-4535-913b-786bde00e516",
   "metadata": {},
   "source": [
    "Last week, we went over the general RNA-seq pipeline and set up our first shell script to begin learning how we can take the commands we input into Terminal and automate it. Like with ChIP-seq, we need to first align our data to a reference genome as a set up for further analysis. \n",
    "\n",
    "Today, we'll be reviewing and performing alignments for our truncated RNA-seq datasets that we generated last week. Everyone will be working with their own group's data for this, so we may all see slightly different outputs/results. As we continue with the RNA-seq module, we'll begin building up our piepline, connecting the lessons from our RNA-seq module together for a more automated RNA-seq analysis workflow.\n",
    "\n",
    "Since we continued with notebook 9-1 into this week, we can merge the processes covered in both 9-1 and 10-1 into a single workflow for our shell script.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "\n",
    "<ul> \n",
    "    <li>Learn how to run HISAT2 for aligning RNA-seq data</li>\n",
    "    <li>Automate 9-1 and 10-1 commands</li>\n",
    "    <li>Review splice-aware alignment</li>\n",
    "    <li>Visualize RNA-seq alignments using IGV</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37fa0705-3bd9-4792-bc48-c8aebd9de331",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">HISAT2 alignment</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df8cd726-68fc-40c8-9ffe-4af4431735ad",
   "metadata": {},
   "source": [
    "<a href=\"https://daehwankimlab.github.io/hisat2/manual/\" rel=\"noopener noreferrer\"><u>Documentation for HISAT2 can be found here.</u></a>\n",
    "\n",
    "Like we did with <code>bowtie2</code>, let's set up our alignment and have that run in the background as we go over the concepts and code while it runs. Since we're finishing up notebook 9-1 with 10-1, we can merge our alignment commands with our 9-1 exercise script. Then we can run everything all together.\n",
    "\n",
    "So first, let's set up our commands including processing the outputs, then once we're done setting up our script, we can run it and go over the concepts and breakdown the code. We'll walk through the set up together while thinking through the logic of the set up."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cce108dd-c934-4171-b4c5-20a8838bd8d9",
   "metadata": {},
   "source": [
    "<h2>1. Prepare directories to handle outputs</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c783ac30-9610-470a-acac-fe8c872aac00",
   "metadata": {},
   "source": [
    "Since we'll be trying to automate more things, we can prepare some directories to handle our outputs. We can merge these commands with our 9-1 script to keep things tidy. Like with our ChIP-seq alignments, we'll have SAM and BAM files, as well as alignment logs. So let's set up our shell script to keep things organized."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86b34fd3-f5ec-465d-a1e0-2b12773f088c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c905202-b790-44cf-ac7e-05ef7789cd89",
   "metadata": {},
   "source": [
    "<h2>2. Alignment using <code>hisat2</code></h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc0364f7-3e36-4101-9e25-9dece6945857",
   "metadata": {},
   "source": [
    "For this part, we can set up the command to run our alignment, and once our shell script is running, we can break down how the input is set up in a later section. For this set up, you can work with your own group's data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78444bfb-2d78-4145-addd-45f5ddaf6b33",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f3f01ee-a084-4a16-a014-5fdde190e3a8",
   "metadata": {},
   "source": [
    "No need to run anything yet, we'll continue with setting up our shell script."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3673810d-0426-480b-b34c-da0e3e3cacde",
   "metadata": {},
   "source": [
    "<h2>3. Check for alignment completion</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7270fa03-acf8-46ae-9b8f-2cfb87d34bc0",
   "metadata": {},
   "source": [
    "Since we're setting up our shell script to be more automated, we can set up our script to check to make sure the alignment was successfully completed before proceeding with the rest of the script. We can set up our script to then abort itself, so that it doesn't proceed with the rest of the script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "403ca0d0-6f4e-4f67-abec-3347b3f3fc28",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e27c5e47-332f-4ac2-a7b9-127fe58b5dd2",
   "metadata": {},
   "source": [
    "<h2>4. Pull out rows from SAM file to look at later</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7868072-0f0a-41b6-ba7a-c2f26513a9d0",
   "metadata": {},
   "source": [
    "Let's also pull out some rows from our SAM file to explore a bit more later once we get the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8512579-9509-4a49-ad98-7c64bcc194a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5e0d952-59e1-463e-9fd1-b93ca13cf9db",
   "metadata": {},
   "source": [
    "<h2>5. Convert SAM to BAM</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "119c1fc3-097d-440b-8d6b-03e0f6ddd2d7",
   "metadata": {},
   "source": [
    "Like with our ChIP-seq alignment outputs, we'll compress our SAM files into BAMs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a18d3e24-3a2d-41c2-ad23-d0be80d3a018",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98cb1288-4324-4a26-a0c5-dfc80c0a170c",
   "metadata": {},
   "source": [
    "<h2>6. Check for successful conversion and delete SAM</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c58e5ed0-666b-4478-a9b9-0ac9633efa18",
   "metadata": {},
   "source": [
    "Then, we can check for successful conversion using the same approach that we took for step 3. This way, we can set up our script to automatically check that we were able to convert our SAM into a BAM before actually deleting our SAM, and we can save server space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96205c0e-0e2c-4036-acf3-6faf44ec1273",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "032a468b-21dc-4486-b16b-bd90b19d1cc7",
   "metadata": {},
   "source": [
    "<h2>7. Sort BAM by position and index</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b6ef64c-369f-4428-86a5-dda7a0876b8a",
   "metadata": {},
   "source": [
    "To visualize our alignment data in IGV, we'll need to sort our reads based on their position and then index them. Later on, once we get our outputs, we can download them to then load into IGV like we did with our ChIP-seq alignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed6a2d85-f005-4359-853f-dd8e91b718a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f206cf7-87b2-4e1f-b439-e64b88d2510b",
   "metadata": {},
   "source": [
    "<h2>8. Sort BAM by name</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bac62ccc-9e6d-477c-bae1-5e3a3488bc5d",
   "metadata": {},
   "source": [
    "This is where things start to differ a bit from our ChIP-seq analysis. On Friday, we'll quantify our reads, and to prepare for that, we'll need another BAM file, where instead of sorting the aligned reads by their chromosomal position, we'll sort them based on their read name/ID. That way, each read of a read pair will appear one after another in the BAM file. This helps to keep memory usage down when we do the counting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9efe709f-7e27-4a51-bc77-d5cbbe7bd4c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebf8c561-c2de-444b-ae06-bb64eeb4e1e6",
   "metadata": {},
   "source": [
    "<h2>9. Check for successful sorting and indexing (and delete unsorted BAM)</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60412ca8-5283-474a-8370-9fb172db2845",
   "metadata": {},
   "source": [
    "We can set up another checkpoint to then check that everything was properly sorted and indexed. To help reduce server space, we can also delete our original unsorted BAM file. That way, we don't have 3 copies of BAMs in various states of sorting. We'll just have two."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df3dd1b9-dbb4-4c87-b769-744c2bf7fcdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1da5ccfb-6469-4f5d-9537-29cd91a632e4",
   "metadata": {},
   "source": [
    "<h2>10. Work in a for loop to automate alignment and processing of both conditions</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1c2dc88-91fa-4e54-af55-f05252a8c847",
   "metadata": {},
   "source": [
    "We have multiple files that we want to process, so we can work everything together into a for loop to automatically work through both sample conditions. We can also consider how to work in our commands from notebook 9-1 as well to automate the entire process so far."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cb6f8fb-5e5b-4c66-9055-9097e0751bf9",
   "metadata": {},
   "source": [
    "<h2>Run your shell script</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81b46877-6091-41d9-b40b-fa1b530950e1",
   "metadata": {},
   "source": [
    "Once we have our completed shell script, let's go ehead and run it. Then we can troubleshoot if needed and review RNA-seq alignments and our <code>hisat2</code> input."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "203fd264-710b-4a1c-9e91-d692d5daa1d9",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Breaking down <code>hisat2</code> command input</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cc28412-8e1c-4e07-ada6-1953d9d3f1f3",
   "metadata": {},
   "source": [
    "While our script runs, let's break down our <code>hisat2</code> input, so we understand how the program operates."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2db02b31-2602-459b-8afa-264276a8dfbc",
   "metadata": {},
   "source": [
    "<h2>Let's break down the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6b41b0e-7787-4cd1-87c0-f3bd1b6bd376",
   "metadata": {},
   "source": [
    "<code>hisat2</code>\n",
    "\n",
    "This calls up the <code>hisat2</code> program.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-x ~/shared/2025-fall/courses/1547808/rna-seq/rna-index/hg19</code>\n",
    "\n",
    "This functions the same as in <code>bowtie2</code>. We indicate with the option <code>-x</code> the path to our reference genome index files along with the basename of those files. In this case, our index files are located in <code>~/shared/2025-fall/courses/1547808/rna-index/</code>, and all the index files contain the same basename <code>hg19</code>.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-1 ~/shared/2025-fall/courses/1547808/rna-seq/truncated/g1/1M_g1_ctrl_r1.fastq.gz</code>\n",
    "\n",
    "This is one key difference between our RNA-seq alignment and ChIP-seq alignment. While the ChIP-seq data was from a single-end (SE) read, our RNA-seq data was obtained via paired-end (PE) sequencing, resulting in two sequence files associated with each replicate. With the <code>-1</code> option, we tell <code>hisat2</code> that we are providing the first mate file for our PE data.\n",
    "\n",
    "If our data were unpaired, then we would (like with <code>bowtie2</code>) use the <code>-U</code> option for unpaired or SE reads.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-2 ~/shared/2025-fall/courses/1547808/rna-seq/truncated/g1/1M_g1_ctrl_r2.fastq.gz</code>\n",
    "\n",
    "With the <code>-2</code> option, we provide <code>hisat2</code> with the second mate file.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>--rna-strandness RF</code>\n",
    "\n",
    "Recall that our RNA-seq library was generated with a specific strandedness with respect to the original mRNA fragment, so we will need to provide this information to the aligner using the <code>--rna-strandness</code> option along with the <code>RF</code> argument to indicate that our library is directional on the first strand.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-p 8</code>\n",
    "\n",
    "Like with <code>bowtie2</code>, we can specify additional performance options. In this case, we'll process our files using the 8 cores of our Biology Hub server to speed up the time it takes for our alignments.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>--mm</code>\n",
    "\n",
    "Because <code>hisat2</code> when aligning for the human genome is expected to use up to 4.5GB with 8 parallel processes, but sometimes it can spike up to or over 5GB during alignment, which will kill our server. So we can try and reduce the memory footprint by specifying that we want <code>hisat2</code> to use a memory-mapped index, so that all alignment processes access the same single instance of our hg19 index in memory rather than separate ones. \n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>-S ~/MCB201B_F2025/Week_10/alignments-temp/sams/1M_g1_ctrl.sam</code>\n",
    "\n",
    "Like with <code>bowtie2</code>, we can specify that we want a <code>.sam</code> output using the <code>-S</code> option followed by the file path and name of our SAM file that we want <code>hisat2</code> to create.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>--summary-file ~/MCB201B_F2025/Week_10/alignment-logs/1M_g1_ctrl-alignment.log</code>\n",
    "\n",
    "The option <code>--summary-file</code> tells <code>hisat2</code> that we would like it to output a file containing alignment statistics, which will be generated once the alignment is completed. The argument after <code>--summary-file</code> is the file path and file name that we want the alignment statistics to be saved to. Like when we ran our ChIP-seq alignments, the summary file signals to us that our alignment was successfully completed, which we can incorporate into our script as a checkpoint to verify successful alignment.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #BBBBBB;\"></hr>\n",
    "\n",
    "<code>--new-summary</code>\n",
    "\n",
    "<p>This option alters the format of the summary file that is generated by the <code>--summary-file</code> to be more human readable.</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3360feaf-21b4-4eb2-8771-c551642e4a2c",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Review of RNA-seq alignment</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53b6a8d6-31b1-4a77-8b2a-71cfcebaa6d7",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 1</strong></h4>\n",
    "<img src=\"./images/10_1_fig_1.png\" style=\"height: 300px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">RNA-seq read alignment strategies (Dr. Ingolia)</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e77bd6f9-29a7-4c22-8cf2-f01892ed3476",
   "metadata": {},
   "source": [
    "Recall from lecture that RNA-seq alignments have special considerations because of the processing of mRNA to remove introns and splice together exons. Three common methods for aligning reads are to align to a reference transcriptome, or reference genome, or to assemble transcripts <i>de novo</i> if a reference is unavailable (<strong>Fig 1</strong>)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1ef12d2-85ba-43e3-9473-afafa49c5c9a",
   "metadata": {},
   "source": [
    "<h4 style=\"text-align: center;\"><strong>Fig 2</strong></h4>\n",
    "<img src=\"./images/10_1_fig_2.png\" style=\"height: 300px; margin: auto;\"/>\n",
    "<p style=\"text-align: center;\">Examples of incorrect read mapping (Kim et al 2013)</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c612e476-2a4d-4288-bd9c-6561b8020e3b",
   "metadata": {},
   "source": [
    "Alignment programs that are unable to account for large gaps in reads that span junctions may incorrectly map reads (<strong>Fig 2</strong>). They may either align the short read to only a single exon along with a penalty for the incorrectly matched basepairs, or they may incorrectly align the read to a pseudogene that doesn't contain an intervening intron, or they may remain unmapped. Splice aware aligners, such as HISAT2, can properly account for junctions, which you will be able to see when we load our alignments into IGV."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f9e3ae2-9a43-4877-b199-161c03ddd944",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Viewing alignment outputs</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3b37f3c-7e9d-426e-88b9-f06e2d931642",
   "metadata": {},
   "source": [
    "<h2>Alignment statistics</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adef2ce5-e86d-4c29-b8f7-cb782f52c1a5",
   "metadata": {},
   "source": [
    "Once your shell script has finished running, you should now have a number of output files to look through. We'll first take a look at our alignment log file containing each sample's alignment statistics. This will give us an idea if there was an issue with how we set up the alignment or potentially an issue with our samples/replicates. Generally, you'll want all your replicates to have similar alignment rates, and if there's a replicate that has drastically lower alignment rates, you may want to take a closer look to see if issues arose during the prep of that replicate."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b80475af-a402-4b8b-be9d-90f55f47932f",
   "metadata": {},
   "source": [
    "<h2>Viewing SAM output</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c6b08d1-3c24-45fd-acd4-d20e4d1a3092",
   "metadata": {},
   "source": [
    "Let's take a look at the 150 rows that we pulled out of our SAM file. At first glance, it will look more or less the same as our SAM files from our ChIP-seq alignment. The notable differences will lie in the FLAG score and the CIGAR string (and the aligner-specific tags)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae0f74ba-f31d-4277-b171-f41064bdb268",
   "metadata": {},
   "source": [
    "<h2>Take a look at read alignments for TAZ and YAP</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a943560-bf89-49b2-84e1-dab69822178c",
   "metadata": {},
   "source": [
    "Let's take a look at our reads using our position-sorted BAM files and the associated index file. Then we can load them into IGV to see how the reads aligned for TAZ, given that we aimed to KO TAZ expression. <a href=\"https://www.ncbi.nlm.nih.gov/gene?Db=gene&Cmd=DetailsSearch&Term=25937\" rel=\"noopener noreferrer\"><u>TAZ (WWTR1) is located at 3q25.1.</u></a>\n",
    "\n",
    "Let's also see how the reads look for YAP, which we didn't KO. <a href=\"https://www.ncbi.nlm.nih.gov/gene?Db=gene&Cmd=DetailsSearch&Term=10413\" rel=\"noopener noreferrer\" target=\"_blank\"><u>YAP1 is located at 11q22.1.</u></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef98846f-44f7-4252-b596-ffb4cedd12be",
   "metadata": {},
   "source": [
    "<h3>Color by first-of-pair strand</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd4aedf7-4b4f-4c62-a651-c482463e20cc",
   "metadata": {},
   "source": [
    "We can also visualize the strandedness of our cDNA library in IGV as well. Right-click the read tracks and select \"first-of-pair strand\" under \"Color alignments by\". You should see the alignments colored either red or blue based on the direction of the cDNA library.\n",
    "\n",
    "A cDNA library that is unstranded will have read alignments for a single gene with a mixed population when colored by direction, whereas a cDNA that is stranded will have read alignments that are all (or nearly all) the same color."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26582650-26fc-46ec-accb-80e980fd721f",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">References</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "134af1d0-bde1-4299-80e0-ca74acd923ab",
   "metadata": {},
   "source": [
    "<p style=\"padding-left: 20px;\"><a href=\"https://www.nature.com/articles/s41587-019-0201-4\" rel=\"noopener noreferrer\"><u>Kim et al 2019 Nat Biotech:</u></a> Graph-based genome alignment and genotyping with HISAT2 and HISAT-genotype</p>\n",
    "\n",
    "<p style=\"padding-left: 20px;\"><a href=\"https://doi.org/10.1093/bioinformatics/btw354\" rel=\"noopener noreferrer\"><u>Ewels et al 2016 Bioinformatics:</u></a> MultiQC: summarize analysis results for multiple tools and samples in a single report</p>\n",
    "\n",
    "<p style=\"padding-left: 20px;\"><a href=\"https://genomebiology.biomedcentral.com/articles/10.1186/gb-2013-14-4-r36\" rel=\"noopener noreferrer\"><u>Kim et al 2013 Genome Biol:</u></a> TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions</p>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
