{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5c0393d9-8331-4577-826b-d130cc5d076c",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">&#127875;10.2 RNA-seq counts</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f40542b8-1e0e-428f-aaa5-e11438341930",
   "metadata": {},
   "source": [
    "In notebook 10-1, we aligned our RNA-seq reads to a reference genome using the splice aware aligner HISAT2, and today, we'll be counting our mapped reads in order to set up for our subsequent differential expression analysis. To do this, we'll make use of a program called HTSeq, which will allow us to count the reads mapped to different genomic features, and in the case of RNA-seq, we're interested in counting reads that have mapped to exons. Each group will then obtain raw counts for their control and TAZ KO replicates. We'll take a look at the data in our Python notebooks to do some basic QC, and then take a look to see if we might have some potentially differentially expressed genes just solely based on one replicate. Like for 9-1 and 10-1, we'll set up our scripts, then we can breakdown the code and concepts while our script runs.\n",
    "\n",
    "We'll then work on a Python script that we can incorporate into the analysis pipeline that we've beeen setting up so far, so our scripts will be able to proceed through all the RNA-seq analysis steps that we've gone through up until this point.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "\n",
    "<ul>\n",
    "    <li>Setting up checkpoints in bash scripts</li>\n",
    "    <li>Incorporate read counting into our scripts</li>\n",
    "    <li>Run a QC on read counts for your group's replicate</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9af86a8-2c3b-429c-9b25-a7900a678c2b",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Counting reads with HTSeq</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1881de1d-aae0-4fc4-bff8-7866dacef44b",
   "metadata": {},
   "source": [
    "<a href=\"https://htseq.readthedocs.io/en/master/index.html\" rel=\"noopener noreferrer\"><u>Documentation for HTSeq can be found here.</u></a>\n",
    "\n",
    "HTSeq (High-throughput sequence analysis in Python) is developed by Fabio Zanini, Simon Anders, Givanna Putri and contributors (<a href=\"https://academic.oup.com/bioinformatics/article/38/10/2943/6551247\" rel=\"noopener noreferrer\"><u>Anders et al 2014</u></a>, <a href=\"https://academic.oup.com/bioinformatics/article/31/2/166/2366196\" rel=\"noopener noreferrer\"><u>Putri et al 2022</u></a>). It's a Python package that acts as a framework for working with high-throughput sequencing data in Python. We will make use of its ability to count reads within features when provided with an alignment file and with a genome annotation file. We'll set up our bash scripts to check for <code>HTSeq</code>, and if needed, install <code>HTSeq</code> and if successful, run <code>htseq-count</code>. After you obtain your group's count data, you'll use this Python notebook to play around with the outputs to run some quick QC."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa249dd-3657-4942-9354-7941d8f04a8d",
   "metadata": {},
   "source": [
    "<h2>1. Create a checkpoint to verify <code>HTSeq</code> installation</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de2815e7-56b1-4360-afcf-ab7bedd595e9",
   "metadata": {},
   "source": [
    "Recall how we've had to manually reinstall due to how our servers reset to a default configuration, so to help our script run in a more automated way, we can set up a checkpoint to verify whether <code>HTSeq</code> is installed, and to install it if needed. Recall from 10-1, when commands are successfully executed, they <code>exit</code> with a value of <code>0</code>, but if they fail to run, they exit with a value that is <code>1</code> or greater. This then allows us to check based on the exit value to verify an installation and install if needed by taking advantage of what we know about conditional statements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "316c9a19-cf4d-4e40-b420-89ad77a5f0de",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "153b2bce-7f4a-4ca1-a107-b75b1e212868",
   "metadata": {},
   "source": [
    "<h2>2. Pull out some rows from our annotation file to look at later</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab8e241b-7303-4768-90ba-a770b919f60e",
   "metadata": {},
   "source": [
    "Genomic features are usually annotated in a GFF or GTF format, which is a fairly large tab-delimited table containing the following information for each annotated genomic feature:\n",
    "\n",
    "<table style=\"text-align: center;\">\n",
    "    <tr style=\"font-size: 15px;\">\n",
    "        <th>seqname</th>\n",
    "        <th>source</th>\n",
    "        <th>feature</th>\n",
    "        <th>start</th>\n",
    "        <th>end</th>\n",
    "        <th>score</th>\n",
    "        <th>strand</th>\n",
    "        <th>frame</th>\n",
    "        <th>attribute</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>Chr name</td>\n",
    "        <td>Program that generated the data or the data source</td>\n",
    "        <td>type of feature (gene, exon, cds, etc)</td>\n",
    "        <td>start position</td>\n",
    "        <td>end position</td>\n",
    "        <td>floating point value</td>\n",
    "        <td>+ or -</td>\n",
    "        <td>codon frame</td>\n",
    "        <td>additional information for each feature</td>\n",
    "    </tr>\n",
    "</table>\n",
    "\n",
    "For this step, we'll incorporate into our shell script a quick operation to pull out some rows from our GTF file so that we can take a look at it later to better understand the information it contains."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec8f1555-884a-45ce-a991-cfff09232c91",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3ba053a-06c1-4d15-8c7b-21cd905fe98d",
   "metadata": {},
   "source": [
    "<h2>3. Set up <code>htseq-count</code> to obtain read counts</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20506b67-6232-4e92-8afc-b7230e73acd8",
   "metadata": {},
   "source": [
    "One of the main things you can do with HTSeq is to determine how many reads overlap with a specific set of genomic features. For example, you can count the reads that overlap with exons in order to begin estimating the levels of gene expression in your cell. HTSeq contains a specific script that performs read quantification with the intention that you will later perform differential expression analysis.\n",
    "\n",
    "We will set up in our RNA-seq script the command line input to take a gene annotation file and alignment files to quantify how many reads map to each annotated gene.\n",
    "\n",
    "<a href=\"https://htseq.readthedocs.io/en/master/htseqcount.html\" rel=\"noopener noreferrer\"><u>Documentation specifically for <code>htseq-count</code> can be found here.</u></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0f7625b-c16a-48b5-a4c2-6c598f46304f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ae9709f-96e8-4d71-94a2-f596a8186580",
   "metadata": {},
   "source": [
    "Now let's run our full script with the FastQC, alignment, and read counting. And we can review the associated concepts and breakdown the code from 9-1, 10-1, and for read counting."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b953eb77-62c6-424f-864c-b8efde128165",
   "metadata": {},
   "source": [
    "<h2>Let's break down the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c288d4f4-184c-4985-ab20-e5638a6ce4c1",
   "metadata": {},
   "source": [
    "<code>htseq-count</code>\n",
    "\n",
    "This is to call up the script to get the counts of reads mapped to a set of genomic features.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-t exon</code>\n",
    "\n",
    "The <code>-t</code> option allows us to specify what type of feature from the <strong>feature</strong> column of the GTF file that we want to obtain read counts for. In this setup, we are looking to determine how many reads overlap with annotated exons.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-i gene_id</code>\n",
    "\n",
    "The <code>-i</code> option provides the labels for each exon that we determine the counts for. Here, we specify that we want it to pull the <strong>gene_id</strong>, and use it as the label for its respective exon.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-r name</code>\n",
    "\n",
    "Here, we specify that our BAM file reads are sorted by name.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-s reverse</code>\n",
    "\n",
    "This option tells <code>htseq-count</code> the strandedness of your cDNA library, and it's an important option because it will have a noticeable impact on the raw counts that you get. If you select the incorrect strandedness, you will end up with inaccurate quantifications that will impact your differential expression analysis. \n",
    "\n",
    "Since our cDNA library is directional on the first strand, so we will need to specify <code>reverse</code> for the strandedness option.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>-f bam</code>\n",
    "\n",
    "We provide the information on the type of alignment file that we are using as the input(s). Our alignment files are BAMs, so we specify <code>bam</code>.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>\"${sorted_bams[@]}\"</code>\n",
    "\n",
    "After we specify our options, we then provide the alignment files that we want to obtain read counts for. Here, we can provide it with our BAM files that were sorted by name by expanding our variable <code>sorted_bams</code>. This will allow us to supply our BAM files for our control and our TAZ KO samples.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>\"${gtf_file}\"</code>\n",
    "\n",
    "Then we specify the annotation file that we want it to use in order to identify how many reads are overlapping with known genomic features. In our case, we provide it with a file containing gene annotations which we've saved to a variable <code>gtf_file</code> that points to the file <code>~/shared/2025-fall/courses/1547808/rna-seq/rna-feature/hg19-refseq.gtf</code>.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>&gt; \"${output_parent}/counts/1M_g1_counts.txt\"</code>\n",
    "\n",
    "We then tell it to redirect the output to a <code>.txt</code> file that we can then play around with later and use for differential expression analysis.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae16d3b1-779f-4275-a4f2-3dcfa064dabd",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise Set 1</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83122006-140e-4716-a1dc-d19131c0403d",
   "metadata": {},
   "source": [
    "Since we provided <code>htseq-count</code> with two BAM files as once, it will provide us with a data matrix containing the read counts for both our control and TAZ KO samples.\n",
    "\n",
    "The first column should contain the <code>gene_id</code> that we instructed <code>htseq-count</code> to use to identify each feature. So we'll have the gene symbols in the first column. The second column should contain your raw counts for your control sample, and the third column should contain your raw counts for your TAZ KO sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f4e20ac-7903-47dc-916b-70b14d7deb37",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f076326f-2396-410f-9407-6c04c2a2c181",
   "metadata": {},
   "source": [
    "<h2>Exercise #1: Import <code>sample-gtf-to-explore.txt</code></h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83355807-2959-4824-9fc7-8626f6607b68",
   "metadata": {},
   "source": [
    "Here, load in the GTF rows that we pulled out from our original GTF file to explore a bit as a group when we reconvene. Recall that the table is tab-delimited.\n",
    "\n",
    "You can add the following headers in the below order:\n",
    "<ol>\n",
    "    <li>seqname</li>\n",
    "    <li>source</li>\n",
    "    <li>feature</li>\n",
    "    <li>start</li>\n",
    "    <li>end</li>\n",
    "    <li>score</li>\n",
    "    <li>strand</li>\n",
    "    <li>frame</li>\n",
    "    <li>attribute</li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bfed856-aa52-4f40-89dd-3e6f04b9ba12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e3831462-f889-43ab-af28-a2b4cd0659e4",
   "metadata": {},
   "source": [
    "<h2>Exercise #2: Load in your group's counts matrix</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5295d3b-82b9-410a-b76b-aabbfd24a2c2",
   "metadata": {},
   "source": [
    "Import the count matrix that you got for your control and TAZ KO samples. The output from <code>htseq-count</code> is tab-delimited without column headers, so you can specify the header names as <code>'gene'</code>, <code>'ctrl'</code>, and <code>'tazko'</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a1a8ff1-db16-4740-aa56-2bb592d5e9fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "91a976fd-173e-4c97-b492-594d2e8ddd52",
   "metadata": {},
   "source": [
    "<h2>Exercise #3: Take a look at the last 10 rows</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f64ec608-4670-4a59-a456-9a943b01ce04",
   "metadata": {},
   "source": [
    "There's additional information about our counts at the last set of rows, so for this exercise, output just the last 10 rows to see what information it contains."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "755f041d-4e5a-4786-ae06-719f0f519aac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f70d489a-9765-4413-ab54-7c767941f1a4",
   "metadata": {},
   "source": [
    "<h2>Exercise #4: Pull out information to perform QC</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e16a7d13-e4ad-46cc-8e93-3350de7bcd1c",
   "metadata": {},
   "source": [
    "Let's take a look at our overall count statistics by pulling out the rows starting with the string <code>__</code> (double underscores) from our DataFrame using the <code>pd.Series.str.contains()</code> function. <a href=\"https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html\" rel=\"noopener noreferrer\"><u>Documentation for <code>pd.Series.str.contains()</code> can be found here.</u></a>\n",
    "\n",
    "The output from <code>pd.Series.str.contains()</code> will return booleans, so see if you can use it to filter your data to just obtain the rows starting with <code>__</code>.\n",
    "\n",
    "You can then save this to a new variable <code>read_stats</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d51a4f0c-1a86-49f0-a478-d5a3cc59a533",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f901cbdd-8a19-4ef5-85e9-de014cd86008",
   "metadata": {},
   "source": [
    "<h2>Exercise #5: Update labels to remove <code>__</code></h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65704c77-fc67-4528-a615-cca5ff183f62",
   "metadata": {},
   "source": [
    "To tidy things up a bit, remove the initial <code>__</code> from the different labels from exercise #4. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f649da32-b6ba-4275-9c0f-6eb5d699f33a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a6fe798e-cbbd-4b96-8f05-9ad5baa70147",
   "metadata": {},
   "source": [
    "<h2>Exercise #6: Determine how many reads overlap with annotation genes</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eb36d1d-3e5a-4094-ba18-b06bf2bc6e69",
   "metadata": {},
   "source": [
    "So you've pulled the reads that were not associated with any annotated genes in our GTF file, so the reads that were counted as part of our read counts are the remaining rows. In other words, the inverse of the rows containing the string <code>__</code> are the ones that contain the reads that were counted. We can continue to make use of the <code>pd.Series.str.contains()</code> function, but this time, see if you can pull the inverse to get all the rows containing gene counts. \n",
    "\n",
    "Then see if you can find the total number of counted reads for each sample condition <code>'ctrl'</code> and <code>'tazko'</code>. Combine the result with your updated <code>read_stats</code> DataFrame from exercise #5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "265f873f-f815-4cdc-82b7-148cf9c30399",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bc15cf12-4abb-41ad-a42d-1188d8d0dd6d",
   "metadata": {},
   "source": [
    "<h2>Exercise #7: Generate a stacked bar plot with your <code>read_stats</code></h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f007ce5-9a5d-406f-879a-a73678ed6363",
   "metadata": {},
   "source": [
    "Now use your <code>read_stats</code> DataFrame to generate a stacked bar plot that visualizes the proportion/percent of each category stacked on top of each other. You can set up your plot so that each sample condition <code>'ctrl'</code> and <code>'tazko'</code> is its own stacked bar.\n",
    "\n",
    "This will allow us to then do a quick QC to see what what the proportion of each category is to get an idea of whether or not something may have gone wrong during counting.\n",
    "\n",
    "Recall back in notebook 2-1 how we set up a stacked bar plot using <code>plt.bar()</code>. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "523f44fb-986a-4bac-8a73-1b2be7cfc4be",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "405b69ed-7a39-488c-8ce2-b099826ce75d",
   "metadata": {},
   "source": [
    "We'll reconvene here and take a look at how everyone's counts data looks like overall."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "179c973d-2d31-4551-b376-6469645c7911",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise Set 2</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f685c572-62d9-478d-a24a-726bf8614c6c",
   "metadata": {},
   "source": [
    "For this exercise set, you'll take a look at your group's data to see how the two sample conditions compare to one another to assess whether or not there is potentially a detectable difference between them. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6edd90cb-2898-4dac-9842-d2da671c8f0a",
   "metadata": {},
   "source": [
    "<h2>Exercise #1: Pull out just reads that got counted</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b22a8cf7-5711-42f3-a96d-135eef6d4a79",
   "metadata": {},
   "source": [
    "For this exercise, filter your original dataset to obtain a data matrix containing just the genes and their respective read counts, and then export the DataFrame as a <code>.csv</code> file with the following naming scheme: <code>1M_g1_gene_counts.csv</code>, where <code>g1</code> is replaced by your group number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6762c4ad-f527-4df1-8ae9-5ec89184798c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4f65dad1-5338-4919-8527-e4a29ac28042",
   "metadata": {},
   "source": [
    "<h2>Exercise #2: Plot sample conditions against each other</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6701462e-9818-4396-b7db-96c1ae13afa1",
   "metadata": {},
   "source": [
    "Now, generate a scatter plot where each dot corresponds to a single gene, and its position along the x-axis is determined by its read count in <code>'ctrl'</code> and its position along the y-axis is determined by its read count in <code>'tazko'</code>. If the data is difficult to distinguish due to the presence of particular outliers, perform a log transformation on your read counts or plot on log axes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e28af56-1d8c-4ae8-904b-dbad3c41be13",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "febf3ae3-827c-486d-8931-927841e17ed4",
   "metadata": {},
   "source": [
    "Consider what it means when a dot (gene) falls along the x=y axis vs showing some particular offset from x=y. What does its position in your scatter plot tell you about its expression in control vs TAZ KO cells?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69fbb5e8-3244-4c33-a930-c426134c6bde",
   "metadata": {},
   "source": [
    "<h2>Exercise #3: Visualize potentially downregulated and upregulated genes</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac89d785-7db4-4510-b5f6-69343471214a",
   "metadata": {},
   "source": [
    "With your single replicate, you can take a quick look to see if there may be differential expression between your sample groups. For this exercise, use the same scatter plot set up you used in exercise 2, but highlight those genes who have a 2-fold difference in their read counts in your TAZ KO sample compared to your control sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eefe071f-939a-47c3-907f-28a547653139",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5bc06db1-5ef1-415c-90a3-1bfa7f9ecba7",
   "metadata": {},
   "source": [
    "Something to note is that these exercise sets are just from a single replicate based off of non-normalized counts, so we can't draw hard conclusions from this visualization, but it can help us get a sense of how our counts went.\n",
    "\n",
    "Next week, we'll pull together our class counts and run a differential expression analysis to explore our results."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90c3408e-1014-4834-8d3e-8ecde32f02ad",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">References</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af046ee0-b260-486a-a5b8-0b88b4b7aee0",
   "metadata": {},
   "source": [
    "<a href=\"https://academic.oup.com/bioinformatics/article/31/2/166/2366196\" rel=\"noopener noreferrer\"><u>Anders et al 2015 Bioinformatics:</u></a> HTSeqâ€”a Python framework to work with high-throughput sequencing data\n",
    "\n",
    "<a href=\"https://academic.oup.com/bioinformatics/article/38/10/2943/6551247\" rel=\"noopener noreferrer\"><u>Putri et al 2022 Bioinformatics:</u></a> Analysing high-throughput sequencing data in Python with HTSeq 2.0"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
