{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cec9421b-d854-42ae-b07e-c0918acc4468",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">11.1 Prepare matrices for DESeq2</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c03d28c-d1cb-49f2-a3d7-878404e7ceec",
   "metadata": {},
   "source": [
    "We'll now prepare our class datasets for us to dive more deeply into our RNA-seq data by first merging each group's count matrices into a single data matrix and creating a conditions matrix that will be the metadata for our class dataset. We'll incorporate what we've covered in notebook 10-2 and this notebook into a Python script that can then be integrated into our RNA-seq analysis pipeline.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "\n",
    "<ul>\n",
    "    <li>Review counts quality control with \"bad\" counts set up</li>\n",
    "    <li>Practice setting up Python scripts</li>\n",
    "    <li>Integrating Python script into your RNA-seq pipeline</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fce055f6-cc5c-4856-9166-72a5882a6a15",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Prepare a set of \"bad\" counts</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7e5cd3f-af0f-46ef-a653-b8b811fa2601",
   "metadata": {},
   "source": [
    "To have an example of when you might notice if something went wrong in how you set up <code>htseq-count</code>, we'll be starting today's lesson by running <code>htseq-count</code> using the incorrect strandedness for our library, which should lead to an unusual count matrix where a large number of reads will be unassigned/not counted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54cdee44-81c5-4cd6-a376-320a77eae32f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ec915e-97d7-496c-af2d-aa7de01c5b3e",
   "metadata": {},
   "source": [
    "We'll then also separately analyze this bad count file to see how the output looks like when we mix up the strandness of our RNA-seq dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1575a33e-5e38-40c8-89b4-02b5450a4e73",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise #1: Work out Python script to QC counts</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5581ad13-0278-4412-9818-ae354d1d26dc",
   "metadata": {},
   "source": [
    "For this first exercise, you'll take what you know about setting up Python scripts and the exercise sets from notebook 10-2 to set up a Python script in the code cell below. For this QC, you'll want to be able to do the following:\n",
    "\n",
    "<ul>\n",
    "    <li>Import any needed packages</li>\n",
    "    <li>Confirm that you're in <code>Week_10</code> directory (to stay consistent with rest of RNA-seq pipeline)</li>\n",
    "    <li>Make a <code>counts-qc</code> directory</li>\n",
    "    <li>Generate a stacked bar plot of count statistics</li>\n",
    "    <li>Create a scatter plot of ctrl vs tazko counts and highlight potential upregulation and downregulation</li>\n",
    "    <ul>\n",
    "        <li>Export the plot as a PDF to <code>counts-qc</code></li>\n",
    "    </ul>\n",
    "    <li>Remove uncounted read statistics from counts file and export as a new <code>1M_*_gene_counts.csv</code></li>\n",
    "    <ul>\n",
    "        <li>The asterisk should be replaced with your group number in the format <code>g1</code></li>\n",
    "        <li>Export the <code>.csv</code> to <code>counts</code> and keep the headers</li>\n",
    "    </ul>\n",
    "</ul>\n",
    "\n",
    "You can use your \"bad\" counts file to test out your code as you're working on it, as we'll also take a look at that output on the side."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b18693fd-9ce0-4619-8499-8b0fb6f0cc98",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Space for working on your Python script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbb6475d-1897-4e04-a9d1-02c5205ac5d5",
   "metadata": {},
   "source": [
    "Once you've confirmed that your script works, we'll reconvene and transfer that code into a new Python file called <code>counts-qc-script.py</code> in our <code>Week_10</code> directory for convenience.\n",
    "\n",
    "With our Python script ready, we'll then integrate it into our RNA-seq script that we've been putting together and give it a test run on our truncated dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a50b4882-4000-4d5e-bddb-2d689afce7cc",
   "metadata": {},
   "source": [
    "<h2>Planned break</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bddd534-5375-4443-8bcc-95070bea8b0f",
   "metadata": {},
   "source": [
    "During this break, email Jack your group's <code>.csv</code> file for your group's read counts. You don't need to send your \"bad\" one."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a27126be-9dee-4bab-88d1-c3675bdc578d",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise #2: Work out Python script to prepare matrices for DESeq2</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ae85075-7928-4a0a-a842-f29047bdc2d5",
   "metadata": {},
   "source": [
    "For this exercise, you'll create another Python script called <code>class-merger.py</code> also in your <code>Week_10</code> directory for simplicity. The goal of this script will be to pull the separate class matrix files from our shared directory and merge the data into a single matrix. We'll also generate a metadata file that contains the information on the experimental setup/condition for each sample.\n",
    "\n",
    "Normally, if you have multiple replicates for <code>htseq-count</code>, you can supply all your alignments at once, and each alignment will have its own column in the counts matrix that <code>htseq-count</code> outputs. Since we each analyzed our own replicates, we'll need to merge them all together into a single matrix. We'll also in the same script create a metadata file that goes with it so that DESeq2 \"knows\" what condition each column belongs to.\n",
    "\n",
    "Then, you will export both matrices as <code>.csv</code> files in a subdirectory called <code>class-set</code> within your <code>counts</code> directory."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b940529-0d2f-426b-a5f3-e2eb57c15538",
   "metadata": {},
   "source": [
    "<h2>Class counts matrix</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2919da57-67ba-45d1-98b5-038049385ba3",
   "metadata": {},
   "source": [
    "Using our Python script, we'll want to merge all our data into a matrix that resembles:\n",
    "\n",
    "<table style=\"text-align:center;\">\n",
    "    <tr>\n",
    "        <th>Gene</th>\n",
    "        <th>ctrl_g1</th>\n",
    "        <th>tazko_g1</th>\n",
    "        <th>ctrl_g2</th>\n",
    "        <th>tazko_g2</th>\n",
    "        <th>...</th>\n",
    "        <th>ctrl_g9</th>\n",
    "        <th>tazko_g9</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>gene name</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>...</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>gene name</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>...</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>gene name</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>...</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "    </tr>    \n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85a9f1fa-1318-4022-82a7-88e54b061fc6",
   "metadata": {},
   "source": [
    "<h2>Conditions matrix</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d7343ed-1c8b-4a97-8471-95b6af4b2255",
   "metadata": {},
   "source": [
    "While we know what condition each column's samples come from, DESeq2 does not, so we need to provide it with information on which condition each sample came from. That way, DESeq2 can assign them to the correct condition (either control or tazko) and make the correct comparisons between conditions.\n",
    "\n",
    "In the same Python script that we use to merge our class data, we'll also simultaneously create a conditions matrix that acts as the metadata for each of our samples, and it will look something like:\n",
    "\n",
    "<table style=\"text-align:center;\">\n",
    "    <tr>\n",
    "        <th>[Index]</th>\n",
    "        <th>condition</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>ctrl_g1</th>\n",
    "        <td>control</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>tazko_g1</th>\n",
    "        <td>tazko</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>control_g2</th>\n",
    "        <td>control</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>tazko_g2</th>\n",
    "        <td>tazko</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>...</th>\n",
    "        <td>...</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>control_g9</th>\n",
    "        <td>control</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>tazko_g9</th>\n",
    "        <td>tazko</td>\n",
    "    </tr>\n",
    "</table>\n",
    "\n",
    "The index values corresponds to the column header of your counts matrix for each sample while the values in the <code>condition</code> column indicates whether that sample is a control sample or a TAZ KO sample, allowing DESeq2 to properly group the data for differential expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce2d7729-cd0f-473c-b2a9-9547e4a1c281",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Space for working on your Python script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6660dabb-a142-4ca5-8332-a6bf70f540f6",
   "metadata": {},
   "source": [
    "With our second Python script ready, let's also integrate it into our RNA-seq script and give it another test run using our truncated dataset."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
