{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cec9421b-d854-42ae-b07e-c0918acc4468",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">11.1 Prepare matrices for DESeq2</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c03d28c-d1cb-49f2-a3d7-878404e7ceec",
   "metadata": {},
   "source": [
    "We'll now prepare our class datasets for us to dive more deeply into our RNA-seq data by first merging each group's count matrices into a single data matrix and creating a conditions matrix that will be the metadata for our class dataset. We'll incorporate what we've covered in notebook 10-2 and this notebook into a Python script that can then be integrated into our RNA-seq analysis pipeline.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "\n",
    "<ul>\n",
    "    <li>Review counts quality control with \"bad\" counts set up</li>\n",
    "    <li>Practice setting up Python scripts</li>\n",
    "    <li>Integrating Python script into your RNA-seq pipeline</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fce055f6-cc5c-4856-9166-72a5882a6a15",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Prepare a set of \"bad\" counts</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7e5cd3f-af0f-46ef-a653-b8b811fa2601",
   "metadata": {},
   "source": [
    "To have an example of when you might notice if something went wrong in how you set up <code>htseq-count</code>, we'll be starting today's lesson by running <code>htseq-count</code> using the incorrect strandedness for our library, which should lead to an unusual count matrix where a large number of reads will be unassigned/not counted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54cdee44-81c5-4cd6-a376-320a77eae32f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#############################\n",
    "#\n",
    "# Just for practice purposes\n",
    "# And to highlight QC\n",
    "#\n",
    "#############################\n",
    "\n",
    "#We'll just make it in our Week_10 directory\n",
    "\n",
    "htseq-count \\\n",
    "-t exon \\\n",
    "-i gene_id \\\n",
    "-r name \\\n",
    "-s yes \\\n",
    "-f bam \\\n",
    "./alignments-temp/bams/1M_g1_ctrl-name.bam \\\n",
    "./alignments-temp/bams/1M_g1_tazko-name.bam \\\n",
    "~/shared/2025-fall/courses/1547808/rna-seq/rna-feature/hg19-refseq.gtf \\\n",
    "> ~/MCB201B_F2025/Week_10/1M_g1_bad_counts.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ec915e-97d7-496c-af2d-aa7de01c5b3e",
   "metadata": {},
   "source": [
    "We'll then also separately analyze this bad count file to see how the output looks like when we mix up the strandness of our RNA-seq dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1575a33e-5e38-40c8-89b4-02b5450a4e73",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise #1: Work out Python script to QC counts</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5581ad13-0278-4412-9818-ae354d1d26dc",
   "metadata": {},
   "source": [
    "For this first exercise, you'll take what you know about setting up Python scripts and the exercise sets from notebook 10-2 to set up a Python script in the code cell below. For this QC, you'll want to be able to do the following:\n",
    "\n",
    "<ul>\n",
    "    <li>Import any needed packages</li>\n",
    "    <li>Confirm that you're in <code>Week_10</code> directory (to stay consistent with rest of RNA-seq pipeline)</li>\n",
    "    <li>Make a <code>counts-qc</code> directory</li>\n",
    "    <li>Generate a stacked bar plot of count statistics</li>\n",
    "    <li>Create a scatter plot of ctrl vs tazko counts and highlight potential upregulation and downregulation</li>\n",
    "    <ul>\n",
    "        <li>Export the plot as a PDF to <code>counts-qc</code></li>\n",
    "    </ul>\n",
    "    <li>Remove uncounted read statistics from counts file and export as a new <code>1M_*_gene_counts.csv</code></li>\n",
    "    <ul>\n",
    "        <li>The asterisk should be replaced with your group number in the format <code>g1</code></li>\n",
    "        <li>Export the <code>.csv</code> to <code>counts</code> and keep the headers</li>\n",
    "    </ul>\n",
    "</ul>\n",
    "\n",
    "You can use your \"bad\" counts file to test out your code as you're working on it, as we'll also take a look at that output on the side."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b18693fd-9ce0-4619-8499-8b0fb6f0cc98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confirming our directory...\n",
      "\n",
      "Current directory is /home/jovyan/MCB201B_F2025/Week_11\n",
      "Changed to confirm that the directory is /home/jovyan/MCB201B_F2025/Week_10\n",
      "Pulled in the file: 1M_g1_counts.txt\n",
      "Setting base name as: 1M_g1\n",
      "Saving a .csv file for 1M_g1...\n",
      "Generating our stacked bar plot for QC...\n",
      "\n",
      "Saving PDF of stacked bar plot for 1M_g1 in /home/jovyan/MCB201B_F2025/Week_10/counts-qc...\n",
      "\n",
      "Saving PDF of scatter plot for 1M_g1 in /home/jovyan/MCB201B_F2025/Week_10/counts-qc...\n",
      "\n",
      "Done with counts QC!\n"
     ]
    }
   ],
   "source": [
    "#Space for working on your Python script\n",
    "\n",
    "#Import any needed packages\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "import seaborn as sns\n",
    "import os\n",
    "\n",
    "#Confirm our directory\n",
    "print('Confirming our directory...\\n')\n",
    "print(f'Current directory is {os.getcwd()}')\n",
    "\n",
    "os.chdir('/home/jovyan/MCB201B_F2025/Week_10')\n",
    "print(f'Changed to confirm that the directory is {os.getcwd()}')\n",
    "\n",
    "#Make our new counts-qc directory\n",
    "try:\n",
    "    os.mkdir('counts-qc')\n",
    "except FileExistsError:\n",
    "    pass\n",
    "\n",
    "#Pull in our file names\n",
    "os.chdir('./counts')\n",
    "file_name = [name for name in os.listdir() if '.txt' in name]\n",
    "print(f'Pulled in the file: {file_name[0]}')\n",
    "\n",
    "#Set up your base name\n",
    "base_name = file_name[0].split('_counts')[0]\n",
    "print(f'Setting base name as: {base_name}')\n",
    "\n",
    "#Load in our file\n",
    "counts = pd.read_csv(file_name[0],\n",
    "                     sep='\\t',\n",
    "                     header=None,\n",
    "                     names=['gene', 'ctrl', 'tazko']\n",
    "                    )\n",
    "\n",
    "#Pull out the read count statistics\n",
    "read_stats = counts[counts['gene'].str.contains('__')].copy()\n",
    "read_stats['gene'] = read_stats['gene'].str.replace('__', '')\n",
    "\n",
    "#Pull out how many genes were counted and set to our read_stats\n",
    "counted_reads = counts[~counts['gene'].str.contains('__')].copy()\n",
    "\n",
    "print(f'Saving a .csv file for {base_name}...')\n",
    "counted_reads.to_csv(f'{base_name}_gene_counts.csv',\n",
    "                     index=False,\n",
    "                    )\n",
    "\n",
    "#Update DataFrame for QC\n",
    "read_stats.loc[len(read_stats)] = ['counted_reads', sum(counted_reads['ctrl']), sum(counted_reads['tazko'])]\n",
    "\n",
    "#Update our DataFrame for plotting\n",
    "read_stats.set_index('gene', inplace=True)\n",
    "read_stats = read_stats.div(read_stats.sum(axis=0), axis=1) * 100\n",
    "read_stats = read_stats.T\n",
    "\n",
    "#Set up our stacked bar plot\n",
    "#See 10-2 notebook\n",
    "#We are just pulling and updating from that code\n",
    "print('Generating our stacked bar plot for QC...\\n')\n",
    "\n",
    "#Set up plot in the usual way\n",
    "#From first exercise set in 10-2, setting up stacked bar plot\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "x = read_stats.index\n",
    "y = read_stats.columns[::-1]\n",
    "\n",
    "bottom = np.zeros(len(x))\n",
    "\n",
    "use_colors = pd.Series(mcolors.CSS4_COLORS)\n",
    "\n",
    "for i in range(0, len(y), 1):\n",
    "    plt.bar(x,\n",
    "            read_stats[y[i]],\n",
    "            label=y[i],\n",
    "            bottom=bottom,\n",
    "            color=use_colors[125+i*4],\n",
    "            lw=0.5,\n",
    "            edgecolor='k',\n",
    "           )\n",
    "    bottom+=read_stats[y[i]]\n",
    "\n",
    "plt.title('Counts QC')\n",
    "plt.ylabel('Percent of total read counts')\n",
    "plt.xticks([0,1],\n",
    "           ['Control', '$TAZ$ KO']\n",
    "          )\n",
    "plt.legend(loc='center',\n",
    "           bbox_to_anchor=(1.3,0.5),\n",
    "           fontsize=4,\n",
    "           edgecolor='w',\n",
    "          )\n",
    "\n",
    "sns.despine()\n",
    "fig.set_size_inches(2, 3)\n",
    "fig.set_dpi(300)\n",
    "\n",
    "#Here we then save our stacked bar plot\n",
    "os.chdir('/home/jovyan/MCB201B_F2025/Week_10/counts-qc')\n",
    "print(f'Saving PDF of stacked bar plot for {base_name} in {os.getcwd()}...\\n')\n",
    "fig.savefig(f'{base_name}_stacked_qc_barplot.pdf', bbox_inches='tight')\n",
    "\n",
    "#Close out the figure\n",
    "plt.close(fig)\n",
    "\n",
    "#Now we set up our scatter plot\n",
    "counted_reads['ratio'] = counted_reads['tazko'] / counted_reads['ctrl']\n",
    "downreg = counted_reads[counted_reads['ratio'] <= 0.5]\n",
    "upreg = counted_reads[counted_reads['ratio'] >= 2]\n",
    "\n",
    "#Create a new figure but with the same variables\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "#We can set up lists to iterate through for our scatterplot\n",
    "datasets = [counted_reads, downreg, upreg]\n",
    "colors_to_use = ['grey', 'b', 'r']\n",
    "labels_to_use = ['no change', 'downregulated', 'upregulated']\n",
    "\n",
    "#Set up for loop to loop through all three lists\n",
    "for i in range(0, len(datasets), 1):\n",
    "    sns.scatterplot(data=datasets[i],\n",
    "                    x='ctrl',\n",
    "                    y='tazko',\n",
    "                    color=colors_to_use[i],\n",
    "                    label=labels_to_use[i],\n",
    "                    s=4,\n",
    "                   )\n",
    "\n",
    "#Pretty up scatter plot\n",
    "plt.loglog()\n",
    "plt.xlabel('Control read counts')\n",
    "plt.ylabel('$TAZ$ KO read counts')\n",
    "\n",
    "plt.legend(loc='center',\n",
    "           bbox_to_anchor=(1.1,0.5),\n",
    "           fontsize=6,\n",
    "           edgecolor='w',\n",
    "          )\n",
    "\n",
    "fig.set_size_inches(3, 3)\n",
    "fig.set_dpi(300)\n",
    "sns.despine()\n",
    "\n",
    "#Save scatterplot\n",
    "print(f'Saving PDF of scatter plot for {base_name} in {os.getcwd()}...\\n')\n",
    "fig.savefig(f'{base_name}_qc_scatterplot.pdf', bbox_inches='tight')\n",
    "\n",
    "#Close out figure\n",
    "plt.close(fig)\n",
    "\n",
    "print('Done with counts QC!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbb6475d-1897-4e04-a9d1-02c5205ac5d5",
   "metadata": {},
   "source": [
    "Once you've confirmed that your script works, we'll reconvene and transfer that code into a new Python file called <code>counts-qc-script.py</code> in our <code>Week_10</code> directory for convenience.\n",
    "\n",
    "With our Python script ready, we'll then integrate it into our RNA-seq script that we've been putting together and give it a test run on our truncated dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a50b4882-4000-4d5e-bddb-2d689afce7cc",
   "metadata": {},
   "source": [
    "<h2>Planned break</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bddd534-5375-4443-8bcc-95070bea8b0f",
   "metadata": {},
   "source": [
    "During this break, email Jack your group's <code>.csv</code> file for your group's read counts. You don't need to send your \"bad\" one."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a27126be-9dee-4bab-88d1-c3675bdc578d",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise #2: Work out Python script to prepare matrices for DESeq2</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ae85075-7928-4a0a-a842-f29047bdc2d5",
   "metadata": {},
   "source": [
    "For this exercise, you'll create another Python script called <code>class-merger.py</code> also in your <code>Week_10</code> directory for simplicity. The goal of this script will be to pull the separate class matrix files from our shared directory and merge the data into a single matrix. We'll also generate a metadata file that contains the information on the experimental setup/condition for each sample.\n",
    "\n",
    "Normally, if you have multiple replicates for <code>htseq-count</code>, you can supply all your alignments at once, and each alignment will have its own column in the counts matrix that <code>htseq-count</code> outputs. Since we each analyzed our own replicates, we'll need to merge them all together into a single matrix. We'll also in the same script create a metadata file that goes with it so that DESeq2 \"knows\" what condition each column belongs to.\n",
    "\n",
    "Then, you will export both matrices as <code>.csv</code> files in a subdirectory called <code>class-set</code> within your <code>counts</code> directory."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b940529-0d2f-426b-a5f3-e2eb57c15538",
   "metadata": {},
   "source": [
    "<h2>Class counts matrix</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2919da57-67ba-45d1-98b5-038049385ba3",
   "metadata": {},
   "source": [
    "Using our Python script, we'll want to merge all our data into a matrix that resembles:\n",
    "\n",
    "<table style=\"text-align:center;\">\n",
    "    <tr>\n",
    "        <th>Gene</th>\n",
    "        <th>ctrl_g1</th>\n",
    "        <th>tazko_g1</th>\n",
    "        <th>ctrl_g2</th>\n",
    "        <th>tazko_g2</th>\n",
    "        <th>...</th>\n",
    "        <th>ctrl_g9</th>\n",
    "        <th>tazko_g9</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>gene name</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>...</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>gene name</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>...</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>gene name</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "        <td>...</td>\n",
    "        <td>count data</td>\n",
    "        <td>count data</td>\n",
    "    </tr>    \n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85a9f1fa-1318-4022-82a7-88e54b061fc6",
   "metadata": {},
   "source": [
    "<h2>Conditions matrix</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d7343ed-1c8b-4a97-8471-95b6af4b2255",
   "metadata": {},
   "source": [
    "While we know what condition each column's samples come from, DESeq2 does not, so we need to provide it with information on which condition each sample came from. That way, DESeq2 can assign them to the correct condition (either control or tazko) and make the correct comparisons between conditions.\n",
    "\n",
    "In the same Python script that we use to merge our class data, we'll also simultaneously create a conditions matrix that acts as the metadata for each of our samples, and it will look something like:\n",
    "\n",
    "<table style=\"text-align:center;\">\n",
    "    <tr>\n",
    "        <th>[Index]</th>\n",
    "        <th>condition</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>ctrl_g1</th>\n",
    "        <td>control</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>tazko_g1</th>\n",
    "        <td>tazko</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>control_g2</th>\n",
    "        <td>control</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>tazko_g2</th>\n",
    "        <td>tazko</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>...</th>\n",
    "        <td>...</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>control_g9</th>\n",
    "        <td>control</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <th>tazko_g9</th>\n",
    "        <td>tazko</td>\n",
    "    </tr>\n",
    "</table>\n",
    "\n",
    "The index values corresponds to the column header of your counts matrix for each sample while the values in the <code>condition</code> column indicates whether that sample is a control sample or a TAZ KO sample, allowing DESeq2 to properly group the data for differential expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ce2d7729-cd0f-473c-b2a9-9547e4a1c281",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting our current directory...\n",
      "\n",
      "Current directory is /home/jovyan/shared/2025-fall/courses/1547808/rna-seq/truncated-counts\n",
      "Changing into our counts file for good measure...\n",
      "\n",
      "Current directory is changed to /home/jovyan/shared/2025-fall/courses/1547808/rna-seq/truncated-counts\n",
      "Pulling in our files...\n",
      "\n",
      "Files that we are merging are in the following order: \n",
      "\n",
      "1M_g1_gene_counts.csv\n",
      "\n",
      "1M_g2_gene_counts.csv\n",
      "\n",
      "1M_g3_gene_counts.csv\n",
      "\n",
      "1M_g4_gene_counts.csv\n",
      "\n",
      "1M_g5_gene_counts.csv\n",
      "\n",
      "1M_g6_gene_counts.csv\n",
      "\n",
      "1M_g7_gene_counts.csv\n",
      "\n",
      "1M_g8_gene_counts.csv\n",
      "\n",
      "1M_g9_gene_counts.csv\n",
      "\n",
      "Going to create our lovely matrices for DESeq2...\n",
      "\n",
      "Saving our class counts matrix and our conditions matrix as .csv files...\n",
      "\n",
      "All done!\n"
     ]
    }
   ],
   "source": [
    "#Space for working on your Python script\n",
    "\n",
    "####################################################################################\n",
    "#\n",
    "# Picked up here Wed 11/12\n",
    "#\n",
    "# This is set up due to the nature of how we analyzed our data as a class\n",
    "# Note that for normal analyses, you can feed all BAMs to htseq at once\n",
    "# For simplicity, we just worked in this notebook without adding to our bash script\n",
    "#\n",
    "####################################################################################\n",
    "\n",
    "#Import our needed packages\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "#Confirm where our script is currently sitting\n",
    "print('Getting our current directory...\\n')\n",
    "print(f'Current directory is {os.getcwd()}')\n",
    "\n",
    "#Change to our usual directory\n",
    "print('Changing into our counts file for good measure...\\n')\n",
    "\n",
    "#Everyone submitted their group's .csv file and Jack uploaded to shared directory\n",
    "os.chdir('/home/jovyan/shared/2025-fall/courses/1547808/rna-seq/truncated-counts')\n",
    "print(f'Current directory is changed to {os.getcwd()}')\n",
    "\n",
    "#Let's pull in our data files to merge together\n",
    "#We use our usual way of pulling names into a list via list comprehension\n",
    "print('Pulling in our files...\\n')\n",
    "data = [name for name in os.listdir() if '.csv' in name]\n",
    "data.sort()\n",
    "\n",
    "#Set up a little for loop to then output file names just for checking\n",
    "print(f'Files that we are merging are in the following order: \\n')\n",
    "for name in data:\n",
    "    print(name, end='\\n\\n')\n",
    "\n",
    "print('Going to create our lovely matrices for DESeq2...\\n')\n",
    "\n",
    "#When pulling objects into a list using list comprehension we can operate on that object\n",
    "#In this case, we're pulling in name and using it to load in the data with pd.read_csv()\n",
    "#This just easily loads in our data as a DataFrame and populates it into a list that we can loop through\n",
    "full_set = [pd.read_csv(name, index_col='gene') for name in data]\n",
    "\n",
    "#One thing to note is that we set the index values based on the 'gene' column using index_col='gene'\n",
    "#So the only remaining columns are 'ctrl' and 'tazko', and the headers are all the same for all replicates\n",
    "#So then we can prepare our DataFrames for joining/merging by first updating all the headers\n",
    "#Essentially just going to replace/tack on additional group/replicate info\n",
    "\n",
    "#Initialize for loop to iterate through our list of DataFrames\n",
    "for i in range(0, len(full_set), 1):\n",
    "    #This then creates a list adaptable to the group number using the position (i) + 1 b/c we don't have a group 0\n",
    "    #Then the elements of that list replaces each replicate's column headers\n",
    "    #So for group 1, i=0 and thus ctrl -> ctrl_g1 and tazko -> tazko_g1\n",
    "    full_set[i].columns = [f'ctrl_g{i+1}', f'tazko_g{i+1}']\n",
    "\n",
    "#Now we use pd.DataFrame.join() to join\n",
    "#If you are a little confused, check the documentation for the join function\n",
    "#It allows us to specify a left-side for our join, which will be our first DataFrame (group 1's)\n",
    "#Then we can give it a list of DataFrames to join to that first one (groups 2 to 9)\n",
    "#So we can pull out first element full_set[0]\n",
    "#And the remaining replicates are encompassed by fulll_set[1:]\n",
    "#Then we can join based on the left (group 1) index values which are gene names\n",
    "#Should be shared for all replicates\n",
    "class_data = full_set[0].join(full_set[1:], how='left')\n",
    "#You'll then have a mega-matrix for the class data\n",
    "\n",
    "#Refer back to text above for how \"metadata\" matrix should look\n",
    "#We create a list with the different sample conditions - either control or tazko\n",
    "#And since we set up our mega-matrix ordered as ctrl and tazko repeated for each replicate for columns\n",
    "#We can just multiply by how many replicates we have to repeat our list elements\n",
    "conditions = ['control', 'tazko']*len(data)\n",
    "\n",
    "#Then create the DataFrame according to the text above\n",
    "#The index will be based on column headers because we need this information for DESeq2\n",
    "#It does not actually \"know\" what ctrl_g1 is and that it is a control condition\n",
    "#So we have to tell it with this \"metadata\" matrix\n",
    "metadata = pd.DataFrame(conditions, index=class_data.columns, columns=['condition'])\n",
    "\n",
    "\n",
    "#Export using our usual way\n",
    "print('Saving our class counts matrix and our conditions matrix as .csv files...\\n')\n",
    "os.chdir('/home/jovyan/MCB201B_F2025/Week_10')\n",
    "try:\n",
    "    os.mkdir('class-set')\n",
    "except FileExistsError:\n",
    "    pass\n",
    "\n",
    "os.chdir('./class-set')\n",
    "\n",
    "\n",
    "#We will proceed with our R notebook and DESeq2 + clustering analyses using these two files\n",
    "class_data.to_csv('1M_class_counts_matrix.csv')\n",
    "metadata.to_csv('1M_class_conditions_matrix.csv')\n",
    "\n",
    "print('All done!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6660dabb-a142-4ca5-8332-a6bf70f540f6",
   "metadata": {},
   "source": [
    "With our second Python script ready, let's also integrate it into our RNA-seq script and give it another test run using our truncated dataset."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
