{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6f424945-4d81-4bb3-b1f2-b2d4a4ac21f6",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Week 8 Assignment</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b80eed2e-4c43-48ab-b8f3-9fcd445345a8",
   "metadata": {},
   "source": [
    "<strong><u>Due: Saturday, October 25 at 11:59PM</u></strong>\n",
    "\n",
    "<strong><u>Total points: 5 pts</u></strong>\n",
    "\n",
    "<strong><u>To turn in your assignment:</u></strong>\n",
    "<ol>\n",
    "    <li>Complete assignment and make sure you have all the outputs requested for the notebook</li>\n",
    "    <li>Go to your server's file directory and locate the <code>week_8_homework</code> directory containing all your outputs from this assignment including this notebook</li>\n",
    "    <li>Right click and select <strong>Download as an Archive</strong> to download the <code>week_8_homework</code> directory as a ZIP folder. Do not mix up with <strong>Download Current Folder as an Archive</strong></li>\n",
    "    <li>Upload your ZIP folder to your individual feedback Google Drive folder under <code>MCB201B</code></li>\n",
    "</ol>\n",
    "\n",
    "<strong><u>How this assignment will be graded:</u></strong>\n",
    "\n",
    "You will be graded on whether or not you have <u>all requested outputs and files</u> and if your output is correct. You will also be graded on if your code is expected to be able to run without errors. For those of you running code in Terminal or on your own computer, you will need to write out your input(s) into the empty code cell for this notebook. You don't need to run them here. If you used line breaks, make sure not to transfer over the <code>&gt;</code> symbol at the start of each new line. There can be more than one correct way to answer a question. If you find that you prefer to separate your answer for a single question into multiple code cells, feel free to do so.\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71b42eb1-9f81-4124-92ff-dbd9ac35c0c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb75922d-9036-4ec6-abb3-faad9c99cf33",
   "metadata": {},
   "source": [
    "<h2>Question 1</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b4be939-5f87-43ad-95f1-7fb603ab7984",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a32d1c7-8f84-4d37-ab65-d94a175748e8",
   "metadata": {},
   "source": [
    "This question is a housekeeping question. Now that we're done with our alignment files, we'll clear them from our servers so that we can help keep storage usage minimal. First run the bash command below using this notebook to output the size of your server, excluding the shared directories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d61665ce-815e-43c0-9a70-0bcd878cff6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd ~\n",
    "\n",
    "du -sh --exclude=./shared --exclude=./_shared ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d82112b7-6944-42a3-ac52-59a691d16e99",
   "metadata": {},
   "source": [
    "Go ahead and delete any SAM and BAM alignment files associated with our ChIP-seq module. We have the <code>macs2</code> outputs now, so we won't need to work with the SAM and BAM files. Once you've deleted those SAM and BAM files, run the code cell below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5cb203e-2f59-4a4b-873d-b56c5aa6774e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd ~\n",
    "\n",
    "du -sh --exclude=./shared --exclude=./_shared --exclude=./homer ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dec30193-ab0b-46ee-a92f-ba1817aaa034",
   "metadata": {},
   "source": [
    "And now run the two code cells below to see if you have any remaining SAMs and BAMs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15c994f7-b8e9-4c26-adc1-b0b53184626e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd ~\n",
    "\n",
    "find . -type f -name \"*.sam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0662ae1-8cbd-4b31-a056-c5748b9cbac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd ~\n",
    "\n",
    "find . -type f -name \"*.bam\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a6fca97-e0f5-471b-b43f-9b8e41563a61",
   "metadata": {},
   "source": [
    "If you have remaining BAMs and SAMs, go ahead and delete them to save space. You can also download them from the Biology Hub if you'd like to keep them as a backup before deleting them from your server. Now do one final check by running the code cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f691c4d4-6866-4139-8329-eab42f815b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd ~\n",
    "\n",
    "find . -type f -name \"*.sam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe44388f-669e-4943-afc7-17cff808645e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd ~\n",
    "\n",
    "find . -type f -name \"*.bam\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1af34334-06e3-49a6-aead-bed8f182e739",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3b6300e-7200-4618-8ef9-ac7de878cc97",
   "metadata": {},
   "source": [
    "<h2>Question 2</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e22ea3f-a1cd-41b4-8612-6cbc44f7af3d",
   "metadata": {},
   "source": [
    "2pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18c883a5-8e59-436b-8305-a143ecd7a148",
   "metadata": {},
   "source": [
    "Using your called peaks from the full ChIP-seq dataset, plot the peak profiles of both replicates of TAZ with both peak profiles on the same plot. Add x-axis and y-axis labels as appropriate and a legend. Output your plot as a PDF file to be included in your assignment submission.\n",
    "\n",
    "Include below the code you used to obtain the data from HOMER."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "758ec43b-06a6-4166-8413-ad032899ebe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Copy over your command line inputs into this code cell. You don't need to run it here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03f45e11-a69b-4ebf-b5b9-5b798620150d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#You can add additional code cells as needed for your Python code to plot the peak profile."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ab6a8e1-0fe8-4653-af90-c7ae0275ae7c",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "241cd165-0308-44b1-9072-2255e200ebeb",
   "metadata": {},
   "source": [
    "<h2>Question 3</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d788ace-463e-44d2-abbf-d8c4f73992ab",
   "metadata": {},
   "source": [
    "2pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba9d5e92-ae2a-403c-85f1-b54165ec1341",
   "metadata": {},
   "source": [
    "Using your called peaks from the full ChIP-seq dataset, determine each peak's distance to the nearest TSS for both replicates of TAZ. Then plot a histogram with the following bins:\n",
    "\n",
    "<ul>\n",
    "    <li>0-1kb</li>\n",
    "    <li>1-10kb</li>\n",
    "    <li>10-100kb</li>\n",
    "    <li>100+ kb</li>\n",
    "</ul>\n",
    "\n",
    "Adjust opacity (alpha) as needed to be able to see the histograms of both replicates on the same plot. Output your plot as a PDF file to be included in your assignment submission."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40fca373-7853-4c46-a966-81697149454c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Copy over your command line inputs into this code cell. You don't need to run it here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68a10951-3c3b-4cda-9bea-e2bc94031b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "#You can add additional code cells as needed for your Python code to plot the histograms."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c38d728a-7107-4724-82aa-c1545cdb0d63",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%px;\"></hr>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
