{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f3e636a9-4d7a-4ff2-adc7-02315b8cdcd3",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">15.1 Multiomics and functional analysis</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5daeaa72-9b01-4518-b820-1bb130bf3e49",
   "metadata": {},
   "source": [
    "In notebook 13-1, we explored our transcriptome data at the level of each gene, and today, we'll take a look at our data with respect to groups of genes through a functional analysis of our differential expression results. While looking at the changes in expression of individual genes can provide insight into the gene-level impacts of experimental manipulations, we can also look for groups of genes that change together through functional analysis. This type of analysis can give us insight into broader biological changes that are occurring as a result of experimental manipulations, such as knocking down TAZ. So rather than asking about what is happening to a specific gene, we are instead asking about what biological processes, signaling pathways, or other functional groups are changing in our cells.\n",
    "\n",
    "Of particular interest to us is also how our transcriptome analysis intersects with our ChIP analysis. Recall that from our ChIP-seq analysis, we identified a set of reproducible peaks where TAZ-TEAD are likely to bind to DNA. Our analysis showed that TAZ-TEAD primarily bind to regions distant from the transcription start site (TSS) rather than the nearby promoter region, suggesting a longer range regulation of target gene expression through distant enhancers. \n",
    "\n",
    "In this notebook, we'll be taking a basic multiomics approach by integrating our two big data analyses together in order to identify overlapping sets of genes from our two analyses. This approach can give us insight into the regulation of TAZ target genes and potentially identify some direct targets of TAZ and determine how TAZ regulates their transcription. We'll then take a look to see what functional groups are directly affected by TAZ KO compared to the overall changes that we see when we look at the full transcriptomic profile.\n",
    "\n",
    "<strong>Learning objectives:</strong>\n",
    "<ul>\n",
    "    <li>Identify intersections between datasets</li>\n",
    "    <li>Explore databases for functional analysis</li>\n",
    "    <li>Review and perform an over-representation analysis</li>\n",
    "    <li>Review and perform gene set enrichment analysis (GSEA)</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f5d97e7-3af0-4a13-8642-6b92d14e7d48",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Load in packages</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed23be18-a132-4055-a651-f162f8a305a1",
   "metadata": {},
   "source": [
    "To visualize intersections, we'll make use of a package called <code>BioVenn</code>, which was developed by Tim Hulsen (<a href=\"https://journals.sagepub.com/doi/10.3233/DS-210032\" rel=\"noopener noreferrer\"><u>Hulsen 2021</u></a>). It allows us to readily pull together our datasets and visualize their intersections as a Venn diagram. Another common way to visualize Venn diagrams is <code>matplotlib-venn</code>, which allows for a bit more customization, but won't simultaneously identify intersections.\n",
    "\n",
    "<a href=\"https://pypi.org/project/BioVenn/\" rel=\"noopener noreferrer\"><u>Documentation for <code>BioVenn</code> can be found here.</u></a>\n",
    "\n",
    "<a href=\"https://pypi.org/project/matplotlib-venn/\" rel=\"noopener noreferrer\"><u>Documentation for <code>matplotlib-venn</code> can be found here.</u></a>\n",
    "\n",
    "We'll be performing both the over-representation analysis and GSEA using Python, and to do this, we'll make use of a package called <code>GSEApy</code>, which was developed by Zhuoqing Fang, Xinyuan Liu, and Gary Peltz (<a href=\"https://academic.oup.com/bioinformatics/article/39/1/btac757/6847088\" rel=\"noopener noreferrer\"><u>Fang et al 2022</u></a>). It also allows us to directly pull into our notebook, a bunch of different databases all containing their own gene sets, which we can then use to identify functional groups that may be impacted when we knocked down TAZ.\n",
    "\n",
    "<a href=\"https://gseapy.readthedocs.io/en/latest/index.html\" rel=\"noopener noreferrer\"><u>Documentation for <code>GSEApy</code> can be found here.</u></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8873f3e4-692b-4a8c-b676-eec5028ff0a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting BioVenn\n",
      "  Using cached BioVenn-1.1.3-py3-none-any.whl.metadata (5.3 kB)\n",
      "Collecting biomart>=0.9.2 (from BioVenn)\n",
      "  Using cached biomart-0.9.2-py3-none-any.whl.metadata (3.3 kB)\n",
      "Requirement already satisfied: matplotlib>=3.3.3 in /srv/conda/lib/python3.11/site-packages (from BioVenn) (3.10.1)\n",
      "Requirement already satisfied: numpy>=1.19.4 in /srv/conda/lib/python3.11/site-packages (from BioVenn) (1.23.5)\n",
      "Requirement already satisfied: requests>=2.2 in /srv/conda/lib/python3.11/site-packages (from biomart>=0.9.2->BioVenn) (2.32.3)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=3.3.3->BioVenn) (1.3.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=3.3.3->BioVenn) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=3.3.3->BioVenn) (4.60.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=3.3.3->BioVenn) (1.4.9)\n",
      "Requirement already satisfied: packaging>=20.0 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=3.3.3->BioVenn) (24.2)\n",
      "Requirement already satisfied: pillow>=8 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=3.3.3->BioVenn) (9.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=3.3.3->BioVenn) (3.2.5)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=3.3.3->BioVenn) (2.9.0.post0)\n",
      "Requirement already satisfied: six>=1.5 in /srv/conda/lib/python3.11/site-packages (from python-dateutil>=2.7->matplotlib>=3.3.3->BioVenn) (1.17.0)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /srv/conda/lib/python3.11/site-packages (from requests>=2.2->biomart>=0.9.2->BioVenn) (3.4.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /srv/conda/lib/python3.11/site-packages (from requests>=2.2->biomart>=0.9.2->BioVenn) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /srv/conda/lib/python3.11/site-packages (from requests>=2.2->biomart>=0.9.2->BioVenn) (2.3.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /srv/conda/lib/python3.11/site-packages (from requests>=2.2->biomart>=0.9.2->BioVenn) (2025.8.3)\n",
      "Using cached BioVenn-1.1.3-py3-none-any.whl (21 kB)\n",
      "Using cached biomart-0.9.2-py3-none-any.whl (12 kB)\n",
      "Installing collected packages: biomart, BioVenn\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2/2\u001b[0m [BioVenn]\n",
      "\u001b[1A\u001b[2KSuccessfully installed BioVenn-1.1.3 biomart-0.9.2\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install BioVenn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ec1b99c8-6aea-4639-8a8c-02e187aa2c0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting gseapy\n",
      "  Using cached gseapy-1.1.11-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (11 kB)\n",
      "Requirement already satisfied: numpy>=1.13.0 in /srv/conda/lib/python3.11/site-packages (from gseapy) (1.23.5)\n",
      "Requirement already satisfied: scipy in /srv/conda/lib/python3.11/site-packages (from gseapy) (1.10.1)\n",
      "Requirement already satisfied: pandas in /srv/conda/lib/python3.11/site-packages (from gseapy) (1.5.3)\n",
      "Requirement already satisfied: matplotlib>=2.2 in /srv/conda/lib/python3.11/site-packages (from gseapy) (3.10.1)\n",
      "Requirement already satisfied: requests in /srv/conda/lib/python3.11/site-packages (from gseapy) (2.32.3)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=2.2->gseapy) (1.3.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=2.2->gseapy) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=2.2->gseapy) (4.60.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=2.2->gseapy) (1.4.9)\n",
      "Requirement already satisfied: packaging>=20.0 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=2.2->gseapy) (24.2)\n",
      "Requirement already satisfied: pillow>=8 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=2.2->gseapy) (9.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=2.2->gseapy) (3.2.5)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /srv/conda/lib/python3.11/site-packages (from matplotlib>=2.2->gseapy) (2.9.0.post0)\n",
      "Requirement already satisfied: six>=1.5 in /srv/conda/lib/python3.11/site-packages (from python-dateutil>=2.7->matplotlib>=2.2->gseapy) (1.17.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /srv/conda/lib/python3.11/site-packages (from pandas->gseapy) (2025.2)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /srv/conda/lib/python3.11/site-packages (from requests->gseapy) (3.4.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /srv/conda/lib/python3.11/site-packages (from requests->gseapy) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /srv/conda/lib/python3.11/site-packages (from requests->gseapy) (2.3.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /srv/conda/lib/python3.11/site-packages (from requests->gseapy) (2025.8.3)\n",
      "Using cached gseapy-1.1.11-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (605 kB)\n",
      "Installing collected packages: gseapy\n",
      "Successfully installed gseapy-1.1.11\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install gseapy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "54273a6d-e7fb-4884-b7c9-a682f1e3394f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import BioVenn as bv\n",
    "import seaborn as sns\n",
    "import scipy.stats as stats\n",
    "import gseapy as gp\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dee5c1bb-6466-4c29-97d7-30e3a7fd1a72",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Load in ChIP-seq and RNA-seq data</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3618f594-a75a-4c96-8ef2-d65f95ec1037",
   "metadata": {},
   "source": [
    "<h2>Load in RNA-seq data</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33ff5baf-daba-444e-aa79-bd87dfb72648",
   "metadata": {},
   "source": [
    "Let's go ahead and load in our RNA-seq differential expression results that we got from <code>DESeq2</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6b7b8c9a-19dc-4106-b647-ea2716675ac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.read_csv('../Week_10/differential-expression-analysis/1M_results.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff48aaf-d2b3-4da4-aebf-d5168a84cb5c",
   "metadata": {},
   "source": [
    "Then now let's take a quick sanity check to make sure it imported properly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6fd1f9a2-282e-485a-9622-59d2b2d3fa5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_cfab2\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_cfab2_level0_col0\" class=\"col_heading level0 col0\" >Unnamed: 0</th>\n",
       "      <th id=\"T_cfab2_level0_col1\" class=\"col_heading level0 col1\" >baseMean</th>\n",
       "      <th id=\"T_cfab2_level0_col2\" class=\"col_heading level0 col2\" >log2FoldChange</th>\n",
       "      <th id=\"T_cfab2_level0_col3\" class=\"col_heading level0 col3\" >lfcSE</th>\n",
       "      <th id=\"T_cfab2_level0_col4\" class=\"col_heading level0 col4\" >stat</th>\n",
       "      <th id=\"T_cfab2_level0_col5\" class=\"col_heading level0 col5\" >pvalue</th>\n",
       "      <th id=\"T_cfab2_level0_col6\" class=\"col_heading level0 col6\" >padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_cfab2_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_cfab2_row0_col0\" class=\"data row0 col0\" >A1BG-AS1</td>\n",
       "      <td id=\"T_cfab2_row0_col1\" class=\"data row0 col1\" >1.717396</td>\n",
       "      <td id=\"T_cfab2_row0_col2\" class=\"data row0 col2\" >1.579446</td>\n",
       "      <td id=\"T_cfab2_row0_col3\" class=\"data row0 col3\" >0.747265</td>\n",
       "      <td id=\"T_cfab2_row0_col4\" class=\"data row0 col4\" >2.113634</td>\n",
       "      <td id=\"T_cfab2_row0_col5\" class=\"data row0 col5\" >0.034547</td>\n",
       "      <td id=\"T_cfab2_row0_col6\" class=\"data row0 col6\" >0.105928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cfab2_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_cfab2_row1_col0\" class=\"data row1 col0\" >A4GALT</td>\n",
       "      <td id=\"T_cfab2_row1_col1\" class=\"data row1 col1\" >10.639844</td>\n",
       "      <td id=\"T_cfab2_row1_col2\" class=\"data row1 col2\" >-0.064188</td>\n",
       "      <td id=\"T_cfab2_row1_col3\" class=\"data row1 col3\" >0.276068</td>\n",
       "      <td id=\"T_cfab2_row1_col4\" class=\"data row1 col4\" >-0.232508</td>\n",
       "      <td id=\"T_cfab2_row1_col5\" class=\"data row1 col5\" >0.816143</td>\n",
       "      <td id=\"T_cfab2_row1_col6\" class=\"data row1 col6\" >0.904300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cfab2_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_cfab2_row2_col0\" class=\"data row2 col0\" >AAAS</td>\n",
       "      <td id=\"T_cfab2_row2_col1\" class=\"data row2 col1\" >28.134083</td>\n",
       "      <td id=\"T_cfab2_row2_col2\" class=\"data row2 col2\" >-0.029254</td>\n",
       "      <td id=\"T_cfab2_row2_col3\" class=\"data row2 col3\" >0.173522</td>\n",
       "      <td id=\"T_cfab2_row2_col4\" class=\"data row2 col4\" >-0.168591</td>\n",
       "      <td id=\"T_cfab2_row2_col5\" class=\"data row2 col5\" >0.866118</td>\n",
       "      <td id=\"T_cfab2_row2_col6\" class=\"data row2 col6\" >0.930518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cfab2_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_cfab2_row3_col0\" class=\"data row3 col0\" >AACS</td>\n",
       "      <td id=\"T_cfab2_row3_col1\" class=\"data row3 col1\" >28.101282</td>\n",
       "      <td id=\"T_cfab2_row3_col2\" class=\"data row3 col2\" >-0.139976</td>\n",
       "      <td id=\"T_cfab2_row3_col3\" class=\"data row3 col3\" >0.158908</td>\n",
       "      <td id=\"T_cfab2_row3_col4\" class=\"data row3 col4\" >-0.880861</td>\n",
       "      <td id=\"T_cfab2_row3_col5\" class=\"data row3 col5\" >0.378393</td>\n",
       "      <td id=\"T_cfab2_row3_col6\" class=\"data row3 col6\" >0.577812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cfab2_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_cfab2_row4_col0\" class=\"data row4 col0\" >AADAT</td>\n",
       "      <td id=\"T_cfab2_row4_col1\" class=\"data row4 col1\" >3.955065</td>\n",
       "      <td id=\"T_cfab2_row4_col2\" class=\"data row4 col2\" >0.101844</td>\n",
       "      <td id=\"T_cfab2_row4_col3\" class=\"data row4 col3\" >0.421201</td>\n",
       "      <td id=\"T_cfab2_row4_col4\" class=\"data row4 col4\" >0.241794</td>\n",
       "      <td id=\"T_cfab2_row4_col5\" class=\"data row4 col5\" >0.808940</td>\n",
       "      <td id=\"T_cfab2_row4_col6\" class=\"data row4 col6\" >0.901120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x79948e58d510>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.head().style"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ac55a76-2993-451a-9ed4-8d65d7feac12",
   "metadata": {},
   "source": [
    "We'll need to use the first column because it contains the gene names that we'll use to identify intersections with our ChIP-seq dataset, so let's go ahead and update the column header to <code>'gene'</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3d20bc27-0d72-41b0-986b-9e90ceb0efa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "res.rename(columns={res.columns[0] : 'gene'},\n",
    "           inplace=True\n",
    "          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6624a0fc-8d60-44cb-b4d6-316c1e696c2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_8fde5\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_8fde5_level0_col0\" class=\"col_heading level0 col0\" >gene</th>\n",
       "      <th id=\"T_8fde5_level0_col1\" class=\"col_heading level0 col1\" >baseMean</th>\n",
       "      <th id=\"T_8fde5_level0_col2\" class=\"col_heading level0 col2\" >log2FoldChange</th>\n",
       "      <th id=\"T_8fde5_level0_col3\" class=\"col_heading level0 col3\" >lfcSE</th>\n",
       "      <th id=\"T_8fde5_level0_col4\" class=\"col_heading level0 col4\" >stat</th>\n",
       "      <th id=\"T_8fde5_level0_col5\" class=\"col_heading level0 col5\" >pvalue</th>\n",
       "      <th id=\"T_8fde5_level0_col6\" class=\"col_heading level0 col6\" >padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_8fde5_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_8fde5_row0_col0\" class=\"data row0 col0\" >A1BG-AS1</td>\n",
       "      <td id=\"T_8fde5_row0_col1\" class=\"data row0 col1\" >1.717396</td>\n",
       "      <td id=\"T_8fde5_row0_col2\" class=\"data row0 col2\" >1.579446</td>\n",
       "      <td id=\"T_8fde5_row0_col3\" class=\"data row0 col3\" >0.747265</td>\n",
       "      <td id=\"T_8fde5_row0_col4\" class=\"data row0 col4\" >2.113634</td>\n",
       "      <td id=\"T_8fde5_row0_col5\" class=\"data row0 col5\" >0.034547</td>\n",
       "      <td id=\"T_8fde5_row0_col6\" class=\"data row0 col6\" >0.105928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_8fde5_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_8fde5_row1_col0\" class=\"data row1 col0\" >A4GALT</td>\n",
       "      <td id=\"T_8fde5_row1_col1\" class=\"data row1 col1\" >10.639844</td>\n",
       "      <td id=\"T_8fde5_row1_col2\" class=\"data row1 col2\" >-0.064188</td>\n",
       "      <td id=\"T_8fde5_row1_col3\" class=\"data row1 col3\" >0.276068</td>\n",
       "      <td id=\"T_8fde5_row1_col4\" class=\"data row1 col4\" >-0.232508</td>\n",
       "      <td id=\"T_8fde5_row1_col5\" class=\"data row1 col5\" >0.816143</td>\n",
       "      <td id=\"T_8fde5_row1_col6\" class=\"data row1 col6\" >0.904300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_8fde5_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_8fde5_row2_col0\" class=\"data row2 col0\" >AAAS</td>\n",
       "      <td id=\"T_8fde5_row2_col1\" class=\"data row2 col1\" >28.134083</td>\n",
       "      <td id=\"T_8fde5_row2_col2\" class=\"data row2 col2\" >-0.029254</td>\n",
       "      <td id=\"T_8fde5_row2_col3\" class=\"data row2 col3\" >0.173522</td>\n",
       "      <td id=\"T_8fde5_row2_col4\" class=\"data row2 col4\" >-0.168591</td>\n",
       "      <td id=\"T_8fde5_row2_col5\" class=\"data row2 col5\" >0.866118</td>\n",
       "      <td id=\"T_8fde5_row2_col6\" class=\"data row2 col6\" >0.930518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_8fde5_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_8fde5_row3_col0\" class=\"data row3 col0\" >AACS</td>\n",
       "      <td id=\"T_8fde5_row3_col1\" class=\"data row3 col1\" >28.101282</td>\n",
       "      <td id=\"T_8fde5_row3_col2\" class=\"data row3 col2\" >-0.139976</td>\n",
       "      <td id=\"T_8fde5_row3_col3\" class=\"data row3 col3\" >0.158908</td>\n",
       "      <td id=\"T_8fde5_row3_col4\" class=\"data row3 col4\" >-0.880861</td>\n",
       "      <td id=\"T_8fde5_row3_col5\" class=\"data row3 col5\" >0.378393</td>\n",
       "      <td id=\"T_8fde5_row3_col6\" class=\"data row3 col6\" >0.577812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_8fde5_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_8fde5_row4_col0\" class=\"data row4 col0\" >AADAT</td>\n",
       "      <td id=\"T_8fde5_row4_col1\" class=\"data row4 col1\" >3.955065</td>\n",
       "      <td id=\"T_8fde5_row4_col2\" class=\"data row4 col2\" >0.101844</td>\n",
       "      <td id=\"T_8fde5_row4_col3\" class=\"data row4 col3\" >0.421201</td>\n",
       "      <td id=\"T_8fde5_row4_col4\" class=\"data row4 col4\" >0.241794</td>\n",
       "      <td id=\"T_8fde5_row4_col5\" class=\"data row4 col5\" >0.808940</td>\n",
       "      <td id=\"T_8fde5_row4_col6\" class=\"data row4 col6\" >0.901120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x79948e289810>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.head().style"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e8b128-78b4-455f-94dc-a504a734dc29",
   "metadata": {},
   "source": [
    "<h2>Load in ChIP-seq data</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26f25467-bab1-482f-a530-d91d342b731a",
   "metadata": {},
   "source": [
    "Now let's load in our ChIP-seq data. Recall that we can use HOMER to annotate peaks based on the nearest TSS to get an idea of what gene each peak may potentially be regulating. In this week's directory, there is a <code>data</code> subdirectory containing an annotated peaks file generated by HOMER for the full ChIP-seq dataset. For this notebook, we'll make use of the full set to be able to work with more peaks identified as significantly enriched when pulling down TAZ.\n",
    "\n",
    "First, let's go ahead and load in our annotated peaks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "feecb972-ab15-4a03-84c3-79d15510feb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks = pd.read_csv('./data/annotated-peaks-full-set.txt',\n",
    "                    sep='\\t',\n",
    "                   )\n",
    "\n",
    "#This dataset was obtained from our full set ChIP-seq analysis not the truncated one\n",
    "#Which gives us (as you probably saw from your homework results) a lot more merged peaks to play with\n",
    "#We can take the merged peaks from the top 2000 or so that are concordant (rather than top 500 like we did with our truncated data)\n",
    "#That gives us a lot of peaks to play with and then we can use HOMER to annotate them based on the nearest TSS\n",
    "#Recall from Dr. Ingolia's Week 15 lecture that annotating peaks based on nearest TSS is a common way to infer what gene is regulated by a ChIP-peak"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "935a71d6-74b6-41f9-97c1-851acbe4e208",
   "metadata": {},
   "source": [
    "Now let's perform a quick sanity check and take a look at the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7d61d26b-1de4-44c3-9d03-2159dc21f567",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PeakID (cmd=annotatePeaks.pl ./top_concordant_peaks_full-set.narrowPeak hg19 -noann)</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Peak Score</th>\n",
       "      <th>Focus Ratio/Region Size</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Detailed Annotation</th>\n",
       "      <th>Distance to TSS</th>\n",
       "      <th>Nearest PromoterID</th>\n",
       "      <th>Entrez ID</th>\n",
       "      <th>Nearest Unigene</th>\n",
       "      <th>Nearest Refseq</th>\n",
       "      <th>Nearest Ensembl</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Alias</th>\n",
       "      <th>Gene Description</th>\n",
       "      <th>Gene Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Merged-chr19-39184227-2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>39184140</td>\n",
       "      <td>39184315</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45938</td>\n",
       "      <td>NM_001322033</td>\n",
       "      <td>81.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_004924</td>\n",
       "      <td>ENSG00000130402</td>\n",
       "      <td>ACTN4</td>\n",
       "      <td>ACTININ-4|FSGS|FSGS1</td>\n",
       "      <td>actinin alpha 4</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Merged-chr10-91038033-2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>91037981</td>\n",
       "      <td>91038085</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-23770</td>\n",
       "      <td>NM_001547</td>\n",
       "      <td>3433.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001547</td>\n",
       "      <td>ENSG00000119922</td>\n",
       "      <td>IFIT2</td>\n",
       "      <td>G10P2|GARG-39|IFI-54|IFI-54K|IFI54|IFIT-2|ISG-...</td>\n",
       "      <td>interferon induced protein with tetratricopept...</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Merged-chr2-65886199-2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>65886111</td>\n",
       "      <td>65886288</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-226466</td>\n",
       "      <td>NM_181784</td>\n",
       "      <td>200734.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_181784</td>\n",
       "      <td>ENSG00000198369</td>\n",
       "      <td>SPRED2</td>\n",
       "      <td>NS14|Spred-2</td>\n",
       "      <td>sprouty related EVH1 domain containing 2</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Merged-chr19-48991373-2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>48991289</td>\n",
       "      <td>48991457</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18747</td>\n",
       "      <td>NM_004228</td>\n",
       "      <td>9266.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_004228</td>\n",
       "      <td>ENSG00000105443</td>\n",
       "      <td>CYTH2</td>\n",
       "      <td>ARNO|CTS18|CTS18.1|PSCD2|PSCD2L|SEC7L|Sec7p-L|...</td>\n",
       "      <td>cytohesin 2</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Merged-chr6-50055081-2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>50055014</td>\n",
       "      <td>50055148</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-37439</td>\n",
       "      <td>NM_001369057</td>\n",
       "      <td>245915.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001037498</td>\n",
       "      <td>ENSG00000180872</td>\n",
       "      <td>DEFB112</td>\n",
       "      <td>DEFB-12</td>\n",
       "      <td>defensin beta 112</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1864</th>\n",
       "      <td>Merged-chr17-855149-2</td>\n",
       "      <td>chr17</td>\n",
       "      <td>855097</td>\n",
       "      <td>855201</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27867</td>\n",
       "      <td>NM_022463</td>\n",
       "      <td>64359.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_022463</td>\n",
       "      <td>ENSG00000167693</td>\n",
       "      <td>NXN</td>\n",
       "      <td>NRX|RRS2|TRG-4</td>\n",
       "      <td>nucleoredoxin</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1865</th>\n",
       "      <td>Merged-chr14-23398869-2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>23398813</td>\n",
       "      <td>23398925</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>51</td>\n",
       "      <td>NR_110002</td>\n",
       "      <td>101926933.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NR_110002</td>\n",
       "      <td>ENSG00000257285</td>\n",
       "      <td>PRMT5-DT</td>\n",
       "      <td>-</td>\n",
       "      <td>PRMT5 divergent transcript</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1866</th>\n",
       "      <td>Merged-chr4-77532537-2</td>\n",
       "      <td>chr4</td>\n",
       "      <td>77532472</td>\n",
       "      <td>77532602</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35833</td>\n",
       "      <td>NR_039655</td>\n",
       "      <td>100616254.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NR_039655</td>\n",
       "      <td>ENSG00000283682</td>\n",
       "      <td>MIR548AH</td>\n",
       "      <td>-</td>\n",
       "      <td>microRNA 548ah</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867</th>\n",
       "      <td>Merged-chr6-37210436-2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>37210351</td>\n",
       "      <td>37210521</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14939</td>\n",
       "      <td>NM_145316</td>\n",
       "      <td>221468.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_145316</td>\n",
       "      <td>ENSG00000172738</td>\n",
       "      <td>TMEM217</td>\n",
       "      <td>C6orf128|dJ355M6.2</td>\n",
       "      <td>transmembrane protein 217</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868</th>\n",
       "      <td>Merged-chr12-66173098-2</td>\n",
       "      <td>chr12</td>\n",
       "      <td>66173044</td>\n",
       "      <td>66173152</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/home/jovyan/shared/course/mcb201b-shared/chip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-45142</td>\n",
       "      <td>NM_001300918</td>\n",
       "      <td>8091.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_003483</td>\n",
       "      <td>ENSG00000149948</td>\n",
       "      <td>HMGA2</td>\n",
       "      <td>BABL|HMGI-C|HMGIC|LIPO|SRS5|STQTL9</td>\n",
       "      <td>high mobility group AT-hook 2</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1869 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     PeakID (cmd=annotatePeaks.pl ./top_concordant_peaks_full-set.narrowPeak hg19 -noann)  \\\n",
       "0                               Merged-chr19-39184227-2                                     \n",
       "1                               Merged-chr10-91038033-2                                     \n",
       "2                                Merged-chr2-65886199-2                                     \n",
       "3                               Merged-chr19-48991373-2                                     \n",
       "4                                Merged-chr6-50055081-2                                     \n",
       "...                                                 ...                                     \n",
       "1864                              Merged-chr17-855149-2                                     \n",
       "1865                            Merged-chr14-23398869-2                                     \n",
       "1866                             Merged-chr4-77532537-2                                     \n",
       "1867                             Merged-chr6-37210436-2                                     \n",
       "1868                            Merged-chr12-66173098-2                                     \n",
       "\n",
       "        Chr     Start       End Strand  Peak Score  \\\n",
       "0     chr19  39184140  39184315      +         0.0   \n",
       "1     chr10  91037981  91038085      +         0.0   \n",
       "2      chr2  65886111  65886288      +         0.0   \n",
       "3     chr19  48991289  48991457      +         0.0   \n",
       "4      chr6  50055014  50055148      +         0.0   \n",
       "...     ...       ...       ...    ...         ...   \n",
       "1864  chr17    855097    855201      +         0.0   \n",
       "1865  chr14  23398813  23398925      +         0.0   \n",
       "1866   chr4  77532472  77532602      +         0.0   \n",
       "1867   chr6  37210351  37210521      +         0.0   \n",
       "1868  chr12  66173044  66173152      +         0.0   \n",
       "\n",
       "                                Focus Ratio/Region Size  Annotation  \\\n",
       "0     /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "1     /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "2     /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "3     /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "4     /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "...                                                 ...         ...   \n",
       "1864  /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "1865  /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "1866  /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "1867  /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "1868  /home/jovyan/shared/course/mcb201b-shared/chip...         NaN   \n",
       "\n",
       "      Detailed Annotation  Distance to TSS Nearest PromoterID    Entrez ID  \\\n",
       "0                     NaN            45938       NM_001322033         81.0   \n",
       "1                     NaN           -23770          NM_001547       3433.0   \n",
       "2                     NaN          -226466          NM_181784     200734.0   \n",
       "3                     NaN            18747          NM_004228       9266.0   \n",
       "4                     NaN           -37439       NM_001369057     245915.0   \n",
       "...                   ...              ...                ...          ...   \n",
       "1864                  NaN            27867          NM_022463      64359.0   \n",
       "1865                  NaN               51          NR_110002  101926933.0   \n",
       "1866                  NaN            35833          NR_039655  100616254.0   \n",
       "1867                  NaN            14939          NM_145316     221468.0   \n",
       "1868                  NaN           -45142       NM_001300918       8091.0   \n",
       "\n",
       "      Nearest Unigene Nearest Refseq  Nearest Ensembl Gene Name  \\\n",
       "0                 NaN      NM_004924  ENSG00000130402     ACTN4   \n",
       "1                 NaN      NM_001547  ENSG00000119922     IFIT2   \n",
       "2                 NaN      NM_181784  ENSG00000198369    SPRED2   \n",
       "3                 NaN      NM_004228  ENSG00000105443     CYTH2   \n",
       "4                 NaN   NM_001037498  ENSG00000180872   DEFB112   \n",
       "...               ...            ...              ...       ...   \n",
       "1864              NaN      NM_022463  ENSG00000167693       NXN   \n",
       "1865              NaN      NR_110002  ENSG00000257285  PRMT5-DT   \n",
       "1866              NaN      NR_039655  ENSG00000283682  MIR548AH   \n",
       "1867              NaN      NM_145316  ENSG00000172738   TMEM217   \n",
       "1868              NaN      NM_003483  ENSG00000149948     HMGA2   \n",
       "\n",
       "                                             Gene Alias  \\\n",
       "0                                  ACTININ-4|FSGS|FSGS1   \n",
       "1     G10P2|GARG-39|IFI-54|IFI-54K|IFI54|IFIT-2|ISG-...   \n",
       "2                                          NS14|Spred-2   \n",
       "3     ARNO|CTS18|CTS18.1|PSCD2|PSCD2L|SEC7L|Sec7p-L|...   \n",
       "4                                               DEFB-12   \n",
       "...                                                 ...   \n",
       "1864                                     NRX|RRS2|TRG-4   \n",
       "1865                                                  -   \n",
       "1866                                                  -   \n",
       "1867                                 C6orf128|dJ355M6.2   \n",
       "1868                 BABL|HMGI-C|HMGIC|LIPO|SRS5|STQTL9   \n",
       "\n",
       "                                       Gene Description       Gene Type  \n",
       "0                                       actinin alpha 4  protein-coding  \n",
       "1     interferon induced protein with tetratricopept...  protein-coding  \n",
       "2              sprouty related EVH1 domain containing 2  protein-coding  \n",
       "3                                           cytohesin 2  protein-coding  \n",
       "4                                     defensin beta 112  protein-coding  \n",
       "...                                                 ...             ...  \n",
       "1864                                      nucleoredoxin  protein-coding  \n",
       "1865                         PRMT5 divergent transcript           ncRNA  \n",
       "1866                                     microRNA 548ah           ncRNA  \n",
       "1867                          transmembrane protein 217  protein-coding  \n",
       "1868                      high mobility group AT-hook 2  protein-coding  \n",
       "\n",
       "[1869 rows x 19 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks\n",
    "#What is particularly important for us is the gene type and the gene name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e1bf8b9-eb23-4d4f-b21e-f9b63b619393",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Identify and visualize intersection between datasets</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94c7980b-5e87-4cd0-af2d-ad8d437e5972",
   "metadata": {},
   "source": [
    "Now that we have both datasets loaded in, let's proceed with identifying their intersections.\n",
    "\n",
    "What we're interested in is the gene names, and recall that our library was based on RNA samples that were enriched for mRNA. So we're particularly interested in the gene names for protein-coding genes since that is what we're analyzing with our RNA-seq data.\n",
    "\n",
    "Let's go ahead and pull out the gene names from our datasets."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4de46a19-8f32-44c4-80a6-f8d87236d3ac",
   "metadata": {},
   "source": [
    "<h2>Pull out gene names for genes that are differentially expressed</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c82b188e-591a-4711-9fd1-496f5fd25ccd",
   "metadata": {},
   "source": [
    "We'll first pull out genes that are significantly differentially expressed based on our RNA-seq analysis p-adjusted values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "692f4124-c121-4aa2-b5de-eae619a0fc2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>1.717396</td>\n",
       "      <td>1.579446</td>\n",
       "      <td>0.747265</td>\n",
       "      <td>2.113634</td>\n",
       "      <td>0.034547</td>\n",
       "      <td>0.105928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>10.639844</td>\n",
       "      <td>-0.064188</td>\n",
       "      <td>0.276068</td>\n",
       "      <td>-0.232508</td>\n",
       "      <td>0.816143</td>\n",
       "      <td>0.904300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>28.134083</td>\n",
       "      <td>-0.029254</td>\n",
       "      <td>0.173522</td>\n",
       "      <td>-0.168591</td>\n",
       "      <td>0.866118</td>\n",
       "      <td>0.930518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AACS</td>\n",
       "      <td>28.101282</td>\n",
       "      <td>-0.139976</td>\n",
       "      <td>0.158908</td>\n",
       "      <td>-0.880861</td>\n",
       "      <td>0.378393</td>\n",
       "      <td>0.577812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>3.955065</td>\n",
       "      <td>0.101844</td>\n",
       "      <td>0.421201</td>\n",
       "      <td>0.241794</td>\n",
       "      <td>0.808940</td>\n",
       "      <td>0.901120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17301</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>6.760829</td>\n",
       "      <td>-0.063447</td>\n",
       "      <td>0.361171</td>\n",
       "      <td>-0.175671</td>\n",
       "      <td>0.860552</td>\n",
       "      <td>0.926733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17302</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>30.511184</td>\n",
       "      <td>0.004179</td>\n",
       "      <td>0.170884</td>\n",
       "      <td>0.024453</td>\n",
       "      <td>0.980491</td>\n",
       "      <td>0.994338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17303</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>144.918149</td>\n",
       "      <td>-0.324147</td>\n",
       "      <td>0.088009</td>\n",
       "      <td>-3.683103</td>\n",
       "      <td>0.000230</td>\n",
       "      <td>0.001677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17304</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>28.852060</td>\n",
       "      <td>0.103463</td>\n",
       "      <td>0.219259</td>\n",
       "      <td>0.471875</td>\n",
       "      <td>0.637016</td>\n",
       "      <td>0.789957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17305</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>65.752887</td>\n",
       "      <td>0.074821</td>\n",
       "      <td>0.125279</td>\n",
       "      <td>0.597231</td>\n",
       "      <td>0.550353</td>\n",
       "      <td>0.725226</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17306 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene    baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "0      A1BG-AS1    1.717396        1.579446  0.747265  2.113634  0.034547   \n",
       "1        A4GALT   10.639844       -0.064188  0.276068 -0.232508  0.816143   \n",
       "2          AAAS   28.134083       -0.029254  0.173522 -0.168591  0.866118   \n",
       "3          AACS   28.101282       -0.139976  0.158908 -0.880861  0.378393   \n",
       "4         AADAT    3.955065        0.101844  0.421201  0.241794  0.808940   \n",
       "...         ...         ...             ...       ...       ...       ...   \n",
       "17301    ZYG11A    6.760829       -0.063447  0.361171 -0.175671  0.860552   \n",
       "17302    ZYG11B   30.511184        0.004179  0.170884  0.024453  0.980491   \n",
       "17303       ZYX  144.918149       -0.324147  0.088009 -3.683103  0.000230   \n",
       "17304     ZZEF1   28.852060        0.103463  0.219259  0.471875  0.637016   \n",
       "17305      ZZZ3   65.752887        0.074821  0.125279  0.597231  0.550353   \n",
       "\n",
       "           padj  \n",
       "0      0.105928  \n",
       "1      0.904300  \n",
       "2      0.930518  \n",
       "3      0.577812  \n",
       "4      0.901120  \n",
       "...         ...  \n",
       "17301  0.926733  \n",
       "17302  0.994338  \n",
       "17303  0.001677  \n",
       "17304  0.789957  \n",
       "17305  0.725226  \n",
       "\n",
       "[17306 rows x 7 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0b14c656-532b-4873-a712-c3b6e5fa373f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We can apply a filter for what we want to focus on, specifically we can pull out significant genes\n",
    "#We can use these gene names to then find which ones are shared with our ChIP-seq dataset\n",
    "de_genes = res.loc[(res['padj'] < 0.05), 'gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "59bd70fc-8b7e-45f0-a8b9-cbd88d0149fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58       ABCG1\n",
       "59       ABCG2\n",
       "84      ABI3BP\n",
       "108      ACADS\n",
       "127      ACKR3\n",
       "160      ACTA2\n",
       "191     ACVR2A\n",
       "202     ADAM12\n",
       "213      ADAM8\n",
       "215    ADAMTS1\n",
       "Name: gene, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Let's take a look to see what is pulled out\n",
    "#These are just genes identified as significant from just RNA-seq differential expression analysis\n",
    "#We haven't identified intersections yet\n",
    "de_genes[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad5d6077-5f0d-4dde-a5b3-7b1f1eb1a559",
   "metadata": {},
   "source": [
    "<h2>Pull out gene names for protein-coding genes</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fa79ff8-1a8b-4cea-b789-0448412c3dae",
   "metadata": {},
   "source": [
    "You can see that the annotations identify the <code>'Gene Type'</code>, which we can use to pull out just the protein-coding genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2519b363-56d2-4ef6-b275-e790d2042842",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Our RNA-seq is on mRNA, so we're focusing on protein-coding genes\n",
    "#So we can also pull those out from our ChIP peak annotation table to get our list of ChIP genes\n",
    "chip_genes = peaks.loc[peaks['Gene Type'] == 'protein-coding', 'Gene Name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "42b6fc9c-7661-469b-9cd4-beb1e3e222c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      ACTN4\n",
       "1      IFIT2\n",
       "2     SPRED2\n",
       "3      CYTH2\n",
       "4    DEFB112\n",
       "5    CCDC115\n",
       "6       SYBU\n",
       "7    TMEM45A\n",
       "8       MYH9\n",
       "9      ILDR2\n",
       "Name: Gene Name, dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chip_genes[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2db0cdac-d3bb-4f0c-9b9b-b1b64361a402",
   "metadata": {},
   "source": [
    "<h2>Generate an empty list</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93e2433c-db50-4918-9d30-ad26c52fe2f0",
   "metadata": {},
   "source": [
    "If we dig into the documentation for <code>BioVenn</code>, you can see that there is a quirk in terms of how the function is set up. It's built to identify and visualize intersections of three datasets, and since we're only interested in comparing our two datasets, we can get around this by feeding it an empty list, which will satisfy it's requirements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2a95798e-c786-4e54-b392-58975ff747bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "the_void = []\n",
    "#The draw_venn() function requires 3 lists, so we can just give it an empty one to satisfy its required parameters\n",
    "#nothing will intersect with this so it's effectively ignored"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcb1bc9d-e7e5-4392-acb9-aac77c930ce7",
   "metadata": {},
   "source": [
    "<h2>Visualize intersections as a Venn diagram</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cb9a503-a194-4f61-b684-d540d77822d9",
   "metadata": {},
   "source": [
    "With our lists of gene names from our datasets (and one empty one), we can generate a Venn diagram using <code>BioVenn</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "899129b8-d3ab-42f1-aca2-40611af0c357",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "x total: 3574\n",
      "y total: 1028\n",
      "z total: 0\n",
      "x only: 3255\n",
      "y only: 709\n",
      "z only: 0\n",
      "x-y total overlap: 319\n",
      "x-z total overlap: 0\n",
      "y-z total overlap: 0\n",
      "x-y only overlap: 319\n",
      "x-z only overlap: 0\n",
      "y-z only overlap: 0\n",
      "x-y-z overlap: 0\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxYAAAOHCAYAAAC6hNVBAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAAknNJREFUeJzs3XeAZFWZ9/HfufdW6DCZPCRBECVJDiJJMiqggiAioKCSXFdAXFwXdRXfVRZYJIgoIEoyIYKSRFB0FEFAAVEECUMY0sROVXXvOe8ft6qma7p6prsr3Arfz27LTHV19emeDvdX53nOY5xzTgAAAABQAy/pBQAAAABofwQLAAAAADUjWAAAAACoGcECAAAAQM0IFgAAAABqRrAAAAAAUDOCBQAAAICaESwAAAAA1IxgAQAAAKBmBAtgCjbccEMZY8Z9+dCHPjThx3r++ecVBMFKH+/qq69e6WOMjIzohhtu0Cc/+Ults802WmeddZTJZNTb26t11llHu+++u04++WRdf/31WrBgQdXHmDlz5krXsOJLOp3WnDlztN122+kTn/iEfv3rX690jS+99JJ+8IMf6LTTTtPee++tLbbYQmuttZbS6bT6+/u1/vrr64ADDtDXvvY1Pf/88xP+/EnS3XffrRNPPFFvectbNGPGDPX29upNb3qTDjvsMH3ve99TGIaTerwwDPX1r39dfX19FR/zcccdt9K3u/DCCyf1OfQ8T9OmTdOGG26oAw88UBdccIFef/31Sa11Rccdd9xK3+f6668/4c+Hc06bbrrpSh9vVZ+TWqzs+6xeCoWCTjvtNM2ZM0frr7++rrjiiro9djvj8wJgKoxzziW9CKDdnHvuuVq4cKHmz5+vH/7wh2Nen0ql9Oyzz2qdddZZ5WN99rOf1Te+8Y0xt0+fPl0nnniiJOnII4/U9ttvP+Y+1lpdcMEF+sY3vqFXXnmlfPumm26qHXfcUbNnz9Yrr7yi++67Ty+99JIkyRijXXfdVSeffHJFAPrCF76g4eFhPf7447r99tvHXcvg4KCeeuop3XfffcrlchX322+//XTjjTdq5syZY9a6xRZb6PHHH5ckeZ6nvffeW5tuuqlef/113X777Vq6dGn5vqlUSp/5zGf01a9+Vb7vj/u5W7x4sT7ykY/olltukST19vbqsMMOU19fn2655Ra9/PLLkqTNN99c1157rbbeeutxH6tk3rx5+sQnPqHHHntszOuOPfbYlYa8++67TzfffLMk6Yorrqj4mCTpgAMO0Oabb64wDPXaa6/p/vvv19NPP11xn+nTp+vyyy/XkUceucq1VnPjjTfqgQcekCT97//+b9X7/PCHP9Thhx++ysf65S9/qYMPPrjq60488URNnz5dO+ywgz74wQ9Oaa2rUvo+u/HGG/XCCy9UvK7ar657771X9957b8Vtxx13nDbccMOVvo/Pf/7z5b8bY/S73/1Ou+66a01rb3d8XgBMiQMwZffcc4+TVPXlP//zP1f59oODg27WrFlV336DDTZY6du++uqrbu+99654mzXXXNPdeuutY+5bKBTcxRdf7DKZTPm+e+yxR9XHveqqqya0lldeecXttddeY+673377VX3czTff3Elys2fPdvfff3/F6xYsWOC23nrrMY915JFHjvvxL1myxG277bbl+/q+X/G4r732mlt33XXLr589e7b7y1/+Mu7jLVq0yH3iE59wxhgnya211lpj1nPssceO+/Yr2mCDDca8/VVXXTXmftdee61LpVIV9zPGuPvuu2/C72s8431t7rbbbhN6+/3222/cx3jmmWdqXt9E7bHHHmPefzXnnHPOmPvdc889K33sd7/73WPe5rzzzmvAR9Fe+LwAmApKoYA6eMtb3jLmtm9/+9tjntFf0TXXXKNFixZps802m9T7Gxoa0oEHHlhRfpTNZnXbbbdVfYY5CAKdcsopuuqqqyb1flZmjTXW0LXXXjtmR+HOO+/Un//853Hf7rzzztOOO+5Ycduaa66p733ve2Pue8MNN+hnP/tZ1cf53Oc+p4ceeqj893322aficVdbbTV98pOfLP994cKF+shHPjJuGdCJJ56oyy+/XJtssonuuusufeITnxj3Y6inD33oQ+XdoBLnnP7nf/6nLo+/xhprjNlB+t3vfqeHH354pW/3xBNP6K677pr012a72WmnnSZ0W7fh8wJgKggWQB3svPPOevvb315x26uvvqobbrhhpW930UUXKZvN6mMf+9ik3t8pp5wy5uL9pJNO0jbbbLPStzvqqKO07777Tup9rczaa6+tzTfffMztf/jDH6reP51O64gjjqj6uq233lpvfetbx9z+gx/8YMxt//znP3X55ZdX3LbtttuOud+Kt/3lL3/RNddcU/X9B0GgL33pS/rrX/+qffbZp+p9GqXa+xvvczhZPT09VfsgLrroopW+3UUXXSTnnE499dS6rKNVffazn9XJJ5+s2bNna+7cubrsssu02267Jb2sxPF5ATAVBAugTk4++eQxt63s4u3OO+/UE088oaOOOkqrrbbahN/PY489VvXi+OMf//iE3v6kk06a8PuaiL6+vjG3DQwMjLntggsu0C233FL1/iXVgsU//vGPMbddffXVstZW3LbuuuuOud9666035rYrr7yy6vu+8sor9V//9V/KZDLjrq9RJvo5nKqTTz55TMPz9ddfr9dee63q/RctWqRrrrlGG2+8sQ488MC6raMVpdNpXXLJJXrjjTf0wgsvVOxydTM+LwCmgmAB1MnRRx89puTkoYce0u9///uq9/+///s/SdJpp502qfdz8cUXj7moXm+99SZcsrLPPvvo0EMPrcuzj9ZaPfXUU2Nu32STTcbctu+++2q//fZb6eP19PRUfR8ruu2228bcNmfOnAndNm/evDFN1eO972Z58sknx9xW7XM4VZtsssmYXZFcLqdvf/vbVe//ne98R0NDQzrllFPkeRP7NbGyE7FGN7x/8YtfHPd+zz777JQ+vtJJWF/60pfGvG6vvfaqeB+l5u6rr7563HV88YtfnNI6qlmyZIkuvPBCHXzwwVp33XXV29urnp4erbvuujrooIN04YUXasmSJVXfds8991zp5+qFF17QSSedpA022EDZbFbrrLOODj30UM2bN6/8GNZafetb39KOO+6oGTNmqK+vT9tuu60uvPBCRVE05n3W8nm57bbbdOyxx5ZPZ0ulUpo1a5a22morHXXUUfrmN79Z9eeFFO9CnnXWWdp55521+uqrK5PJqK+vTxtuuKF23HFHHXPMMTr//PP14IMPTvyTD6DpgqQXAHSK3t5eHXfccbrwwgsrbr/ooov0jne8o+K2f/7zn7rtttv0jne8Q9tss43+8pe/TPj93HHHHWNuq1aONJ5p06bppptumvD9V+a73/3umGe9N9lkE7373e+e0uO98cYbY27bYostKv6ez+fLp0uN1t/fP6HbnHN65JFHtPvuu09pjfU2MDCgiy++eMztn/nMZ+r6fk455RTdddddFbdddtllOuussxQEy38VRFGkSy65RH19fTr++OO1ePHiCT3+dtttp9NPP73qqWKj7brrrjr99NOrnvQ0VQceeKBWW201zZs3b0wJ2RFHHFGxc1X68xZbbKHTTz993JPd6uHaa6/VySefXA6yM2fO1JFHHiljjH72s5/ptttu02233aZzzjlHl156qY4++ugJP/ZDDz2kE044QYsWLSrf9vLLL+vmm2/WL3/5S/3whz/UAQccoEMOOUR33nlnxds+/PDDevjhh3X//ffr+uuvr3jdVD4v8+fP11FHHVXxJMr06dN12GGHaY011tBTTz2lH/7wh+XS0Msvv7xih/Xb3/62TjnllHL/U29vrw4++GDNnTtXCxcu1B/+8Ac98MAD+sEPfqC5c+fW7esGQAMk3DwOtLXSqVCl04KefPLJ8qlCpZcgCNwLL7xQ8Xannnqqk+RuuOEG59zkTmJa8X6S3Cc/+cm6fUyrWsvAwIB78MEH3b//+7+7IAgq7rfTTju5f/3rX1N6v9Zat+aaa4553yuecvXUU09V/RzcddddYx4zl8tVve93v/vdVa6n2glD9ToVqlAouBdffNHdcMMNbsstt6y4T09PT91O3xn9bxeGoVt//fXHrKn0NVjy4x//uOJr6plnnpnUqVDVvn6qnYZV7aSn8R6zkadCVTvZ7Zxzzlnp20zEJZdcUvGYvb297sknnyy//qmnnnJ9fX0V97n00ksrHuO5555zTzzxhDv00EPHrLG/v999+tOfdg8++KD76U9/6tZYY42K16+11lru8MMPd9ttt52788473bx589x73/veMY9z++231/R5WbBgwZiv9U033dS99NJLFfebN2+ey2azTpK74IILyrc//vjjFT9Hpk2bNubrIAxDd9ZZZzlJbu7cuZP7hwDQVJRCAXVUreQkDENdeuml5b8vWbJEV199tdZZZx29//3vn9Tjv/rqq1VvnzZt2uQXOwnPPfdcuRyiv79f22+/vS644AKFYajp06fr6KOP1m9/+1v98Y9/1Jve9KYpvY9f//rXFbM4JOmQQw4Zc8rVeGUj1eZdjH4mfiKP0WjHH3+8jDFKpVKaO3eujjzySD366KOS4ub1Cy64QM8//7xOP/30ur9v3/ernnS1Yh/QVEv0sNw//vEP/du//VvFbYcddlhFedvGG2+sQw89tOI+n/rUpyrK4tZff31tttlmmjFjxpj3sffee+uCCy7Qdtttp8MOO0yf+tSnKl6/YMEC3Xzzzbr11lu17777apdddqn4OVRS627Nqaeequeee67itq9//etae+21K27bZZddqn5d/+hHP6o4qW3TTTcdM3fE93197Wtf0wYbbFDTWgE0HsECqLNTTjllzG3f/va3NTIyIiluEh4YGNBJJ5007oXveMa7IM5ms5Nf6CRMnz5dp59+uk4//XSddNJJ2n///dXb2ytJWrp0qa699lodffTROvvss6uWM61KFEU6++yzK27bfffdde21146579DQUNXHqDaNebwJzYODg5NeYz0ccMABOv300/XpT39axxxzTEVfzF/+8hd99rOf1Yc+9KGq5W71cMIJJyidTlfcNm/evPIJYw8//LDuu+8+vetd79Lb3va2hqyhG3zjG98Yc6zxDjvsMOZ+Kx67HIZh1WGZ1Rx//PEVf6/WY3XggQdqrbXWKv997ty5Y8oDS8F2Kv71r3/pJz/5ScVtqVRq3F6qakdhl4ZYljzyyCP68Y9/PGYAojFG559//pifEwBaC8ECqLN3v/vdWn/99Stue/3113XdddfJWquLL75Y6XR6wqc4jVbtmUtJ5dDSKLNmzdJ5552n8847T5deeqluv/12vfTSSxXHmM6fP19f+9rXtMUWW6xyRsKKzjjjDP3pT38q//3YY4/VHXfcUfW0pFKgWdGKFyLj3SZVP4WpGT74wQ/qvPPO0wUXXKBrrrlGTzzxhO68806tscYakqRCoaC77rpLBxxwgE466aRx1z9Va6yxhj7wgQ+Mub20S8FuRXX33nvvuA3NKzaFS/HE8hVVO6Fs7ty5Y26rdjBBNRtttFHF36dPn77K+1S73+gejcm68847x3yNrr/++uMehLDTTjvp5ZdfrvjZt+LPyiiKdPjhh2vdddfVcccdp2uuuaa8k/m+972v6ul7AFoHwQKos/FKTr75zW/qlltu0b/+9S998IMfLF9MTsZ4b7Ns2bJJP1atZsyYoSuvvFJ77rlnxe0LFizQe97zngkdl+qc05lnnllueJ8+fbquvPJKXX311ePuwowXrqqdcDPeMLzxHiMJ++67b9V5J9/61rfKF/r1VG1H7cYbb9Sjjz6qG264QRtssIHe85731P39douBgYExz8JL1csVqx0u8OKLL467KzfaigEhlUqt8j7V7lcoFFb5vsbzz3/+c8xtKzs62/M8rbXWWhVPDhx22GFV1/7SSy/pe9/7no499litvfba2meffRq2kwegfggWQANUKzl55JFHys8ET3Xo2BprrFG1znjFGudmMcZUfQbxxRdfXGXt9uDgoA4//HCdd955kuJjcP/617+OKfFY0XrrrVf1QqTalPPxJp9vvPHGK30fzbbXXntVLT06//zz6/6+dt111zHDHPP5vA4++GDlcrlJHTGLscYrV6xW9ljt61jShE/iSlq1Y5tX/Lm3Km9961t18cUXj/u5kOInIO6+++5yKSGA1sVvD6ABxis5mT9/vnbaaacxtdWTsf/++4+5rdrxq+NZtmyZnn32WT377LMaHh6e8jpKxqvFX1k51COPPKLtt99eP/nJTzRjxgxdccUVuuuuu8aEphtvvFHf+c53Km5Lp9NjjqCVqg+Uq3abMWbMhXUrqPZ5nD9//pR6VlalWhicP3++enp6Jj0FfqrqXebVKsbbDau2ezbebsGK83BaVbUdkXw+P+nH+fjHP66///3vOuOMM1Y5j+f888+vKDsD0FoIFkCDVCs5kWqvXz/11FPHNCU///zzVSdUV3PGGWfoTW96kzbaaKO6XLRO5tlt55zOP/987bzzzvr73/+ugw8+WI8//rhOOOGEqvc/66yzdOKJJ465/YADDhhzW7WPZeHChWNu22WXXVqqFKqkmbsE1YY5lm6fPXt2TY9d7Zn5akMOG90XVE977rmnnHMrfSmVBPb392udddYZ8xjVyhWrBd+5c+eO20fUat785jePuW28ae6rstFGG+kb3/iGnnjiCb3wwgu6+uqrdeihh1b9vqjWwwKgNRAsgAapVnKy5ppr6vDDD6/pcbfccksdc8wxY26/4oorVvm2y5Yt04033igpLj1ad911a1qLVL3OWhp70fHcc89pr7320umnn66+vj5dc801uvXWW6s2sK5KadLyaNWGZlW77aMf/eik318zVPs8zpw5s+r08FqVhjmuaKoleqPNmjVrzG3VdsZeeumlmt/XilqlhOuggw4ac9tEvz4PPPDAhqypEfbbb7+qT3KM1yNy1VVXabPNNtMHP/jB8m2XX365dt5554rPxdy5c3Xsscfqpptuqtp/lNRx0QBWrTV+CgMdasWSk0984hOTrkGu5tJLL9U222wz5rZVTfD+3Oc+pyVLlsjzPH35y1+ueR1S9UATBIEOO+ywitu22mor/eY3v5EU7yR85CMfWeVJO+P1jmy66aZjdjIeeuihMfdb8bYtt9xSH/nIRyb18TXDgw8+qEceeWTM7UcccUTD3ufJJ59ccVG4++67a+utt675cas9xoIFCyr+Pn/+/IZMT67WID26NOf973+/jDENm7RdcuaZZ47ZuXnggQfG3G/F24Ig0JlnntnQtdXTxhtvPOb7PAzDcSev33DDDfrHP/5R8YTL/Pnzdf/99+unP/1p1bepFrRWnHMBoHUQLIAGOuGEEzQ8PFx+Oeecc+ryuH19fbrtttu0++67l28bHh7WAQccUPW4ypGREZ155pnlAVlf/epXtfPOO9e0hsHBQX3qU5/SrbfeOuZ1X/7yl8ccI1mt0bMWX//61ysuUH71q19VXKgtXLhQl19+efnvs2bN0jXXXLPSJtEkzJs3T+973/vG9BzMnTtXX/nKVxr2fjfZZJOKr8277767Lo+77rrrarfddqu47e677644tateoXZF1Xpvnn32WUlxP0Pp62P0oLpG2HTTTcec6HXTTTfpqaeeKv/96aef1s9+9rOK+1x00UXadNNNG7q2erv44ovHHKX7uc99bsywy29/+9u68847NXfuXJ100kljHufss8/WPffcM+b20g5rSTabrdjxANBaJjedC4Ak6dxzz9XChQs1f/58SfEzzmeccYYk6fOf/3y5HMQYs9LhdaW3qdZ8vWjRovLrjzzySG2//fYVr19zzTX161//Wuedd57OP/98vfrqq1qwYIEOOuggveUtb9FOO+2kmTNnasGCBbr77rv1xhtvaPr06fr6179e9TjcL3zhCxoeHl7lWoaHh/Xss8/qvvvuG1M3PmfOHH3lK1/RJz/5yXE/5nqZMWOG7r77bh1zzDH65S9/qSiKtOeee+p973uf+vr6dOutt+rFF1+UFA8Pu+6661batP3000/rsssuK/993rx5Y+4z+t9ZiktBRg8Du++++3TzzTdLqj4f4MYbb9Rjjz2mKIq0aNEi/fnPf9Zjjz025n7777+/vvvd72r11Vdf9SeiyvsYHbBW9nWUyWTGfZzLLrtMTz/99LiB8Nxzz9X06dO1ww47jLnQu/jii7X77ruX3/YPf/iDdt55Z+2444565JFH1NPTox122GHMM/alxyx9D5W+z55++ukx77/0MZVOFZPiadRve9vb9Le//a1829lnn62HHnpIDz74oObPn693vOMd2mabbfTggw/qhhtuKH8Pj3bnnXdqYGBgzL/vZJx88smaPn26TjnlFC1dulSDg4Pacccdddhhh8kYo5tuuqk8qHH69Om65JJL9OEPf7jiMUr/Bg8++OCYxz/33HO14YYb6uyzz9YZZ5yx0o/jpJNO0oMPPqgHHnhgzNfl6K+P8847b9Kfl7XXXlv33XefPvShD5W/Z/75z39qs80208EHH6xp06bpoYce0p/+9CettdZa+vnPf161v2dwcFB77723dt55Z22++eYKgkCPPvpoxfdhb2+vrr766qrzOQC0CAdg0jbYYAMnqerLM888M+HHGe8xVny56qqrVvo4w8PD7rrrrnMnnnii22qrrdxaa63l0um0y2azbu7cue6AAw5w559/vnvttdfGfYwZM2ZMeD2e57m+vj639tpru1122cV97GMfc9dff71btmxZzR9rtZdVueuuu9xHP/pRt8kmm7hp06a5bDbr1l9/fffe977XXXXVVS6fz6/yMe65555Jr+ucc86peIwLLrhgUm+fTqfdzJkz3Wabbebe/e53u//+7/92f/nLX1a51pU59thjp/x1NNoee+wxoY/h2GOPrfr2TzzxhDvqqKPcmmuu6YIgcLNnz3a77babu+SSS1wYhit9/NL30Mq+z8b72njllVfcpz/9abfFFlu4/v5+53me6+3tdW9729vcmWee6RYtWuScc+6qq66a9L/vVCxevNhdcMEF7oADDnDrrLOOy2azLpvNunXWWaf8fVla02T/DTbYYAPn3Kq/t+65556Vfl2M/lzW8nm5/fbb3fHHH+/e+ta3uhkzZjjf992MGTPcTjvt5L7yla9U/Tiff/55d9VVV7nTTjvNvfOd73Qbb7yxmzVrlvN93/X09Lj11lvP7bPPPu7cc891L7/8cs3/HgAayzjXoWf+AQAAAGgaeiwAAAAA1IxgAQAAAKBmBAsAAAAANSNYAAAAAKgZwQIAAABAzQgWAAAAAGpGsAAAAABQM4IFAAAAgJoRLAAAAADUjGABAAAAoGYECwAAAAA1I1gAAAAAqBnBAgAAAEDNCBYAAAAAakawAAAAAFAzggUAAACAmhEsAAAAANSMYAEAAACgZgQLAAAAADUjWAAAAACoGcECAAAAQM0IFgAAAABqRrAAAAAAUDOCBQAAAICaESwAAAAA1IxgAQAAAKBmBAsAAAAANSNYAAAAAKgZwQIAAABAzQgWAAAAAGpGsAAAAABQM4IFAAAAgJoRLAAAAADUjGABAAAAoGYECwAAAAA1I1gAAAAAqBnBAgAAAEDNCBYAAAAAakawAAAAAFAzggUAAACAmhEsAAAAANSMYAEAAACgZgQLAAAAADUjWAAAAACoGcECAAAAQM0IFgAAAABqRrAAAAAAUDOCBQAAAICaESwAAAAA1IxgAQAAAKBmBAsAAAAANSNYAAAAAKgZwQIAAABAzQgWAAAAAGpGsAAAAABQM4IFAAAAgJoRLAAAAADUjGABAAAAoGYECwAAAAA1I1gAAAAAqBnBAgAAAEDNCBYAAAAAakawAAAAAFAzggUAAACAmhEsAAAAANSMYAEAAACgZgQLAAAAADUjWAAAAACoGcECAAAAQM0IFgAAAABqRrAAAAAAUDOCBQAAAICaESwAAAAA1IxgAQAAAKBmBAsAAAAANSNYAAAAAKgZwQIAAABAzQgWAAAAAGpGsAAAAABQM4IFAAAAgJoRLAAAAADUjGABAAAAoGYECwAAAAA1I1gAAAAAqBnBAgAAAEDNCBYAAAAAakawAAAAAFAzggUAAACAmhEsAAAAANSMYAEAAACgZgQLAAAAADULkl4AAKzIOSc5K+ciyVrJRXLOytlIzkaSC4v/tcX7OUlOclLxfyTn5Ip/NqXHLf7XqBonyaxwH1f13is+jiu/xsgYs/y1ZtRtxpOMJ2N8yQtkPF/G80fdVvyv8UY9BgAA7cO4+DcyADRFOTTYULJhZWCwoZwLJRsHCTkn55xMOSLE1+rOGI29iB/151EX9ZIafqG+/Mfo6HCzPOCUXlcKQKZ4Wym2uHL4MDLGkzxfxgSjAognGV/GC2S8gPABAGhJBAsADeFGhwcbyUWF+MXmR+02uIo9geWBwSsGBVP+b6deSC/fbRn1X2fjKOVcMXiUcpIp73AYPy35KXmlsFH8r/GocAUAJINgAWDKnHPx7oINR70U5KK8XBQWy5SspOU7DaZYEiTjqZMDQ70tDyC2/Hk1btROTqnUygtk/LSMnyrudowKHnyuAQANRLAAMCHLQ0ReLirIRnm5MCcXxeVMpe4EJ1WGB8p2Gq5UXja652T5v0cc5oznywRZeUFaxkvJ+CnCBgCgrggWAMZYeYiIyvcrNR2z+9Cayj/eXVTsa6n8tyNsAADqiWABIC5hiiYaInwuPNtcvMMx0bCRLjaT828OAFg5ggXQhZyN5MKcbJiTC0dko9yYC0xCRHcZN2x4caO4F2TlBVmZIM3XBQCgKoIF0AXiU5lysoUqQcKY4tGmXCyiUilsOBvJFJvwRdAAAIyDYAF0oFUHCZ9aekzaeEHD89MyQY+8IEPQAIAuRrAAOoBzrljaNBwHiTAfz5CQCBJomNFBQ86WJ4yXg0YqKxNk4lPCAAAdj2ABtKlSn0QUDsvlh+L5Ec6WB6jFpU1c0KF5xgQNGRk/kEn1yUsVS6c8P+llAgAahGABtIn4CNhQNhyRLcQ7E86G8aTm4o5EvDvBrgRag3NWsqFUKsPzgrg3I90rL8iyiwYAHYZgAbQw51x8DGw4IlsYki3k4pN7pDhIsCuBNlEKxiqX6PkyQVpeqjcumfIzhAwAaHMEC6DFxP0S8a6ELQzJRYV4noTxZLyAxli0veUlU/GOmzFGxk/JpPrkp7IyQZbADABtiGABtIDSzoTND8kWBmTDgiQ3Kkx4hAl0pDhk2LhcykVykjw/LS/dJy/VW2z+5msfANoBwQJISLlnojCkKD8YT7t2Nn6m1k/xjC26knNWikI5F8oYT8bPLA8ZfoqQAQAtjGABNJmzYVzmlB+ULYwUy5yMjJdiZwIYxdlSuVQcuL1UVl66X17QI+MHSS8PALACggXQBM5a2TDumbD5oeUNrH6KnglgFUb3ZBjn4qF8qd7i6VI9HGELAC2CYAE0SKkJO8oPyRUG5aJC/Aov4JhNYIpWPF0qbvrulZ/ulQl6+L4CgAQRLIA6czaK+yZyy+TCkXjOhFeafE3fBFAvccgoSDaSk5EXZORnpslL98aHHgAAmopgAdRB6VSnKD8gl493J1yxb4IyDaDxnLNyUUHGRfH3XbpffrqPU6UAoIkIFkANnLNx30RuIJ6G7aL4mVKP02uAJJRKpZwNRzV8T5OXohcDABqNYAFMgYsKivIDsrkBuSgf30gjNtAySvMxXFSQkZPxU/GJUul+jq0FgAYhWAATFDdjDyvKDcjmh+LdCWZOAC2v1IsR72LEJ0r5mX6ZVA/fuwBQRwQLYBXiZuxBRSPLZMOcjJzkpSSP3QmgnZSOrVVUiGfH+Gl56WlxyKBMCgBqRrAAxuFsGO9O5JbJRfm4GdtP8wwn0AEqmr3LAWMag/cAoAYEC2AFLioUy52WyUaFYrlTmt0JoAONLpPyvJRMZpqCzDQZP5X00gCg7RAsgCIb5RXllsUN2TaMSyM43QnoCstPkyrIeIG8dH9xB4MnFQBgoggW6Grl+RO5pbK5wVHHxTIZG+hGFQHD+PLSffIz05mHAQATQLBAV4pPeBqJdyjyg3LOEigAlFU2envxSVLZaTJBDz8jAGAcBAt0FeecXGE43qEoDEnOMX8CwEo5G0lRXpKRl+6Rl5kuL9XLzwwAWAHBAl2hvEMxskRRfkhGiudPcMQkgAkqBwxj5KV65GdnsIMBAKMQLNDxbJhTNLJENj8oORuf8ESgADBFccAoSEbyUn3FgEEPBgAQLNCxXFRQOLJENjcQ10r7qbiPAgDqYPkOhicv0y8/M0NekE56WQCQGIIFOo6zoaKRpYpyS+UspzwBaJxSk7eLiqdIZaYpyE5nDgaArkSwQMdwNirOoVgaD7YjUABoktHH1HpeSl5menyKFLukALoIwQJtzzkrmx9QNLxELspLxo/LnggUAJqsYpK3n44DRmYafV0AugLBAm3LOSebH1Q0skQuHJEzHlNyAbQE51zcf+GsTJCRn50hL90nY7yklwYADUOwQFuyYU7R8CLZwpCcVAwU/MIG0Fqcc3JRXsbZeMhez0yZIMsTIAA6EsECbcXZSOHIYtmRZcWTnjg6FkDrc85KYa54gtQ0BdmZMj79FwA6C8ECbSEuexpQNLxYNsrJeCkaswG0HWdDuSgf919kZ8T9F+y2AugQBAu0PBuOxIEiP0gfBYC2R3kUgE5FsEDLcjaMB9yNLGViNoCOU1ke1a8gO4vyKABtjWCBllNZ9pRnHgWAjlZRHpWZEc+/oDwKQBsiWKCl2MKIopFFsvkhyp4AdI2K42lTvQoojwLQhggWaAnLT3ui7AlA94rLo/KSMXF5VM8spncDaBsECyTKOScXDiscWigbjnDaEwBoVHlUkJHfM1teqpefiwBaHsECiXE2Uji8WDZX3KVg2x8AyuLTo3IyUjz7gt0LAC2OYIGmG7tLkeYkFAAYB7sXANoFwQJNxS4FAEweuxcA2gHBAk3BLgUA1I7dCwCtjGCBhmOXAgDqh90LAK2KYIGGYZcCABqH3QsArYZggYZglwIAGo/dCwCthGCBurNhTuHQG7KFIXYpAKAJlu9eZBX0riYvlU16SQC6EMECdeOcU5RbJju8SM6GUpCRMV7SywKAruCck8KcZDz5PTPlZ2ewUwygqQgWqIu49GmhbG6ZJCP5aX6hAUCTOeckW5BsJC89TUHvbHaNATQNwQI1s4WRuPQpHJHx0zKen/SSAKCrORvJhTl5qUyxNKon6SUB6AIEC0yZc07RyFJFI4skG9GgDQAtJC6NGhlVGjWTn9EAGopggSlxNoyPkc0tkzxf8lL8wgKAFhOXRoWSDeWl+xT0zpHxU0kvC0CHIlhg0mxheFTpU4bSJwBocc5GclFOnp+Jw0WqhyeDANQdwQITFpc+LVE0vFhylD4BQDtxzklRTpKRn50hv2cmJ/cBqCuCBSbE2VDh4Buy+QHJCyQvIFQAQBtyUYHSKAANQbDAKtkwp3DwdU59AoAO4ZyNT40K0pwaBaBuCBYYl3NOtjCkaOh1uSik9AkAOkj51CjPl98zW35mGj/jAdSEYIGqnHOKhhcrGlksJ8U7FfzCAYCOEvdd5CXn5PXMVEDfBYAaECwwhrNR8SjZpXEvBfW3ANDRXBRKtiAv3a+gb46Mx7RuAJNHsEAFFxVUGHxdtjDIUbIA0EXKR9IGPQr6VpMXZJJeEoA2Q7BAmS0MKxx8XS7KS0GG7XAA6DKlvgvjpxT0zpGX7kt6SQDaCMECcT9Fbpmi4YWSZT4FAHQz55xclJORkd8zS352Br8TAEwIwaLLOWcVDi2SzS2R5El+il8gAIDl8y6y0xX0zJHx2MUGsHIEiy5WMfTOT9GsBwCo4GwoFxXkp3sV9K7GYR4AVopg0aVslFc48JpsYVgmoEkbAFBdPExvRF6QVdC3Ok3dAMZFsOhCNhxROPiaXJiTgh5KnwAAK1XR1N23OpO6AVRFsOgytjBUPPmpQJM2AGDCyuHCC+T3rSafE6MArIBg0UWi3IDCodcla4vHyRIqAAATF58YlZeR5PfOkZ+Zxu8SAGUEiy7gnFM0slTR8EI5ScZP84sAADBlLspLzhaPo53J7xQAkggWHc85p2h4saKRRZLxZPx00ksCAHQAFxUkF8nLzFDQO5twAYBg0cnKMypGFktewDGBAIC6cjaUokI866J3joxh1gXQzRhc0KGctQqHXpfNLWNGBQCgIYwXyMkoGlkqWaugbzWOLwe6GFebHSgefPe6ovyAjM+MCgBA48S/YzKK8sskVwwX7JADXYlSqA7jooIKg6/JFoZkgizb0gCAplg+SK+nOEiPnj6g2xAsOkg8TftV2TAXT9MmVAAAmmj5IL20gv41mNINdBmuPDtEKVS4cISdCgBAIowxUpCVG/VEF4DuwdVnB7BhKVTkpKCHI/8AAIkZGy5Gkl4SgCYhWLQ5G+YUDr5SDBVZQgUAIHFjwkWBcAF0A4JFG7NhrrhTkSdUAABayvJwESocJFwA3YBg0abKoSIiVAAAWlMcLjJyUaEYLoaTXhKABiJYtCFCBQCgXSzfuSgoHHyNnQuggxEs2kzcU0GoAAC0j8pwQVkU0KkIFm3Ehvk4VNBTAQBoM2N6LjgtCug4BIs2YSNCBQCgvVX0XAy8xpwLoMMQLNrA8uF3HCkLAGhvDNEDOhfBosW5KFQ48JpcOEKoAAB0hIpwMfiqbJRPekkA6oBg0cKcjRQOvSYbDjNRGwDQUcrhIswrHHxNLgqTXhKAGhEsWpRzVuHg64ryQzLsVAAAOlApXNjCiMKh1+RslPSSANSAYNGCnHMKh95QlB+Q8dMyhn8mAEBnMsbIBBlF+UGFg6/LOZv0kgBMEVesLcY5p3B4oezI0jhUeH7SSwIAoKGM8WT8jGx+mcKhN+ScS3pJAKaAYNFCnHOKRhbLjiyR/BShAgDQNYznS35admSpwuFFhAugDREsWkiUW6ZoeJFkfBkvSHo5AAA0lfECyUvJjixWNLKEcAG0GYJFi4jyg4qG3pDkyfippJcDAEAijB9Ixlc0vEg2P5D0cgBMAsGiBdjCsKLB1+XkJEIFAKDLlZ5gCwffkM0PJbwaABNFsEiYDXPx+d02lPEzHCsLAIAk+Wk52XieU2Ek6dUAmACCRYJcVCgOBSowVRsAgFGMMTJ+Ri4KFQ6+Jhvmkl4SgFUgWCTE2UiFwdfjH5SECgAAxihP547y8YwLy3RuoJURLBJQGoBnC4MyAeVPAACMpzydOxxhgB7Q4ggWTeacUzS8WDa3rNhTwT8BAAArUyqLsvlBhUPMuABaFVe1TWbzg4pGFktewAA8AAAmyHie5Kdkc0sU5ZYmvRwAVRAsmsiGI4qG35CTmFUBAMAkxcNjPUVDi2QLw0kvB8AKCBZNEp9q8bpcFMr46aSXAwBAe/JTkosUDr4uG+WTXg2AUQgWTeCcVTj0umw4wglQAADUoPKkqDfkbJT0kgAUESwaLD4BaqFsfkCGUAEAQM3icJGRLQwqHFpIMzfQIggWDRaNLJXNLZX8NCdAAQBQJ8Z48UlRuaWKhhcTLoAWwJVuA9nCkKLhRZLxig1nAACgXoznS16gaGSxbGEw6eUAXY9g0SC2WPspZyWPE6AAAGgE46fkJEVDb8iGuaSXA3Q1gkUDOBsVT4DKSUzWBgCgoYyfLp6++JqcDZNeDtC1CBZ15pxTOLxQtjDECVAAADRB6aQoWxhROEgzN5AUgkWd2fyAbG6ZDM3aAAA0jTFGJsjI5pcpGlmS9HKArsSVbx3ZMKdoeKGcDM3aAAA0WdzM7SsaXixbGEl6OUDXIVjUibORwqE3mKwNAECSvJTkbPw7meF5QFMRLOog7qtYLFsYplkbAIAElYfnhcMMzwOajGBRBzY/KJtbIuOn6KsAACBhxph4eF5+maLc0qSXA3QNroJrZKM8fRUAALQY4/mSPEXDi2RD+i2AZiBY1MA5q2joDbkoT18FAACtxk9JpR5I+i2AhiNYTFG5ryI/yLwKAABaULnfojCscJh+C6DRCBZTZAtDsiNLJI++CgAAWpUxnoyfls0tk80PJL0coKNxRTwFLiooGlooJyfjp5JeDgAAWAnjBXIyioYXyoa5pJcDdCyCxSS54tnYNsrL+JmklwMAACbA+Gm5KCz2W9iklwN0JILFJEUjS2XzgzI+8yoAAGgXpX4LVxhWNLI46eUAHYlgMQk2zMU/jDxfxuNTBwBAOzHGk7xA0cgS2QJH0AL1xtXxBDlnFQ0vlGwkefRVAADQlrxAcjY+JYqSKKCuCBYTFJdADUkBJVAAALQrSqKAxiFYTEBFCRRHywIA0NYoiQIag6vkVaAECgCADkRJFFB3BItVoAQKAIDOQ0kUUH8Ei5WgBAoAgM5FSRRQX1wtj4MSKAAAugAlUUDdECzGQQkUAACdj5IooH4IFlVQAgUAQPegJAqoD66aV0AJFAAAXYiSKKBmBIsVRCPLKIECAKDLLC+JGlI0siTp5QBtiWAxiosKsrklkvEogQIAoMuUS6JyS2SjfNLLAdoOV89FzjmFI4vjHyR+OunlAACAJHgpORspGlos51zSqwHaCsGiyIUjsrkBGT9NCRQAAF3KGCPjpxXlB2QLQ0kvB2grBAvFDdvh8GLJWRkvSHo5AAAgQcbzZSRFw4tp5AYmgWAhKcotkyvEDdsAAAAK0nLhsKIcjdzARHV9sHBRKDuyRI6GbQAAUFRq5LYjS2nkBiaoq6+kRzdsGxq2AQDAaF5K1oZxSRSN3MAqdXewCEdkc8to2AYAAGOUGrltbiAumQawUl0bLOIJ24tp2AYAAOMyni9JCmnkBlapa4NFlFsWHyNHwzYAAFiZIC0XjijKLU16JUBL68pgQcM2AACYqLiR25cdYSI3sDJdd1VNwzYAAJg0LyVrCzRyAyvRfcEiyscN216Khm0AADAh5Ubu/IBcOJL0coCW1FXBwjmnaHiJ5CKJhm0AADAJxgsk5xSNsGsBVNNdwSIckS0MSBwvCwAApsJPyxaG4wNgAFTommDhnFM0skRyjuNlAQDAlBjPl5wUjSyRcxw/C4zWNcHCFobiZxdo2AYAALUI0nKFYdn8YNIrAVpKVwSLyt0KP+nlAACANmaMJ2dMvGvB0DygrCuChc0PyBWGGYYHAADqwvhpuTCnKLcs6aUALaPjg4VzNn5GwRiG4QEAgLqIdy082dxSORslvRygJXT8lXY0skwuzDEMDwAA1JXx07JRTtHI0qSXArSEjg4WzkbxMwnGY7cCAADUlTFGxgtk80vlokLSywES19FX21FuqWyUZ7cCAAA0hpeSjQoKc+xaAB0bLFwUyuaWyXg+w/AAAEBDxLsWKdncMtkwn/RygER1bLAIi7sV8lJJLwUAAHQyL5CzoaLcEjnnkl4NkJiODBYuKsjmlsp4KXYrAABAQxlj4kbu3IBcxK4FuldHBosot0zOhpIXJL0UAADQDYwv5yLmWqCrdVywcDaUzQ/IeAG7FQAAoCnKvRb5gbgUG+hCHRcsotyAbFSgtwIAADRXuddiIOmVAInoqGARz61YJmM8disAAEBTleZauNwyuShMejlA03VUsLD5wbhpirkVAAAgCV5K1hbotUBX6phg4ZxVVJ6yzW4FAABoPmOMjAlk88vkbJT0coCm6phgYfODcmFOxqe3AgAAJMhPyUV5di3QdToiWDjnFI0slTNGxnTEhwQAANqUMUbO+LK5ZXLWJr0coGk64ircFobkopwMvRUAAKAFmOKuhS0MJr0UoGnaPlg452RHlkrOsVsBAABagjGenDFxRYVj1wLdoe2vxF1hWFFhmJOgAABASzF+Wi4ckc0PJb0UoCnaOlg45xTllsrIyXh+0ssBAAAoM8aTk4lPrXQu6eUADdfewSLMybJbAQAAWlRp18IVhpNeCtBwbR0sovyA5Cy7FQAAoCUZz5OcU5QbYNcCHa9tg4WLQtn8oOQFSS8FAABgfH4qPsHSFpJeCdBQbRssovygnA0JFgAAoLUZX86FsjmOnkVna8tg4ZyTzQ9IxpMxJunlAAAAjMsYI2MC2fwAA/PQ0dozWBSGZcOcjJ9KeikAAACr5gfFgXkcPYvO1ZbBIsoPxEfMMhAPAAC0gdI1i80vo4kbHavtrsxtlI8HzbBbAQAA2omfli2MyEW5pFcCNET7BYvcoJwLJcMRswAAoI0YT85ZRTRxo0O1VbBw1srmB2RMQNM2AABoK8YYGS+QKwzEJ1sCHaatgoUtDMlFecnniFkAANCGvKA4i4smbnSetgkW8RGzyySJpm0AANCWjDFyxiiiiRsdqG2u0OMj2kYkP530UgAAAKbMeGnZQk4uHEl6KUBdtU2wiPIDci6S2K0AAABtzHiejJyi/EDSSwHqqi2u0p2N5PKDMh5N2wAAoAN4gWx+UC4qJL0SoG7aIljYcDj+xvOYXQEAADqAF0g2kg2Hk14JUDftESzy8XnP7FYAAIBOUGritvlBmrjRMVo+WLgolC0Mx8keAACgQxgvKE7iphwKnaHlg4UNhyQbESwAAEBnMb6cs7IFZlqgM7R0sIhnVwzKGUMZFAAA6CjGGMl4lEOhY7R2sIgKsoURGXYrAABABzJeIBfl5MJc0ksBatbSwcIWhuSclYyf9FIAAADqznh+XKFBORQ6QMsGi1IZlIxHGRQAAOhYxvOL5VA26aUANWndYBHm5KIcZVAAAKCzeYGcLcgVRpJeCVCTlg0WpTIo41EGBQAAOpcxnuScosJg0ksBatKSwcI5K5sflDHsVgAAgC7gBXKFITkbJb0SYMpaM1gURuRsQfIJFgAAoAt4wfKhwECbaslgERUGJefirUEAAIAOVzqoxuYph0L7arkrd2cjucIwk7YBAEB38VKy4bBcFCa9EmBKWi9YhDk5G0o0bQMAgG7i+ZKNZENOh0J7arlgEYXDlEEBAICuUy6HCumzQHtqqat355xcYYhJ2wAAoDt5gVxhWM4yLA/tp7WCRZSTiwqUQQEAgO7k+bJRQS6iHArtp6WChS2MxOPsKYMCAABdyBhPRlLEFG60oZa5gnfOyRaGJOOVawwBAAC6jvHkCoNyziW9EmBSWidY2IJcmJPhmFkAANDNvEAuKshFuaRXAkxK6wSLMCc5S+M2AADobsYrVnJQDoX20jLBwuaHJIkyKAAA0NWMMZIx8elQlEOhjbREsHClYTCUQQEAAMh4gWyYkyxTuNE+WiJY2HAk/sYhWAAAAMSl4Y4p3GgvrREsCvGEScqgAAAARk3hLgwlvBJg4hIPFs7ZeNo2Q/EAAACW84J4xpeNkl4JMCHJB4swLxdRBgUAAFDBCyQbylEOhTaRfLCIcnJyMkzbBgAAKCuXQ4XMs0B7SPxqPj6jmd4KAACAFTnjyYYjHDuLtpBosHDWyoYjMvRXAAAAjGE8Xy7KS/RZoA0kGyyiXNyQRLAAAAAYy/iSjeQiyqHQ+hIOFnmJ/goAAICq6LNAO0n0it4WRiRmVwAAAIyLPgu0i8SChbPxNEljKIMCAAAYD30WaBfJBYsoT38FAADAqhT7LCx9FmhxiQYL+isAAABWrtRn4eizQItL7Kqe/goAAICJoc8C7SCRYEF/BQAAwMQZz5cL85INk14KMK5kggX9FQAAABNnfMlFslE+6ZUA40osWNBfAQAAMDH0WaAdJHJlbwvD9FcAAABMQtxnMUyfBVpW04OFc1Y2zNFfAQAAMAnxPIuC5JhngdbU/GARFSRnJcqgAAAAJs54ctbG11JACyJYAAAAtIG4N9URLNCyEggW8WkGhh4LAACASTESJ0OhZSUSLByhAkCH+Mr/O1+77f1u7bLHgVrvzVtrg0230WEfPFYP/PlhSdLjf/u7jvrIx7XVDntoy+1319yNttTu+7xH1934k4rH+c1985SZMbfqy4JXXi3f74STPl31Pu9+39FN/bgBJMMZTy7M0cCNlhQ08505Z+WiPI3bADrGjT/+mT7zqZN0/EeOUj6f15Ef+bh+cdtd+s198/S3h3+vx//2D/1u3v2ad+8vtd66c3XzLbfpiA+foPsfeEjPPve8zv7sv5cfK5PJqL+vd8z78LzK54CmT5+mVFD543vG9GmN+QABtBTj+XK22MBtmnoZB6xSc4NFVCgOxuMbAUBn+MoXz9a7D9xXkpROp7Xvu/bUL267S4ODQ3ri709q2vR+nfKJj2q9dedKkg55z4Ha6E0b6l/PPKtvXfG9imBxxPvfq+9cduEq3+ePr7tSe7xz14Z8PABanPHkolAuKshwPYUW0/RgQeM2gE5yyLsPKP950aLF+slNt0iSNt1kY+24/bbq6+vVgfu9q+JtVpszS/965lkNDQ9X3D5//os64aRP6+G/PKZ8Pq/ddtlJZ53xKW24wXoV97v1tjt13oWX6h9PPqU111hN7z/sPTr1kx9TEHCRAXQ6Yzy5UgN3qifp5QAVmnqFT+M2gE518r99VnM33kr3/f6P2nXnHXTbzTeor0pZk3NOzz43X5K0+267lG9PpVJavGSpzvn8mbr/t7dr7z3fqSuvuU7v2Oug8v0lqb+/X5J0041X6547btJrry/UWZ//so756MkN/ggBtAoauNGqjGti909h2QJFhSF5QbZZ7xIAmub1Nxbq6OM+qXt/+3ttuMH6uu/uW7TG6qtV3OdnP/+lPnjMiVpj9dX06ztu0iYbbzTuY6278VZyzumjxx6tyy76etX7XXjx5Trr81+WJN1758+0y0471PeDAtBybJiT56eVmr4OT9aipTRtx4LGbQCdbrU5s/WNc8+RJD373PO6/Dvfq3j9E//4p079989pqy3fpnvvvHncUFF6rNmzZkqS7n/gz+Peb5M3L3+MP/5p/PsB6BwVDdxAC2lesCg1btNfAaBDWGv14ksvV9y26SYbl//85D+fLv/5nt/8Tvsc+D4decRh+t3dt2rjjTbUE39/svz6199YqJGRkYrHKp0GFUXLLx5eePGlqveJ72dr+GgAtA0mcKNFNTVY0LgNoJMsXbpMe+1/aMVtL760oPzn1VdfTdZanfv1C/SJU0/X96+8VOd97UvKZDKSpEMOP6Z838/955d12x13l/++ZMlSvbFwkSRp6y03L9/+9p32qnh//3rmufKfR98PQOdiAjdaVRODBY3bADrPc8+/oKuuuV6SVCgU9OVzz5Mk9fb26ITjjtZ3rvqBvvTV8/TGwkX68PEnaZ03bVF+mf9C5e7Dlddcr1wuJ0k678JLZK1VT09WZ/77KeX7LFs2oO9f90NJ8SlUV1z5fUnSO3bZUfvsvXvDP14ArYEGbrSipjVv07gNoNPkcjl97gtf0R/++IDCKNSCBa+qr69Pu+68vT77mdP01s021UWXXqEz/+OL4z/GkhclSTfferu+e/W1en7+CxoYGNDSZQPaZaftdc7ZZ2rbbbYq3/+zn/+S7vvdHxVGoZ559nnNmjlDh773IJ1z9pnq7+9r9IcMoEXYMCcvSCs9fW7SSwHKmhIsnHMqLJkv56yMn270uwMAAOhopTKo9Mz1iqVRQPKa85XoIjn6KwAAAOrDeJKzcjZMeiVAWVOu9J0NadwGAACoF+PFT9oSLNBCmhQsIsk5xa1GAAAAqEXpMBxnmWWB1tGcLQQbyokToQAAAOrFSBw5i5bSlGBh+aIHAACoK2cMwQItpTk7FlGe/goAAIA6MsaTs3k1aXIAsEoNv9p3zslGBY5CAwAAqCfjFftYbdIrASQ1Y8eCE6EAAAAagCNn0Voav2Nhw3iLjmABAABQP8ZIzhEs0DKaEiyMqP0DAACoJ2OKV1gEC7SIJgSL+HxljpoFAACoLyPJRgQLtIbGB4soL8dgPAAAgLpzxotP3wRaQBOCRYHdCgAAgAYwxsjZAkfOoiU0NFi40kkFNG4DAADUn/HknJVclPRKgAbvWFgrORefWgAAAIA6K50MxSwLJK/BOxaRJGZYAAAANAQ7Fmghjb3id1Gx5o8dCwAAgEYwisvPgaQ1dsfCWhlx1CwAAEAjLJ9lwY4Fktf4HYuGvgMAAIDuFu9YECyQvIafCsVeBQAAQOM4LR9IDCSpwaVQITsWAAAAjWSMZJm+jeQ1thQqYoYFAABAIxl57FigJTS4FCqkcRsAAKCRjImrRJi+jYQ1LFjEU7etOGoWAACggYyR5JhlgcQ1bseCqdsAAABNwPRttIYG7lgUp243uI0DAACgqxkvLoNixwIJa9xVv7PxFzk7FgAAAA1l5Ji+jcQ1bsfCRnRXAAAANBjTt9EqGrtjIXEqFAAAQIPF07fZsUCyGtgAwZFnAAAAzcO1F5LVwOZtRykUAABAEziJU6GQuIaWQgEAAKBZuPZCsho7IK9RDw4AAIDljOFJXSSugTsWTkzdBgAAaAYTDycGEtTAYBGRKwAAAJrEUQqFhLFjAQAA0OaMMcVrLyA5De2xMAQLAACAJoh7LBzhAglq7KlQ5AoAAICmiEMFwQLJaUiwcM4Vpz+SLAAAABrOGEmOcigkqkE7FnxRAwAANJUTwQKJakywKG3FGXYsAAAAGq+4Y8GTu0hQg4KFLX5dEywAAAAaL77mcgzJQ4Ia02NBYgYAAGgeU2zephQKCaIUCgAAoO1RCoXkNe64WQCYgvt+/0dts/PeysyYq8yMuVpt3bdoy+131+x1NtGmW+6sfzvj83r9jYWSpA8ff5LW3XgrZWbM1Z77HaJCoVDxWMeecKrW3XgrrbvxVvrw8SeNeV93/OoeZWbM1Xve/+GmfGwAAHQyggWAlvLOd+ysh//46/Lf99z9HXr0wd/q0Qd/qzAs6FtXXK0jjzlRkvSDqy7TJ044VpL0h/sf1Jlnf6nisb73nYv1iROO1SdOOFY/uOqyMe/r+9f+UJJ09z2/1UsvL2jUhwQAQFcgWABoC3PXWVuHHXKwpHhXY8Err465z2XfvkrX//CnE3q8xYuX6LY775YkRVGk6274Sf0WCwBAF2KOBYC2EYZR+c9uVIPi+w45WH19vZKkk//ts3rs8SdW+Vg//MnN+vBRh2utNdeQJH3/+h9NeB3fvOw72mqHPbTpljtp+3fso132OFD/+3/Ld0ReePElfeRjp2jDt2yrt779Hdp59wP0o5/cXPEYt915t7bZeW+ttu5b9N4PHKN/O+Pz5fKvbXbeW/l8fsLrAYDluAZDchq3Y8HXNYA6euIf/yxfnB/23oO09lprll/3tre+RZf+39clSUNDw/rgMSdqyZKlK32871/3Qx334SP1wcMPlST9/R//1J8efGiV67jjV/fojM+dozXXWF3/+Osf9eDvf6WPHXe0bi/ufgwMDGqfgz6gG3/8M13w9f/WXx+4V0uXLdOHP3qybrr5F5KkZ5+bryOOPkF/e+IfuvqKb+rmH12jx//29/L7ePiPv1Y6nZ74JwcAgBbQoFOhyBUA6uM3983TmzffQW/fcU9lsxl957ILdN33Lh9zvyMPP0wnffx4SdJTTz+jj5306YpdjdH+/uRTGhwa1jZv31IfPvID5duvKfZcrMxjj8cB4J9P/Uu33Xm3RkZG9OGjPqB//9QnJUk3/OgmPfPsc5KknXfaXqlUSttsvaUk6fyL4l2NK678vvL5vIwx2u9de8oYowP332einxIAGB8XYEgQPRYAWtoe79xVjz74W+2843Z68aUFuuqa68cNDN849xzttMO2kqRbfnGHvnH+xVXv9/1rf6gPH3W4JGmrLTfXllu8VZL0o5/8XLlcTpL03PMvqHf2+hUvzz3/gt6xyw4yxujlBa/osCOO1dpv2kIfP+V0vfUtm0qS/vLo4+X3c/BhH9IOu+2rB/78iNZcY3UtLu6iPPnU05Kk/v6+8s7EnNmzavo8AQCQNIIFgJbX09OjL//X5yRJv//Dn/SD639c9X6pVErXfe9yrb7aHEnSF7/6Df3mvnkV97HW6oYf/VQXX3aFNt1yJ2265U564YWXJUmLlyzRz2+9XVLcwxFFUcWLc04777i9bv/5jXrPwfurt7dHQ0PDuvHHP9MB7/2gnHPKZjPl9/WLm67TA7+7S08++kc9/89H9OiDvy0/NgAAnYZgAaAt7PHOXbXDdttIks678BJZa6veb9256+ia714i3/cVRZHu+/0fK15/169/oy02f6v+9cSf9eSj9+vJR+/XXx/8jYIgkCRdc13cxL3hBuspt+TFipcNN1hP9/729+rv69OPr7tSLz/zWPm422efe15vLFykzd+6Wfl9vfDiS+U//+L2u3T65/5LkvSWTd4sKe7HKDVpv7FwUc2fIwAAktSYYGHi+Y8AUE9nfPpkSdKT/3y63Ahdzd57vlPnnH1G1dd9/9of6tijP1hx2xqrr6YD9ttb0qpnWjz0yKP6r//+Hw0PDyubzWrLzeMyqg3WX1erzZmtIw8/VBtusL6kuJdCigPE/5x3kXbZcXtJ0sc/9hGl02k553Tn3ffKOac77rp7Ip8CAFg5LsCQoAbtWJAsAExNafJ2yb2//b123v0ASdJ7332ANnnzRpKkk//tLO3/niN0+Xe+p8u/8z0de8KpFY/z2dNP07sP2q/itpNOO1M/u+U2nfOVr2veHx8o337djT/RAw8+LCmeabHHvu/VPb/5XdX1bbn5W5XL5bTtLvtoqx320H9/7X918IH76qYbvydJymaz+uXPrteh7zlIN99yuzbdcicdfNhR+uixR+sD73uvpDiE/PDa7+htb32Ljj3hVB182Ie08UYb1fJpA4AiLsCQHOMaUOxrw5wKS1+U/LSModoKAFblmmtv1Iknf0aSlFvyYsKrAdBunHNy4YjS09eRl+pJejnoUlz1AwAAAKgZwQIAEvadq36gL37l6+W/M3kbANCOGlQKlY9LobxAxvPr/fAAAAAYxTkrF+aLpVDZpJeDLtWQHQtjaBwCAABoJmNK/wMko0HHzZriFzZDoAAAABrOOcWnchIskJzGHTcrQ64AAABoGiPDcbNIUIN2LEoPS7IAAABoPFd8XpdzeZCchn31GUqhAAAAmoNSKLSAxjVvG49cAQAA0FQECySngTsWnhzJAgAAoAmcjPE4mROJalwhnvHElgUAAEATOCd59FcgWQ0OFgAAAGi0uMOCay8kq7HBov5DvQEAADCG40ldJK6BwYKTlAEAAJrCiVIoJK6hzdsAAABohuJxs0CCGnwqFAAAABotno3nJ70MdLkGbiuQmgEAAJqHay8kix4LAACANuckZlggcQ0shfLlJDlOhgIAAGgoI0mGUigkq3E7Fl5p+iPBAgAAoFGcc/GOBadCIWEN3bGQDLMsAAAAGsrFT+ayY4GENbDHwi8OybMNexcAAABdz1lJXvFJXSA5jW3ephQKAACgsZyTjGFAHhLXwFIoI+MFNG8DAAA0lCsOJuZUKCSrsdHWC+ixAAAAaCDnnIwXcNwsEtfgYOHLiB4LAACAhnFW8oOkVwE0Nlh4XiDHthwAAEDDGInGbbSExu5YGI9YAQAA0EDxDAuCBZLX0GDB9G0AAIDGYuo2WkWDeyyYvg0AANAoTN1GK2n4jgXTtwEAABqFqdtoHQ3vsYinbxMsAAAA6q48dZsdCySv8c3bxpM4chYAAKD+ylO32bFA8hpcCmVk/LScI1gAAADUm3NWxk+xY4GW0PivQj8lQykUAABA/RWDBdAKGh4s4iF5AAAAaATjESzQGhoeLIwXj5hnlgUAAED9OOfiqdvFay0gaY0vhfKCuO6PPgsAAIA6cnIyBAu0jObsWBAsAAAA6stZGeMRLNAymhAsil/wBAsAAID6cTaeuM1Rs2gRTTmbLD5ylh4LAACAenHOyXipePI20AKaFCxSMpwNBQAAUDfGWclPJ70MoKw5waK4RceuBQAAQH04SZ5PfwVaR5OCRaD4QDQAAADUqnTUrGjcRgtpWrAwxtDADQAAUBdOMhw1i9bSlGAhjpwFAACoH2cljppFi2nOjoUx8vyUHMECAACgds7KGD9+4hZoEc37avQz8ekFAAAAqImzkUyQ4ahZtJSmBQsviI9D42QoAACA2nlBJuklABWaFiyMl5JkJOZZAAAATFnpRKj42gpoHc0LFn5K8jzJRs16lwAAAJ3HRXHjtk+wQGtpXo+F8WT8NA3cAAAAtXA2Pg2KE6HQYpq3Y2GMDA3cAAAANaFxG62qqWeU0cANAABQOxq30YqaGixo4AYAAJg6GrfRypobLMoN3JRDAQAATFpp4jaN22hBzR3XWG7g5mQoAACASXMRjdtoWc3dsaCBGwAAYMpo3EYra+6OhWjgBgAAqAWN22hVTQ8WNHADAABMHo3baHXNDxY0cAMAAEwejdtocU0PFjRwAwAATAGN22hxzd+xMEZekKWBGwAAYDJsJJPK0riNltX8HQtJpth0RAM3AADAqpWumbwgm/BKgPElEiw8PyN5vkQ5FAAAwKqV+ys4EQqtK5FgIc+P+ywswQIAAGCViv0VNG6jlSVTCkWfBQAAwMTZSCbVQ38FWloyOxaizwIAAGAilvdXUAaF1pZYsKDPAgAAYALor0CbSCxY0GcBAAAwAZb+CrSH5Eqh6LMAAABYNUd/BdpDcjsWWl4rSJ8FAADAWPRXoJ0kGiwMfRYAAADjo78CbSTRYEGfBQAAwErYSManvwLtIdkdC/osAAAAxucimYD+CrSHZHcsJHlBVhJ9FgAAAKMt76/IJrwSYGISDxYmyEheINkw6aUAAAC0DhdJxqdxG20j+WDh+XESJ1gAAACUORvKBOn4CVigDSQeLCTJS/dKohwKAABAKl4TOScv1Ut/BdpGSwQLE2Qk43HsLAAAgCQ5Gx9yk6K/Au2jNYKFl5IJMhw7CwAAIBWPmU0xvwJtpTWChTHyUr2SiyiHAgAAcJFMqo8yKLSVlggWkuSlsjLGiydMAgAAdCnnrJwknzIotJmWCRbGz8h4KYlyKAAA0M1sJM9PyfgEC7SX1gkWxsike2ngBgAA3c2GMqkeGa9lLtOACWmpr1g/6JGMkaMcCgAAdKHl07Z7El4JMHktFSxMkJHxAsqhAABAd3KRVBoeDLSZ1goWni8T9DCFGwAAdCVnQ3lBRsZn2jbaT0sFC0nyUvHWH8fOAgCAbrJ82nZf0ksBpqT1gkWQkTyfJm4AANBdnJUxnkzAUDy0p5YLFvKCuK4wohwKAAB0D2dDmSAt46eTXgowJS0XLIwx8tJ9khzlUAAAoCvEZVCRPKZto421XLCQin0WXkATNwAA6A4ukjGBvHRv0isBpqwlg4XxAnmpXoIFAADoCi4K5aWyMl4q6aUAU9aSwUJSObEzLA8AAHQy55yMnLw0ZVBob60bLIIeGT/FrgUAAOhsNoyH4qWYto321rLBwnieTKqPKdwAAKCz2TBu2vYYiof21rLBQpL8dK+cDOVQAACgI5WucWjaRido6WBhgqy8IC0XFZJeCgAAQP3ZUMZPxTO8gDbX2sHCmHisvYuYaQEAADqPjWRSfTKen/RKgJq1dLCQ4q1BY3yJcigAANBBnLNyMvIpg0KHaPlgYfy0TJCRs5RDAQCADhIV5AUpGcqg0CFaP1gYIz/dLzlHORQAAOgIzjk5Z+Wl+pldgY7R8sFCkrxUT7EciqNnAQBAB3BWxnicBoWO0hbBQl4QD42JGJYHAADan7MFeUFWxk8nvRSgbtoiWBhj5GX6JSNmWgAAgLbmnJNxTl6GMih0lrYIFlKxHMpnpgUAAGhzthDPrkhRBoXO0jbBIq5D7JecpYkbAAC0JeecnI3kpfuZXYGO0zbBQpL8dHGAjKXXAgAAtCEXLX+yFOgwbRUsjJ+Sl+4jWAAAgPYUFeSlemX8VNIrAequrYKFpHimhTFylqNnAQBA+3DOSsUDaWjaRidqu2BhgqxMkGUSNwAAaCsuKsj46fgIfaADtV+wMEZ+ZpoMk7gBAECbcM5Jrti0bdru8guYkLb8yi7XJrJrAQAA2oENZbxAfrov6ZUADdOWwcJ4vky6T86G7FoAAIDWZ0N56T6attHR2jJYSHETtzG+5GjiBgAArcvZSDImPoAG6GBtGyzKzU9M4gYAAC3M2UL58Bmgk7VvsCge1yYVj28DAABoMc45Gefig2c4YhYdrm2DhVRs4g4ycuxaAACAVlQ+YrY36ZUADdfWwcIYT352uoyzNHEDAICW4pyTc5G8zDQZz096OUDDtXWwkCQv1Sfjp+WifNJLAQAAWM4W5Pkp+RmattEd2j5YGM+Xl5kmsWsBAABahHNOzobxQDwvSHo5QFO0fbCQJD8zTR4D8wAAQKsoDcTLTEt6JUDTdESwMJ4vLz2NgXkAACBx8W5FQV56GgPx0FU6IlhIkp/pl+elJBsmvRQAANDNbCRj2K1A9+mYYGH8lEymn10LAACQGOdc3LSd6ZMXpJNeDtBUHRMspLjXwng+uxYAACAZLpKMx24FulJHBQvPT8tL97FrAQAAkhEV4gG+fibplQBN11HBQpL8zHQZ48XPGAAAADSJs5FkTDy815iklwM0XccFC1PctVDE0bMAAKCJony8WxFkk14JkIjOCxbGxHWNxoufOQAAAGgwZ228W5GZxm4FulbHBQtJMkFWXqpXivL0WgAAgMaLcvKCHplUb9IrARLTmcHCGPnZGRK9FgAAoMHiCon42oPdCnSzjgwWkmSCjLxMv1xUYNcCAAA0TpSXl+6VSfUkvRIgUUHSC2iUuNdihmxuMJ5r4aeSXhLGUwp+bvRLfJsp/X1ljORM/F/JxP8t/10Szx4BABrE2TCeW5GdyW4Ful7HBgtJ8oK0vMw0RSOLJC/gG77ZnJOsJOvigGBdHBKsZKyVouLr7ApvI1WECbPC3yuYUa8a/c9b/Ld2RpJn4r05z8j5xeDhmcrXGRFAAACT4pyL51ZkpssEzK0AOjpYSFKQnS6XH5CzBclPJ72czuOKYSEqBgTrZCInhW75bXLLdyLMqP9K8cX86Ot5b9Rfin+syBSli/9q5W1u7J9La5KLw41bMTsUA4XzJAVGzjOSH7/EfxaBAwBQnQ0l4zO3Aijq+GBh/JS8zHSFw29IXopv/FoUdyBMVAwNoZMKNr54LwWM0UxxN6BcnlTHz321x1rpw5vK5a1QfmUiSaGNd1bKoSd+Py4wcegIijsevqkMQACAruOck7Oh/OxMeexWAJK6IFhIkp+dJptfJhflJb75J87GOw9mxRBhR9cpLS8taqtyotI6K5Y76i9u+S6LyVspP+q13gphIyiGjXb52AEAtbMFeV6gIDs96ZUALaMrgoXxAnmZ6YqG3pBzjl2L8ZSCROhkCjYOE1GVEBF0wUX06BItf1T5VbWwYRSXTaU9uRRBAwA6XXm3ome2DIfDAGVdESwkyc/EuxY2ytNgVeJGBYl8lSDhme4IERO1krBhrKThSGZYcdDwjZQqBo2UV2wQ5/MIAB0hKsjz0/LZrQAqdE2wMJ4vPztTbuBVOWdlTMeO8Fg5Wyxryts4TESjeiMIEpNXLgWT4kThRvVsjAoaKU+uuKPB5xgA2pdzTnKhvMxqMl7XXEYBE9JV3xFeuk8mlZUt5GRS2aSX0zxRsbSpYGXyo3YljGm/3ohWVy1o2Lh0yuRteTfDpb24dCqgERwA2omL8vKCjPzMtKSXArScrgoWpjjAxoWvyNlIxvOTXlLjWCeTdzL5KA4TpYZrdiWayxSPrPUrdzPMUCQNRXHJVMaTTXtSin8XAGhlzlkZZ+NheJ18DQFMUVcFC0nyUr3yUj2y+SE5k+2sRm7nZApOJld8djwiTLSUKrsZJpI0FMkfiuITpjJxyRT/XgDQgsK44sFL9yW9EqAldV2wMMbIz86SLYzEg23a/TSHUgN23saBIiz2TBAmWp8xxe9AUx4saAaLOxkpb3nI8Pk3BICkORtKxijIzurePk1gFbouWEiSl8rGx8+OLJa8oD13LWxxd2IkincnrAgT7cwr9lqUyqVKPRlevIths+xiAEBSnHNyUV5+ZoZMqifp5QAtyzjn3Krv1nmcDVVY+rKsDdtrYmboZHLR8t0JqTgzQVx0dppi47esW36yVLa4i0HDNwA0jYvyMjJKTV+HuRXASnTljoUUD83ze9rk+Fk3eneieKoTuxOdb3Tjt3XlXQznF3sxsn78NQAAaBjnrGQjeb1zCBXAKnTtjoUUb22GAwtk80NS0IKN3K54stNwsdxJKpbMiEDRrZyTIsW7GJ7ksj5lUgDQQK4wLJPKKjVt7dZ+EhJoAV27YyG1cCO3K57sNGxlCk6Si5+1pvwFpYZvp3g+xnAkfyQq9mH4HFkLAHVEwzYwOV0dLCTJBJliI/ei5Bu57ahAERZ3KHwjefwwwwpKZVKeigHDys/ZeLp31pdLEzAAoBbLG7an07ANTFBXl0KVxI3cL8naKJlG7nKgiOIdCiNKWzB5dvlUdZf25Hp9OXYwAGBKSg3bwfS15fnppJcDtIWu37GQRjdyv9bcRu5SydNQcYfCiFIWTF3pyNrRjd4ZT7bXJ6gCwCSMbtgmVAATR7Ao8tLT5KUH44ncjW7krmjKLvZQcOGHehkdMEas/LyVy3qyPb4UUFYHAKsU5mWCrPzs9KRXArQVgkVRRSO3iyTTgE+Nc1LByRuO51BIoocCjeMZKaXlPRgjTq6nGDCY5g0AVTkbSUYKemjYBiaLYDFK3Mg9LZ7Ibfz67lpETt5QJDMSxSf6cMoTmmHFJu/BSH7OyvZ4cj0+u2QAMErcsJ2jYRuYIoLFKMYYBT0z5QrDslFOJsjW/qCuWPI0XJyU7RuGmqH5RgeMSPIGIrmck+ujwRsAyqK8PD9d3K3g5yIwWezxrcB4gfzeWZIrnl89VS5uoPWWhPIGIhmruDGbEhQkyRSDbWBkClbekoK8gbB8mhQAdCtnI8lZedmZTNgGpogdiyq8VJ/87DRFI0umVhK1YtkTjdloNcbEQTdyMkNWfr5AeRSArlWeWZHul5+ZlvRygLZFsKgiLomaJVcYmVxJVOn42MGIsie0B99UlkflnWyfL6XYzATQRaK8PD8lv5cSKKAWXD2MY9IlUZGTtyyUtzSSsY6yJ7SP0eVReSt/cSgzGManmAFAhxtdAsXMCqA2BIuVKJVEuSivcQeUOyczEslbXJAZtnGDbOBRToL2UwoYJt698JaEUsEmvSoAaJhSCZRHCRRQF5RCrcQqS6IiJ28wlBlxknFMzUb7G3V6lMlb+QUn2+vJ9dJ7AaADUQIF1BU7FqtQtSSKXQp0umq7FyG7FwA6ByVQQP2xYzEBFadEOU/+kJUZjuJXskuBTjV69yJn5UdxY7fLEKIBtDdOgQIagx2LCSiVRHkuLW9xTmYoiqdmp7jAQhcoHU1rnbylkbyBSLI0dgNoY5RAAQ1BsJgI52SGCgqGApmC5ALHiU/oLsbE5X6+ZIZo7AbQviiBAhqHYLEqkZUWLpPeWCIjTyabkmTlxDO26EJevHthClb+klBmiGNpAbQPToECGosei5XJF6SFS6WRvJQKZHxfngtkbRQ/4+Hx6UMXMib+yRFJ3oCVi5xsXxCHDgBoYY4SKKCh2LEYz3BOem2xNFKQMmnJ9yVJxnjyUr2SJOeiBBcIJKh0apQvmSErb2kohexcAGhdzoYycvJ7ZlMCBTQIwWJFzklLB+NQEUZSJjW2QdsP5AVZyTlKotDdSqVReStvaUHK03cBoPU4Z+WigrzMDHnpvqSXA3Qs48YdKd2FrJUWDUgDQ5LnSanxS52cnGx+UC4qSJ4vI7ZU0cWci3csPCPb58llGagHoDU456RwRCbVo1T/WjIez6kCjcJ3V0kYSa8vkZYNSoG/0lAhSUZGXqpHxniSpSQKXa5UGuUkb1kkbzCiqRtAS3BRXsYPFPTOIVQADcZ3mCTlCnHp09CIlE6V+ylWxRhfJtUjiX4LoBwuPCMzGMV9F8y7AJCgir6KIJP0coCOR7AYyUmvL47DRSYdl0BNgvFTMkFWchxBC0iKZ7wERmak2NQd8X0BoPnivoq8vMw0een+pJcDdIXuDhZDI9LrS8dv0p6AuCQqK+OnJBsSLgBpeVN3jhOjADRf3FeRk5fqUdAzm6NlgSbpzmDhnLRsKO6piGxc/lTDD504XPTKGJ9+C6DErHBiFJO6ATRL8WCVuK9iYuXNAGrXfcGidJzswqXxhc8UdypWVNlvwQUUIGl5uAidvKWhDMfRAmgwZyNJVn7PrPhoeABN013Bwjlp8UD8sorjZKci7rfI0G8BjFZs6jZRMVyMEC4ANIZzTi7KyUtPk5+ZnvRygK7TPcHCOWnRsni3wl/1cbJTYWTkBVkZL5BsRLgASkonRlknb1lBZoSSQQD1VZpX4QU9Cnpn0VcBJKA7goVz0sJiqAj8+KVBjPGYbwFUUwoXkrxlIeECQH1V9FXU/8lDAKvW+cHCFkPFssF4l2KCMypqYbyA+RZANcZIgSfJxOFimO8PALVzNlTcVzFbXoq+CiApnR0sSuVPywal1MQH39WD8VPxDzfnaOYGVhQYSUbeADsXAGrjbCRFBXnZGfIz05JeDtDVOjdYuBV3Kpr7oRoZmSArE6QlR78FMEYpXFAWBWCKljdr9yvooa8CSFpnBouKnYrmlD9VEzdz98h4KZq5gWoqei7Y2QMwcRXN2n1z4t5GAInqvO9C56Qlg9LSISlILlSUGOPJS/eWm7kJF8AKfCM5xWVRzLkAMEEuysn4KQV9q9GsDbSIzgsWy4biYNHg058mwxhfXro3bly1XDgBFUYdRWuWhUzoBrBKLirIyMjvnSMvyCS9HABFnRUsBkeKw+9My4SKEuOl5AU9kmjmBsYYPURvWSiF7OwBqM7ZSLKh/J5Z8tN9SS8HwCidEyyGc9LCpfGfGzD8rh5MkF4+mZtwAVQqhYtCMVxEhAsAlZyzcbN2drr87IyklwNgBZ0RLHKFOFRY27KhQio2c6d6ZHyauYGqjJFSRiZv43Bh+R4BEHPOyYU5+ek+BT2zOQEKaEHtHywKofTGEqkQSelUfGHSwuJw0SvjB4QLoJpSuMgVw4XjewTodstPgEor6F1NxmutcmcAsfYOFpGNdyryBSnT+qGixBgvDhecFAVUVyqLGrHyBiPCBdDlXJSXPD8OFX4q6eUAGEf7BovSrIrhnJROt02oKDFeIC/VI4mTooCqPCP5RmYoYoAe0MVclJeRFPTOKf7eBNCq2jNYOCctHZQGhqVUKr4AaUPGTxd/SDo5x4UTMIZvJGPkDVqJGRdA13FRQXJWfs8seen+pJcDYBXaM1gM5eJZFb4Xv7QxExTDheMYWqAqX5J18gY4hhboJs6G8bGy2ZnyszNo1gbaQPtdlecKcQmU1NInQE2UkZEJMhxDC4yn1G8RFsMFJ0UBHc/ZSIry8bGyPbMIFUCbaK9gEUZxs3YYdUSoKCkfQxukJRcRLoAVlcJFzsob5KQooJOVZ1WkpyvonUOoANpI+wQL56TFy6Rcvq1OgJqo5cfQpuOdC06KAiqVwsWwkxmmJwnoRM7Z5bMq+ubEpycCaBvt8x27bEgaGImbtTssVJRUzrgICRfAijwjGckbslKBnT2gk8SzKnLyUlkFfaszqwJoQ+0RLEbyHdOsvSrxjIs+GY9wAVRVbuaO6LcAOkRpAJ4JMkr1rRH/DgTQdlr/Kj2K4mZta6WgO569MMaTl+6Ln61hgB5QqVQSlWd4HtAJyqHCT8U7FQzAA9pWawcL56RFA3FfRbpzS6CqMcaPdy6Yzg2MVe63sDI5SqKAduaivIwXKOhbXV6QSXo5AGrQ2sFiYEgaHO7ovoqVMV4gL90bf+yEC6CSZyTj4l2LkHABtCMXxlO1/b7VmKoNdIDWDRb5grR4UDKd31exMsZLyUv1SiJcAGP4RgrptwDakYvykpz83jny031JLwdAHbTmFbt10uIBKbJSqjv6KlbG89PxzoUkWY7ZBMqMkVLxfAuOoAXah4vykrPye2fLz0xLejkA6qQ1g8XAkDSck9JBV5ZAVROHi/gZHWfDhFcDtBBjJM9wBC3QJsqhomeO/Mx0BuABHaT1gkW+EB8t63nxC8pG71w4di6A5XxJrthvQUkU0LJcVCiGitnys4QKoNO01pX76BKoLjladrI8P1PsuXByjnABSFp+ShQlUUDLclFBzkbFUDGDUAF0oNYKFpRATYgJ0nG4cIQLoIySKKBlxTsVkYJeQgXQyVonWFACNWFGpjJcUBYFxEaXRDE4D2gJLspL7FQAXaE1ruCdi4+WpQRqwoyMvGBUWRQN3QAlUUCLqTj9iVABdLzWCBZDOWl4REpRAjVZXpAZ1dAdMucCKJVEDVsp4vsBSErF6U+ECqArJB8sIistGZBkunoQXi08P1M+ipYheoDikqjQyRsKKYkCEhCHClcMFZz+BHSL5K/klw3G/RXpIOmVtLXlcy6Y0A3IGMk3MiNWpsD3AtAszjm5MBeHil6OlAW6TbLBIl+Qlg1Lvk8JVB2U5lwYQ7gA5BvJSYZGbqApnHNyUU6SUdC3GsPvgC6UXLCgYbshSjsXxniECyAwMgUauYFGc85J4Yg8z1eqfw35mWmECqALJRcsaNhuGOOl4nDh+YQLdDdjJEMjN9BIpVBh/LSCvjXLB4oA6D7JBAtrpaWD8Z9p2G4I4wXFcBFIEadFoYuVGrnZtQDqzjkrF47IBBkF/WvKS2WTXhKABCVzVT84EvdXpFKJvPtuYYwfh4sgLdlQzjGNGF2o1Mg9bKWQ7wGgXkqhwkv1KtW/prwgnfSSACSs+cEistKyofJZ82gsYzx5qV4ZPyM5S7hAd/IUT+QeppEbqAdnI7kwJz/dr1T/GjI+TxQCSCJYDAxL+TDurUBTGOPFp0UFpXBBSQi6TPGJDDNipZBgAdTC2VCK8vIz0xT0rRH38wGAmh0swkgaGIr7KmjYbiojIy/VIy/ISs4RLtB9PElW8obYtQCmykUFKSrIy85U0LeajEefJIDlmvsTYWBIKkQcL5sQIyOTyspL9cThwtLUjS5iTHz8bI6hecBUuCgv2Uh+z2wFvbPjY80BYJTm/VQohMVheOxWJMnIyASZeEo3g/TQbYp9XYZdC2DCnHOy4UhxmvYc+T0zmVEBoKrmBYuBYSlit6IVGBl5flp+eZAeOxfoIr6RyVuZPF/zwKosH3wXKOhfQ0HPDEIFgHE1J1iEUXzErO+zW9FCjJeSl+mX8VLxzkWXnhj15Qsv1C6HHKKd3vMerbP99lp3hx10yMc+pj898kjF/Z574QUd9JGPKNhwQwUbbqhn588f81jf+9GPtMPBB2uDnXfWm3fbTe//+Mf196eeatJHggkp7VqMsGsBrIyzkVxhWCbIKuhfU366L+klAWhxzQkWg8NSGLJb0YLKsy78dNeeGHXjz3+uEz/0Id1/yy16dt48bb/11vrF3Xdr3w99SAtefVWSdMPNN2vPI47Q0oGBcR/n8h/8QB8780w9/fzzevj22/XHn/9cd//+99r9Ax/QiwsWNOvDwUQUdy1ErwVQ1fKTn/qVmramvCCT9JIAtIHGB4soisug2K1oWaXjaMsnRnVZadS5Z52lYz/wAUlSOp3WfrvvLkkaHBrS34q7DalUSvf//Ofa953vHPdxLvne9yRJ22y+uWbPnKnVZs/WNptvroWLF+sb3/pWgz8KTIpn4hOi2LUAKjjn5MK8ZMP45Kf+NWQ8jocHMDGN/2kxOBKfBJVheE4rMzJSKivP8+UKQ/EWuOfHt3e4Q/bfv/znRUuW6Ee33ipJestGG2mnt79dkvT+gw5a5eM8+8ILkqT+vuXlAtP6+yVJv7rvvnotF/XixydEqcdJqc7/OgdWpdRPIc+X3zNHfmY6/RQAJqWxOxbWxrsVnmG3og2Umrq9dH888KjLdi5O+o//0Frbbqv7/vQn7br99rrj2mvV19s74bdff511JElLli0r31b683MvvljfxaJ2niQneSPd2VsEjOaclcIRGT+tVP+aCrI0aQOYvMYGi6GclC8wZbvNGC+Iw4Wf7qqm7su+9jW9+MAD2mvXXTXvwQe11xFH6NXXX5/w25987LGSpIcfe0wvvfKKXnj5ZT306KOSpDDqvt6Vlsc0bkBSsUk7HJFJ9SqYtmY86wgApqBxwcK5eCCeYW5FO4r7Lvpkgky5qbsbdi9Wmz1b5/3nf0qSnpk/X5d9//sTftuTjjlGl/+//6ctN9tM+x51lD7z5S/r4He9S5I0e8aMhqwXNfIkWSeTI/ihO8WTtPPyMzOU6l9Dnp9OekkA2ljjthJG8lKO3Yp2ZmTkpXrkjCdbGJZc5/VdWGv18quvau5aa5Vve8tGG5X//OS//jXhxzLG6GNHHqmPHXlk+bbPfPnLkqQtNtusDqtF3RkjeZIZsXI9rnwULdDpnHNyUV5Gkt8zm6F3AOqicTsWgyOSk+Q1bwYf6s/IyAuy8tP9xWF6nbVzsXTZMu3+/vdX3PbiK6+U/7zGnDkTfqx5f/6zHnrssYrbHnn8cUnS4QcfXMMq0VC+ZCIXHz8LdIG4n2JYnucr6F+dUAGgbhpz1V8IpeEccys6iPGLw/T80jC9zikdee7FF3XljTdKkgqFgr54/vmSpN6eHp3woQ9N+HHumTdPp33hCxocGpIk/fS223Tfn/6kXbbbTscdfnj9F476MKWBeZajZ9HxnA2lMCcv1aegf63ik0aECgD1YZxrwG/SJQPSooH4iFl+YHUUJydXGJENR+Ib2rw0KpfL6ayvfU3zHnxQYRjq5ddeU39vr3bdfnuddfLJetsmm0iSfnTrrfrKRRfp1Tfe0GtvvCFJ2nSjjZQKAj1w661Kp9O65a679JWLLtIrr78u3/Pk+77ef9BB+vxpp1UcQYsWZJ1knaKZKSnFLis6T1z6lJOR5GVmKOidFe9CA0Ad1T9YWCstWCiFkZRmdkWnclFBtjAU71wYn19QaH8FK9fjy06jLwydxTkrF+bk+Sn5vXPkpXrZpQDQEPX/DTqSj0uhUoSKTmb8lDyvX7YwLBcV4r4L47X17gW6nFccmNfrJJ+vY3QGFxUkG8pP98nvnc2pTwAaqr5PMzsXN21LnK7SBYzx5aX74jPPnTqusRtdpnT0LE3c6ADOOblwRHJWfs8sBRwlC6AJ6rtjEUbxjoVP03a3MDIyQVbG8+PdCxvKURqFdmSMJCeTs3JZ5u+gfTkbyUU5eX5GQe8cmVQPpU8AmqK+wWIkL0WRlOFZkW5jvJS8dDFcRHlKo9CefCNTKE7iTvG1i/binJNsWCx96o9DhU9ZMoDmqd/Tys5JQyNM2u5i8bTu3rg0SqI0Cu0n3rSgHAptxzknlUqfemcr6F+TUAGg6eq3Y5EP40nbzK7oastLo4L41CgbyRnD7gXagzGSKZZD9TqeJEFbcDaUi/LygqyC3jnLn9wBgCarX7AYzsVHzaY4qhGS8QJ56X65MBfPvHCRXJvPvECX8IxM6KSCk9J8vaJ1xbsUOckY+dmZCnpmyXg8uQcgOfVJAc5JwyOSRxkUljPGi5sG/aDY2M3uBdqAZ6TIyctb2TSHEKA1VexS9MymQRtAS6hPsMgV4lIoditQRamxm90LtA1jpNJMC47ORgthlwJAK6tPEhjJxbsWHs/uoTp2L9BWfMmETiZ0cpRDoUWwSwGg1dUeLJyL+ysIFZgAdi/QFoozLZS3EuVQSBi7FADaRe3BIh9KhYiheJgwdi/QFoyRydv4oo5nhZEQdikAtJPag0UuL1mGSWHy2L1AS/MkEzmG5SER7FIAaEe1BYtSGZQxPKOHKWH3Ai2rNCyv4OSYM4YmKU3PdjaUF2TYpQDQVmoLFmEUl0L51CCjNsZLycsEcmFOLhyJt/+NL2P42kJCin0WJm/lejhKG43nbCRFecnzFfTMkp+dwS4FgLZSW7DIFaQokjLpOi0H3aw0tdv5KdnCiFyUl3NWojwKSfGMTMFKVhLXd2gQ55xclJOR5KX75ffMlBdkkl4WAExabcFiJE8ZFOrOGF9euleK0rIh5VFIkCcplEzBynFABepsxbInv2emvFQfZU8A2tbUg4V1ceM2x8yiAYyM5Kfk+ZRHIUGlcqiCk8smvRh0ksqyp9nys9MpewLQ9qYeLAph3GPBtG00EOVRSJwxUsHGh1XwTDJqNLbsaZa8gHJiAJ1h6qkgX+AXLZqG8igkpnTsbOSkgK81TE1c9lSQs1Gx7GmWvFQvZU8AOsrUg0UuL4n+CjTPysqjZAwBA41ROnY2dHJs0GIKnA2lqEDZE4CON7Vfk9bFJ0JxzCwSUFEeFY7IhXnJSc549F+g/uizwBSV+yiMJy8zTX52JmVPADra1IJFoUB/BRJnjC8v1Sv5mThgRIW4/4KAgXqjzwKT4JyVwpycjPx0XzyPIshS9gSg400tGeRDfsGiJRgZyQvkpfskGxYDRkiDN+qLPgtMgHNWLsrLOCeT6lEqO0OGPgoAXWRqwSJXEP0VaCVxwEjJSwfxzkU4Ejd4SwQM1I4+C6xEfNJTXsbZuDE7O1NemnkUALrP5H9FOhefCEV/BVqQkZHx03J+Kj6aNswRMFA7YyTn4mCR9FrQMipOevLT8rLT5WemUYoJoGtNPliEkRRFDMZDS4sDRkbOS8tFuWLACOWMxxG1mBojKSRWYIWJ2X5Kfu/MOFBw0hOALjfFYOGkgGCB1mdM6QSpdBwuIgIGpsgzMqGLT8Xz+LrpRssDRUHGCxT0zJKfmS7jUx8HANJUgkUhlAyN22gvxngyqR65IC0b5qQoz5A9TI5R/KRKRLDoNstLnkIZL5CfnSU/O02ez9GxADDa5INFPpQcv1TRnozx5ad65YKMXJiP+zDowcBEFL80TOTkUskuBc3hnJOigpwrljz1zI5Lnny+AACgmskFCxq30SGM8Ys7GBmavDExNHB3jfjY2IKMszJ+Sn5mjvz0NEqeAGAVJvdTksZtdBhjvOU9GLZAwMDK0cDd0UbPofCCtLz0dPmZfpqyAWCCJh8sLI3b6DzGePEpUn66OAej2OQtMckby5UauBkQ2lGcjaSoIMnJC7Lys9PlpfpkeBINACZlcsEispITv1DRsSrnYBTiU6RKk7yNJxnDLkY3M4qfXIk01fGiaBHOOclZKcrH39epHvmZafLSvTyRAABTNMkdi7BBywBay+iAIRvKhjm5qCA5cZJUNytN4LZOjn//tjR6BoUxnrx0X3xkbKqHSdkAUKPJBYtCJH6XopsYGclLyUsHko2KuxijTpKiTKq7mGKyiOizaDfOWjkb908YP5CXnSk/0yfjZwgUAFAnEw8WzhVnWHARhe4TB4xAxgvik6RsIT6u1oZyLiqWSbGL0RWcCBZtorQ7IRvKychLZeSni+VOHrVsAFBvE//Jal3cY8FgKHS50Y3esqFslI/LpDhNqjuYUikUWlXpuFi5SMYL5GWny0/3ywRZdicAoIEmHiyiKN61YIYFIGl5mZTvpeSC+JjKFcukaPbuQMbER85yMlRLiZuxS6c7qXhc7Cz56T4G2gFAk0w8WIQ23rXgFykwRnkeRpCRolC2eJqUXBQ3+bKL0TmMZKyKJ+QlvRg4Z6UoLkksNWN7mWnyUj30PwFAk01ux8KIYAGshJGR/JQ8P5CcLfZhjN7F4ESptlc6ctZK4ro1EaN7J2SMjJeSn54uL90n46cpdwKAhEyux8JRVQxMhJGRjC+T6pFTttiLUZCiAiGj3RUPhpJly6KZ4jARSTYudTJ+SiY7XX6qTyaVZXcCAFrA5HYs+CUKTNqYXgwbxsfW2lEN3/RjtB3jRAN3g5WG2DlbkHFO8nx56f643CnVI+P5SS8RADDKJEuhuOgBahGfKJWW/HR8co0tzsWIQskVh64RMlpbaZaFJVY0irORnA0lZ2WMJz/oWR4maMQGgJY1ueZtrnOAuqk4trZ4PGbpVCmVng0nZLQuckXdlHYmZEM5Z2WMkedn5KV75aV66ZsAgDYxiR0Ly44F0ADlfozAj0+VKk34toViyGAno9UYJ5mIWRa1KB0PG+9MuGLQTslPFcNEwERsAGg3k5u8zQ95oKEqJnwrG4cMGy4PGdbS+N0CXOlkKExKxWlOkmT8YplTbzy8jp0JAGhrkwsWHqduAM1SGTIyxSbWqDjlO1x+upRMMWiwm9E0xsTHzWKVXLHESTaKb/ACeZl+ealeeUE2/honTABAR2DHAmgD5XIp3y83fstGsrZ4hK2zlEw1W+kIbn4uVlhe4hQVm6/jORMmO01+KhvvTHCaEwB0pIkHC4nmbaBFGONJviffT8ml3PKSqahQvqArF+oQNOrPxH0WqAwSxhW3cYwvL0jLC3rjk5yCNHMmAKALTG7HAkDLqSiZCjLxzkWx9CTuzbDFoFEc6EbQqJ8unJE3bpDw0zKprLwgK89PU+IEAF1okjsW/JIAWpkp9lvEszIUN4AXezPGBo3SGxE0Jq3UvN0Fz7eUgoSKO2GSVggSGXl+hiABAJhMsOAXBtBuKnszKoNGHDaq7WiY8n8JG6vQYcHCOad4+F8Uz5MoB4n4KFgT9MhLESQAANVNbscCQFtbWdAo92m44jPTo5+QJ2yMo31rocYPEaZ4DGxGKpY1eQFBAgCwagQLTFoul9OlV1+pG27+qZ5/8UWtPmeOJOntm2+h9x/8Hi0bGNDnzv2yXnz5ZUmSe/mNcR/r0Sf+pv2P+oC22+rt+vn3ruXCpclWDBqS4p2L4sUmYWMcRvEnok12LCYTIoyfik9x4nhxAMAkESwwKUuXLdU+R7xPDzzysI467P2658c3q7e3V6+9/rqOOvlEHXr8MXrmTw/rK2edreM/fdoqH++Oe3+tl195RbfedYcWLlqkObNnN+GjwMqUm8GlyYcNqasCh2mxNou4H8KWX5xzMqUVlkNEVgoyhAgAQN0RLDApp33+c3rgkYc1c8YMXXHeBert7ZUkrb7aavrBxd/SuttuOanHO+I9h+oXv7pL22/9dkJFC1tl2HA2Dhqj/r6ywFF+zDaW1OrLuw/lz7uVca78uTbGK/ZEpEeFh4AQAQBouEn8hmml5+WQhNffeEPX/vTHkqS9dt1Nfb19Fa9fa401deGXv6qZ02dU3P6NS7+pnQ/eT71vWlcHHHW4Xn39NUnSb/8wT3u87z26d97vdN5lF+vZ+c/rmeef01t221Fm7Tkya8/Rv//X57XX+w9R30brafv999bv7v/jKtd5/U0/0dv32UPrb7eV3r7PHtr6XbvrjC/9V/n1i5cs0Sc/e7rW3XYLvXmX7bXlXrvpkqu+U7xgi93/0IPa6aB91fumdfWuww/VWV/5YnlNb9ltRz3z/HNT/jx2CiMj4wXFGvys/FSf/Mx0+dkZ8rPT5Wf65aXj6crGTxWDRBxGyrseo19cHEpcsZHctcPPnDov0TkXf/yj55KEOdlwRK4wLFsYlsIRKcoXh8/58lN98npmKdW3utLT1lZq+lylZ6yn9Iy5SvWvrqBnZvxvEWQIFQCAhuK3DCbsT488pCiKJEkbb/imqvc59aMnauaMymAxa8ZM/fEXd2rfPfbQHff+Wl88738kSbvvsquuuvCbFfd90/ob6I7rf1z++2tvvK57fnKzLj73/+nPf/2LDvrwB/X6G+P3bDz+j7/rw6d+UoNDQ3rqDw/qkV/9Rt/4wpf041t/Lkmy1uqADx2uy79/tf794yfpqT88qNVmz9GpZ5+l8y+/VJK0bGCZDv7wUfrTww/pf/7zv3T3j36mlxYsKL+PO67/sd60/gYT/bR1HSMjY3wZLyXPz8hL9chP98vLTo+DR2aavHRfcXBapvhMur88eBTLraqGjxUDiHMa/X9NNYF3FweFUlioDAx2VGAYHRpcmJdsWDza1cgE6Tg8ZGcq1beagmlrlcNDasa6Sk1bU6ne2fKz04tBLk2AAAAkYhK/fdq7bAG1W7R4cfnPPdnshN/ufQe9W5K05WZvkyT98c9/nvDbHnLAQZKkg/fZT5K0bGBA373+B+Pe/29P/kPWWr20YIFuuu0XWjawTPvtuZfO/+J/S5J+9dvf6P6H4ve/6/Y7SpJ23GYbSdI3Lr1YknTtT36sNxYtlCQduPc+kqT37n/ghNeM6uLA4a2wy9ErP90vPzN9efDIxuHDT/fFF8orCyCl+QorCyJVX6JiQJnCS/GoXuesbJSTLQaDUjgY/XcXjhRf8pJdHhiMMfJGBQa/d45S/WsomLaW0tPXUWpGcddh5npKT58bh4e+OfFuULqvuPvgc9gBAKCl0GOBCZs1c2b5z8MjI5N+uyCIv9wWLVk84bedM2tW8b/L+y/+8fRT8eOtu0bFfe/+4U3a4e3bKJPJaGh4SEd+8gSlUintv+fe+u/P/ock6ZHHHy3f/7hPn6KebI8GBge15upryBijZQPLyo8vSbNnlt7/rAmvGVNjRvVfqPI/FZycVO4zcOVb4/93Wt7cEf+5XOJW8TaVbzfBBcb/U/yviZy8VEbKBpI8Gc8rfgxeuZ/EjPqzjCn3P8h4hAIAQMchWGDCdnz7tvI8T9Za/eu5Z6ve5/6HHtTGG1SWSa14ATW6l6EWpbKs8uPKacP11tcfbrld37jsYt32619p8ZIluvWuO/S7P/1Rz9z/sLKZTPn+V194iXbZfocxj1uv9aExqgWQ5a9rEuckl5fXM1vqzaz6/gAAdAGatzFhq82Zo6Pf9wFJ0j3zfqehoaGK1//xzw9ol3cfoIGhwbq9z4XF8qtSaZIkbfbmTSTF8zFGv+y56276y+OP6dXXX9d1l35brz32pM79j/+UFDdsP/XMv7TFZm8tP878l14s//nPf3lEH/zExyoeP37/i4rvf1HdPiZ0CkOFKAAAo9Dhh0n55lf/R9tv/XYtWrxYHz/z38vh4qUFL+ukz52pUz96gjZcb/26vb+f33GbJOkXv7pTkjStv18fPfLoce//z2f+pTP/+xwtWrxYQRBo2y23Lr/dphtvrL3e8c7yLsV3rv2+wjBUoVDQV//vfG2/9dslSUe//wPl0qvbfv0rSdKtd91Rt48JAACgE9G8jUmZMX267vvZL/S/5/y3nvzX09po5+209bt213s+crSOPfxIXfClr+qnv7hF//k/55bfZs/3vVc//cUtuvTqKyVJL72yQHu+77367R/mVQzR2/+oD+gfT/2z4v3NmTVbe73/EJ3yH2dpu6221i9/cKNWK076rmaj9TfQGqutpu3231tv230XHftvp2jv3d6pX3z/Bk2fNl3GGP3squ/r2COO1COPP6b1tttK7zz0YO24zbY646RTJUnT+qfpFz+4Xjtus63O+sqXtdf7Dyn3WgAAAKA64yZaUP7cAimTbvByAOnZ+c/rTTvGJzXd85ObteeuuyW8Iuneeb/TXu8/RJL0zJ8eruuuDNqQdVKhIK01m5+LAAAU0WMBAJPm4gZyTnYCAKBs4sHC8QsUjffM889p/6M+UP778Z8+TTff/ssEVxT3V6xYssXk7S5Xep6FYAEAQNnES6GeXSBl2fIHAEVRXA619hwp8JNeDQAALWFyp0Jxvj8AsGMBAEAVHDcLAFNhRLAAAGCUiQcLI/q3AUAq7t4yIA8AgNEmESxIFgAgKf5R6HMqFAAAo00uWNBjAQDxz0Kfpm0AAEajFAoAJotgAQDAGBMPFr7HjgUAlAScfQEAwGiTCBZ+fG47AEDyCBYAAIw2uWDBjgUASHIECwAAVjDx34xs+wNA/ASLMfGpUAAAoGziaYFn5wBgebDgZyIAABUmHywohwLQzQgWAABUNblSKI6cBdDtrJM8E5+UBwAAyibXvO15krMNXA4AtDjrpFTA1G0AAFYwuTkWnuHIWQDdzdk4WAAAgAqTmLxt4l+mBAsAXc1IAVO3AQBY0eSKhFMBpVAAupdzca+ZT7AAAGBFkwsWga/4tyoAdKHSiVDM9QEAYIzJ/XYsPUvHkbMAupEtDcdjxwIAgBVNcseCBm4AXczaeOfWY+cWAIAVTTJYBPHpUJY+CwBdyDkpk+KoWQAAqphcsPA4GQpAF3PiqFkAAMYx+Q7EdIqToQB0n9KJUAQLAACqmnywSAWSDA3cALpLZCXPY4YFAADjmEKw8GngBtB9So3bPkfNAgBQzeR/Q9LADaAb0bgNAMBKTT5Y0MANoBvRuA0AwEpNbU8/k6LHAkD3sJbGbQAAVmFqwSKdin/JEi4AdIOo2F+RJlgAADCeKQaLID4dJaLPAkAXsDZ+QsWjcRsAgPFM7bek78e/ZAkWALqBk5RNJ70KAABa2tSffsvSZwGgC9BfAQDAhEw9WNBnAaAb0F8BAMCE1BYs6LMA0OnorwAAYEKm/pvS9+JjZwkWADqVc/RXAAAwQbU9BZdNF3/xUg4FoANZFw8FTaeSXgkAAC2vtmCRSce/dJnCDaATRZGUor8CAICJqC1YpIP4l24U1Wk5ANBCrJV6MpIxSa8EAICWV1uwMCb+pWvpswDQYayLf8Zl6K8AAGAiaj/mJJOOf/lSDgWgk0RRfMxshv4KAAAmog7BIhX/8qUcCkAniayUzXDMLAAAE1T7b0zPi3/5Eiza3oY7vF1m7TlVXz79hbMlSUNDQzrzy+do/e220oY7vF1v2W1H/d8Vl495rKtvvE7b7ruX1t12C71px2102PHH6O//fLLZHxIwNaWT7jhmFgCACavPU3HZtCTDsbMdzCs+a/vBT56g8y67WJ877d/07AOP6KC999Wn/+tsfel/v16+77e+d5WO//Rpevq5Z/TXu+/TA7f9Sr+67zd6xyEH6cWXX0rqQwAmLrLLZ/UAAIAJqV+wCHwpZNei3b322JMVL3+49XYZY3TwPvvq3nm/06133SFJOuhd+0qS3r3vfpKkcy+6QG8sXChJ+uaVV0iStt1ya82eNUurzZmjbbfcWgsXLdL/XHJRAh8VMElhJPUUf64BAIAJqU+w8D2pN8MU7jb30aOO1mpz5lS8XHXD9dpmi630rnfuoV/86q7yfVefM0eStMZqq0uS8vm87rj315KkZ+fPlyT19/WV7z+tv1+SdNdv7m3GhwJMnXOSkdSbTXolAAC0lfp1JfZk4l/GnA7Vtv7rM2dW/P3V11/TNT++UWed+ilJ0pP/eqr8up5sjySpt6enfNuT/3pakrT+3LmSpCVLl5ZfV/rzcy+80ICVA3UUFk+Dor8CAIBJqV+wyKSlVEA5VAf5vysu19y11tb7D36PJGlgcLD8ulLPhTfqxJzS60/96AmSpIce/ateWvCyXnjpRf35r3+RJIVR2JS1A1MWFYficRoUAACTEtTtkTwTl0MtHiyWEjCptp0NDA7osmuu0lc/93n5flxnPrq0yVorz/NkRw1HLL3+5OM+psz/b+9eWiQ76ziO/06dS9WpS3fPJHkBKooBdacLJYSEJCAayUKDGgKJDKgExSAoqGQREERBXYqCC0URL6iJGATNIkgQBV3oexDJzZnp7uqqc57zuPif012TmZjuup1LfT9Q9NiNWpPpTNev/rekrx/+7Ce696MP6V3vvFMfvv8B/eLZ3+rywaXt/kaAiygKq7ym/bqfCQAArbO+YCHZD+Nrx9YOFRIs2uz7P/6R4ijWYw9/4vRzb3/L205/PT2ZajQc6Xg6Pf3cO95qXw+CQFceeVRXHnn09GtPPvVVSdK777xz008dWF7urPLKtW0AAC5svbX+JJYS2qHaLssyffcH39PnPnVF6cIMxYfuu//01y+98kr58WVJUpIkeuDueyRJL/7tr/p72fpU+ce//ilJevjBhzb51IHleW8Vi+HAKrAAAOBC1hssgsB+KPuCmxYt9tNf/1KvXf2vnnj8yg2fv+cDd+mD994nSXru+T9Kkn7/J/v45Sc+r9vLTVHP//kFPfGVL+no2GYufvW7Z/TCX17U+9/7Pj3+8U9u67cBXExR2FzFkDYoAACWEXi/5gSQO+nf9m624vV2WmHzvPd6z7136b677tZ3nv76TV8/Oj7SU9/8hn7+7G8UhZHiONZnHn1MT376swrKuZpn/vCcnv72t/Sfl19S2AsVhqE+9uBH9LUvfFHj0XjbvyXgfGZzKR1Id+wzIwYAwBLWHywk6dVr0rUjacA7fwBaoPBSlkl3HHC/AgCAJW1mn+JwYC0FHMwD0AZ5bhVWblcAALC0zQSLfmyPnJsFABquGtoeDbhdAQDACjbzUzQIpFG5TYghbgBN5gopDGmBAgBgRZt7ey7tS1HI6lkAzZY72wTFsgkAAFayuWAR9qxq4RxVCwDN5MpL2yOqFQAArGqzDcWjgbUYULUA0ER5btVVLm0DALCyzQaLOJLGqb0rSNUCQJMU5da6ccrdCgAA1mDzK1BGAykMWD0LoFmy3LbXpdzbAQBgHTYfLJJYGqbWckDVAkATFOXfRZMh1QoAANZkO0vbx+V++IKqBYAGyHJ70yNlaBsAgHXZTrBIYlvnmDHEDaBm3ku+kCap1KNaAQDAumwnWASBNB7aD3FmLQDUqapWcBAPAIC12k6wkGxIcti3H+rMWgCoQ+HtsTe09kwAALA22/vJGgTSZETVAkB9skwaUK0AAGATtvuWXRLZNW42RAHYtmp5xGREtQIAgA3Y7k/XILD1jlzjBrBtWS4NEmvJBAAAa7f9t+2S6hq3o2oBYDuq9su9EXcrAADYkHr6ASZDKYqoWgDYjiy3mxWDpO5nAgBAZ9UTLKLQdshTtQCwabmzpRF7XNkGAGCT6ptgHA+lfiLNs9qeAoCO896WRYxTW3kNAAA2pr5gEfak/ZEk1s8C2JAsl+KI2QoAALag3p2Lab88mpfREgVgvYri7BheFNb9bAAA6Lx6g0UQWNWC9bMA1m2eS2kijYZ1PxMAAHZC/Veikti2RDHIDWBdqoHt/bF9BAAAG1d/sJAsWPRjBrkBrK4a2B4xsA0AwDY1I1iEPXtnUbLKBQAsqxrY3mdgGwCAbWpGsJBskHuc2osCWqIALMMV9vfHwZiBbQAAtqw5wSIo+6GT2IYuAeAivLcNc6OBNBzU/WwAANg5zQkWkr3DuD+W5GmJAnAxVQvUwYQWKAAAatCsYCHZXQtaogBcBC1QAADUrnnBgpYoABdBCxQAAI3QvGAh0RIF4PzmtEABANAEzQwWUtkSNaQlCsAby50kb6GCFigAAGrV3GARBNYvPUikWUa4AHCjojyEtzeyNyIAAECtmhssJDucd2nPPma0RAEoeS/NM7t/wyE8AAAaodnBQpL6sXQwknxhm18AIMut9enSROo1/68xAAB2QTt+Io+H0ii1zS+0RAG7zTnJy1olk7juZwMAAErtCBZBIF0aS/3E2h8IF8Bu8t6qFePU1ssCAIDGaEewkKRwoe0h474FsHO8t0UOg8SqFcxVAADQKO0JFtLZCwrvyzWTAHZCNawdh9LlfVvoAAAAGqV9P53Hqa2XzHOpYJgb2Am5swrF5T0piep+NgAA4BbaFyyCwK5yj1LmLYBd4Jy9iXAwtvWyAACgkdoXLCSpF0iXJxzPA7quKGymajK0BwAAaKx2BgvJhrkv71nP9ZxhbqBzqrmK4YBhbQAAWqC9wUKyHfaXJlIgNkUBXVJtgOon9gYCR/AAAGi89v+0Hg7sxkVR2EA3gHZb3AB1255d2AYAAI3X/mAh2WXu/bHkCtbQAm3mvbU2hqF02z6XtQEAaJFuBIsgkPZHNtyZ5xYwALRP5qy1sVrOAAAAWqMbwUKycHEwsepFlnHjAmibLJd8YXNTw0HdzwYAAFxQd4KFdLaGdljeuChYQwu0Qr5wq2Kc1v1sAADAEroVLCTbHnPbxA5pcUAPaD7nLFjsjezBWlkAAFqpe8FCOhv8HMTSbE64AJrKOWuB2htyqwIAgJbrZrCQbEXl7QfldW7CBdA4rrqqPSrv0RAqAABos+4GC6kMF/uEC6BpnLMlC5ORzUURKgAAaL1uBwtJiiLCBdAkVfsTlQoAADol8H5HXmnnTnr5qnQyt6NbPV7MAFuXOwsWkyGhAgCAjul+xaJStUWliTSfc+cC2DZCBQAAnbY7FYuKc9Ir16XjqRTHUrg72QqoTZZbmN8fS/uslAUAoIt2L1hI9gLn1evS4bHNYERh3c8I6CbvLVRItk52MiRUAADQUbsZLCS7yn31ULp2ZHcvCBfAenlvRyp7PenynjQa1P2MAADABkV1P4Ha9IKzg1xXD+1FULy7/ziAtfJemmVSHFqoSPt1PyMAALBhu1uxqHhvLVGvHUpeUhLRqgGsovC2ICGJpdv2pX5c9zMCAABbQLCoHJ9Ir12XMmcvhAgXwMW5wg7fpX2rVFAFBABgZxAsFs0z6dVrC7cu2BgFnFuWW7AYp7ZOlo1rAADsFILF6zlnG6OOpmyMAs7De2meS4Fslewe62QBANhFBItbWdwYFQTWzsELJeBm1ZB2FEqXJ9YCxb8rAADsJILFG/FeOpxK/z20uxcJcxfADap5in5i8xQMaQMAsNMIFm/mZG5D3bM5l7oByUJ37sp5ioF0MKFlEAAAECzOxTlbR3s0lYKe7eaneoFdtHj0bn/EJW0AAHCKYHFei61RrmAlLXbPYuvTpYk0SOp+RgAAoEEIFhc1y6TXypW0cSSFtICg47y3+y6+XCV7MOb7HgAA3IRgsQxXWOXi8FhSwLVudFdR2CrZKLTWp3HK9zoAALglgsWyvJeOZ7aWdp5RvUC3VAPaRWErZA/GthkNAADgDRAsVpU7CxeHU1G9QCdUVYqwHNAeD6Ue39MAAOD/I1isA9ULdAFVCgAAsAKCxTrdUL0QR/XQHq6QMqoUAABgeQSLdfNems6kq0d2VC8MbfCVgIEm8t7anuSldCAdjKhSAACApRAsNsUV0vVjezjH1W40y+n1bGdBYn8sDfsEYAAAsDSCxabNM+nakXQ0k+Rpj0L9FtueJqk0GRF6AQDAyggW20B7FJqg8HY5W5KGA5uloO0JAACsCcFim1xhR/WuT6U8J2BgO7y3CoUvK2Z7I9qeAADA2hEs6pA7m704nFqPexRZKwov9LBO3ktZuT42iaS9oVUqerQ9AQCA9SNY1GmeW8A4mlqbShzR647VnQ5mF1Ic2urYccr3FgAA2CiCRd2qdZ/Xj+zIXuHtxWCPCgYuaHHTUxjaYPZ4aO12AAAAG0awaArvpVlmMxjHM2tfoUUK57EYKKJIGg+kUWoVMAAAgC0hWDRNVcE4nErHU8l5CxcMeeP1qqHsqso1Ti1QUKEAAAA1IFg02Ty3+YtqyJstUpCsmpU5SeVcThUomKEAAAA1Ili0QZZLRycWMnJnn4sY9N4p3tswdu6kQFI/tvmJYZ8tTwAAoBEIFm3iCulkZhWMWWbvXFPF6LbC282T6s962LeVsYOEP3MAANAoBIs2qga9j0/skTspKOcwegEvONuuqk64sjoVR9JoYIGCgWwAANBQBIu2y91ZwJiXg7y9hZCBdvC+rE44yZfViUFigWKQ0O4EAAAaj2DRFd5L80yazmxdbZZLXmyUarqinJsoCgsP/fis1Yk/NwAA0CIEiy4qvDSbW8iYzs42CIWhPahk1KeqTLgyTASBlEQWJtK+tToRJgAAQAsRLLrOFdLJvHzMylabsl0qZCZjK05nJoryn31gdyfSvjToW5WCPwMAANByBItdUhQ29F1VMzJn754HKqsZXPleC+/PwkRRhokwlJLYtjr1YyoTAACgcwgWu6q68D0rqxnV+love8Eb9gga57UYJFwhqaoI9aR+IqWJhYmQmQkAANBdBAsYV9jw9zyTTsqPp0FD9kK516N1SirnJIpyVqIMEkFgw9b9MkQkZVWCeRYAALAjCBa4NeesojHPrJqR5WVrj9fpC+ldCBu3ChEKpDA4a2/qxzaAHccECQAAsLMIFjif6gJ0vhA4qrDhvb3elrdDfVXQ6PWs2tH00FG1MhXlw5eVGqms1iyEiCSySkQc0toEAACwgGCB5Z2GjfJKdO4sbGTlob5i4QV6JahCx8KvFdgL+Orr61B9W3udBYfTR/V1f/b/Vz2XqCdFZXCIIpuTIEQAAAC8KYIF1u90kNlZ6CjKRxVAnCsrHbrxxf7Z/8DZi3i/8DlJZwnk9f/5df996ebQUAWHsHd20yMMpF5Yfp71uwAAAMsiWAAAAABYWa/uJwAAAACg/QgWAAAAAFZGsAAAAACwMoIFAAAAgJURLAAAAACsjGABAAAAYGUECwAAAAArI1gAAAAAWBnBAgAAAMDK/gfCUqoCWhiPmgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 1000x1000 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Now we can draw our Venn diagram to visualize the intersection\n",
    "#The draw_venn() function will output a visualization as well as a dictionary of results corresponding to different intersections\n",
    "venn = bv.draw_venn(list_x=de_genes,                 #We give it our first list - our RNAseq hits\n",
    "                    list_y=chip_genes,               #and then our second one - our ChIPseq hits\n",
    "                    list_z=the_void,                 #and our empty list\n",
    "                    title='MCB201B Multi-omics',     #update title\n",
    "                    subtitle=None,                   #remove the default subtitle\n",
    "                    xtitle='RNA-seq',                #update label for our list_x (rnaseq)\n",
    "                    ytitle='Chip-seq',               #update the label for our list_y (chipseq)\n",
    "                    x_c='#F5E4CC',                   #update color for list_x circle in Venn diagram\n",
    "                    y_c='#FFD1DC',                   #update color for list_y circle in Venn diagram\n",
    "                   )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f95805ce-c184-4b9d-abea-19c4cbb1c4f1",
   "metadata": {},
   "source": [
    "<h2>Identify the overlapping genes</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2538a0eb-4e61-48ce-a8ed-de613dd956c6",
   "metadata": {},
   "source": [
    "Let's go ahead and identify the differentially expressed genes that also correspond to a peak based on our analyses.\n",
    "\n",
    "While the documentation is a bit limited, we can see from the paper <a href=\"https://journals.sagepub.com/doi/10.3233/DS-210032\" rel=\"noopener noreferrer\"><u>Hulsen 2021</u></a> that the returned objects for different set operations can be pulled out of our <code>venn</code> variable, and specifically we can pull out the intersection between <code>list_x</code> and <code>list_y</code> as <code>xy_only</code>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e9ef2e64-b2e0-4d03-9a56-85ccac0c2bfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ADAM19',\n",
       " 'TENM2',\n",
       " 'LRP3',\n",
       " 'RGMB',\n",
       " 'SLC1A3',\n",
       " 'ADGRF5',\n",
       " 'ARHGAP24',\n",
       " 'ATP8B1',\n",
       " 'PALLD',\n",
       " 'ZBTB18']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#The output is a dictionary that we can then pull out what we're interested using their key\n",
    "venn['xy_only'][0:10]\n",
    "#Here we pull out the intersection between list_x and list_y, which is our rna-seq and our chip-seq datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4da3cec8-dc0d-400c-9b7c-9187b81ec3fb",
   "metadata": {},
   "source": [
    "We can also do it ourselves using set operators to identify the intersection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9637c4a4-a8b8-4b8d-8c7a-dc1212b72a53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.series.Series"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(de_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "94fff99d-63d8-41ac-b802-59a894b57605",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SLC1A3',\n",
       " 'ADGRF5',\n",
       " 'ZBTB18',\n",
       " 'DEK',\n",
       " 'PLXNA1',\n",
       " 'MTUS1',\n",
       " 'NRP1',\n",
       " 'NSD1',\n",
       " 'LIMA1',\n",
       " 'CYTH3']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We can use set operators to do the same thing that draw_venn() does when it generates the Venn diagram\n",
    "list(set(de_genes) & set(chip_genes))[0:10]\n",
    "#The extra list notation is just to pull out the first 10, so it's not outputting everything\n",
    "#Recall from lecture that the intersection tells us that these differentially expressed genes are likely direct targets of TAZ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "841f850d-6028-43c7-baa5-a7639896908c",
   "metadata": {},
   "source": [
    "What can we infer about this set of overlapping genes in terms of how TAZ is regulating them?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7955261-48fd-4989-9b72-b91ee34b0fe3",
   "metadata": {},
   "source": [
    "<h2>Identify the differentially expressed genes without a peak</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c5ae1bc-aecf-4d46-8dc7-30c6db3c5650",
   "metadata": {},
   "source": [
    "Let's go ahead and also pull out the differentially expressed genes that don't correspond to a ChIP-seq peak. In other words, we would be identifying the genes that are present from our RNA-seq analysis but not in our ChIP-seq annotation. We can pull these out from our <code>venn</code> object, or we can identify these genes using set operators."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bd5440fb-f720-428e-bf7a-7b68a733da92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GAPVD1',\n",
       " 'TEPP',\n",
       " 'CHCHD3',\n",
       " 'DDX39A',\n",
       " 'TUBB6',\n",
       " 'GLYR1',\n",
       " 'YWHAE',\n",
       " 'CYP1B1',\n",
       " 'TFAP2C',\n",
       " 'GPRC5A']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We can pull out the difference as well, specifically just our RNA-seq hits that DO NOT intersect with our ChIP peaks\n",
    "venn['x_only'][0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2cd4acc9-3abf-4a69-8b6a-d05379a23972",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GAPVD1',\n",
       " 'TEPP',\n",
       " 'CHCHD3',\n",
       " 'DDX39A',\n",
       " 'TUBB6',\n",
       " 'GLYR1',\n",
       " 'YWHAE',\n",
       " 'CYP1B1',\n",
       " 'TFAP2C',\n",
       " 'UROS']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We can also do the same with set operators to find the difference\n",
    "list(set(de_genes) - set(chip_genes))[0:10]\n",
    "\n",
    "#These genes are likely indirectly regulated by TAZ\n",
    "#Since they don't intersect with a ChIP peak, it's possible that things directly regulated by TAZ\n",
    "#or other secondary changes within the cell drive the changes in gene expression that we see with these genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f63d8bac-ca88-4908-85b6-5c47a442f67d",
   "metadata": {},
   "source": [
    "How are these genes different than the overlapping set of genes in terms of how TAZ is regulating them?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4df9d3dc-ff79-494e-ae03-841ae05afbd5",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise: Compare expression changes between directly and indirectly regulated genes</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03f95dd1-33a5-4dc3-9134-98859997083a",
   "metadata": {},
   "source": [
    "For this exercise, take a look at the expression changes between genes directly regulated by TAZ and those that have their expression impacted based on more indirect regulation/effects."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38cc489c-bbc5-4ac3-96e7-56d073ba25a3",
   "metadata": {},
   "source": [
    "<h2>1. Create DataFrames for overlapping and non-overlapping sets containing <code>DESeq2</code> results</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d402077-461c-4b60-a7a0-dd486dd79c32",
   "metadata": {},
   "source": [
    "For this part, you can set up two DataFrames, one for the overlapping set and the other for the non-overlapping set (RNA-seq only) with the <code>DESeq2</code> results pulled into them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cb2f17ca-79b4-4182-ba20-079e1cf5ec97",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We can then use the pd.Series.isin() function to pull out the information just for genes contained within the intersection\n",
    "overlap = res.loc[res.gene.isin(venn['xy_only']), :]\n",
    "#So this would just be information on genes that are directly regulated by TAZ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f298142e-7218-4b02-af14-87fce386500b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_07311\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_07311_level0_col0\" class=\"col_heading level0 col0\" >gene</th>\n",
       "      <th id=\"T_07311_level0_col1\" class=\"col_heading level0 col1\" >baseMean</th>\n",
       "      <th id=\"T_07311_level0_col2\" class=\"col_heading level0 col2\" >log2FoldChange</th>\n",
       "      <th id=\"T_07311_level0_col3\" class=\"col_heading level0 col3\" >lfcSE</th>\n",
       "      <th id=\"T_07311_level0_col4\" class=\"col_heading level0 col4\" >stat</th>\n",
       "      <th id=\"T_07311_level0_col5\" class=\"col_heading level0 col5\" >pvalue</th>\n",
       "      <th id=\"T_07311_level0_col6\" class=\"col_heading level0 col6\" >padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_07311_level0_row0\" class=\"row_heading level0 row0\" >84</th>\n",
       "      <td id=\"T_07311_row0_col0\" class=\"data row0 col0\" >ABI3BP</td>\n",
       "      <td id=\"T_07311_row0_col1\" class=\"data row0 col1\" >2.217591</td>\n",
       "      <td id=\"T_07311_row0_col2\" class=\"data row0 col2\" >-3.911127</td>\n",
       "      <td id=\"T_07311_row0_col3\" class=\"data row0 col3\" >0.806348</td>\n",
       "      <td id=\"T_07311_row0_col4\" class=\"data row0 col4\" >-4.850422</td>\n",
       "      <td id=\"T_07311_row0_col5\" class=\"data row0 col5\" >0.000001</td>\n",
       "      <td id=\"T_07311_row0_col6\" class=\"data row0 col6\" >0.000015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_07311_level0_row1\" class=\"row_heading level0 row1\" >97</th>\n",
       "      <td id=\"T_07311_row1_col0\" class=\"data row1 col0\" >ABTB2</td>\n",
       "      <td id=\"T_07311_row1_col1\" class=\"data row1 col1\" >24.241325</td>\n",
       "      <td id=\"T_07311_row1_col2\" class=\"data row1 col2\" >0.442388</td>\n",
       "      <td id=\"T_07311_row1_col3\" class=\"data row1 col3\" >0.171200</td>\n",
       "      <td id=\"T_07311_row1_col4\" class=\"data row1 col4\" >2.584034</td>\n",
       "      <td id=\"T_07311_row1_col5\" class=\"data row1 col5\" >0.009765</td>\n",
       "      <td id=\"T_07311_row1_col6\" class=\"data row1 col6\" >0.039476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_07311_level0_row2\" class=\"row_heading level0 row2\" >115</th>\n",
       "      <td id=\"T_07311_row2_col0\" class=\"data row2 col0\" >ACAT2</td>\n",
       "      <td id=\"T_07311_row2_col1\" class=\"data row2 col1\" >58.251661</td>\n",
       "      <td id=\"T_07311_row2_col2\" class=\"data row2 col2\" >-0.310600</td>\n",
       "      <td id=\"T_07311_row2_col3\" class=\"data row2 col3\" >0.113016</td>\n",
       "      <td id=\"T_07311_row2_col4\" class=\"data row2 col4\" >-2.748280</td>\n",
       "      <td id=\"T_07311_row2_col5\" class=\"data row2 col5\" >0.005991</td>\n",
       "      <td id=\"T_07311_row2_col6\" class=\"data row2 col6\" >0.026466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_07311_level0_row3\" class=\"row_heading level0 row3\" >173</th>\n",
       "      <td id=\"T_07311_row3_col0\" class=\"data row3 col0\" >ACTN1</td>\n",
       "      <td id=\"T_07311_row3_col1\" class=\"data row3 col1\" >527.742339</td>\n",
       "      <td id=\"T_07311_row3_col2\" class=\"data row3 col2\" >-0.292663</td>\n",
       "      <td id=\"T_07311_row3_col3\" class=\"data row3 col3\" >0.062124</td>\n",
       "      <td id=\"T_07311_row3_col4\" class=\"data row3 col4\" >-4.710941</td>\n",
       "      <td id=\"T_07311_row3_col5\" class=\"data row3 col5\" >0.000002</td>\n",
       "      <td id=\"T_07311_row3_col6\" class=\"data row3 col6\" >0.000029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_07311_level0_row4\" class=\"row_heading level0 row4\" >175</th>\n",
       "      <td id=\"T_07311_row4_col0\" class=\"data row4 col0\" >ACTN4</td>\n",
       "      <td id=\"T_07311_row4_col1\" class=\"data row4 col1\" >1001.076407</td>\n",
       "      <td id=\"T_07311_row4_col2\" class=\"data row4 col2\" >-0.293635</td>\n",
       "      <td id=\"T_07311_row4_col3\" class=\"data row4 col3\" >0.055564</td>\n",
       "      <td id=\"T_07311_row4_col4\" class=\"data row4 col4\" >-5.284628</td>\n",
       "      <td id=\"T_07311_row4_col5\" class=\"data row4 col5\" >0.000000</td>\n",
       "      <td id=\"T_07311_row4_col6\" class=\"data row4 col6\" >0.000002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x79948ded6710>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap.head().style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1aa4fcb9-7c42-494c-9060-4990e1674ad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We can use the same setup to also pull those that are indirect targets\n",
    "no_overlap = res.loc[res.gene.isin(venn['x_only']), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0b0bfd4f-e88e-42fa-93b7-d241eed19348",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_502cb\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_502cb_level0_col0\" class=\"col_heading level0 col0\" >gene</th>\n",
       "      <th id=\"T_502cb_level0_col1\" class=\"col_heading level0 col1\" >baseMean</th>\n",
       "      <th id=\"T_502cb_level0_col2\" class=\"col_heading level0 col2\" >log2FoldChange</th>\n",
       "      <th id=\"T_502cb_level0_col3\" class=\"col_heading level0 col3\" >lfcSE</th>\n",
       "      <th id=\"T_502cb_level0_col4\" class=\"col_heading level0 col4\" >stat</th>\n",
       "      <th id=\"T_502cb_level0_col5\" class=\"col_heading level0 col5\" >pvalue</th>\n",
       "      <th id=\"T_502cb_level0_col6\" class=\"col_heading level0 col6\" >padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_502cb_level0_row0\" class=\"row_heading level0 row0\" >11</th>\n",
       "      <td id=\"T_502cb_row0_col0\" class=\"data row0 col0\" >AARS1</td>\n",
       "      <td id=\"T_502cb_row0_col1\" class=\"data row0 col1\" >75.605433</td>\n",
       "      <td id=\"T_502cb_row0_col2\" class=\"data row0 col2\" >-0.433385</td>\n",
       "      <td id=\"T_502cb_row0_col3\" class=\"data row0 col3\" >0.100656</td>\n",
       "      <td id=\"T_502cb_row0_col4\" class=\"data row0 col4\" >-4.305609</td>\n",
       "      <td id=\"T_502cb_row0_col5\" class=\"data row0 col5\" >0.000017</td>\n",
       "      <td id=\"T_502cb_row0_col6\" class=\"data row0 col6\" >0.000164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_502cb_level0_row1\" class=\"row_heading level0 row1\" >21</th>\n",
       "      <td id=\"T_502cb_row1_col0\" class=\"data row1 col0\" >ABCA1</td>\n",
       "      <td id=\"T_502cb_row1_col1\" class=\"data row1 col1\" >41.210361</td>\n",
       "      <td id=\"T_502cb_row1_col2\" class=\"data row1 col2\" >-0.653427</td>\n",
       "      <td id=\"T_502cb_row1_col3\" class=\"data row1 col3\" >0.163851</td>\n",
       "      <td id=\"T_502cb_row1_col4\" class=\"data row1 col4\" >-3.987933</td>\n",
       "      <td id=\"T_502cb_row1_col5\" class=\"data row1 col5\" >0.000067</td>\n",
       "      <td id=\"T_502cb_row1_col6\" class=\"data row1 col6\" >0.000564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_502cb_level0_row2\" class=\"row_heading level0 row2\" >26</th>\n",
       "      <td id=\"T_502cb_row2_col0\" class=\"data row2 col0\" >ABCA2</td>\n",
       "      <td id=\"T_502cb_row2_col1\" class=\"data row2 col1\" >65.941915</td>\n",
       "      <td id=\"T_502cb_row2_col2\" class=\"data row2 col2\" >0.342784</td>\n",
       "      <td id=\"T_502cb_row2_col3\" class=\"data row2 col3\" >0.119841</td>\n",
       "      <td id=\"T_502cb_row2_col4\" class=\"data row2 col4\" >2.860329</td>\n",
       "      <td id=\"T_502cb_row2_col5\" class=\"data row2 col5\" >0.004232</td>\n",
       "      <td id=\"T_502cb_row2_col6\" class=\"data row2 col6\" >0.020035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_502cb_level0_row3\" class=\"row_heading level0 row3\" >27</th>\n",
       "      <td id=\"T_502cb_row3_col0\" class=\"data row3 col0\" >ABCA3</td>\n",
       "      <td id=\"T_502cb_row3_col1\" class=\"data row3 col1\" >32.013130</td>\n",
       "      <td id=\"T_502cb_row3_col2\" class=\"data row3 col2\" >0.697068</td>\n",
       "      <td id=\"T_502cb_row3_col3\" class=\"data row3 col3\" >0.193700</td>\n",
       "      <td id=\"T_502cb_row3_col4\" class=\"data row3 col4\" >3.598708</td>\n",
       "      <td id=\"T_502cb_row3_col5\" class=\"data row3 col5\" >0.000320</td>\n",
       "      <td id=\"T_502cb_row3_col6\" class=\"data row3 col6\" >0.002232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_502cb_level0_row4\" class=\"row_heading level0 row4\" >31</th>\n",
       "      <td id=\"T_502cb_row4_col0\" class=\"data row4 col0\" >ABCA7</td>\n",
       "      <td id=\"T_502cb_row4_col1\" class=\"data row4 col1\" >10.881120</td>\n",
       "      <td id=\"T_502cb_row4_col2\" class=\"data row4 col2\" >-0.842159</td>\n",
       "      <td id=\"T_502cb_row4_col3\" class=\"data row4 col3\" >0.288254</td>\n",
       "      <td id=\"T_502cb_row4_col4\" class=\"data row4 col4\" >-2.921590</td>\n",
       "      <td id=\"T_502cb_row4_col5\" class=\"data row4 col5\" >0.003483</td>\n",
       "      <td id=\"T_502cb_row4_col6\" class=\"data row4 col6\" >0.017020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x79948e092e90>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_overlap.head().style"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4680e6f3-a6d2-4a38-ba7a-cf1fe5edf1d4",
   "metadata": {},
   "source": [
    "<h2>2. Visualize distribution of <code>'log2FoldChange'</code> for both sets</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "413af805-4da1-4429-a5d1-7673e71d9b31",
   "metadata": {},
   "source": [
    "Once you have the <code>DESeq2</code> results corresponding to each set, visualize the distribution of the expression changes between the two to see if there may be differences in how the loss of TAZ affects their expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "896f5ea1-a7e7-469d-81e6-b4049fb1320e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='log2FoldChange', ylabel='Proportion'>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkkAAAGwCAYAAAC99fF4AAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAARutJREFUeJzt3XtYVWX+///XVk6KSp4CKUQyT6SZYSkY2kExstKmAzqFNql9GDNFahod6+NhKsspJTM1/ZjmVak1ajUjhTiTZ7REaPJQWuqgCBGW4CFB4P7+4c/9a7sXCBtkAz4f17Wui33ve933e7Fo93KttdeyGWOMAAAA4KCBuwsAAACojQhJAAAAFghJAAAAFghJAAAAFghJAAAAFghJAAAAFghJAAAAFjzcXUBdVVpaqmPHjqlp06ay2WzuLgcAAFSAMUYnT55UYGCgGjQo/1gRIclFx44dU1BQkLvLAAAALjhy5IiuvfbacvsQklzUtGlTSed/yc2aNXNzNQAAoCIKCgoUFBRk//94eQhJLrpwiq1Zs2aEJAAA6piKXCrDhdsAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWCEkAAAAWPNxdAAC4w6ixE5SVd8Kh7ZpWV+n/5s52T0EAah1CEoArUlbeCfnd+aRj278XuqkaALURp9sAAAAsEJIAAAAsEJIAAAAsEJIAAAAsEJIAAAAsEJIAAAAsuD0kzZs3TyEhIfLx8VFYWJg2b95cZt/Vq1drwIABat26tZo1a6bw8HAlJyc79Vu1apVCQ0Pl7e2t0NBQrVmzpkrzAgCAK49bQ9LKlSsVHx+vyZMnKz09XZGRkYqOjlZmZqZl/02bNmnAgAFKSkpSWlqa7rjjDt13331KT0+390lNTVVMTIxiY2P19ddfKzY2Vo888oh27Njh8rwAAODKYzPGGHdN3qtXL918882aP3++va1Lly4aMmSIZsyYUaExbrjhBsXExOh///d/JUkxMTEqKCjQZ599Zu9z9913q3nz5lq+fLnL8xYWFqqwsND+uqCgQEFBQcrPz1ezZs0qvtEAaoXooX9wuplk/r8X6rMVS9xUEYCaUFBQID8/vwr9/9ttR5KKioqUlpamqKgoh/aoqCht27atQmOUlpbq5MmTatGihb0tNTXVacyBAwfax3R13hkzZsjPz8++BAUFVahGAABQN7ktJOXl5amkpET+/v4O7f7+/srJyanQGK+//rpOnz6tRx55xN6Wk5NT7piuzjtp0iTl5+fblyNHjlSoRgAAUDe5/dltNpvN4bUxxqnNyvLlyzV16lR98sknuvrqqys9ZmXn9fb2lre39yXrAgAA9YPbQlKrVq3UsGFDp6M3ubm5Tkd5LrZy5UqNHDlSH330kfr37+/wXkBAQLljVmVeAABw5XDb6TYvLy+FhYUpJSXFoT0lJUURERFlrrd8+XI9/vjj+uCDDzRo0CCn98PDw53GXLdunX1MV+cFAABXFreebktISFBsbKx69uyp8PBwLVy4UJmZmYqLi5N0/jqgrKwsLVu2TNL5gDR8+HC98cYb6t27t/1oUKNGjeTn5ydJGj9+vPr27atXX31VgwcP1ieffKL169dry5YtFZ4XAADArSEpJiZGx48f1/Tp05Wdna2uXbsqKSlJwcHBkqTs7GyHexe9/fbbKi4u1lNPPaWnnnrK3j5ixAgtXbpUkhQREaEVK1bo+eef1wsvvKD27dtr5cqV6tWrV4XnBQAAcOt9kuqyytxnAUDtw32SgCtTnbhPEgAAQG1GSAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALBASAIAALDg9pA0b948hYSEyMfHR2FhYdq8eXOZfbOzs/X73/9enTp1UoMGDRQfH+/U5/bbb5fNZnNaBg0aZO8zdepUp/cDAgIux+YBAIA6yq0haeXKlYqPj9fkyZOVnp6uyMhIRUdHKzMz07J/YWGhWrdurcmTJ6t79+6WfVavXq3s7Gz7snv3bjVs2FAPP/ywQ78bbrjBod8333xT7dsHAADqLg93Tj5r1iyNHDlSo0aNkiQlJiYqOTlZ8+fP14wZM5z6t2vXTm+88YYk6Z133rEcs0WLFg6vV6xYocaNGzuFJA8PD44eAQCAMrntSFJRUZHS0tIUFRXl0B4VFaVt27ZV2zyLFy/W0KFD5evr69B+4MABBQYGKiQkREOHDtXBgwfLHaewsFAFBQUOCwAAqL/cFpLy8vJUUlIif39/h3Z/f3/l5ORUyxxffvmldu/ebT9SdUGvXr20bNkyJScna9GiRcrJyVFERISOHz9e5lgzZsyQn5+ffQkKCqqWGgEAQO3k9gu3bTabw2tjjFObqxYvXqyuXbvq1ltvdWiPjo7Wgw8+qG7duql///5au3atJOndd98tc6xJkyYpPz/fvhw5cqRaagQAALWT265JatWqlRo2bOh01Cg3N9fp6JIrzpw5oxUrVmj69OmX7Ovr66tu3brpwIEDZfbx9vaWt7d3lesCAAB1g9uOJHl5eSksLEwpKSkO7SkpKYqIiKjy+B9++KEKCwv12GOPXbJvYWGh9u3bpzZt2lR5XgAAUD+49dttCQkJio2NVc+ePRUeHq6FCxcqMzNTcXFxks6f4srKytKyZcvs62RkZEiSTp06pZ9++kkZGRny8vJSaGiow9iLFy/WkCFD1LJlS6d5n332Wd13331q27atcnNz9eKLL6qgoEAjRoy4fBsLAADqFLeGpJiYGB0/flzTp09Xdna2unbtqqSkJAUHB0s6f/PIi++Z1KNHD/vPaWlp+uCDDxQcHKzDhw/b2/fv368tW7Zo3bp1lvMePXpUw4YNU15enlq3bq3evXtr+/bt9nkBAABsxhjj7iLqooKCAvn5+Sk/P1/NmjVzdzkAKil66B/kd+eTDm35/16oz1YscVNFAGpCZf7/7fZvtwEAANRGhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALhCQAAAALHu4uAADc4cfdW3X22F6Htvyff3FTNQBqI0ISgCuSr85o7t3+Dm1xH5xxUzUAaiNOtwEAAFggJAEAAFggJAEAAFggJAEAAFjgwm0AcNHEsU/ozPEsh7bGLa/RK3PfcVNFAKoTIQlAvTZq7ARl5Z1waj99purfZDtzPEtzBno6tI1LziqjN4C6hpAEoF7LyjshvzufdGrP3v5xzRcDoE7hmiQAAAALbg9J8+bNU0hIiHx8fBQWFqbNmzeX2Tc7O1u///3v1alTJzVo0EDx8fFOfZYuXSqbzea0nD171uV5AQDAlcetIWnlypWKj4/X5MmTlZ6ersjISEVHRyszM9Oyf2FhoVq3bq3Jkyere/fuZY7brFkzZWdnOyw+Pj4uzwsAAK48bg1Js2bN0siRIzVq1Ch16dJFiYmJCgoK0vz58y37t2vXTm+88YaGDx8uPz+/Mse12WwKCAhwWKoyLwAAuPK4LSQVFRUpLS1NUVFRDu1RUVHatm1blcY+deqUgoODde211+ree+9Venp6lectLCxUQUGBwwIAAOovt4WkvLw8lZSUyN/f8QGT/v7+ysnJcXnczp07a+nSpfr000+1fPly+fj4qE+fPjpw4ECV5p0xY4b8/PzsS1BQkMs1AgCA2s/tF27bbDaH18YYp7bK6N27tx577DF1795dkZGR+vDDD9WxY0e9+eabVZp30qRJys/Pty9HjhxxuUYAAFD7ue0+Sa1atVLDhg2djt7k5uY6HeWpigYNGuiWW26xH0lydV5vb295e3tXW10AAKB2c9uRJC8vL4WFhSklJcWhPSUlRREREdU2jzFGGRkZatOmTY3OCwAA6ja33nE7ISFBsbGx6tmzp8LDw7Vw4UJlZmYqLi5O0vlTXFlZWVq2bJl9nYyMDEnnL87+6aeflJGRIS8vL4WGhkqSpk2bpt69e6tDhw4qKCjQnDlzlJGRobfeeqvC8wIAALg1JMXExOj48eOaPn26srOz1bVrVyUlJSk4OFjS+ZtHXnzvoh49eth/TktL0wcffKDg4GAdPnxYknTixAk9+eSTysnJkZ+fn3r06KFNmzbp1ltvrfC8AAAAbn9225gxYzRmzBjL95YuXerUZowpd7zZs2dr9uzZVZoXAADA7d9uAwAAqI0ISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABY83F0AANQn/9m9R+OGDXRqb9zyGr0y9x03VATAVYQkAKhGXqZQcwZ6OrWPS85yQzUAqoLTbQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABbcHpLmzZunkJAQ+fj4KCwsTJs3by6zb3Z2tn7/+9+rU6dOatCggeLj4536LFq0SJGRkWrevLmaN2+u/v3768svv3ToM3XqVNlsNoclICCgujcNAADUYS6FpNOnT+uFF15QRESErr/+el133XUOS0WtXLlS8fHxmjx5stLT0xUZGano6GhlZmZa9i8sLFTr1q01efJkde/e3bLPhg0bNGzYMH3xxRdKTU1V27ZtFRUVpawsx0cC3HDDDcrOzrYv33zzTcV/AQAAoN5z6dlto0aN0saNGxUbG6s2bdrIZrO5NPmsWbM0cuRIjRo1SpKUmJio5ORkzZ8/XzNmzHDq365dO73xxhuSpHfesX5Q5Pvvv+/wetGiRfr73/+uf/3rXxo+fLi93cPDg6NHAACgTC6FpM8++0xr165Vnz59XJ64qKhIaWlpmjhxokN7VFSUtm3b5vK4Fztz5ozOnTunFi1aOLQfOHBAgYGB8vb2Vq9evfTyyy+XexSssLBQhYWF9tcFBQXVViMAAKh9XDrd1rx5c6fQUVl5eXkqKSmRv7+/Q7u/v79ycnKqNPZvTZw4Uddcc4369+9vb+vVq5eWLVum5ORkLVq0SDk5OYqIiNDx48fLHGfGjBny8/OzL0FBQdVWIwAAqH1cCkl//etf9b//+786c+ZMlQu4+FSdMcbl03cXmzlzppYvX67Vq1fLx8fH3h4dHa0HH3xQ3bp1U//+/bV27VpJ0rvvvlvmWJMmTVJ+fr59OXLkSLXUCAAAaieXTre9/vrr+uGHH+Tv76927drJ09PT4f1du3ZdcoxWrVqpYcOGTkeNcnNznY4uueK1117Tyy+/rPXr1+vGG28st6+vr6+6deumAwcOlNnH29tb3t7eVa4LAADUDS6FpCFDhlR5Yi8vL4WFhSklJUUPPPCAvT0lJUWDBw+u0th/+9vf9OKLLyo5OVk9e/a8ZP/CwkLt27dPkZGRVZoXAADUHy6FpClTplTL5AkJCYqNjVXPnj0VHh6uhQsXKjMzU3FxcZLOn+LKysrSsmXL7OtkZGRIkk6dOqWffvpJGRkZ8vLyUmhoqKTzp9heeOEFffDBB2rXrp39SFWTJk3UpEkTSdKzzz6r++67T23btlVubq5efPFFFRQUaMSIEdWyXQAAoO5zKSRdkJaWpn379slmsyk0NFQ9evSo1PoxMTE6fvy4pk+fruzsbHXt2lVJSUkKDg6WdP7mkRffM+m3c6SlpemDDz5QcHCwDh8+LOn8zSmLior00EMPOaw3ZcoUTZ06VZJ09OhRDRs2THl5eWrdurV69+6t7du32+cFAABwKSTl5uZq6NCh2rBhg6666ioZY5Sfn6877rhDK1asUOvWrSs81pgxYzRmzBjL95YuXerUZowpd7wLYak8K1asqEhpAADgCubSt9uefvppFRQUaM+ePfr555/1yy+/aPfu3SooKNC4ceOqu0YAAIAa59KRpM8//1zr169Xly5d7G2hoaF66623FBUVVW3FAQAAuItLR5JKS0udvvYvSZ6eniotLa1yUQAAAO7mUki68847NX78eB07dszelpWVpQkTJuiuu+6qtuIAAADcxaWQNHfuXJ08eVLt2rVT+/btdf311yskJEQnT57Um2++Wd01AgAA1DiXrkkKCgrSrl27lJKSom+//VbGGIWGhjo8Hw0AAKAuq9J9kgYMGKABAwZUVy0AAAC1RoVD0pw5c/Tkk0/Kx8dHc+bMKbcvtwEAAAB1XYVD0uzZs/Xoo4/Kx8dHs2fPLrOfzWYjJAEAgDqvwiHp0KFDlj8DAADURy59u2369Ok6c+aMU/uvv/6q6dOnV7koAAAAd3MpJE2bNk2nTp1yaj9z5oymTZtW5aIAAADczaWQZIyRzWZzav/666/VokWLKhcFAADgbpW6BUDz5s1ls9lks9nUsWNHh6BUUlKiU6dOKS4urtqLBAAAqGmVCkmJiYkyxuiJJ57QtGnT5OfnZ3/Py8tL7dq1U3h4eLUXCQAAUNMqFZJGjBih4uJiSVL//v117bXXXpaiAAAA3K3S1yR5eHhozJgxKikpuRz1AAAA1AouXbjdq1cvpaenV3ctAAAAtYZLz24bM2aMnnnmGR09elRhYWHy9fV1eP/GG2+sluIAoKp+3L1VZ4/tdWpvUPyrG6oBUJe4FJJiYmIkOT6jzWaz2W8NwKk4ALWFr85o7t3+Tu1/mG/cUA2AusSlkMRjSQAAQH3nUkgKDg6u7joAAABqFZdCkiT98MMPSkxM1L59+2Sz2dSlSxeNHz9e7du3r876AAAA3MKlb7clJycrNDRUX375pW688UZ17dpVO3bs0A033KCUlJTqrhEAAKDGuXQkaeLEiZowYYJeeeUVp/Y///nPGjBgQLUUBwAA4C4uHUnat2+fRo4c6dT+xBNPaO9e56/aAgAA1DUuhaTWrVsrIyPDqT0jI0NXX311VWsCAABwO5dOt40ePVpPPvmkDh48qIiICNlsNm3ZskWvvvqqnnnmmequEQAAoMa5FJJeeOEFNW3aVK+//romTZokSQoMDNTUqVMdbjAJAABQV7kUkmw2myZMmKAJEybo5MmTkqSmTZtWa2EAAADu5PJ9kiQpNzdX3333nWw2mzp16qTWrVtXV10AAABu5dKF2wUFBYqNjVVgYKD69eunvn37KjAwUI899pjy8/Oru0YAAIAa51JIGjVqlHbs2KG1a9fqxIkTys/P1z//+U/t3LlTo0ePrtRY8+bNU0hIiHx8fBQWFqbNmzeX2Tc7O1u///3v1alTJzVo0EDx8fGW/VatWqXQ0FB5e3srNDRUa9asqdK8AADgyuNSSFq7dq3eeecdDRw4UM2aNVPTpk01cOBALVq0SGvXrq3wOCtXrlR8fLwmT56s9PR0RUZGKjo6WpmZmZb9CwsL1bp1a02ePFndu3e37JOamqqYmBjFxsbq66+/VmxsrB555BHt2LHD5XkBAMCVx6WQ1LJlS/n5+Tm1+/n5qXnz5hUeZ9asWRo5cqRGjRqlLl26KDExUUFBQZo/f75l/3bt2umNN97Q8OHDLeeXpMTERA0YMECTJk1S586dNWnSJN11111KTEx0eV4AAHDlcSkkPf/880pISFB2dra9LScnR3/605/0wgsvVGiMoqIipaWlKSoqyqE9KipK27Ztc6UsSeePJF085sCBA+1jujpvYWGhCgoKHBYAV7a93+7X13v2OSynz5xxd1kAqolL326bP3++vv/+ewUHB6tt27aSpMzMTHl7e+unn37S22+/be+7a9cuyzHy8vJUUlIif39/h3Z/f3/l5OS4Upak82GtvDFdnXfGjBmaNm2ay3UBqH8Kz52TR/MQhzZTml1GbwB1jUshaciQIdVWgM1mc3htjHFquxxjVnbeSZMmKSEhwf66oKBAQUFBVaoTAADUXi6FpClTplR54latWqlhw4ZOR29yc3OdjvJURkBAQLljujqvt7e3vL29Xa4LAADULS5dk3RBWlqa3nvvPb3//vtKT0+v1LpeXl4KCwtTSkqKQ3tKSooiIiJcrik8PNxpzHXr1tnHvFzzAgCA+sWlI0m5ubkaOnSoNmzYoKuuukrGGOXn5+uOO+7QihUrKnzn7YSEBMXGxqpnz54KDw/XwoULlZmZqbi4OEnnT3FlZWVp2bJl9nUyMjIkSadOndJPP/2kjIwMeXl5KTQ0VJI0fvx49e3bV6+++qoGDx6sTz75ROvXr9eWLVsqPC8AAIBLIenpp59WQUGB9uzZoy5dukiS9u7dqxEjRmjcuHFavnx5hcaJiYnR8ePHNX36dGVnZ6tr165KSkpScHCwpPM3j7z43kU9evSw/5yWlqYPPvhAwcHBOnz4sCQpIiJCK1as0PPPP68XXnhB7du318qVK9WrV68KzwsAAOBSSPr888+1fv16e0CSpNDQUL311ltOX62/lDFjxmjMmDGW7y1dutSpzRhzyTEfeughPfTQQy7PCwAA4NI1SaWlpfL09HRq9/T0VGlpaZWLAgAAcDeXQtKdd96p8ePH69ixY/a2rKwsTZgwQXfddVe1FQcAAOAuLoWkuXPn6uTJk2rXrp3at2+v66+/XiEhITp58qTefPPN6q4RAACgxrl0TVJQUJB27dqllJQUffvttzLGKDQ0VP3796/u+gAAANyi0iGpuLhYPj4+ysjI0IABAzRgwIDLURcAAIBbVfp0m4eHh4KDg1VSUnI56gEAAKgVXLom6fnnn9ekSZP0888/V3c9AAAAtYJL1yTNmTNH33//vQIDAxUcHCxfX1+H93ft2lUtxQEAALiLSyFpyJAhstlsFbqxIwAAQF1UqZB05swZ/elPf9LHH3+sc+fO6a677tKbb76pVq1aXa76AAAA3KJS1yRNmTJFS5cu1aBBgzRs2DCtX79ef/zjHy9XbQAAAG5TqSNJq1ev1uLFizV06FBJ0qOPPqo+ffqopKREDRs2vCwFAgAAuEOljiQdOXJEkZGR9te33nqrPDw8HB5PAgAAUB9UKiSVlJTIy8vLoc3Dw0PFxcXVWhQAAIC7Vep0mzFGjz/+uLy9ve1tZ8+eVVxcnMNtAFavXl19FQIAALhBpULSiBEjnNoee+yxaisGAACgtqhUSFqyZMnlqgMAAKBWcemxJAAAAPUdIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMACIQkAAMCC20PSvHnzFBISIh8fH4WFhWnz5s3l9t+4caPCwsLk4+Oj6667TgsWLHB4//bbb5fNZnNaBg0aZO8zdepUp/cDAgIuy/YBAIC6ya0haeXKlYqPj9fkyZOVnp6uyMhIRUdHKzMz07L/oUOHdM899ygyMlLp6en6y1/+onHjxmnVqlX2PqtXr1Z2drZ92b17txo2bKiHH37YYawbbrjBod8333xzWbcVAADULR7unHzWrFkaOXKkRo0aJUlKTExUcnKy5s+frxkzZjj1X7Bggdq2bavExERJUpcuXbRz50699tprevDBByVJLVq0cFhnxYoVaty4sVNI8vDwqNTRo8LCQhUWFtpfFxQUVHhdAABQ97jtSFJRUZHS0tIUFRXl0B4VFaVt27ZZrpOamurUf+DAgdq5c6fOnTtnuc7ixYs1dOhQ+fr6OrQfOHBAgYGBCgkJ0dChQ3Xw4MFy650xY4b8/PzsS1BQ0KU2EQAA1GFuC0l5eXkqKSmRv7+/Q7u/v79ycnIs18nJybHsX1xcrLy8PKf+X375pXbv3m0/UnVBr169tGzZMiUnJ2vRokXKyclRRESEjh8/Xma9kyZNUn5+vn05cuRIRTcVAADUQW493SZJNpvN4bUxxqntUv2t2qXzR5G6du2qW2+91aE9Ojra/nO3bt0UHh6u9u3b691331VCQoLlvN7e3vL29i5/YwAAQL3htiNJrVq1UsOGDZ2OGuXm5jodLbogICDAsr+Hh4datmzp0H7mzBmtWLHC6SiSFV9fX3Xr1k0HDhyo5FYAAID6ym0hycvLS2FhYUpJSXFoT0lJUUREhOU64eHhTv3XrVunnj17ytPT06H9ww8/VGFhoR577LFL1lJYWKh9+/apTZs2ldwKAABQX7n1dFtCQoJiY2PVs2dPhYeHa+HChcrMzFRcXJyk89cBZWVladmyZZKkuLg4zZ07VwkJCRo9erRSU1O1ePFiLV++3GnsxYsXa8iQIU5HmCTp2Wef1X333ae2bdsqNzdXL774ogoKCjRixIjLu8EA6r3ikhJ9vWefU/veb0vdUA2AqnBrSIqJidHx48c1ffp0ZWdnq2vXrkpKSlJwcLAkKTs72+GeSSEhIUpKStKECRP01ltvKTAwUHPmzLF//f+C/fv3a8uWLVq3bp3lvEePHtWwYcOUl5en1q1bq3fv3tq+fbt9XgCoCo/mgU5thecOuaESAFXh9gu3x4wZozFjxli+t3TpUqe2fv36adeuXeWO2bFjR/sF3VZWrFhRqRoBAMCVx+2PJQEAAKiNCEkAAAAW3H66DQCqw6ixE5SVd8Kp/fSZMzVfDIB6gZAEoF7IyjshvzufdGrP3v5xzRcDoF7gdBsAAIAFQhIAAIAFTrcBQA04ceKExg0b6NTeuOU1emXuO26oCMClEJIAoAY09SjVnIGeTu3jkrPcUA2AiuB0GwAAgAVCEgAAgAVCEgAAgAVCEgAAgAVCEgAAgAVCEgAAgAVCEgAAgAXukwSgXvhx91adPbbXqb1B8a9uqAZAfUBIAlAv+OqM5t7t79T+h/nGDdUAqA843QYAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGCBkAQAAGDB7SFp3rx5CgkJkY+Pj8LCwrR58+Zy+2/cuFFhYWHy8fHRddddpwULFji8v3TpUtlsNqfl7NmzVZoXAABcWdwaklauXKn4+HhNnjxZ6enpioyMVHR0tDIzMy37Hzp0SPfcc48iIyOVnp6uv/zlLxo3bpxWrVrl0K9Zs2bKzs52WHx8fFyeFwAAXHncGpJmzZqlkSNHatSoUerSpYsSExMVFBSk+fPnW/ZfsGCB2rZtq8TERHXp0kWjRo3SE088oddee82hn81mU0BAgMNSlXkBAMCVx20hqaioSGlpaYqKinJoj4qK0rZt2yzXSU1Ndeo/cOBA7dy5U+fOnbO3nTp1SsHBwbr22mt17733Kj09vUrzSlJhYaEKCgocFgAAUH+5LSTl5eWppKRE/v7+Du3+/v7KycmxXCcnJ8eyf3FxsfLy8iRJnTt31tKlS/Xpp59q+fLl8vHxUZ8+fXTgwAGX55WkGTNmyM/Pz74EBQVVepsBAEDd4fYLt202m8NrY4xT26X6/7a9d+/eeuyxx9S9e3dFRkbqww8/VMeOHfXmm29Wad5JkyYpPz/fvhw5cuTSGwcAAOosD3dN3KpVKzVs2NDp6E1ubq7TUZ4LAgICLPt7eHioZcuWlus0aNBAt9xyi/1IkivzSpK3t7e8vb0vuV0AAKB+cNuRJC8vL4WFhSklJcWhPSUlRREREZbrhIeHO/Vft26devbsKU9PT8t1jDHKyMhQmzZtXJ4XAABcedx2JEmSEhISFBsbq549eyo8PFwLFy5UZmam4uLiJJ0/xZWVlaVly5ZJkuLi4jR37lwlJCRo9OjRSk1N1eLFi7V8+XL7mNOmTVPv3r3VoUMHFRQUaM6cOcrIyNBbb71V4XkBAADcGpJiYmJ0/PhxTZ8+XdnZ2eratauSkpIUHBwsScrOzna4d1FISIiSkpI0YcIEvfXWWwoMDNScOXP04IMP2vucOHFCTz75pHJycuTn56cePXpo06ZNuvXWWys8LwAAgFtDkiSNGTNGY8aMsXxv6dKlTm39+vXTrl27yhxv9uzZmj17dpXmBQAAcPu32wAAAGojQhIAAIAFQhIAAIAFt1+TBABlGTV2grLyTji0HT6wX+06dHTqe/rMmRqqyjXFJSX6es8+p/a935a6oRoAFUFIAlBrZeWdkN+dTzq0nfgm3qlNkrK3f1xDVbnOo3mgU1vhuUNuqARARXC6DQAAwAIhCQAAwAIhCQAAwALXJAGoU7xPHNKP78U7tTco/rXmiwFQrxGSANQpV3me09y7vZza/zDfuKEaAPUZp9sAAAAsEJIAAAAsEJIAAAAsEJIAAAAsEJIAAAAsEJIAAAAsEJIAAAAscJ8kAHCjEydOaNywgU7tjVteo1fmvuOGigBcQEgCADdq6lGqOQM9ndrHJWe5oRoAv8XpNgAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAs8uw1AjRk1doKy8k44tV/T6ir939zZNV8QAJSDkASgxmTlnZDfnU86t/97oRuqAYDyuf1027x58xQSEiIfHx+FhYVp8+bN5fbfuHGjwsLC5OPjo+uuu04LFixweH/RokWKjIxU8+bN1bx5c/Xv319ffvmlQ5+pU6fKZrM5LAEBAdW+bQAAoO5y65GklStXKj4+XvPmzVOfPn309ttvKzo6Wnv37lXbtm2d+h86dEj33HOPRo8erffee09bt27VmDFj1Lp1az344IOSpA0bNmjYsGGKiIiQj4+PZs6cqaioKO3Zs0fXXHONfawbbrhB69evt79u2LDh5d9gALXaiRMnNG7YQKf2xi2v0Stz33FDRQDcya0hadasWRo5cqRGjRolSUpMTFRycrLmz5+vGTNmOPVfsGCB2rZtq8TERElSly5dtHPnTr322mv2kPT+++87rLNo0SL9/e9/17/+9S8NHz7c3u7h4VGpo0eFhYUqLCy0vy4oKKjwugDqhqYepZoz0NOpfVxylhuqAeBubgtJRUVFSktL08SJEx3ao6KitG3bNst1UlNTFRUV5dA2cOBALV68WOfOnZOnp/OH25kzZ3Tu3Dm1aNHCof3AgQMKDAyUt7e3evXqpZdfflnXXXddmfXOmDFD06ZNq+jmAaiEPbu/UfTQPzi179+WpLbH9jq0NSj+tabKAnCFc1tIysvLU0lJifz9/R3a/f39lZOTY7lOTk6OZf/i4mLl5eWpTZs2TutMnDhR11xzjfr3729v69Wrl5YtW6aOHTvqxx9/1IsvvqiIiAjt2bNHLVu2tJx70qRJSkhIsL8uKChQUFBQhbcXQNnOmYaWF3Q33f6x5t7t5dD2h/mmpsoCcIVz+7fbbDabw2tjjFPbpfpbtUvSzJkztXz5cm3YsEE+Pj729ujoaPvP3bp1U3h4uNq3b693333XIQj9lre3t7y9vS+9QQAAoF5wW0hq1aqVGjZs6HTUKDc31+lo0QUBAQGW/T08PJyOAL322mt6+eWXtX79et14443l1uLr66tu3brpwIEDLmwJAACoj9x2CwAvLy+FhYUpJSXFoT0lJUURERGW64SHhzv1X7dunXr27OlwPdLf/vY3/fWvf9Xnn3+unj17XrKWwsJC7du3z/J0HQAAuDK59XRbQkKCYmNj1bNnT4WHh2vhwoXKzMxUXFycpPPXAWVlZWnZsmWSpLi4OM2dO1cJCQkaPXq0UlNTtXjxYi1fvtw+5syZM/XCCy/ogw8+ULt27exHnpo0aaImTZpIkp599lndd999atu2rXJzc/Xiiy+qoKBAI0aMqOHfAOBe3AHbUXFJib7es8+pfe+3pfVqTgAV49aQFBMTo+PHj2v69OnKzs5W165dlZSUpODgYElSdna2MjMz7f1DQkKUlJSkCRMm6K233lJgYKDmzJlj//q/dP7mlEVFRXrooYcc5poyZYqmTp0qSTp69KiGDRumvLw8tW7dWr1799b27dvt8wJXCu6A7cyjeaBTW+G5Q/VuTgCX5vYLt8eMGaMxY8ZYvrd06VKntn79+mnXrl1ljnf48OFLzrlixYqKlgcAbmF1Y0tuagnULLeHJACAM6sbW3JTS6Bmuf3ZbQAAALURIQkAAMACp9sAVEl1fEPO+8Qh/fhevFM7jyAB4E6EJABVUh3fkLvK85zT40ckHkECwL043QYAAGCBkAQAAGCBkAQAAGCBa5KAeqS6HjOyZ/c3ih76hyqNAQB1HSEJqEeq6zEj50xDp3Gu5EeVALgyEZIAoI74z+49To8qkXhcCXC5EJIAoI7wMoVOjyqReFwJcLkQkgDgEqweNitJRafy3VANgJpCSAKuAFYXYkvSvm/3q/edNTdGXWX1sFlJikgtcUM1AGoKIQm4AlhdiC1Jhd/EW/a3ekxIg18y5XfnmxUeAwDqOkISACdWjwl5dOE5N1UDAO5BSAJQIYW//mr5EFrvE4cs+/+4e6vTdTw/7j5keUQLAGojQhKACmnqZSwfQlvWESZfnXG6jifymzOXpTYAuBx4LAkAAIAFjiQB1aC6HgcClOf0mTP6es8+p/a935a6oRqg/iMkAdWguh4HUt+dOHFCHhbXNTUo/rXmi6mDTKmRR/NAp/bCc9bXhQGoGkISgCop64JuqxstNvUotbyu6Q/zzeUoDQCqhJAEXMGs7ockVe7ITlkXdI9eyI0WAdRthCTgClBWGGphK+DIDgCUgZAEXEZlPcqjOi7otrpYvKxHhFjdHFIiDFVUcUmJ5QXTJaWX72iZ1ZxlzVfWs+Uat7xGr8x9p9pr44sKuFIQkoDLqKzHgVTHBd1WF4vziJDLx+qCaWOO1uicZc1X1rPlxiVnObVNHPuEzhx3bq9MoOKLCrhSEJKAeqQ6rjFC/XbmeJZloLr99XU1ejQKqAsISUAd9ePurTp7bK9DG9cYwVVeprDCR6OAKwUhCaijfHVGc+/2d2gjDOG3/rN7j9PRoQPf7pEG3uSegoA6hpCEWqWyF4TWpwtIy9qWwwf2q12Hjk7tp8/U7uegVebCY1RNWReWnzv9i9PRoT47frHsW5m/J6ujmJKU//MvFR4DqAsISahVKntBaH26gLSsbTnxTbxle/b2j2ugqqqp6IXHqDrLC8tLsy3arO/abdW3LFZHMSUp7oPaHdyBynJ7SJo3b57+9re/KTs7WzfccIMSExMVGRlZZv+NGzcqISFBe/bsUWBgoJ577jnFxcU59Fm1apVeeOEF/fDDD2rfvr1eeuklPfDAA1WaF3BFWRdS5xz8VuOGHXNo+3H3IcswxMXYAOAebg1JK1euVHx8vObNm6c+ffro7bffVnR0tPbu3au2bds69T906JDuuecejR49Wu+99562bt2qMWPGqHXr1nrwwQclSampqYqJidFf//pXPfDAA1qzZo0eeeQRbdmyRb169XJpXtQP1fHV57LGOLxluwIsTj+UdSH16IXnnE6DdNt8jBs+otaxuq7J6pEzrrD674lv06E2cWtImjVrlkaOHKlRo0ZJkhITE5WcnKz58+drxowZTv0XLFigtm3bKjExUZLUpUsX7dy5U6+99po9JCUmJmrAgAGaNGmSJGnSpEnauHGjEhMTtXz5cpfmRf1QHV99LmuMiNSq36yR55qhNrL61ltEauVuarn3wCGFdghxaj/w7R59Nv4mhza+TYfaxG0hqaioSGlpaZo4caJDe1RUlLZt22a5TmpqqqKiohzaBg4cqMWLF+vcuXPy9PRUamqqJkyY4NTnQrByZV5JKiwsVGFhof11fv75f0kVFBSUv6F13JRnx+jXn52vVWjUoo2mvTav2ucrPlekc7+edmo/9vUmxT10l3P7nsNqHP6o5TgX75uic8UqsDhD1aD4V73Y17n9njc/d5rzh/37VNC3m/N8pUanzhY7tZeU0V5calTw67kqjWHVXpm+7hi7toxxpdRXmb8zq76SVFxSWuExGjUosfxvaUjGcb3Y1zkkDcn41WnsonPFl+1ztaY/z1A7Xfj7MqYC/wA1bpKVlWUkma1btzq0v/TSS6Zjx46W63To0MG89NJLDm1bt241ksyxY8eMMcZ4enqa999/36HP+++/b7y8vFye1xhjpkyZYiSxsLCwsLCw1IPlyJEjl8wqbr9w22azObw2xji1Xar/xe0VGbOy806aNEkJCQn216Wlpfr555/VsmXLctdzRUFBgYKCgnTkyBE1a9asWseuLdjG+oFtrB/YxvqBbawYY4xOnjypwEDnb3lezG0hqVWrVmrYsKFycnIc2nNzc+Xv7/zVUkkKCAiw7O/h4aGWLVuW2+fCmK7MK0ne3t7y9vZ2aLvqqqvK3sBq0KxZs3r7h34B21g/sI31A9tYP7CNl+bn51ehfg1cnqGKvLy8FBYWppSUFIf2lJQURUREWK4THh7u1H/dunXq2bOnPD09y+1zYUxX5gUAAFcet55uS0hIUGxsrHr27Knw8HAtXLhQmZmZ9vseTZo0SVlZWVq2bJkkKS4uTnPnzlVCQoJGjx6t1NRULV682P6tNUkaP368+vbtq1dffVWDBw/WJ598ovXr12vLli0VnhcAAMBtF25f8NZbb5ng4GDj5eVlbr75ZrNx40b7eyNGjDD9+vVz6L9hwwbTo0cP4+XlZdq1a2fmz5/vNOZHH31kOnXqZDw9PU3nzp3NqlWrKjWvu509e9ZMmTLFnD171t2lXDZsY/3ANtYPbGP9wDZWP5sxFfkOHAAAwJXFbdckAQAA1GaEJAAAAAuEJAAAAAuEJAAAAAuEJDd46aWXFBERocaNG5d5Q8rMzEzdd9998vX1VatWrTRu3DgVFRWVO25hYaGefvpptWrVSr6+vrr//vt19OjRy7AFlbNhwwbZbDbL5auvvipzvccff9ypf+/evWuw8spp166dU70XPyPwYsYYTZ06VYGBgWrUqJFuv/127dmzp4YqrrzDhw9r5MiRCgkJUaNGjdS+fXtNmTLlkn+btX1fzps3TyEhIfLx8VFYWJg2b95cbv+NGzcqLCxMPj4+uu6667RgwYIaqrTyZsyYoVtuuUVNmzbV1VdfrSFDhui7774rd52y/pv99ttva6jqypk6dapTrQEBAeWuU5f2oWT9+WKz2fTUU09Z9q8L+3DTpk267777FBgYKJvNpo8//tjhfVc/H1etWqXQ0FB5e3srNDRUa9ascblGQpIbFBUV6eGHH9Yf//hHy/dLSko0aNAgnT59Wlu2bNGKFSu0atUqPfPMM+WOGx8frzVr1mjFihXasmWLTp06pXvvvVclJdZP7K4pERERys7OdlhGjRqldu3aqWfPnuWue/fddzusl5SUVENVu2b69OkO9T7//PPl9p85c6ZmzZqluXPn6quvvlJAQIAGDBigkydP1lDFlfPtt9+qtLRUb7/9tvbs2aPZs2drwYIF+stf/nLJdWvrvly5cqXi4+M1efJkpaenKzIyUtHR0crMzLTsf+jQId1zzz2KjIxUenq6/vKXv2jcuHFatWpVDVdeMRs3btRTTz2l7du3KyUlRcXFxYqKitLp084Pkr7Yd99957DPOnToUAMVu+aGG25wqPWbb74ps29d24eS9NVXXzls34UbIj/88MPlrleb9+Hp06fVvXt3zZ071/J9Vz4fU1NTFRMTo9jYWH399deKjY3VI488oh07drhWZI3caACWlixZYvz8/Jzak5KSTIMGDUxWVpa9bfny5cbb29vk5+dbjnXixAnj6elpVqxYYW/LysoyDRo0MJ9//nm1114VRUVF5uqrrzbTp08vt9+IESPM4MGDa6aoahAcHGxmz55d4f6lpaUmICDAvPLKK/a2s2fPGj8/P7NgwYLLUOHlMXPmTBMSElJun9q8L2+99VYTFxfn0Na5c2czceJEy/7PPfec6dy5s0Pb//zP/5jevXtfthqrU25urpFU7r3hvvjiCyPJ/PLLLzVXWBVMmTLFdO/evcL96/o+NMaY8ePHm/bt25vS0lLL9+vaPpRk1qxZY3/t6ufjI488Yu6++26HtoEDB5qhQ4e6VBdHkmqh1NRUde3a1eHhewMHDlRhYaHS0tIs10lLS9O5c+cUFRVlbwsMDFTXrl21bdu2y15zZXz66afKy8vT448/fsm+GzZs0NVXX62OHTtq9OjRys3NvfwFVsGrr76qli1b6qabbtJLL71U7mmoQ4cOKScnx2GfeXt7q1+/frVun5UnPz9fLVq0uGS/2rgvi4qKlJaW5rAPJCkqKqrMfZCamurUf+DAgdq5c6fOnTt32WqtLvn5+ZJUoX3Wo0cPtWnTRnfddZe++OKLy11alRw4cECBgYEKCQnR0KFDdfDgwTL71vV9WFRUpPfee09PPPHEJR+wXpf24W+5+vlY1r519TOVkFQL5eTkOD1st3nz5vLy8nJ6MO9v1/Hy8lLz5s0d2v39/ctcx10WL16sgQMHKigoqNx+0dHRev/99/Xvf/9br7/+ur766ivdeeedKiwsrKFKK2f8+PFasWKFvvjiC40dO1aJiYkaM2ZMmf0v7JeL93Vt3Gdl+eGHH/Tmm29e8pE+tXVf5uXlqaSkpFL7wOq/T39/fxUXFysvL++y1VodjDFKSEjQbbfdpq5du5bZr02bNlq4cKFWrVql1atXq1OnTrrrrru0adOmGqy24nr16qVly5YpOTlZixYtUk5OjiIiInT8+HHL/nV5H0rSxx9/rBMnTpT7D826tg8v5urnY1n71tXPVLc+u60+mTp1qqZNm1Zun6+++uqS1+BcYPWvA2PMJf/VUB3rVJQr23z06FElJyfrww8/vOT4MTEx9p+7du2qnj17Kjg4WGvXrtXvfvc71wuvhMps44QJE+xtN954o5o3b66HHnrIfnSpLBfvn8u5z8riyr48duyY7r77bj388MMaNWpUuevWhn1ZnsruA6v+Vu21zdixY/Wf//zH4VmWVjp16qROnTrZX4eHh+vIkSN67bXX1Ldv38tdZqVFR0fbf+7WrZvCw8PVvn17vfvuu0pISLBcp67uQ+n8PzSjo6MdzjZcrK7tw7K48vlYnZ+phKRqMnbsWA0dOrTcPu3atavQWAEBAU4Xmf3yyy86d+6cU0L+7TpFRUX65ZdfHI4m5ebmKiIiokLzVpYr27xkyRK1bNlS999/f6Xna9OmjYKDg3XgwIFKr+uqquzXC9/e+v777y1D0oVv3+Tk5KhNmzb29tzc3DL38+VS2e08duyY7rjjDvsDoivLHfvSSqtWrdSwYUOnf2WWtw8CAgIs+3t4eJQbht3t6aef1qeffqpNmzbp2muvrfT6vXv31nvvvXcZKqt+vr6+6tatW5l/X3V1H0rSf//7X61fv16rV6+u9Lp1aR+6+vlY1r519TOVkFRNWrVqpVatWlXLWOHh4XrppZeUnZ1t/+NYt26dvL29FRYWZrlOWFiYPD09lZKSokceeUSSlJ2drd27d2vmzJnVUtfFKrvNxhgtWbJEw4cPl6enZ6XnO378uI4cOeLwH8zlVpX9mp6eLkll1hsSEqKAgAClpKSoR48eks5fa7Bx40a9+uqrrhXsospsZ1ZWlu644w6FhYVpyZIlatCg8mft3bEvrXh5eSksLEwpKSl64IEH7O0pKSkaPHiw5Trh4eH6xz/+4dC2bt069ezZ06W/68vNGKOnn35aa9as0YYNGxQSEuLSOOnp6W7fXxVVWFioffv2KTIy0vL9urYPf2vJkiW6+uqrNWjQoEqvW5f2oaufj+Hh4UpJSXE4sr9u3TrXDxa4dLk3quS///2vSU9PN9OmTTNNmjQx6enpJj093Zw8edIYY0xxcbHp2rWrueuuu8yuXbvM+vXrzbXXXmvGjh1rH+Po0aOmU6dOZseOHfa2uLg4c+2115r169ebXbt2mTvvvNN0797dFBcX1/g2Wlm/fr2RZPbu3Wv5fqdOnczq1auNMcacPHnSPPPMM2bbtm3m0KFD5osvvjDh4eHmmmuuMQUFBTVZdoVs27bNzJo1y6Snp5uDBw+alStXmsDAQHP//fc79PvtNhpjzCuvvGL8/PzM6tWrzTfffGOGDRtm2rRpUyu30Zjz35i8/vrrzZ133mmOHj1qsrOz7ctv1aV9uWLFCuPp6WkWL15s9u7da+Lj442vr685fPiwMcaYiRMnmtjYWHv/gwcPmsaNG5sJEyaYvXv3msWLFxtPT0/z97//3V2bUK4//vGPxs/Pz2zYsMFhf505c8be5+JtnD17tlmzZo3Zv3+/2b17t5k4caKRZFatWuWOTbikZ555xmzYsMEcPHjQbN++3dx7772madOm9WYfXlBSUmLatm1r/vznPzu9Vxf34cmTJ+3//5Nk/wz973//a4yp2OdjbGyswzdRt27daho2bGheeeUVs2/fPvPKK68YDw8Ps337dpdqJCS5wYgRI4wkp+WLL76w9/nvf/9rBg0aZBo1amRatGhhxo4da86ePWt//9ChQ07r/Prrr2bs2LGmRYsWplGjRubee+81mZmZNbhl5Rs2bJiJiIgo831JZsmSJcYYY86cOWOioqJM69atjaenp2nbtq0ZMWJErdqe30pLSzO9evUyfn5+xsfHx3Tq1MlMmTLFnD592qHfb7fRmPNfc50yZYoJCAgw3t7epm/fvuabb76p4eorbsmSJZZ/uxf/e6uu7cu33nrLBAcHGy8vL3PzzTc7fD1+xIgRpl+/fg79N2zYYHr06GG8vLxMu3btzPz582u44oora3/99u/w4m189dVXTfv27Y2Pj49p3ry5ue2228zatWtrvvgKiomJMW3atDGenp4mMDDQ/O53vzN79uyxv1/X9+EFycnJRpL57rvvnN6ri/vwwm0KLl5GjBhhjKnY52O/fv3s/S/46KOPTKdOnYynp6fp3LlzlYKhzZj/72o1AAAA2HELAAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAAAAuEJAAuu/322xUfH+/uMirl8ccf15AhQ8rt48p2HT58WDabTRkZGS7XBqB2ISQBqDVWr16tAQMGqHXr1mrWrJnCw8OVnJzs0Gfq1Kmy2WxOy/r16y9rbd9//73+8Ic/6Nprr5W3t7dCQkI0bNgw7dy587LOC8B9CEkAao1NmzZpwIABSkpKUlpamu644w7dd999Sk9Pd+h3ww03KDs722Hp27fvZatr586dCgsL0/79+/X2229r7969WrNmjTp37qxnnnnmss0LwL0ISQCqxS+//KLhw4erefPmaty4saKjo3XgwAGHPosWLVJQUJAaN26sBx54QLNmzdJVV11lfz8xMVHPPfecbrnlFnXo0EEvv/yyOnTooH/84x8O43h4eCggIMBh8fLykiR98803uvPOO9WoUSO1bNlSTz75pE6dOlVm3adPn9bw4cPVpEkTtWnTRq+//rrD+8YYPf744+rQoYM2b96sQYMGqX379rrppps0ZcoUffLJJw79Dx48qDvuuEONGzdW9+7dlZqaan/v+PHjGjZsmK699lo1btxY3bp10/Llyx3Wv/322zVu3Dg999xzatGihQICAjR16lSHPt9++61uu+02+fj4KDQ0VOvXr5fNZtPHH39s75OVlaWYmBg1b95cLVu21ODBg3X48OEyfw8AnBGSAFSLxx9/XDt37tSnn36q1NRUGWN0zz336Ny5c5KkrVu3Ki4uTuPHj1dGRoYGDBigl156qdwxS0tLdfLkSbVo0aJCNZw5c0Z33323mjdvrq+++kofffSR1q9fr7Fjx5a5zp/+9Cd98cUXWrNmjdatW6cNGzYoLS3N/n5GRob27NmjZ555Rg0aOH9k/jbkSdLkyZP17LPPKiMjQx07dtSwYcNUXFwsSTp79qzCwsL0z3/+U7t379aTTz6p2NhY7dixw2GMd999V76+vtqxY4dmzpyp6dOnKyUlxf47GTJkiBo3bqwdO3Zo4cKFmjx5stPv4Y477lCTJk20adMmbdmyRU2aNNHdd9+toqKiCv0uAUgyAOCifv36mfHjx5v9+/cbSWbr1q329/Ly8kyjRo3Mhx9+aIwxJiYmxgwaNMhh/UcffdT4+fmVOf7MmTNNixYtzI8//mhvmzJlimnQoIHx9fW1L7fccosxxpiFCxea5s2bm1OnTtn7r1271jRo0MDk5OQYY4wZMWKEGTx4sDHGmJMnTxovLy+zYsUKe//jx4+bRo0amfHjxxtjjFm5cqWRZHbt2lXu7+LQoUNGkvm///s/e9uePXuMJLNv374y17vnnnvMM888Y3/dr18/c9tttzn0ueWWW8yf//xnY4wxn332mfHw8DDZ2dn291NSUowks2bNGmOMMYsXLzadOnUypaWl9j6FhYWmUaNGJjk5udztAPD/83BrQgNQL+zbt08eHh7q1auXva1ly5bq1KmT9u3bJ0n67rvv9MADDzisd+utt+qf//yn5ZjLly/X1KlT9cknn+jqq692eK9Tp0769NNP7a+9vb3tdXTv3l2+vr729/r06aPS0lJ999138vf3dxjnhx9+UFFRkcLDw+1tLVq0UKdOneyvjTGSJJvNdulfhKQbb7zR/nObNm0kSbm5uercubNKSkr0yiuvaOXKlcrKylJhYaEKCwsd6r14jAvj5ObmSjr/ewwKClJAQID9/VtvvdWhf1pamr7//ns1bdrUof3s2bP64YcfKrQdACRCEoAquxAkrNovhIvf/nyp9VauXKmRI0fqo48+Uv/+/Z3e9/Ly0vXXX1/ufBezai9r/t/q2LGjpPMB7Kabbrpkf09PT6c5S0tLJUmvv/66Zs+ercTERHXr1k2+vr6Kj493OgX22zEujHNhjPK28YLS0lKFhYXp/fffd3qvdevWl9wGAOdxTRKAKgsNDVVxcbHDtTXHjx/X/v371aVLF0lS586d9eWXXzqsZ/X1+eXLl+vxxx/XBx98oEGDBlW6joyMDJ0+fdretnXrVjVo0MAedn7r+uuvl6enp7Zv325v++WXX7R//37765tuukmhoaF6/fXX7UHlt06cOFHh+jZv3qzBgwfrscceU/fu3XXdddc5Xdx+KZ07d1ZmZqZ+/PFHe9tXX33l0Ofmm2/WgQMHdPXVV+v66693WPz8/Co1H3AlIyQBqLIOHTpo8ODBGj16tLZs2aKvv/5ajz32mK655hoNHjxYkvT0008rKSlJs2bN0oEDB/T222/rs88+czgqsnz5cg0fPlyvv/66evfurZycHOXk5Cg/P79CdTz66KPy8fHRiBEjtHv3bn3xxRd6+umnFRsb63SqTZKaNGmikSNH6k9/+pP+9a9/affu3Xr88ccdLtC22WxasmSJ9u/fr759+yopKUkHDx7Uf/7zH7300kv27auI66+/XikpKdq2bZv27dun//mf/1FOTk6F15ekAQMGqH379hoxYoT+85//aOvWrfYLty/8Lh999FG1atVKgwcP1ubNm3Xo0CFt3LhR48eP19GjRys1H3AlIyQBqBZLlixRWFiY7r33XoWHh8sYo6SkJPupoz59+mjBggWaNWuWunfvrs8//1wTJkyQj4+PfYy3335bxcXFeuqpp9SmTRv7Mn78+ArV0LhxYyUnJ+vnn3/WLbfcooceekh33XWX5s6dW+Y6f/vb39S3b1/df//96t+/v2677TaFhYU59Ln11lu1c+dOtW/fXqNHj1aXLl10//33a8+ePUpMTKzw7+iFF17QzTffrIEDB+r2229XQEDAJe/+fbGGDRvq448/1qlTp3TLLbdo1KhRev755yXJ/rts3LixNm3apLZt2+p3v/udunTpoieeeEK//vqrmjVrVqn5gCuZzVTkpDwAXAajR4/Wt99+q82bN7u7lDpt69atuu222/T999+rffv27i4HqDe4cBtAjXnttdc0YMAA+fr66rPPPtO7776refPmubusOmfNmjVq0qSJOnTooO+//17jx49Xnz59CEhANSMkAagxX375pWbOnKmTJ0/quuuu05w5czRq1Ch3l1XnnDx5Us8995yOHDmiVq1aqX///k53CgdQdZxuAwAAsMCF2wAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABYISQAAABb+H//JumIDT32IAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Let's take a look at the distribution of the expression changes for direct vs indirect targets\n",
    "#We can plot a histogram for the direct targets' log2FoldChange\n",
    "sns.histplot(overlap.log2FoldChange,\n",
    "             bins=np.arange(-10, 10, 0.25),\n",
    "             stat='proportion',\n",
    "             label='overlapping',\n",
    "            )\n",
    "\n",
    "#We can also do the same for the indirect targets' log2FoldChange\n",
    "sns.histplot(no_overlap.log2FoldChange,\n",
    "             bins=np.arange(-10, 10, 0.25),\n",
    "             stat='proportion',\n",
    "             label='non-overlapping',\n",
    "            )\n",
    "\n",
    "#We set stat='proportion' to more easily compare the distributions visually"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4ef72a7-6a46-4ce4-a01b-52dd4b429184",
   "metadata": {},
   "source": [
    "<h2>3. Determine if there is a statistically significant difference in their distribution</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f561666-db9b-44b1-93d9-267dcd2946c1",
   "metadata": {},
   "source": [
    "See if you can set up a Mann-Whitney U-test to determine if there is a significant difference between their distributions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "15dc7be8-7939-45c9-9ca3-c1c602f6548f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MannwhitneyuResult(statistic=397436.0, pvalue=4.468576593091868e-12)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Can see maybe there is a difference in the distribution\n",
    "#Let's check statistically if there is a difference\n",
    "stats.mannwhitneyu(overlap.log2FoldChange,\n",
    "                   no_overlap.log2FoldChange,\n",
    "                  )\n",
    "\n",
    "#We can see in this case that there is"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ff3c646-6e43-4577-b403-4fdc145eee7e",
   "metadata": {},
   "source": [
    "What does the distribution tell us about how TAZ loss impacts the genes that are directly regulated vs indirectly regulated?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9cc8cc3-4860-41c7-a038-61f33e4a5002",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Functional analysis of RNA-seq results</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0598441-b138-4cb0-b649-403df2d5be58",
   "metadata": {},
   "source": [
    "Here, we'll proceed with a functional analysis of our RNA-seq results, and then we'll also run the analysis on our set of genes identified as potentially regulated directly by TAZ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba04aca0-dc66-4f94-98c9-de380a513a8e",
   "metadata": {},
   "source": [
    "<h2>Explore databases</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6858cc36-6176-47ad-9783-f23b4013d170",
   "metadata": {},
   "source": [
    "Recall from Dr. Ingolia's lecture that there are a large number of different databases curated by different organizations and people. Each database will group genes a little differently depending on their own classification scheme, and so selecting a database can affect how you interpret your functional analysis. Some may allow you to gain insight into the regulation of different pathways and others may provide information on how biological processes are altered in your experimental manipulation.\n",
    "\n",
    "Let's briefly take some time to take a look at some databases to see how they organize their gene sets."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2ebf267-0671-4841-aa88-09dabd3dc8c3",
   "metadata": {},
   "source": [
    "<h2>Pull in Molecular Signatures Database (MSigDB)</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41bb4013-f426-4c3b-831d-0e9d80d32900",
   "metadata": {},
   "source": [
    "One of the major databases is the <a href=\"https://www.gsea-msigdb.org/gsea/msigdb\" rel=\"noopener noreferrer\"><u>Molecular Signatures Database (MSigDB)</u></a>, which contains annotated gene sets for functional analysis. MSigDB is a joint project between UC San Diego and the Broad Institute (<a href=\"https://www.pnas.org/doi/10.1073/pnas.0506580102\" rel=\"noopener norferrer\"><u>Subramanian, Tamayo et al 2005</u></a>, <a href=\"https://pmc.ncbi.nlm.nih.gov/articles/PMC4707969/\" rel=\"noopener norferrer\"><u>Liberzon et al 2015</u></a>) and funded by the National Cancer Institute's Informatics Technology for Cancer Research.\n",
    "\n",
    "<code>GSEApy</code> allows us to pull in information on the current <a href=\"https://data.broadinstitute.org/gsea-msigdb/msigdb/release/\" rel=\"noopener noreferrer\"><u>MSigDB database from the Broad Institute.</u></a>\n",
    "\n",
    "To do this, we'll make use of the <code>gp.Msigdb()</code> function from the GSEApy package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a64d0737-6c43-4cf6-8683-c1b43a7bcfad",
   "metadata": {},
   "outputs": [],
   "source": [
    "#To better understand these parts, explore the Molecular Signatures DataBase (the first and last links in the above markdown cell)\n",
    "msig = gp.Msigdb()\n",
    "#Here, we're pulling in information of the directories for the MsSigDB database releases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ae171c5c-5e99-4c01-b5e4-198e11370af4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<gseapy.msigdb.Msigdb at 0x79948dd24ad0>"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We can take a look at the object\n",
    "msig\n",
    "#Which isn't the most informative for us"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62b97967-98a6-447d-b207-92fe2a16dd53",
   "metadata": {},
   "source": [
    "We can then take a look to see what directories the server currently has:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3159593d-1840-4256-96ab-8a05148c50cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Last modified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6.0</td>\n",
       "      <td>2018-12-17 16:29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6.1</td>\n",
       "      <td>2018-12-17 16:29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6.2</td>\n",
       "      <td>2018-12-17 16:29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7.0</td>\n",
       "      <td>2020-01-27 04:43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7.1</td>\n",
       "      <td>2020-03-30 16:55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7.2</td>\n",
       "      <td>2020-09-18 23:33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7.3</td>\n",
       "      <td>2021-03-22 15:09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7.4</td>\n",
       "      <td>2021-04-02 14:38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7.5.1</td>\n",
       "      <td>2023-03-09 19:38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7.5</td>\n",
       "      <td>2022-01-13 01:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2022.1.Hs</td>\n",
       "      <td>2023-03-09 19:37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2022.1.Mm</td>\n",
       "      <td>2023-03-09 19:37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2023.1.Hs</td>\n",
       "      <td>2023-08-19 15:08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2023.1.Mm</td>\n",
       "      <td>2023-08-19 15:09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2023.2.Hs</td>\n",
       "      <td>2023-10-21 01:29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2023.2.Mm</td>\n",
       "      <td>2023-10-20 08:32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2024.1.Hs</td>\n",
       "      <td>2024-09-03 14:38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2024.1.Mm</td>\n",
       "      <td>2024-09-03 14:38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2025.1.Hs</td>\n",
       "      <td>2025-06-08 14:03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2025.1.Mm</td>\n",
       "      <td>2025-06-08 14:03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Name     Last modified\n",
       "0         6.0  2018-12-17 16:29\n",
       "1         6.1  2018-12-17 16:29\n",
       "2         6.2  2018-12-17 16:29\n",
       "3         7.0  2020-01-27 04:43\n",
       "4         7.1  2020-03-30 16:55\n",
       "5         7.2  2020-09-18 23:33\n",
       "6         7.3  2021-03-22 15:09\n",
       "7         7.4  2021-04-02 14:38\n",
       "8       7.5.1  2023-03-09 19:38\n",
       "9         7.5  2022-01-13 01:00\n",
       "10  2022.1.Hs  2023-03-09 19:37\n",
       "11  2022.1.Mm  2023-03-09 19:37\n",
       "12  2023.1.Hs  2023-08-19 15:08\n",
       "13  2023.1.Mm  2023-08-19 15:09\n",
       "14  2023.2.Hs  2023-10-21 01:29\n",
       "15  2023.2.Mm  2023-10-20 08:32\n",
       "16  2024.1.Hs  2024-09-03 14:38\n",
       "17  2024.1.Mm  2024-09-03 14:38\n",
       "18  2025.1.Hs  2025-06-08 14:03\n",
       "19  2025.1.Mm  2025-06-08 14:03"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#But we can pull out the information within the object\n",
    "msig.list_dbver()\n",
    "#We can see the directories containing the different MSigDB releases\n",
    "#Can see that MSigDB contains curated gene sets for mice (Mus musculus (Mm)) and humans (Homo sapiens (Hs))\n",
    "#To better understand, go into the releases for MSigDB using the link contained in the link for \"MSigDB databse from the Broad Institute\"\n",
    "#You'll be able to manually navigate through the different files\n",
    "#Information on the annotations like c5 or h are better explained in the link for \"Molecular Signatures Database (MSigDB)\" link\n",
    "#h stands for Hallmark gene sets\n",
    "#c5 is the gene ontology one\n",
    "#Both are commonly used and other ones are also commonly used like KEGG"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50ce1288-fad5-45cf-8614-0732fab3caac",
   "metadata": {},
   "source": [
    "To get the <code>gmt</code> file, we will make use of the function <code>msig.get_gmt()</code>, where <code>msig</code> is the <code>msigdb</code> object that we created earlier with the <code>Msigdb()</code> function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2cb00fa7-41b1-498b-a287-450920787625",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Here we'll pull in all the Hallmark gene sets contained within the 2025 release for humans\n",
    "#Category specifies essentially what is the gmt file we want and dbver specifies the release version\n",
    "hallmark_db = msig.get_gmt(category='h.all', dbver='2025.1.Hs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "6b027699-738a-4df8-bc21-4e54b0621ab6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Can see that we get a dictionary\n",
    "#The dictionary contains key:value pairs\n",
    "#the keys are the different Hallmark gene sets like HALLMARK_ADIPOGENESIS\n",
    "#the values are all the genes annotated to that specific grouping\n",
    "#When we run our functional analysis, we'll be taking a look to see if there is enrichment within our results compared to what would be expected by chance\n",
    "type(hallmark_db)\n",
    "#The functional groupings then give us an idea of what biological process or moelcular function or signaling pathway etc\n",
    "#is being changed following TAZ KO\n",
    "#Which can form the basis for potential further questions to explore (like for your final project)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd8ddbe6-906b-45f8-9f46-83bde52d8055",
   "metadata": {},
   "source": [
    "<h2>Let's break down the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cf5d648-e3ca-4c2d-aabc-b2557e831012",
   "metadata": {},
   "source": [
    "<code>hallmark_db</code>\n",
    "\n",
    "This is the variable that we will save our database. The database is imported into Python as a <code>dict</code> (dictionary) object, so it has <code>key:value</code> pairs, where the key is the category or group that a gene or genes belong to, and the value is the gene name. \n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>&equals;</code>\n",
    "\n",
    "Our assignment operator.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>msig.get_gmt()</code>\n",
    "\n",
    "This function allows us to pull in the database that we want to use for our functional analysis. The files are usually in a <a href=\"https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29\" rel=\"noopener noreferrer\"><u>.gmt (Gene Matrix Transposed) format</u></a>.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>category='h.all'</code>\n",
    "\n",
    "This is the database that we want to pull in, since if we take a look at the directory, it contains a bunch of different .gmt files. Here, we're pulling in the MSigDB Hallmark database.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>dbver='2025.1.Hs'</code>\n",
    "\n",
    "This is the database version that we're pulling in. As you saw when we took a look at the <code>msigdb</code> object, there were different versions of the MSigDB for both humans and mice, and the <code>dbver</code> parameter allows us to select which version we want to pull into our notebook.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5f74e40-ea6d-46c7-ac9d-893f4f0c1329",
   "metadata": {},
   "source": [
    "Since the object is a dictionary, we can pull all the values associated with a specific key. In other words, pull all the genes associated with a specific MSigDB Hallmark category."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "0d3a5138-da28-4e4f-bdc9-e0072c3c14a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ACHE',\n",
       " 'ADGRG1',\n",
       " 'AMOT',\n",
       " 'CDK5R1',\n",
       " 'CDK6',\n",
       " 'CELSR1',\n",
       " 'CNTFR',\n",
       " 'CRMP1',\n",
       " 'DPYSL2',\n",
       " 'ETS2']"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hallmark_db['HALLMARK_HEDGEHOG_SIGNALING'][0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aadc73ca-54a1-4578-b675-eb86b29c486b",
   "metadata": {},
   "source": [
    "What you should see is a list of genes that have been categorized as being involved in adipogenesis by MSigDB. We can also take a look at the whole database by converting the dictionary to a DataFrame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2d7650cc-0e9e-4ca0-b66f-cad7a6c4a1cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HALLMARK_ADIPOGENESIS</th>\n",
       "      <th>HALLMARK_ALLOGRAFT_REJECTION</th>\n",
       "      <th>HALLMARK_ANDROGEN_RESPONSE</th>\n",
       "      <th>HALLMARK_ANGIOGENESIS</th>\n",
       "      <th>HALLMARK_APICAL_JUNCTION</th>\n",
       "      <th>HALLMARK_APICAL_SURFACE</th>\n",
       "      <th>HALLMARK_APOPTOSIS</th>\n",
       "      <th>HALLMARK_BILE_ACID_METABOLISM</th>\n",
       "      <th>HALLMARK_CHOLESTEROL_HOMEOSTASIS</th>\n",
       "      <th>HALLMARK_COAGULATION</th>\n",
       "      <th>...</th>\n",
       "      <th>HALLMARK_PROTEIN_SECRETION</th>\n",
       "      <th>HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY</th>\n",
       "      <th>HALLMARK_SPERMATOGENESIS</th>\n",
       "      <th>HALLMARK_TGF_BETA_SIGNALING</th>\n",
       "      <th>HALLMARK_TNFA_SIGNALING_VIA_NFKB</th>\n",
       "      <th>HALLMARK_UNFOLDED_PROTEIN_RESPONSE</th>\n",
       "      <th>HALLMARK_UV_RESPONSE_DN</th>\n",
       "      <th>HALLMARK_UV_RESPONSE_UP</th>\n",
       "      <th>HALLMARK_WNT_BETA_CATENIN_SIGNALING</th>\n",
       "      <th>HALLMARK_XENOBIOTIC_METABOLISM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCA1</td>\n",
       "      <td>AARS1</td>\n",
       "      <td>ABCC4</td>\n",
       "      <td>APOH</td>\n",
       "      <td>ACTA1</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ADD1</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABCA2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>...</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABCC1</td>\n",
       "      <td>ACE</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ALDH18A1</td>\n",
       "      <td>ABCC1</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ADAM17</td>\n",
       "      <td>ABCC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCB8</td>\n",
       "      <td>ABCE1</td>\n",
       "      <td>ABHD2</td>\n",
       "      <td>APP</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>ADIPOR2</td>\n",
       "      <td>AIFM3</td>\n",
       "      <td>ABCA2</td>\n",
       "      <td>ACAT2</td>\n",
       "      <td>ACOX2</td>\n",
       "      <td>...</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ATOX1</td>\n",
       "      <td>ACRBP</td>\n",
       "      <td>APC</td>\n",
       "      <td>ACKR3</td>\n",
       "      <td>ARFGAP1</td>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>AXIN1</td>\n",
       "      <td>ABCC3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAA2</td>\n",
       "      <td>ABI1</td>\n",
       "      <td>ACSL3</td>\n",
       "      <td>CCND2</td>\n",
       "      <td>ACTC1</td>\n",
       "      <td>AFAP1L2</td>\n",
       "      <td>ANKH</td>\n",
       "      <td>ABCA3</td>\n",
       "      <td>ACSS2</td>\n",
       "      <td>ADAM9</td>\n",
       "      <td>...</td>\n",
       "      <td>ANP32E</td>\n",
       "      <td>CAT</td>\n",
       "      <td>ACRV1</td>\n",
       "      <td>ARID4B</td>\n",
       "      <td>AREG</td>\n",
       "      <td>ASNS</td>\n",
       "      <td>ADD3</td>\n",
       "      <td>AGO2</td>\n",
       "      <td>AXIN2</td>\n",
       "      <td>ABCD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACADL</td>\n",
       "      <td>ACHE</td>\n",
       "      <td>ACTN1</td>\n",
       "      <td>COL3A1</td>\n",
       "      <td>ACTG1</td>\n",
       "      <td>AKAP7</td>\n",
       "      <td>ANXA1</td>\n",
       "      <td>ABCA4</td>\n",
       "      <td>ACTG1</td>\n",
       "      <td>ANG</td>\n",
       "      <td>...</td>\n",
       "      <td>AP1G1</td>\n",
       "      <td>CDKN2D</td>\n",
       "      <td>ACTL7B</td>\n",
       "      <td>BCAR3</td>\n",
       "      <td>ATF3</td>\n",
       "      <td>ATF3</td>\n",
       "      <td>ADGRL2</td>\n",
       "      <td>ALAS1</td>\n",
       "      <td>CCND2</td>\n",
       "      <td>ABHD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACADM</td>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>ADAMTS1</td>\n",
       "      <td>COL5A2</td>\n",
       "      <td>ACTG2</td>\n",
       "      <td>APP</td>\n",
       "      <td>APP</td>\n",
       "      <td>ABCA5</td>\n",
       "      <td>ADH4</td>\n",
       "      <td>ANXA1</td>\n",
       "      <td>...</td>\n",
       "      <td>AP2B1</td>\n",
       "      <td>EGLN2</td>\n",
       "      <td>ADAD1</td>\n",
       "      <td>BMP2</td>\n",
       "      <td>ATP2B1</td>\n",
       "      <td>ATF4</td>\n",
       "      <td>ADORA2B</td>\n",
       "      <td>ALDOA</td>\n",
       "      <td>CSNK1E</td>\n",
       "      <td>ACO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>UQCR11</td>\n",
       "      <td>UBE2D1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>VWF</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>UPB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>UQCRC1</td>\n",
       "      <td>UBE2N</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>WASL</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>YRDC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>UPP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>UQCRQ</td>\n",
       "      <td>WARS1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>WNK4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ZBTB10</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>VNN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>VEGFB</td>\n",
       "      <td>WAS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>YWHAH</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ZC3H12A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>VTN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>YWHAG</td>\n",
       "      <td>ZAP70</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ZFP36</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XDH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    HALLMARK_ADIPOGENESIS HALLMARK_ALLOGRAFT_REJECTION  \\\n",
       "0                   ABCA1                        AARS1   \n",
       "1                   ABCB8                        ABCE1   \n",
       "2                   ACAA2                         ABI1   \n",
       "3                   ACADL                         ACHE   \n",
       "4                   ACADM                       ACVR2A   \n",
       "..                    ...                          ...   \n",
       "195                UQCR11                       UBE2D1   \n",
       "196                UQCRC1                        UBE2N   \n",
       "197                 UQCRQ                        WARS1   \n",
       "198                 VEGFB                          WAS   \n",
       "199                 YWHAG                        ZAP70   \n",
       "\n",
       "    HALLMARK_ANDROGEN_RESPONSE HALLMARK_ANGIOGENESIS HALLMARK_APICAL_JUNCTION  \\\n",
       "0                        ABCC4                  APOH                    ACTA1   \n",
       "1                        ABHD2                   APP                     ACTB   \n",
       "2                        ACSL3                 CCND2                    ACTC1   \n",
       "3                        ACTN1                COL3A1                    ACTG1   \n",
       "4                      ADAMTS1                COL5A2                    ACTG2   \n",
       "..                         ...                   ...                      ...   \n",
       "195                       None                  None                      VWF   \n",
       "196                       None                  None                     WASL   \n",
       "197                       None                  None                     WNK4   \n",
       "198                       None                  None                    YWHAH   \n",
       "199                       None                  None                      ZYX   \n",
       "\n",
       "    HALLMARK_APICAL_SURFACE HALLMARK_APOPTOSIS HALLMARK_BILE_ACID_METABOLISM  \\\n",
       "0                    ADAM10               ADD1                         ABCA1   \n",
       "1                   ADIPOR2              AIFM3                         ABCA2   \n",
       "2                   AFAP1L2               ANKH                         ABCA3   \n",
       "3                     AKAP7              ANXA1                         ABCA4   \n",
       "4                       APP                APP                         ABCA5   \n",
       "..                      ...                ...                           ...   \n",
       "195                    None               None                          None   \n",
       "196                    None               None                          None   \n",
       "197                    None               None                          None   \n",
       "198                    None               None                          None   \n",
       "199                    None               None                          None   \n",
       "\n",
       "    HALLMARK_CHOLESTEROL_HOMEOSTASIS HALLMARK_COAGULATION  ...  \\\n",
       "0                              ABCA2                  A2M  ...   \n",
       "1                              ACAT2                ACOX2  ...   \n",
       "2                              ACSS2                ADAM9  ...   \n",
       "3                              ACTG1                  ANG  ...   \n",
       "4                               ADH4                ANXA1  ...   \n",
       "..                               ...                  ...  ...   \n",
       "195                             None                 None  ...   \n",
       "196                             None                 None  ...   \n",
       "197                             None                 None  ...   \n",
       "198                             None                 None  ...   \n",
       "199                             None                 None  ...   \n",
       "\n",
       "    HALLMARK_PROTEIN_SECRETION HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY  \\\n",
       "0                        ABCA1                                    ABCC1   \n",
       "1                       ADAM10                                    ATOX1   \n",
       "2                       ANP32E                                      CAT   \n",
       "3                        AP1G1                                   CDKN2D   \n",
       "4                        AP2B1                                    EGLN2   \n",
       "..                         ...                                      ...   \n",
       "195                       None                                     None   \n",
       "196                       None                                     None   \n",
       "197                       None                                     None   \n",
       "198                       None                                     None   \n",
       "199                       None                                     None   \n",
       "\n",
       "    HALLMARK_SPERMATOGENESIS HALLMARK_TGF_BETA_SIGNALING  \\\n",
       "0                        ACE                       ACVR1   \n",
       "1                      ACRBP                         APC   \n",
       "2                      ACRV1                      ARID4B   \n",
       "3                     ACTL7B                       BCAR3   \n",
       "4                      ADAD1                        BMP2   \n",
       "..                       ...                         ...   \n",
       "195                     None                        None   \n",
       "196                     None                        None   \n",
       "197                     None                        None   \n",
       "198                     None                        None   \n",
       "199                     None                        None   \n",
       "\n",
       "    HALLMARK_TNFA_SIGNALING_VIA_NFKB HALLMARK_UNFOLDED_PROTEIN_RESPONSE  \\\n",
       "0                              ABCA1                           ALDH18A1   \n",
       "1                              ACKR3                            ARFGAP1   \n",
       "2                               AREG                               ASNS   \n",
       "3                               ATF3                               ATF3   \n",
       "4                             ATP2B1                               ATF4   \n",
       "..                               ...                                ...   \n",
       "195                            VEGFA                               None   \n",
       "196                             YRDC                               None   \n",
       "197                           ZBTB10                               None   \n",
       "198                          ZC3H12A                               None   \n",
       "199                            ZFP36                               None   \n",
       "\n",
       "    HALLMARK_UV_RESPONSE_DN HALLMARK_UV_RESPONSE_UP  \\\n",
       "0                     ABCC1                   ABCB1   \n",
       "1                    ACVR2A                   ACAA1   \n",
       "2                      ADD3                    AGO2   \n",
       "3                    ADGRL2                   ALAS1   \n",
       "4                   ADORA2B                   ALDOA   \n",
       "..                      ...                     ...   \n",
       "195                    None                    None   \n",
       "196                    None                    None   \n",
       "197                    None                    None   \n",
       "198                    None                    None   \n",
       "199                    None                    None   \n",
       "\n",
       "    HALLMARK_WNT_BETA_CATENIN_SIGNALING HALLMARK_XENOBIOTIC_METABOLISM  \n",
       "0                                ADAM17                          ABCC2  \n",
       "1                                 AXIN1                          ABCC3  \n",
       "2                                 AXIN2                          ABCD2  \n",
       "3                                 CCND2                          ABHD6  \n",
       "4                                CSNK1E                           ACO2  \n",
       "..                                  ...                            ...  \n",
       "195                                None                           UPB1  \n",
       "196                                None                           UPP1  \n",
       "197                                None                           VNN1  \n",
       "198                                None                            VTN  \n",
       "199                                None                            XDH  \n",
       "\n",
       "[200 rows x 50 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#This is just to let us more easily see all the Hallmark gene sets at once\n",
    "#Keep in mind that gene sets can have different numbers of genes annotated to them\n",
    "#So it's not required that every gene set has the same number of genes\n",
    "#You can also see that a single gene can be annotated to more than one gene set as well\n",
    "#Which isn't too suprising since we know that a gene can be involved in multiple things\n",
    "pd.DataFrame.from_dict(hallmark_db, orient='index').T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a284c216-5f6f-4a2e-9914-751d9b840a23",
   "metadata": {},
   "source": [
    "For this version of the MSigDB Hallmark human database, you can see that there are 50 different categories ranging from adipogenesis to xenobiotic metabolism with varying numbers of genes inside of them. <a href=\"https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=H\" rel=\"noopener noreferrer\"><u>We can also find more information on these gene sets by diving into the MSigDB page on their Hallmark database.</u></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b66b992-401a-4919-bb99-6892f1e6651b",
   "metadata": {},
   "source": [
    "<h2>Use GSEApy to pull Enrichr library of databases</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfe5b54e-7e4c-411d-bbcf-0b02c296ac81",
   "metadata": {},
   "source": [
    "GSEApy also functions as an Enrichr wrapper, meaning that it allows you to use the Enrichr web program through Python. <a href=\"https://maayanlab.cloud/Enrichr/\" rel=\"noopener noreferrer\"><u>Enrichr is developed and maintained by the Ma'ayan lab at the Icahn School of Medicine, Mount Sinai, Center for Bioinformatics</u></a> (<a href=\"https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-128\" rel=\"noopener noreferrer\"><u>Chen et al 2013</u></a>), and it's one of many web-based tools/programs that allow you to perform functional analyses, such as the ones we're doing in this lesson.\n",
    "\n",
    "To take a look at what human-specific databases we can work with for functional analysis, we can make use of the <code>gp.get_library_name()</code> function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4c5320fd-e608-42dd-af9a-42859ef6b7e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This example is another way to pull in other databases too\n",
    "#Here we'll pull from Enrichr - see the Enrichr link in the Markdown cell above\n",
    "databases = gp.get_library_name(organism='Human')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdd8a8c3-631e-4ba8-8877-01d9fa0766da",
   "metadata": {},
   "source": [
    "Let's take a look at what we pulled into our notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d763481e-d6be-41d8-8272-7d6d8ba82dba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ARCHS4_Cell-lines',\n",
       " 'ARCHS4_IDG_Coexp',\n",
       " 'ARCHS4_Kinases_Coexp',\n",
       " 'ARCHS4_TFs_Coexp',\n",
       " 'ARCHS4_Tissues',\n",
       " 'Achilles_fitness_decrease',\n",
       " 'Achilles_fitness_increase',\n",
       " 'Aging_Perturbations_from_GEO_down',\n",
       " 'Aging_Perturbations_from_GEO_up',\n",
       " 'Allen_Brain_Atlas_10x_scRNA_2021']"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Just outputing the first 10 databases that we pulled in from Enrichr\n",
    "databases[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31532e7b-13c4-4183-b79a-160ab77bc939",
   "metadata": {},
   "source": [
    "You should see that we now have a long list of different human-specific databases that have categorized genes in a variety of different ways.\n",
    "\n",
    "Some commonly used ones that Dr. Ingolia mentioned in lecture are those databases from the <a href=\"https://geneontology.org/\" rel=\"noopener noreferrer\"><u>Gene Ontology (GO) resource</u></a>, the MSigDB, and the <a href=\"https://www.genome.jp/kegg/\" rel=\"noopener noreferrer\"><u>Kyoto Encyclopedia of Genes and Genomes (KEGG)</u></a>. Recall that GO is at the highest level of grouping, divided into three major categories: biological process, cell component, and molecular function, and you should be able to see in this list of databases, these three categories as their own database for functional analysis.\n",
    "\n",
    "Note that these are pulled from the Enrichr library, and that they might not always be the most up-to-date. Note the date for their repository of the MSigDB Hallmark database, <code>MSigDB_Hallmark_2020</code>, while we can directly download or pull the most up-to-date version directly from MSigDB."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e8d4000-464b-4f8e-a7f3-95e01342a325",
   "metadata": {},
   "source": [
    "<h2>Using databases</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f7824ed-1357-4db8-a396-ea98570ae7b9",
   "metadata": {},
   "source": [
    "These databases often can be used through web or desktop programs that have been developed, and we'll explore some of them towards the end of this lesson if we have time.\n",
    "\n",
    "To use these databases via <code>GSEApy</code>, we can just pass the name or the variable for the gene set that we want to use for our functional analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dc64840-e0b2-4b43-96e1-b0fab51c52a2",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Load in and prepare data</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c8dec67-558c-479d-89a4-c38e247338c2",
   "metadata": {},
   "source": [
    "Let's go ahead and load in our RNA-seq data that we'll use for this part of our notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "41b1ebe9-39e9-45e0-9aea-530f55ba1259",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Let's now go ahead and load in our data so we can play around with some functional analyses\n",
    "#We'll be focusing on over-representation analysis and gene set enrichment analysis (GSEA)\n",
    "#These two are fairly common approaches to look for enrichment of functional groups or processes etc\n",
    "#We'll use normalized counts for GSEA and shrunken logfoldchange for over-representation analysis\n",
    "norm_counts = pd.read_csv('../Week_10/differential-expression-analysis/1M_class_norm_counts.csv')\n",
    "shrinklfc = pd.read_csv('../Week_10/differential-expression-analysis/1M_results_lfcShrink.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02b3c2bc-6429-41b0-8868-aec672ed137d",
   "metadata": {},
   "source": [
    "Let's take a look at our imported data as a sanity check."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "250edb59-6d96-44e8-a834-f04b4dbe5c5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_9260a\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_9260a_level0_col0\" class=\"col_heading level0 col0\" >Unnamed: 0</th>\n",
       "      <th id=\"T_9260a_level0_col1\" class=\"col_heading level0 col1\" >baseMean</th>\n",
       "      <th id=\"T_9260a_level0_col2\" class=\"col_heading level0 col2\" >log2FoldChange</th>\n",
       "      <th id=\"T_9260a_level0_col3\" class=\"col_heading level0 col3\" >lfcSE</th>\n",
       "      <th id=\"T_9260a_level0_col4\" class=\"col_heading level0 col4\" >pvalue</th>\n",
       "      <th id=\"T_9260a_level0_col5\" class=\"col_heading level0 col5\" >padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_9260a_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_9260a_row0_col0\" class=\"data row0 col0\" >A1BG-AS1</td>\n",
       "      <td id=\"T_9260a_row0_col1\" class=\"data row0 col1\" >1.717396</td>\n",
       "      <td id=\"T_9260a_row0_col2\" class=\"data row0 col2\" >0.530432</td>\n",
       "      <td id=\"T_9260a_row0_col3\" class=\"data row0 col3\" >0.753164</td>\n",
       "      <td id=\"T_9260a_row0_col4\" class=\"data row0 col4\" >0.034547</td>\n",
       "      <td id=\"T_9260a_row0_col5\" class=\"data row0 col5\" >0.105928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_9260a_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_9260a_row1_col0\" class=\"data row1 col0\" >A4GALT</td>\n",
       "      <td id=\"T_9260a_row1_col1\" class=\"data row1 col1\" >10.639844</td>\n",
       "      <td id=\"T_9260a_row1_col2\" class=\"data row1 col2\" >-0.039800</td>\n",
       "      <td id=\"T_9260a_row1_col3\" class=\"data row1 col3\" >0.218284</td>\n",
       "      <td id=\"T_9260a_row1_col4\" class=\"data row1 col4\" >0.816143</td>\n",
       "      <td id=\"T_9260a_row1_col5\" class=\"data row1 col5\" >0.904300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_9260a_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_9260a_row2_col0\" class=\"data row2 col0\" >AAAS</td>\n",
       "      <td id=\"T_9260a_row2_col1\" class=\"data row2 col1\" >28.134083</td>\n",
       "      <td id=\"T_9260a_row2_col2\" class=\"data row2 col2\" >-0.023496</td>\n",
       "      <td id=\"T_9260a_row2_col3\" class=\"data row2 col3\" >0.155790</td>\n",
       "      <td id=\"T_9260a_row2_col4\" class=\"data row2 col4\" >0.866118</td>\n",
       "      <td id=\"T_9260a_row2_col5\" class=\"data row2 col5\" >0.930518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_9260a_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_9260a_row3_col0\" class=\"data row3 col0\" >AACS</td>\n",
       "      <td id=\"T_9260a_row3_col1\" class=\"data row3 col1\" >28.101282</td>\n",
       "      <td id=\"T_9260a_row3_col2\" class=\"data row3 col2\" >-0.117385</td>\n",
       "      <td id=\"T_9260a_row3_col3\" class=\"data row3 col3\" >0.146797</td>\n",
       "      <td id=\"T_9260a_row3_col4\" class=\"data row3 col4\" >0.378393</td>\n",
       "      <td id=\"T_9260a_row3_col5\" class=\"data row3 col5\" >0.577812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_9260a_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_9260a_row4_col0\" class=\"data row4 col0\" >AADAT</td>\n",
       "      <td id=\"T_9260a_row4_col1\" class=\"data row4 col1\" >3.955065</td>\n",
       "      <td id=\"T_9260a_row4_col2\" class=\"data row4 col2\" >0.042448</td>\n",
       "      <td id=\"T_9260a_row4_col3\" class=\"data row4 col3\" >0.272361</td>\n",
       "      <td id=\"T_9260a_row4_col4\" class=\"data row4 col4\" >0.808940</td>\n",
       "      <td id=\"T_9260a_row4_col5\" class=\"data row4 col5\" >0.901120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x79948bad56d0>"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shrinklfc.head().style"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3442f23e-d75f-4098-aa45-fbaaab82a573",
   "metadata": {},
   "source": [
    "Now let's quickly update the column header for the first row to make things a little easier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c3006add-f2ee-42bc-8887-b8dba80651c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_counts.rename(columns={norm_counts.columns[0] : 'gene'},\n",
    "                   inplace=True\n",
    "                  )\n",
    "\n",
    "shrinklfc.rename(columns={shrinklfc.columns[0] : 'gene'},\n",
    "                 inplace=True\n",
    "                )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "482435b7-7c0c-4fe9-9634-667efcf488fb",
   "metadata": {},
   "source": [
    "Then let's pull just the gene names for the significant hits from our differential expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "f5a10912-241f-44b1-90e3-a92dc222d510",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We are the ones that decide what is \"interesting\" for our over-representation analysis\n",
    "#This is where its a bit subjective and what you set for a threshold of \"interesting\" will dictate how you interpret your results\n",
    "#For example, here we're pulling all significant hits that are 2-fold different from control (so both upreg and downreg)\n",
    "sig_genes = shrinklfc.loc[(shrinklfc.padj < 0.05) & (shrinklfc.log2FoldChange.abs() >= 1), :]\n",
    "\n",
    "#However, you can also consider only the signficantly upregulated genes as \"interesting\"\n",
    "#Or just the significantly downregulated genes as \"interesting\"\n",
    "#Or even just only those that are significantly and super highly upregulated based on some threshold like 10X greater or w/e\n",
    "\n",
    "#In any case, you'll need to keep in mind how it then affects your interpretation of the enriched gene sets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5158cce-3705-4fce-8484-5382b11a6010",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Over-representation analysis</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "239f8caa-ccf2-4007-bd03-6b58b8dffe92",
   "metadata": {},
   "source": [
    "This type of functional analysis asks if there is an over-representation of sets of genes among the differentially expressed genes you identified through your RNA-seq analysis. The statistical test used by this type of analysis is a hypergeometric test which tells us if the level of enrichment we see is true or just due to random chance. \n",
    "\n",
    "Gene sets that are enriched would then suggest that the biological process, molecular function, or other functional group that they belong to is being broadly affected by the experimental manipulation that was done."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "936a9961-2683-4209-8330-2ab2b94c4142",
   "metadata": {},
   "source": [
    "<h2>Deciding on background for analysis</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33aeabec-b617-4e87-9d61-79b1f017172a",
   "metadata": {},
   "source": [
    "Recall from Dr. Ingolia's lecture that an important thing to consider for your over-representation analysis is \"What is the universe of genes that you are looking in?\". In other words, what is the background set of genes that will form your reference point? While the human genome has ~25,000 genes, a typical cell will usually express maybe ~10,000 to ~12,000 genes. So you'll want to take into account the fact that your cells probably aren't going to be expressing all genes in the genome by limiting your \"universe of genes\" to the data that you have rather than all the genes in the genome. \n",
    "\n",
    "What that means for our experiment is that we'll generate a list of all expressed genes based on our RNA-seq data and that will be the background for our over-representation analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "6a3d1df2-119c-4f66-be35-5c12ffc39404",
   "metadata": {},
   "outputs": [],
   "source": [
    "#An important consideration is what the \"background\" enrichment level is since that is what will be used to determine if we do have enrichment\n",
    "#So here, we'll use only the expressed genes that were also factored into our DESeq2 analysis\n",
    "#Recall weird things that were omitted automatically by DESeq2 due to the presence of weird outliers\n",
    "#Those were indicated by a padj of NA\n",
    "#We can use pd.isna() to identify them and then the bitwise inversion ~ to pull just those that DO NOT have padj of NA\n",
    "bg_genes = shrinklfc.loc[~pd.isna(shrinklfc.padj), 'gene'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2f6355d3-bbda-4927-a0d3-50d5086afcff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A1BG-AS1',\n",
       " 'A4GALT',\n",
       " 'AAAS',\n",
       " 'AACS',\n",
       " 'AADAT',\n",
       " 'AAGAB',\n",
       " 'AAK1',\n",
       " 'AAMDC',\n",
       " 'AAMP',\n",
       " 'AAR2']"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bg_genes[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fa88027-18a2-476b-aa43-df2b0d145e4a",
   "metadata": {},
   "source": [
    "<h2>Run an over-representation analysis</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b3e236-5749-425c-ade6-979ed111bd2b",
   "metadata": {},
   "source": [
    "<code>GSEApy</code> contains a couple of ways to run an over-representation analysis, either through Enrichr web program or within Python. We'll perform the hypergeometric test in Python using the function <code>gp.enrich()</code>. <a href=\"https://gseapy.readthedocs.io/en/latest/run.html#gseapy.enrich\" rel=\"noopener noreferrer\"><u>Documentation for <code>gp.enrich()</code> can be found here.</u></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "a3bcf627-7c6b-460d-a1d4-2a55069326a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now we can run our over-representation analysis\n",
    "#To better understand, dig into the documentation for this function\n",
    "#We feed it lists for our \"interesting\" genes and the background\n",
    "#And we provide it with the dict of our Hallmark gene set\n",
    "ora = gp.enrich(gene_list=sig_genes.gene.to_list(),\n",
    "                gene_sets=hallmark_db,\n",
    "                background=bg_genes,\n",
    "                outdir=None,\n",
    "               )\n",
    "\n",
    "#You can also refer to detailed code breakdown below."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e5cbf4e-78c4-43e4-9161-37ea1400cb58",
   "metadata": {},
   "source": [
    "<h2>Let's break down the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d2299bc-647b-43f6-8969-1d969abb938c",
   "metadata": {},
   "source": [
    "<code>ora</code>\n",
    "\n",
    "Our variable that will hold the results of the over-representation analysis.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>&equals;</code>\n",
    "\n",
    "Assignment operator.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>gp.enrich()</code>\n",
    "\n",
    "We're calling up this function to perform an over-representation analysis.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>gene_list=sig_genes.gene.to_list()</code>\n",
    "\n",
    "Here, we pass to the function our list of significant genes as a list that we pulled out of our <code>sig_genes</code> DataFrame.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>gene_sets=hallmark_db</code>\n",
    "\n",
    "This is the database that we want it to use when performing the functional analysis.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>background=bg_genes</code>\n",
    "\n",
    "Here, we provide the list of background genes. In our case, we pulled all the expressed genes from our RNA-seq analysis. Recall that we filtered out genes that weren't expressed (mean == 0) prior to performing our differential expression analysis. So all the output genes showed some level of expression in our cells.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>outdir=None</code>\n",
    "\n",
    "We specify that we don't want the function to output the results of the analysis to a file yet.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e44d705c-3457-4dad-a180-8f4ab8325094",
   "metadata": {},
   "source": [
    "Now let's take a look at our output object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "bd39ad70-1dd8-42a5-9fe7-efa2d11f20da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<gseapy.enrichr.Enrichr at 0x79948badf690>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ora\n",
    "#The object itself is not super informative, so we'll have to pull out our results as attributes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44f39cdb-b7dd-42a5-a500-799763a136e6",
   "metadata": {},
   "source": [
    "You can see that the results are stored in an Enrichr object, and if we dig into <a href=\"https://gseapy.readthedocs.io/en/latest/run.html#gseapy.enrich\" rel=\"noopener noreferrer\"><u>the documentation</u></a>, we find that there are two ways to pull out the results. We can either pull it out using the <code>res2d</code> attribute or <code>results</code> attribute. The documentation notes that <code>res2d</code> will pull the most recent query while <code>results</code> will pull all queries.\n",
    "\n",
    "Let's pull our most recent query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "48bd5e8b-3cc8-4c31-9fc0-b28a6e2ed83f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_e3828\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_e3828_level0_col0\" class=\"col_heading level0 col0\" >Gene_set</th>\n",
       "      <th id=\"T_e3828_level0_col1\" class=\"col_heading level0 col1\" >Term</th>\n",
       "      <th id=\"T_e3828_level0_col2\" class=\"col_heading level0 col2\" >Overlap</th>\n",
       "      <th id=\"T_e3828_level0_col3\" class=\"col_heading level0 col3\" >P-value</th>\n",
       "      <th id=\"T_e3828_level0_col4\" class=\"col_heading level0 col4\" >Adjusted P-value</th>\n",
       "      <th id=\"T_e3828_level0_col5\" class=\"col_heading level0 col5\" >Odds Ratio</th>\n",
       "      <th id=\"T_e3828_level0_col6\" class=\"col_heading level0 col6\" >Combined Score</th>\n",
       "      <th id=\"T_e3828_level0_col7\" class=\"col_heading level0 col7\" >Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row0\" class=\"row_heading level0 row0\" >13</th>\n",
       "      <td id=\"T_e3828_row0_col0\" class=\"data row0 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row0_col1\" class=\"data row0 col1\" >HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION</td>\n",
       "      <td id=\"T_e3828_row0_col2\" class=\"data row0 col2\" >32/157</td>\n",
       "      <td id=\"T_e3828_row0_col3\" class=\"data row0 col3\" >0.000000</td>\n",
       "      <td id=\"T_e3828_row0_col4\" class=\"data row0 col4\" >0.000000</td>\n",
       "      <td id=\"T_e3828_row0_col5\" class=\"data row0 col5\" >4.132167</td>\n",
       "      <td id=\"T_e3828_row0_col6\" class=\"data row0 col6\" >85.615963</td>\n",
       "      <td id=\"T_e3828_row0_col7\" class=\"data row0 col7\" >ABI3BP;COL12A1;OXTR;CXCL1;PDGFRB;FZD8;ITGB3;SERPINE2;MEST;MATN2;SAT1;ACTA2;TNC;FLNA;TIMP3;THBS2;MYLK;COL6A2;NNMT;COL1A2;GJA1;INHBA;PMEPA1;DKK1;THBS1;VCAM1;PLAUR;FSTL1;ITGA2;WNT5A;ADAM12;CDH11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row1\" class=\"row_heading level0 row1\" >14</th>\n",
       "      <td id=\"T_e3828_row1_col0\" class=\"data row1 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row1_col1\" class=\"data row1 col1\" >HALLMARK_ESTROGEN_RESPONSE_EARLY</td>\n",
       "      <td id=\"T_e3828_row1_col2\" class=\"data row1 col2\" >30/162</td>\n",
       "      <td id=\"T_e3828_row1_col3\" class=\"data row1 col3\" >0.000000</td>\n",
       "      <td id=\"T_e3828_row1_col4\" class=\"data row1 col4\" >0.000001</td>\n",
       "      <td id=\"T_e3828_row1_col5\" class=\"data row1 col5\" >3.661751</td>\n",
       "      <td id=\"T_e3828_row1_col6\" class=\"data row1 col6\" >62.946727</td>\n",
       "      <td id=\"T_e3828_row1_col7\" class=\"data row1 col7\" >ELF3;ESRP2;UGCG;FHL2;SCNN1A;SYT12;CELSR1;CALCR;RAB17;TIAM1;RETREG1;SLC1A1;DHRS3;CLDN7;RPS6KA2;TPD52L1;CALB2;KRT18;PTGES;PAPSS2;CLIC3;KRT8;MUC1;KAZN;INHBB;GJA1;KRT19;SLC2A1;DHCR7;MAPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row2\" class=\"row_heading level0 row2\" >28</th>\n",
       "      <td id=\"T_e3828_row2_col0\" class=\"data row2 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row2_col1\" class=\"data row2 col1\" >HALLMARK_KRAS_SIGNALING_UP</td>\n",
       "      <td id=\"T_e3828_row2_col2\" class=\"data row2 col2\" >23/132</td>\n",
       "      <td id=\"T_e3828_row2_col3\" class=\"data row2 col3\" >0.000004</td>\n",
       "      <td id=\"T_e3828_row2_col4\" class=\"data row2 col4\" >0.000064</td>\n",
       "      <td id=\"T_e3828_row2_col5\" class=\"data row2 col5\" >3.390333</td>\n",
       "      <td id=\"T_e3828_row2_col6\" class=\"data row2 col6\" >42.145458</td>\n",
       "      <td id=\"T_e3828_row2_col7\" class=\"data row2 col7\" >KCNN4;ST6GAL1;ADAM8;SERPINA3;NGF;EMP1;SATB1;CFH;LCP1;TSPAN13;CAB39L;ANGPTL4;PLEK2;TNFRSF1B;CMKLR1;ITGB2;INHBA;PLAUR;GALNT3;GPRC5B;ITGA2;GFPT2;ETV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row3\" class=\"row_heading level0 row3\" >5</th>\n",
       "      <td id=\"T_e3828_row3_col0\" class=\"data row3 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row3_col1\" class=\"data row3 col1\" >HALLMARK_APICAL_SURFACE</td>\n",
       "      <td id=\"T_e3828_row3_col2\" class=\"data row3 col2\" >9/27</td>\n",
       "      <td id=\"T_e3828_row3_col3\" class=\"data row3 col3\" >0.000018</td>\n",
       "      <td id=\"T_e3828_row3_col4\" class=\"data row3 col4\" >0.000186</td>\n",
       "      <td id=\"T_e3828_row3_col5\" class=\"data row3 col5\" >8.030370</td>\n",
       "      <td id=\"T_e3828_row3_col6\" class=\"data row3 col6\" >87.567233</td>\n",
       "      <td id=\"T_e3828_row3_col7\" class=\"data row3 col7\" >AKAP7;PLAUR;MDGA1;CX3CL1;GATA3;SRPX;ATP8B1;AFAP1L2;PCSK9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row4\" class=\"row_heading level0 row4\" >24</th>\n",
       "      <td id=\"T_e3828_row4_col0\" class=\"data row4 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row4_col1\" class=\"data row4 col1\" >HALLMARK_INFLAMMATORY_RESPONSE</td>\n",
       "      <td id=\"T_e3828_row4_col2\" class=\"data row4 col2\" >22/135</td>\n",
       "      <td id=\"T_e3828_row4_col3\" class=\"data row4 col3\" >0.000019</td>\n",
       "      <td id=\"T_e3828_row4_col4\" class=\"data row4 col4\" >0.000186</td>\n",
       "      <td id=\"T_e3828_row4_col5\" class=\"data row4 col5\" >3.126780</td>\n",
       "      <td id=\"T_e3828_row4_col6\" class=\"data row4 col6\" >33.928454</td>\n",
       "      <td id=\"T_e3828_row4_col7\" class=\"data row4 col7\" >ICOSLG;LAMP3;CX3CL1;SLC4A4;RNF144B;EDN1;F3;TNFSF10;NAMPT;ADM;ITGB3;ICAM1;IL18;BST2;TNFRSF1B;CMKLR1;TNFSF15;INHBA;CD14;ACVR2A;PLAUR;HAS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row5\" class=\"row_heading level0 row5\" >26</th>\n",
       "      <td id=\"T_e3828_row5_col0\" class=\"data row5 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row5_col1\" class=\"data row5 col1\" >HALLMARK_INTERFERON_GAMMA_RESPONSE</td>\n",
       "      <td id=\"T_e3828_row5_col2\" class=\"data row5 col2\" >24/166</td>\n",
       "      <td id=\"T_e3828_row5_col3\" class=\"data row5 col3\" >0.000062</td>\n",
       "      <td id=\"T_e3828_row5_col4\" class=\"data row5 col4\" >0.000428</td>\n",
       "      <td id=\"T_e3828_row5_col5\" class=\"data row5 col5\" >2.712373</td>\n",
       "      <td id=\"T_e3828_row5_col6\" class=\"data row5 col6\" >26.260184</td>\n",
       "      <td id=\"T_e3828_row5_col7\" class=\"data row5 col7\" >OASL;DDX60;GBP4;ISG15;IFI27;CFH;TNFSF10;NAMPT;IFIH1;OAS2;HLA-A;ICAM1;BST2;RSAD2;IFI44;CMKLR1;IFIT2;HERC6;VCAM1;SLAMF7;ST8SIA4;TNFAIP2;IFIT1;IFIT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row6\" class=\"row_heading level0 row6\" >15</th>\n",
       "      <td id=\"T_e3828_row6_col0\" class=\"data row6 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row6_col1\" class=\"data row6 col1\" >HALLMARK_ESTROGEN_RESPONSE_LATE</td>\n",
       "      <td id=\"T_e3828_row6_col2\" class=\"data row6 col2\" >23/155</td>\n",
       "      <td id=\"T_e3828_row6_col3\" class=\"data row6 col3\" >0.000059</td>\n",
       "      <td id=\"T_e3828_row6_col4\" class=\"data row6 col4\" >0.000428</td>\n",
       "      <td id=\"T_e3828_row6_col5\" class=\"data row6 col5\" >2.796739</td>\n",
       "      <td id=\"T_e3828_row6_col6\" class=\"data row6 col6\" >27.255656</td>\n",
       "      <td id=\"T_e3828_row6_col7\" class=\"data row6 col7\" >SCNN1A;SERPINA3;CKB;CALCR;TIAM1;MEST;CDC20;RPS6KA2;TPD52L1;TSPAN13;PTGES;PAPSS2;LLGL2;CLIC3;GPER1;CDH1;TRIM29;KRT19;SERPINA1;AGR2;DHCR7;PERP;MAPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row7\" class=\"row_heading level0 row7\" >19</th>\n",
       "      <td id=\"T_e3828_row7_col0\" class=\"data row7 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row7_col1\" class=\"data row7 col1\" >HALLMARK_HEDGEHOG_SIGNALING</td>\n",
       "      <td id=\"T_e3828_row7_col2\" class=\"data row7 col2\" >8/27</td>\n",
       "      <td id=\"T_e3828_row7_col3\" class=\"data row7 col3\" >0.000139</td>\n",
       "      <td id=\"T_e3828_row7_col4\" class=\"data row7 col4\" >0.000834</td>\n",
       "      <td id=\"T_e3828_row7_col5\" class=\"data row7 col5\" >6.807732</td>\n",
       "      <td id=\"T_e3828_row7_col6\" class=\"data row7 col6\" >60.458321</td>\n",
       "      <td id=\"T_e3828_row7_col7\" class=\"data row7 col7\" >ETS2;NRCAM;NRP2;CELSR1;ADGRG1;VLDLR;AMOT;SHH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row8\" class=\"row_heading level0 row8\" >42</th>\n",
       "      <td id=\"T_e3828_row8_col0\" class=\"data row8 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row8_col1\" class=\"data row8 col1\" >HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "      <td id=\"T_e3828_row8_col2\" class=\"data row8 col2\" >23/180</td>\n",
       "      <td id=\"T_e3828_row8_col3\" class=\"data row8 col3\" >0.000556</td>\n",
       "      <td id=\"T_e3828_row8_col4\" class=\"data row8 col4\" >0.002966</td>\n",
       "      <td id=\"T_e3828_row8_col5\" class=\"data row8 col5\" >2.348164</td>\n",
       "      <td id=\"T_e3828_row8_col6\" class=\"data row8 col6\" >17.598221</td>\n",
       "      <td id=\"T_e3828_row8_col7\" class=\"data row8 col7\" >ICOSLG;NR4A2;CXCL1;EDN1;F3;NAMPT;SAT1;TNC;IFIH1;PLPP3;ICAM1;IL18;TSC22D1;ACKR3;IFIT2;INHBA;PMEPA1;ETS2;KYNU;PLAUR;PLK2;GFPT2;TNFAIP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row9\" class=\"row_heading level0 row9\" >1</th>\n",
       "      <td id=\"T_e3828_row9_col0\" class=\"data row9 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row9_col1\" class=\"data row9 col1\" >HALLMARK_ALLOGRAFT_REJECTION</td>\n",
       "      <td id=\"T_e3828_row9_col2\" class=\"data row9 col2\" >16/109</td>\n",
       "      <td id=\"T_e3828_row9_col3\" class=\"data row9 col3\" >0.000850</td>\n",
       "      <td id=\"T_e3828_row9_col4\" class=\"data row9 col4\" >0.004079</td>\n",
       "      <td id=\"T_e3828_row9_col5\" class=\"data row9 col5\" >2.767177</td>\n",
       "      <td id=\"T_e3828_row9_col6\" class=\"data row9 col6\" >19.565548</td>\n",
       "      <td id=\"T_e3828_row9_col7\" class=\"data row9 col7\" >ELF4;IL18;ICOSLG;ACVR2A;TGFB2;ST8SIA4;SRGN;ITGB2;INHBB;RPL39;TLR6;FLNA;HLA-A;IL11;ICAM1;INHBA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row10\" class=\"row_heading level0 row10\" >47</th>\n",
       "      <td id=\"T_e3828_row10_col0\" class=\"data row10 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row10_col1\" class=\"data row10 col1\" >HALLMARK_XENOBIOTIC_METABOLISM</td>\n",
       "      <td id=\"T_e3828_row10_col2\" class=\"data row10 col2\" >19/146</td>\n",
       "      <td id=\"T_e3828_row10_col3\" class=\"data row10 col3\" >0.001309</td>\n",
       "      <td id=\"T_e3828_row10_col4\" class=\"data row10 col4\" >0.005239</td>\n",
       "      <td id=\"T_e3828_row10_col5\" class=\"data row10 col5\" >2.400693</td>\n",
       "      <td id=\"T_e3828_row10_col6\" class=\"data row10 col6\" >15.936955</td>\n",
       "      <td id=\"T_e3828_row10_col7\" class=\"data row10 col7\" >CDA;ITIH4;AOX1;PDK4;CYP1A1;RAP1GAP;PTGES;PAPSS2;FBP1;MAN1A1;ATOH8;PC;VTN;ETS2;KYNU;TGFB2;VNN1;CYB5A;SPINT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row11\" class=\"row_heading level0 row11\" >25</th>\n",
       "      <td id=\"T_e3828_row11_col0\" class=\"data row11 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row11_col1\" class=\"data row11 col1\" >HALLMARK_INTERFERON_ALPHA_RESPONSE</td>\n",
       "      <td id=\"T_e3828_row11_col2\" class=\"data row11 col2\" >14/93</td>\n",
       "      <td id=\"T_e3828_row11_col3\" class=\"data row11 col3\" >0.001369</td>\n",
       "      <td id=\"T_e3828_row11_col4\" class=\"data row11 col4\" >0.005239</td>\n",
       "      <td id=\"T_e3828_row11_col5\" class=\"data row11 col5\" >2.856094</td>\n",
       "      <td id=\"T_e3828_row11_col6\" class=\"data row11 col6\" >18.833073</td>\n",
       "      <td id=\"T_e3828_row11_col7\" class=\"data row11 col7\" >OASL;DDX60;HERC6;GBP4;LAMP3;BST2;RSAD2;IFI44;ISG15;SAMD9;IFIH1;IFIT2;IFI27;IFIT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row12\" class=\"row_heading level0 row12\" >32</th>\n",
       "      <td id=\"T_e3828_row12_col0\" class=\"data row12 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row12_col1\" class=\"data row12 col1\" >HALLMARK_MYOGENESIS</td>\n",
       "      <td id=\"T_e3828_row12_col2\" class=\"data row12 col2\" >17/125</td>\n",
       "      <td id=\"T_e3828_row12_col3\" class=\"data row12 col3\" >0.001419</td>\n",
       "      <td id=\"T_e3828_row12_col4\" class=\"data row12 col4\" >0.005239</td>\n",
       "      <td id=\"T_e3828_row12_col5\" class=\"data row12 col5\" >2.529276</td>\n",
       "      <td id=\"T_e3828_row12_col6\" class=\"data row12 col6\" >16.586601</td>\n",
       "      <td id=\"T_e3828_row12_col7\" class=\"data row12 col7\" >FHL1;EPHB3;GSN;MYLK;COL6A2;PC;SPDEF;ADAM12;ERBB3;VIPR1;FST;TPD52L1;PRNP;CKB;WWTR1;PPFIA4;SCD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row13\" class=\"row_heading level0 row13\" >9</th>\n",
       "      <td id=\"T_e3828_row13_col0\" class=\"data row13 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row13_col1\" class=\"data row13 col1\" >HALLMARK_COAGULATION</td>\n",
       "      <td id=\"T_e3828_row13_col2\" class=\"data row13 col2\" >13/84</td>\n",
       "      <td id=\"T_e3828_row13_col3\" class=\"data row13 col3\" >0.001550</td>\n",
       "      <td id=\"T_e3828_row13_col4\" class=\"data row13 col4\" >0.005315</td>\n",
       "      <td id=\"T_e3828_row13_col5\" class=\"data row13 col5\" >2.954866</td>\n",
       "      <td id=\"T_e3828_row13_col6\" class=\"data row13 col6\" >19.116250</td>\n",
       "      <td id=\"T_e3828_row13_col7\" class=\"data row13 col7\" >F3;THBS1;GSN;ITGB3;ITGA2;CRIP2;F2RL2;THBD;F8;SERPINA1;CFH;TIMP3;FYN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row14\" class=\"row_heading level0 row14\" >4</th>\n",
       "      <td id=\"T_e3828_row14_col0\" class=\"data row14 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row14_col1\" class=\"data row14 col1\" >HALLMARK_APICAL_JUNCTION</td>\n",
       "      <td id=\"T_e3828_row14_col2\" class=\"data row14 col2\" >19/152</td>\n",
       "      <td id=\"T_e3828_row14_col3\" class=\"data row14 col3\" >0.002105</td>\n",
       "      <td id=\"T_e3828_row14_col4\" class=\"data row14 col4\" >0.006735</td>\n",
       "      <td id=\"T_e3828_row14_col5\" class=\"data row14 col5\" >2.291710</td>\n",
       "      <td id=\"T_e3828_row14_col6\" class=\"data row14 col6\" >14.125159</td>\n",
       "      <td id=\"T_e3828_row14_col7\" class=\"data row14 col7\" >CX3CL1;CDH4;ICAM5;IKBKG;CLDN7;LIMA1;CALB2;COL17A1;ICAM1;CDH1;VCAM1;ITGA2;MYH10;CDH11;GRB7;FSCN1;PPP2R2C;INSIG1;ARHGEF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row15\" class=\"row_heading level0 row15\" >8</th>\n",
       "      <td id=\"T_e3828_row15_col0\" class=\"data row15 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row15_col1\" class=\"data row15 col1\" >HALLMARK_CHOLESTEROL_HOMEOSTASIS</td>\n",
       "      <td id=\"T_e3828_row15_col2\" class=\"data row15 col2\" >11/69</td>\n",
       "      <td id=\"T_e3828_row15_col3\" class=\"data row15 col3\" >0.002738</td>\n",
       "      <td id=\"T_e3828_row15_col4\" class=\"data row15 col4\" >0.008214</td>\n",
       "      <td id=\"T_e3828_row15_col5\" class=\"data row15 col5\" >3.072049</td>\n",
       "      <td id=\"T_e3828_row15_col6\" class=\"data row15 col6\" >18.126713</td>\n",
       "      <td id=\"T_e3828_row15_col7\" class=\"data row15 col7\" >PLAUR;FADS2;GSTM2;ALDOC;LSS;MAL2;NSDHL;TRIB3;SCD;DHCR7;ANTXR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row16\" class=\"row_heading level0 row16\" >21</th>\n",
       "      <td id=\"T_e3828_row16_col0\" class=\"data row16 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row16_col1\" class=\"data row16 col1\" >HALLMARK_HYPOXIA</td>\n",
       "      <td id=\"T_e3828_row16_col2\" class=\"data row16 col2\" >20/169</td>\n",
       "      <td id=\"T_e3828_row16_col3\" class=\"data row16 col3\" >0.003125</td>\n",
       "      <td id=\"T_e3828_row16_col4\" class=\"data row16 col4\" >0.008823</td>\n",
       "      <td id=\"T_e3828_row16_col5\" class=\"data row16 col5\" >2.151336</td>\n",
       "      <td id=\"T_e3828_row16_col6\" class=\"data row16 col6\" >12.409633</td>\n",
       "      <td id=\"T_e3828_row16_col7\" class=\"data row16 col7\" >NDRG1;PGF;ALDOC;SRPX;F3;ADM;HS3ST1;SAP30;PLAC8;EDN2;PPFIA4;FBP1;TMEM45A;ANGPTL4;SELENBP1;ACKR3;SLC2A1;PLAUR;MT1E;VLDLR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row17\" class=\"row_heading level0 row17\" >44</th>\n",
       "      <td id=\"T_e3828_row17_col0\" class=\"data row17 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row17_col1\" class=\"data row17 col1\" >HALLMARK_UV_RESPONSE_DN</td>\n",
       "      <td id=\"T_e3828_row17_col2\" class=\"data row17 col2\" >16/133</td>\n",
       "      <td id=\"T_e3828_row17_col3\" class=\"data row17 col3\" >0.006614</td>\n",
       "      <td id=\"T_e3828_row17_col4\" class=\"data row17 col4\" >0.017638</td>\n",
       "      <td id=\"T_e3828_row17_col5\" class=\"data row17 col5\" >2.197803</td>\n",
       "      <td id=\"T_e3828_row17_col6\" class=\"data row17 col6\" >11.029745</td>\n",
       "      <td id=\"T_e3828_row17_col7\" class=\"data row17 col7\" >F3;ACVR2A;ITGB3;GJA1;EFEMP1;MT1E;FHL2;COL1A2;VLDLR;INSIG1;MGLL;PLPP3;LTBP1;PDGFRB;HAS2;FYN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row18\" class=\"row_heading level0 row18\" >2</th>\n",
       "      <td id=\"T_e3828_row18_col0\" class=\"data row18 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row18_col1\" class=\"data row18 col1\" >HALLMARK_ANDROGEN_RESPONSE</td>\n",
       "      <td id=\"T_e3828_row18_col2\" class=\"data row18 col2\" >11/90</td>\n",
       "      <td id=\"T_e3828_row18_col3\" class=\"data row18 col3\" >0.019859</td>\n",
       "      <td id=\"T_e3828_row18_col4\" class=\"data row18 col4\" >0.050171</td>\n",
       "      <td id=\"T_e3828_row18_col5\" class=\"data row18 col5\" >2.256835</td>\n",
       "      <td id=\"T_e3828_row18_col6\" class=\"data row18 col6\" >8.844727</td>\n",
       "      <td id=\"T_e3828_row18_col7\" class=\"data row18 col7\" >NDRG1;SAT1;KRT8;SPDEF;DHCR24;INSIG1;TSC22D1;KRT19;ADAMTS1;SCD;PMEPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row19\" class=\"row_heading level0 row19\" >6</th>\n",
       "      <td id=\"T_e3828_row19_col0\" class=\"data row19 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row19_col1\" class=\"data row19 col1\" >HALLMARK_APOPTOSIS</td>\n",
       "      <td id=\"T_e3828_row19_col2\" class=\"data row19 col2\" >15/139</td>\n",
       "      <td id=\"T_e3828_row19_col3\" class=\"data row19 col3\" >0.021250</td>\n",
       "      <td id=\"T_e3828_row19_col4\" class=\"data row19 col4\" >0.051001</td>\n",
       "      <td id=\"T_e3828_row19_col5\" class=\"data row19 col5\" >1.945041</td>\n",
       "      <td id=\"T_e3828_row19_col6\" class=\"data row19 col6\" >7.491099</td>\n",
       "      <td id=\"T_e3828_row19_col7\" class=\"data row19 col7\" >BIK;IL18;TNFSF10;GSN;TGFB2;SAT1;LEF1;PDGFRB;ERBB3;KRT18;TIMP3;EMP1;SATB1;BCAP31;CD14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row20\" class=\"row_heading level0 row20\" >18</th>\n",
       "      <td id=\"T_e3828_row20_col0\" class=\"data row20 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row20_col1\" class=\"data row20 col1\" >HALLMARK_GLYCOLYSIS</td>\n",
       "      <td id=\"T_e3828_row20_col2\" class=\"data row20 col2\" >17/176</td>\n",
       "      <td id=\"T_e3828_row20_col3\" class=\"data row20 col3\" >0.038519</td>\n",
       "      <td id=\"T_e3828_row20_col4\" class=\"data row20 col4\" >0.088043</td>\n",
       "      <td id=\"T_e3828_row20_col5\" class=\"data row20 col5\" >1.713621</td>\n",
       "      <td id=\"T_e3828_row20_col6\" class=\"data row20 col6\" >5.580600</td>\n",
       "      <td id=\"T_e3828_row20_col7\" class=\"data row20 col7\" >BIK;ARTN;NDST3;ANGPTL4;TGFA;SAP30;ELF3;CHST4;G6PD;CYB5A;PC;FUT8;PRPS1;EGLN3;VLDLR;NSDHL;PPFIA4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row21\" class=\"row_heading level0 row21\" >10</th>\n",
       "      <td id=\"T_e3828_row21_col0\" class=\"data row21 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row21_col1\" class=\"data row21 col1\" >HALLMARK_COMPLEMENT</td>\n",
       "      <td id=\"T_e3828_row21_col2\" class=\"data row21 col2\" >14/141</td>\n",
       "      <td id=\"T_e3828_row21_col3\" class=\"data row21 col3\" >0.046824</td>\n",
       "      <td id=\"T_e3828_row21_col4\" class=\"data row21 col4\" >0.102162</td>\n",
       "      <td id=\"T_e3828_row21_col5\" class=\"data row21 col5\" >1.774132</td>\n",
       "      <td id=\"T_e3828_row21_col6\" class=\"data row21 col6\" >5.431244</td>\n",
       "      <td id=\"T_e3828_row21_col7\" class=\"data row21 col7\" >F3;KYNU;PLAUR;PRSS3;PCSK9;C4BPB;GATA3;CXCL1;CTSL;CDA;F8;SERPINA1;CFH;FYN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row22\" class=\"row_heading level0 row22\" >27</th>\n",
       "      <td id=\"T_e3828_row22_col0\" class=\"data row22 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row22_col1\" class=\"data row22 col1\" >HALLMARK_KRAS_SIGNALING_DN</td>\n",
       "      <td id=\"T_e3828_row22_col2\" class=\"data row22 col2\" >9/80</td>\n",
       "      <td id=\"T_e3828_row22_col3\" class=\"data row22 col3\" >0.052350</td>\n",
       "      <td id=\"T_e3828_row22_col4\" class=\"data row22 col4\" >0.109251</td>\n",
       "      <td id=\"T_e3828_row22_col5\" class=\"data row22 col5\" >2.069164</td>\n",
       "      <td id=\"T_e3828_row22_col6\" class=\"data row22 col6\" >6.103645</td>\n",
       "      <td id=\"T_e3828_row22_col7\" class=\"data row22 col7\" >TENM2;NR4A2;RSAD2;TGFB2;C5;SLC16A7;EDN2;TCF7L1;EDN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row23\" class=\"row_heading level0 row23\" >36</th>\n",
       "      <td id=\"T_e3828_row23_col0\" class=\"data row23 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row23_col1\" class=\"data row23 col1\" >HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "      <td id=\"T_e3828_row23_col2\" class=\"data row23 col2\" >3/15</td>\n",
       "      <td id=\"T_e3828_row23_col3\" class=\"data row23 col3\" >0.058475</td>\n",
       "      <td id=\"T_e3828_row23_col4\" class=\"data row23 col4\" >0.116950</td>\n",
       "      <td id=\"T_e3828_row23_col5\" class=\"data row23 col5\" >4.348766</td>\n",
       "      <td id=\"T_e3828_row23_col6\" class=\"data row23 col6\" >12.346818</td>\n",
       "      <td id=\"T_e3828_row23_col7\" class=\"data row23 col7\" >FOXA2;SRPRB;FOXO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row24\" class=\"row_heading level0 row24\" >3</th>\n",
       "      <td id=\"T_e3828_row24_col0\" class=\"data row24 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row24_col1\" class=\"data row24 col1\" >HALLMARK_ANGIOGENESIS</td>\n",
       "      <td id=\"T_e3828_row24_col2\" class=\"data row24 col2\" >4/26</td>\n",
       "      <td id=\"T_e3828_row24_col3\" class=\"data row24 col3\" >0.069259</td>\n",
       "      <td id=\"T_e3828_row24_col4\" class=\"data row24 col4\" >0.127863</td>\n",
       "      <td id=\"T_e3828_row24_col5\" class=\"data row24 col5\" >3.107608</td>\n",
       "      <td id=\"T_e3828_row24_col6\" class=\"data row24 col6\" >8.296997</td>\n",
       "      <td id=\"T_e3828_row24_col7\" class=\"data row24 col7\" >THBD;SLCO2A1;VTN;FSTL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row25\" class=\"row_heading level0 row25\" >22</th>\n",
       "      <td id=\"T_e3828_row25_col0\" class=\"data row25 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row25_col1\" class=\"data row25 col1\" >HALLMARK_IL2_STAT5_SIGNALING</td>\n",
       "      <td id=\"T_e3828_row25_col2\" class=\"data row25 col2\" >14/149</td>\n",
       "      <td id=\"T_e3828_row25_col3\" class=\"data row25 col3\" >0.068054</td>\n",
       "      <td id=\"T_e3828_row25_col4\" class=\"data row25 col4\" >0.127863</td>\n",
       "      <td id=\"T_e3828_row25_col5\" class=\"data row25 col5\" >1.668331</td>\n",
       "      <td id=\"T_e3828_row25_col6\" class=\"data row25 col6\" >4.483559</td>\n",
       "      <td id=\"T_e3828_row25_col7\" class=\"data row25 col7\" >DHRS3;TNFSF10;NDRG1;GBP4;BATF3;TNFRSF1B;F2RL2;MUC1;PRNP;SPRY4;IRF6;EMP1;MAP6;TIAM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row26\" class=\"row_heading level0 row26\" >33</th>\n",
       "      <td id=\"T_e3828_row26_col0\" class=\"data row26 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row26_col1\" class=\"data row26 col1\" >HALLMARK_NOTCH_SIGNALING</td>\n",
       "      <td id=\"T_e3828_row26_col2\" class=\"data row26 col2\" >4/29</td>\n",
       "      <td id=\"T_e3828_row26_col3\" class=\"data row26 col3\" >0.095627</td>\n",
       "      <td id=\"T_e3828_row26_col4\" class=\"data row26 col4\" >0.170003</td>\n",
       "      <td id=\"T_e3828_row26_col5\" class=\"data row26 col5\" >2.741363</td>\n",
       "      <td id=\"T_e3828_row26_col6\" class=\"data row26 col6\" >6.434814</td>\n",
       "      <td id=\"T_e3828_row26_col7\" class=\"data row26 col7\" >SAP30;FZD1;WNT5A;FZD7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row27\" class=\"row_heading level0 row27\" >7</th>\n",
       "      <td id=\"T_e3828_row27_col0\" class=\"data row27 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row27_col1\" class=\"data row27 col1\" >HALLMARK_BILE_ACID_METABOLISM</td>\n",
       "      <td id=\"T_e3828_row27_col2\" class=\"data row27 col2\" >7/71</td>\n",
       "      <td id=\"T_e3828_row27_col3\" class=\"data row27 col3\" >0.137655</td>\n",
       "      <td id=\"T_e3828_row27_col4\" class=\"data row27 col4\" >0.227843</td>\n",
       "      <td id=\"T_e3828_row27_col5\" class=\"data row27 col5\" >1.807404</td>\n",
       "      <td id=\"T_e3828_row27_col6\" class=\"data row27 col6\" >3.584088</td>\n",
       "      <td id=\"T_e3828_row27_col7\" class=\"data row27 col7\" >HSD3B1;DIO2;FADS2;CH25H;DHCR24;PXMP2;PHYH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row28\" class=\"row_heading level0 row28\" >46</th>\n",
       "      <td id=\"T_e3828_row28_col0\" class=\"data row28 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row28_col1\" class=\"data row28 col1\" >HALLMARK_WNT_BETA_CATENIN_SIGNALING</td>\n",
       "      <td id=\"T_e3828_row28_col2\" class=\"data row28 col2\" >4/33</td>\n",
       "      <td id=\"T_e3828_row28_col3\" class=\"data row28 col3\" >0.136896</td>\n",
       "      <td id=\"T_e3828_row28_col4\" class=\"data row28 col4\" >0.227843</td>\n",
       "      <td id=\"T_e3828_row28_col5\" class=\"data row28 col5\" >2.368910</td>\n",
       "      <td id=\"T_e3828_row28_col6\" class=\"data row28 col6\" >4.710649</td>\n",
       "      <td id=\"T_e3828_row28_col7\" class=\"data row28 col7\" >LEF1;FZD1;DKK1;FZD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row29\" class=\"row_heading level0 row29\" >41</th>\n",
       "      <td id=\"T_e3828_row29_col0\" class=\"data row29 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row29_col1\" class=\"data row29 col1\" >HALLMARK_TGF_BETA_SIGNALING</td>\n",
       "      <td id=\"T_e3828_row29_col2\" class=\"data row29 col2\" >5/51</td>\n",
       "      <td id=\"T_e3828_row29_col3\" class=\"data row29 col3\" >0.194927</td>\n",
       "      <td id=\"T_e3828_row29_col4\" class=\"data row29 col4\" >0.311883</td>\n",
       "      <td id=\"T_e3828_row29_col5\" class=\"data row29 col5\" >1.836613</td>\n",
       "      <td id=\"T_e3828_row29_col6\" class=\"data row29 col6\" >3.003102</td>\n",
       "      <td id=\"T_e3828_row29_col7\" class=\"data row29 col7\" >THBS1;LTBP2;CDH1;WWTR1;PMEPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row30\" class=\"row_heading level0 row30\" >35</th>\n",
       "      <td id=\"T_e3828_row30_col0\" class=\"data row30 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row30_col1\" class=\"data row30 col1\" >HALLMARK_P53_PATHWAY</td>\n",
       "      <td id=\"T_e3828_row30_col2\" class=\"data row30 col2\" >13/177</td>\n",
       "      <td id=\"T_e3828_row30_col3\" class=\"data row30 col3\" >0.276709</td>\n",
       "      <td id=\"T_e3828_row30_col4\" class=\"data row30 col4\" >0.422635</td>\n",
       "      <td id=\"T_e3828_row30_col5\" class=\"data row30 col5\" >1.274940</td>\n",
       "      <td id=\"T_e3828_row30_col6\" class=\"data row30 col6\" >1.638029</td>\n",
       "      <td id=\"T_e3828_row30_col7\" class=\"data row30 col7\" >NDRG1;TGFA;SAT1;SLC7A11;PLK2;TPD52L1;INHBB;PRMT2;TSC22D1;TRIB3;PERP;H2AJ;IRAK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row31\" class=\"row_heading level0 row31\" >30</th>\n",
       "      <td id=\"T_e3828_row31_col0\" class=\"data row31 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row31_col1\" class=\"data row31 col1\" >HALLMARK_MTORC1_SIGNALING</td>\n",
       "      <td id=\"T_e3828_row31_col2\" class=\"data row31 col2\" >14/193</td>\n",
       "      <td id=\"T_e3828_row31_col3\" class=\"data row31 col3\" >0.281757</td>\n",
       "      <td id=\"T_e3828_row31_col4\" class=\"data row31 col4\" >0.422635</td>\n",
       "      <td id=\"T_e3828_row31_col5\" class=\"data row31 col5\" >1.255001</td>\n",
       "      <td id=\"T_e3828_row31_col6\" class=\"data row31 col6\" >1.589724</td>\n",
       "      <td id=\"T_e3828_row31_col7\" class=\"data row31 col7\" >PHGDH;NAMPT;FADS2;SLC7A11;G6PD;ITGB2;EGLN3;DHCR24;INSIG1;VLDLR;TRIB3;SCD;DHCR7;SLC2A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row32\" class=\"row_heading level0 row32\" >20</th>\n",
       "      <td id=\"T_e3828_row32_col0\" class=\"data row32 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row32_col1\" class=\"data row32 col1\" >HALLMARK_HEME_METABOLISM</td>\n",
       "      <td id=\"T_e3828_row32_col2\" class=\"data row32 col2\" >11/164</td>\n",
       "      <td id=\"T_e3828_row32_col3\" class=\"data row32 col3\" >0.409142</td>\n",
       "      <td id=\"T_e3828_row32_col4\" class=\"data row32 col4\" >0.595116</td>\n",
       "      <td id=\"T_e3828_row32_col5\" class=\"data row32 col5\" >1.162043</td>\n",
       "      <td id=\"T_e3828_row32_col6\" class=\"data row32 col6\" >1.038509</td>\n",
       "      <td id=\"T_e3828_row32_col7\" class=\"data row32 col7\" >MPP1;TNS1;PDZK1IP1;SLC7A11;UCP2;PC;SELENBP1;RAP1GAP;FN3K;ABCG2;SLC2A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row33\" class=\"row_heading level0 row33\" >23</th>\n",
       "      <td id=\"T_e3828_row33_col0\" class=\"data row33 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row33_col1\" class=\"data row33 col1\" >HALLMARK_IL6_JAK_STAT3_SIGNALING</td>\n",
       "      <td id=\"T_e3828_row33_col2\" class=\"data row33 col2\" >4/58</td>\n",
       "      <td id=\"T_e3828_row33_col3\" class=\"data row33 col3\" >0.470391</td>\n",
       "      <td id=\"T_e3828_row33_col4\" class=\"data row33 col4\" >0.664082</td>\n",
       "      <td id=\"T_e3828_row33_col5\" class=\"data row33 col5\" >1.279741</td>\n",
       "      <td id=\"T_e3828_row33_col6\" class=\"data row33 col6\" >0.965168</td>\n",
       "      <td id=\"T_e3828_row33_col7\" class=\"data row33 col7\" >ITGB3;TNFRSF1B;CXCL1;CD14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row34\" class=\"row_heading level0 row34\" >38</th>\n",
       "      <td id=\"T_e3828_row34_col0\" class=\"data row34 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row34_col1\" class=\"data row34 col1\" >HALLMARK_PI3K_AKT_MTOR_SIGNALING</td>\n",
       "      <td id=\"T_e3828_row34_col2\" class=\"data row34 col2\" >5/89</td>\n",
       "      <td id=\"T_e3828_row34_col3\" class=\"data row34 col3\" >0.633543</td>\n",
       "      <td id=\"T_e3828_row34_col4\" class=\"data row34 col4\" >0.868860</td>\n",
       "      <td id=\"T_e3828_row34_col5\" class=\"data row34 col5\" >1.007666</td>\n",
       "      <td id=\"T_e3828_row34_col6\" class=\"data row34 col6\" >0.459926</td>\n",
       "      <td id=\"T_e3828_row34_col7\" class=\"data row34 col7\" >TIAM1;CAB39L;NGF;TRIB3;SLC2A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row35\" class=\"row_heading level0 row35\" >16</th>\n",
       "      <td id=\"T_e3828_row35_col0\" class=\"data row35 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row35_col1\" class=\"data row35 col1\" >HALLMARK_FATTY_ACID_METABOLISM</td>\n",
       "      <td id=\"T_e3828_row35_col2\" class=\"data row35 col2\" >7/128</td>\n",
       "      <td id=\"T_e3828_row35_col3\" class=\"data row35 col3\" >0.664444</td>\n",
       "      <td id=\"T_e3828_row35_col4\" class=\"data row35 col4\" >0.885926</td>\n",
       "      <td id=\"T_e3828_row35_col5\" class=\"data row35 col5\" >0.955188</td>\n",
       "      <td id=\"T_e3828_row35_col6\" class=\"data row35 col6\" >0.390485</td>\n",
       "      <td id=\"T_e3828_row35_col7\" class=\"data row35 col7\" >IDH3G;VNN1;TP53INP2;CYP1A1;DHCR24;MGLL;NSDHL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row36\" class=\"row_heading level0 row36\" >40</th>\n",
       "      <td id=\"T_e3828_row36_col0\" class=\"data row36 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row36_col1\" class=\"data row36 col1\" >HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY</td>\n",
       "      <td id=\"T_e3828_row36_col2\" class=\"data row36 col2\" >2/44</td>\n",
       "      <td id=\"T_e3828_row36_col3\" class=\"data row36 col3\" >0.755365</td>\n",
       "      <td id=\"T_e3828_row36_col4\" class=\"data row36 col4\" >0.972599</td>\n",
       "      <td id=\"T_e3828_row36_col5\" class=\"data row36 col5\" >0.910354</td>\n",
       "      <td id=\"T_e3828_row36_col6\" class=\"data row36 col6\" >0.255404</td>\n",
       "      <td id=\"T_e3828_row36_col7\" class=\"data row36 col7\" >PRNP;G6PD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row37\" class=\"row_heading level0 row37\" >0</th>\n",
       "      <td id=\"T_e3828_row37_col0\" class=\"data row37 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row37_col1\" class=\"data row37 col1\" >HALLMARK_ADIPOGENESIS</td>\n",
       "      <td id=\"T_e3828_row37_col2\" class=\"data row37 col2\" >9/180</td>\n",
       "      <td id=\"T_e3828_row37_col3\" class=\"data row37 col3\" >0.769974</td>\n",
       "      <td id=\"T_e3828_row37_col4\" class=\"data row37 col4\" >0.972599</td>\n",
       "      <td id=\"T_e3828_row37_col5\" class=\"data row37 col5\" >0.855870</td>\n",
       "      <td id=\"T_e3828_row37_col6\" class=\"data row37 col6\" >0.223723</td>\n",
       "      <td id=\"T_e3828_row37_col7\" class=\"data row37 col7\" >IDH3G;ANGPTL4;LAMA4;MYLK;DHCR7;UCP2;MGLL;FZD4;PHYH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row38\" class=\"row_heading level0 row38\" >34</th>\n",
       "      <td id=\"T_e3828_row38_col0\" class=\"data row38 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row38_col1\" class=\"data row38 col1\" >HALLMARK_OXIDATIVE_PHOSPHORYLATION</td>\n",
       "      <td id=\"T_e3828_row38_col2\" class=\"data row38 col2\" >5/200</td>\n",
       "      <td id=\"T_e3828_row38_col3\" class=\"data row38 col3\" >0.994432</td>\n",
       "      <td id=\"T_e3828_row38_col4\" class=\"data row38 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row38_col5\" class=\"data row38 col5\" >0.431733</td>\n",
       "      <td id=\"T_e3828_row38_col6\" class=\"data row38 col6\" >0.002410</td>\n",
       "      <td id=\"T_e3828_row38_col7\" class=\"data row38 col7\" >IDH3G;CYB5A;PDK4;SLC25A5;PHYH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row39\" class=\"row_heading level0 row39\" >29</th>\n",
       "      <td id=\"T_e3828_row39_col0\" class=\"data row39 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row39_col1\" class=\"data row39 col1\" >HALLMARK_MITOTIC_SPINDLE</td>\n",
       "      <td id=\"T_e3828_row39_col2\" class=\"data row39 col2\" >7/196</td>\n",
       "      <td id=\"T_e3828_row39_col3\" class=\"data row39 col3\" >0.956918</td>\n",
       "      <td id=\"T_e3828_row39_col4\" class=\"data row39 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row39_col5\" class=\"data row39 col5\" >0.609141</td>\n",
       "      <td id=\"T_e3828_row39_col6\" class=\"data row39 col6\" >0.026825</td>\n",
       "      <td id=\"T_e3828_row39_col7\" class=\"data row39 col7\" >PCNT;GSN;MYH10;FSCN1;FLNA;TIAM1;SORBS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row40\" class=\"row_heading level0 row40\" >17</th>\n",
       "      <td id=\"T_e3828_row40_col0\" class=\"data row40 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row40_col1\" class=\"data row40 col1\" >HALLMARK_G2M_CHECKPOINT</td>\n",
       "      <td id=\"T_e3828_row40_col2\" class=\"data row40 col2\" >5/196</td>\n",
       "      <td id=\"T_e3828_row40_col3\" class=\"data row40 col3\" >0.993343</td>\n",
       "      <td id=\"T_e3828_row40_col4\" class=\"data row40 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row40_col5\" class=\"data row40 col5\" >0.440890</td>\n",
       "      <td id=\"T_e3828_row40_col6\" class=\"data row40 col6\" >0.002945</td>\n",
       "      <td id=\"T_e3828_row40_col7\" class=\"data row40 col7\" >H2BC12;CDC20;SAP30;HMGB3;MYBL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row41\" class=\"row_heading level0 row41\" >37</th>\n",
       "      <td id=\"T_e3828_row41_col0\" class=\"data row41 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row41_col1\" class=\"data row41 col1\" >HALLMARK_PEROXISOME</td>\n",
       "      <td id=\"T_e3828_row41_col2\" class=\"data row41 col2\" >2/84</td>\n",
       "      <td id=\"T_e3828_row41_col3\" class=\"data row41 col3\" >0.966759</td>\n",
       "      <td id=\"T_e3828_row41_col4\" class=\"data row41 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row41_col5\" class=\"data row41 col5\" >0.467498</td>\n",
       "      <td id=\"T_e3828_row41_col6\" class=\"data row41 col6\" >0.015804</td>\n",
       "      <td id=\"T_e3828_row41_col7\" class=\"data row41 col7\" >DHRS3;DHCR24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row42\" class=\"row_heading level0 row42\" >39</th>\n",
       "      <td id=\"T_e3828_row42_col0\" class=\"data row42 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row42_col1\" class=\"data row42 col1\" >HALLMARK_PROTEIN_SECRETION</td>\n",
       "      <td id=\"T_e3828_row42_col2\" class=\"data row42 col2\" >2/95</td>\n",
       "      <td id=\"T_e3828_row42_col3\" class=\"data row42 col3\" >0.981514</td>\n",
       "      <td id=\"T_e3828_row42_col4\" class=\"data row42 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row42_col5\" class=\"data row42 col5\" >0.412141</td>\n",
       "      <td id=\"T_e3828_row42_col6\" class=\"data row42 col6\" >0.007690</td>\n",
       "      <td id=\"T_e3828_row42_col7\" class=\"data row42 col7\" >KRT18;BNIP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row43\" class=\"row_heading level0 row43\" >12</th>\n",
       "      <td id=\"T_e3828_row43_col0\" class=\"data row43 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row43_col1\" class=\"data row43 col1\" >HALLMARK_E2F_TARGETS</td>\n",
       "      <td id=\"T_e3828_row43_col2\" class=\"data row43 col2\" >5/199</td>\n",
       "      <td id=\"T_e3828_row43_col3\" class=\"data row43 col3\" >0.994177</td>\n",
       "      <td id=\"T_e3828_row43_col4\" class=\"data row43 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row43_col5\" class=\"data row43 col5\" >0.433987</td>\n",
       "      <td id=\"T_e3828_row43_col6\" class=\"data row43 col6\" >0.002534</td>\n",
       "      <td id=\"T_e3828_row43_col7\" class=\"data row43 col7\" >HMGB3;CDC20;MYBL2;CDCA8;PRPS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row44\" class=\"row_heading level0 row44\" >43</th>\n",
       "      <td id=\"T_e3828_row44_col0\" class=\"data row44 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row44_col1\" class=\"data row44 col1\" >HALLMARK_UNFOLDED_PROTEIN_RESPONSE</td>\n",
       "      <td id=\"T_e3828_row44_col2\" class=\"data row44 col2\" >2/109</td>\n",
       "      <td id=\"T_e3828_row44_col3\" class=\"data row44 col3\" >0.991370</td>\n",
       "      <td id=\"T_e3828_row44_col4\" class=\"data row44 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row44_col5\" class=\"data row44 col5\" >0.358072</td>\n",
       "      <td id=\"T_e3828_row44_col6\" class=\"data row44 col6\" >0.003103</td>\n",
       "      <td id=\"T_e3828_row44_col7\" class=\"data row44 col7\" >IFIT1;SRPRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row45\" class=\"row_heading level0 row45\" >11</th>\n",
       "      <td id=\"T_e3828_row45_col0\" class=\"data row45 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row45_col1\" class=\"data row45 col1\" >HALLMARK_DNA_REPAIR</td>\n",
       "      <td id=\"T_e3828_row45_col2\" class=\"data row45 col2\" >3/145</td>\n",
       "      <td id=\"T_e3828_row45_col3\" class=\"data row45 col3\" >0.993973</td>\n",
       "      <td id=\"T_e3828_row45_col4\" class=\"data row45 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row45_col5\" class=\"data row45 col5\" >0.377589</td>\n",
       "      <td id=\"T_e3828_row45_col6\" class=\"data row45 col6\" >0.002283</td>\n",
       "      <td id=\"T_e3828_row45_col7\" class=\"data row45 col7\" >UPF3B;BCAP31;CDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row46\" class=\"row_heading level0 row46\" >45</th>\n",
       "      <td id=\"T_e3828_row46_col0\" class=\"data row46 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row46_col1\" class=\"data row46 col1\" >HALLMARK_UV_RESPONSE_UP</td>\n",
       "      <td id=\"T_e3828_row46_col2\" class=\"data row46 col2\" >4/132</td>\n",
       "      <td id=\"T_e3828_row46_col3\" class=\"data row46 col3\" >0.962605</td>\n",
       "      <td id=\"T_e3828_row46_col4\" class=\"data row46 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row46_col5\" class=\"data row46 col5\" >0.539615</td>\n",
       "      <td id=\"T_e3828_row46_col6\" class=\"data row46 col6\" >0.020566</td>\n",
       "      <td id=\"T_e3828_row46_col7\" class=\"data row46 col7\" >C4BPB;CYP1A1;HTR7;ICAM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e3828_level0_row47\" class=\"row_heading level0 row47\" >31</th>\n",
       "      <td id=\"T_e3828_row47_col0\" class=\"data row47 col0\" >gs_ind_0</td>\n",
       "      <td id=\"T_e3828_row47_col1\" class=\"data row47 col1\" >HALLMARK_MYC_TARGETS_V1</td>\n",
       "      <td id=\"T_e3828_row47_col2\" class=\"data row47 col2\" >2/196</td>\n",
       "      <td id=\"T_e3828_row47_col3\" class=\"data row47 col3\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row47_col4\" class=\"data row47 col4\" >0.999940</td>\n",
       "      <td id=\"T_e3828_row47_col5\" class=\"data row47 col5\" >0.196548</td>\n",
       "      <td id=\"T_e3828_row47_col6\" class=\"data row47 col6\" >0.000012</td>\n",
       "      <td id=\"T_e3828_row47_col7\" class=\"data row47 col7\" >CDC20;VBP1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x79948e559650>"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We can pull out using ora.res2d\n",
    "#Here we also simultaneously sort by the Adjusted P-value and output a stylized table\n",
    "ora.res2d.sort_values(by='Adjusted P-value').style\n",
    "#We'll then get an output that is sorted with most significant enrichment up at the top"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1826cb2a-7a18-4a0d-a549-ef131965c086",
   "metadata": {},
   "source": [
    "We can also sort by the significance when looking at our output results table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb8dde80-d65d-461c-8a3a-d50393febc60",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Refer to above code cell since we did it simultaneously"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "744786f2-c488-4c65-ba17-3c14bb4fb856",
   "metadata": {},
   "source": [
    "We can also update the values in the <code>Gene_set</code> column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "de754375-e6af-4516-92e7-49e608c5a4af",
   "metadata": {},
   "outputs": [],
   "source": [
    "ora.res2d.replace(to_replace={'gs_ind_0' : 'MSigDB Hallmark'},\n",
    "                  inplace=True\n",
    "                 )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d88a71c-df82-4cc6-baf5-4d66bf60c29b",
   "metadata": {},
   "source": [
    "Let's check how it updated:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "581cebe9-4623-4ffb-9986-d9b50d95b277",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_c5e8f\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_c5e8f_level0_col0\" class=\"col_heading level0 col0\" >Gene_set</th>\n",
       "      <th id=\"T_c5e8f_level0_col1\" class=\"col_heading level0 col1\" >Term</th>\n",
       "      <th id=\"T_c5e8f_level0_col2\" class=\"col_heading level0 col2\" >Overlap</th>\n",
       "      <th id=\"T_c5e8f_level0_col3\" class=\"col_heading level0 col3\" >P-value</th>\n",
       "      <th id=\"T_c5e8f_level0_col4\" class=\"col_heading level0 col4\" >Adjusted P-value</th>\n",
       "      <th id=\"T_c5e8f_level0_col5\" class=\"col_heading level0 col5\" >Odds Ratio</th>\n",
       "      <th id=\"T_c5e8f_level0_col6\" class=\"col_heading level0 col6\" >Combined Score</th>\n",
       "      <th id=\"T_c5e8f_level0_col7\" class=\"col_heading level0 col7\" >Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_c5e8f_row0_col0\" class=\"data row0 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row0_col1\" class=\"data row0 col1\" >HALLMARK_ADIPOGENESIS</td>\n",
       "      <td id=\"T_c5e8f_row0_col2\" class=\"data row0 col2\" >9/180</td>\n",
       "      <td id=\"T_c5e8f_row0_col3\" class=\"data row0 col3\" >0.769974</td>\n",
       "      <td id=\"T_c5e8f_row0_col4\" class=\"data row0 col4\" >0.972599</td>\n",
       "      <td id=\"T_c5e8f_row0_col5\" class=\"data row0 col5\" >0.855870</td>\n",
       "      <td id=\"T_c5e8f_row0_col6\" class=\"data row0 col6\" >0.223723</td>\n",
       "      <td id=\"T_c5e8f_row0_col7\" class=\"data row0 col7\" >IDH3G;ANGPTL4;LAMA4;MYLK;DHCR7;UCP2;MGLL;FZD4;PHYH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_c5e8f_row1_col0\" class=\"data row1 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row1_col1\" class=\"data row1 col1\" >HALLMARK_ALLOGRAFT_REJECTION</td>\n",
       "      <td id=\"T_c5e8f_row1_col2\" class=\"data row1 col2\" >16/109</td>\n",
       "      <td id=\"T_c5e8f_row1_col3\" class=\"data row1 col3\" >0.000850</td>\n",
       "      <td id=\"T_c5e8f_row1_col4\" class=\"data row1 col4\" >0.004079</td>\n",
       "      <td id=\"T_c5e8f_row1_col5\" class=\"data row1 col5\" >2.767177</td>\n",
       "      <td id=\"T_c5e8f_row1_col6\" class=\"data row1 col6\" >19.565548</td>\n",
       "      <td id=\"T_c5e8f_row1_col7\" class=\"data row1 col7\" >ELF4;IL18;ICOSLG;ACVR2A;TGFB2;ST8SIA4;SRGN;ITGB2;INHBB;RPL39;TLR6;FLNA;HLA-A;IL11;ICAM1;INHBA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_c5e8f_row2_col0\" class=\"data row2 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row2_col1\" class=\"data row2 col1\" >HALLMARK_ANDROGEN_RESPONSE</td>\n",
       "      <td id=\"T_c5e8f_row2_col2\" class=\"data row2 col2\" >11/90</td>\n",
       "      <td id=\"T_c5e8f_row2_col3\" class=\"data row2 col3\" >0.019859</td>\n",
       "      <td id=\"T_c5e8f_row2_col4\" class=\"data row2 col4\" >0.050171</td>\n",
       "      <td id=\"T_c5e8f_row2_col5\" class=\"data row2 col5\" >2.256835</td>\n",
       "      <td id=\"T_c5e8f_row2_col6\" class=\"data row2 col6\" >8.844727</td>\n",
       "      <td id=\"T_c5e8f_row2_col7\" class=\"data row2 col7\" >NDRG1;SAT1;KRT8;SPDEF;DHCR24;INSIG1;TSC22D1;KRT19;ADAMTS1;SCD;PMEPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_c5e8f_row3_col0\" class=\"data row3 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row3_col1\" class=\"data row3 col1\" >HALLMARK_ANGIOGENESIS</td>\n",
       "      <td id=\"T_c5e8f_row3_col2\" class=\"data row3 col2\" >4/26</td>\n",
       "      <td id=\"T_c5e8f_row3_col3\" class=\"data row3 col3\" >0.069259</td>\n",
       "      <td id=\"T_c5e8f_row3_col4\" class=\"data row3 col4\" >0.127863</td>\n",
       "      <td id=\"T_c5e8f_row3_col5\" class=\"data row3 col5\" >3.107608</td>\n",
       "      <td id=\"T_c5e8f_row3_col6\" class=\"data row3 col6\" >8.296997</td>\n",
       "      <td id=\"T_c5e8f_row3_col7\" class=\"data row3 col7\" >THBD;SLCO2A1;VTN;FSTL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_c5e8f_row4_col0\" class=\"data row4 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row4_col1\" class=\"data row4 col1\" >HALLMARK_APICAL_JUNCTION</td>\n",
       "      <td id=\"T_c5e8f_row4_col2\" class=\"data row4 col2\" >19/152</td>\n",
       "      <td id=\"T_c5e8f_row4_col3\" class=\"data row4 col3\" >0.002105</td>\n",
       "      <td id=\"T_c5e8f_row4_col4\" class=\"data row4 col4\" >0.006735</td>\n",
       "      <td id=\"T_c5e8f_row4_col5\" class=\"data row4 col5\" >2.291710</td>\n",
       "      <td id=\"T_c5e8f_row4_col6\" class=\"data row4 col6\" >14.125159</td>\n",
       "      <td id=\"T_c5e8f_row4_col7\" class=\"data row4 col7\" >CX3CL1;CDH4;ICAM5;IKBKG;CLDN7;LIMA1;CALB2;COL17A1;ICAM1;CDH1;VCAM1;ITGA2;MYH10;CDH11;GRB7;FSCN1;PPP2R2C;INSIG1;ARHGEF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row5\" class=\"row_heading level0 row5\" >5</th>\n",
       "      <td id=\"T_c5e8f_row5_col0\" class=\"data row5 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row5_col1\" class=\"data row5 col1\" >HALLMARK_APICAL_SURFACE</td>\n",
       "      <td id=\"T_c5e8f_row5_col2\" class=\"data row5 col2\" >9/27</td>\n",
       "      <td id=\"T_c5e8f_row5_col3\" class=\"data row5 col3\" >0.000018</td>\n",
       "      <td id=\"T_c5e8f_row5_col4\" class=\"data row5 col4\" >0.000186</td>\n",
       "      <td id=\"T_c5e8f_row5_col5\" class=\"data row5 col5\" >8.030370</td>\n",
       "      <td id=\"T_c5e8f_row5_col6\" class=\"data row5 col6\" >87.567233</td>\n",
       "      <td id=\"T_c5e8f_row5_col7\" class=\"data row5 col7\" >AKAP7;PLAUR;MDGA1;CX3CL1;GATA3;SRPX;ATP8B1;AFAP1L2;PCSK9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row6\" class=\"row_heading level0 row6\" >6</th>\n",
       "      <td id=\"T_c5e8f_row6_col0\" class=\"data row6 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row6_col1\" class=\"data row6 col1\" >HALLMARK_APOPTOSIS</td>\n",
       "      <td id=\"T_c5e8f_row6_col2\" class=\"data row6 col2\" >15/139</td>\n",
       "      <td id=\"T_c5e8f_row6_col3\" class=\"data row6 col3\" >0.021250</td>\n",
       "      <td id=\"T_c5e8f_row6_col4\" class=\"data row6 col4\" >0.051001</td>\n",
       "      <td id=\"T_c5e8f_row6_col5\" class=\"data row6 col5\" >1.945041</td>\n",
       "      <td id=\"T_c5e8f_row6_col6\" class=\"data row6 col6\" >7.491099</td>\n",
       "      <td id=\"T_c5e8f_row6_col7\" class=\"data row6 col7\" >BIK;IL18;TNFSF10;GSN;TGFB2;SAT1;LEF1;PDGFRB;ERBB3;KRT18;TIMP3;EMP1;SATB1;BCAP31;CD14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row7\" class=\"row_heading level0 row7\" >7</th>\n",
       "      <td id=\"T_c5e8f_row7_col0\" class=\"data row7 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row7_col1\" class=\"data row7 col1\" >HALLMARK_BILE_ACID_METABOLISM</td>\n",
       "      <td id=\"T_c5e8f_row7_col2\" class=\"data row7 col2\" >7/71</td>\n",
       "      <td id=\"T_c5e8f_row7_col3\" class=\"data row7 col3\" >0.137655</td>\n",
       "      <td id=\"T_c5e8f_row7_col4\" class=\"data row7 col4\" >0.227843</td>\n",
       "      <td id=\"T_c5e8f_row7_col5\" class=\"data row7 col5\" >1.807404</td>\n",
       "      <td id=\"T_c5e8f_row7_col6\" class=\"data row7 col6\" >3.584088</td>\n",
       "      <td id=\"T_c5e8f_row7_col7\" class=\"data row7 col7\" >HSD3B1;DIO2;FADS2;CH25H;DHCR24;PXMP2;PHYH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row8\" class=\"row_heading level0 row8\" >8</th>\n",
       "      <td id=\"T_c5e8f_row8_col0\" class=\"data row8 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row8_col1\" class=\"data row8 col1\" >HALLMARK_CHOLESTEROL_HOMEOSTASIS</td>\n",
       "      <td id=\"T_c5e8f_row8_col2\" class=\"data row8 col2\" >11/69</td>\n",
       "      <td id=\"T_c5e8f_row8_col3\" class=\"data row8 col3\" >0.002738</td>\n",
       "      <td id=\"T_c5e8f_row8_col4\" class=\"data row8 col4\" >0.008214</td>\n",
       "      <td id=\"T_c5e8f_row8_col5\" class=\"data row8 col5\" >3.072049</td>\n",
       "      <td id=\"T_c5e8f_row8_col6\" class=\"data row8 col6\" >18.126713</td>\n",
       "      <td id=\"T_c5e8f_row8_col7\" class=\"data row8 col7\" >PLAUR;FADS2;GSTM2;ALDOC;LSS;MAL2;NSDHL;TRIB3;SCD;DHCR7;ANTXR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row9\" class=\"row_heading level0 row9\" >9</th>\n",
       "      <td id=\"T_c5e8f_row9_col0\" class=\"data row9 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row9_col1\" class=\"data row9 col1\" >HALLMARK_COAGULATION</td>\n",
       "      <td id=\"T_c5e8f_row9_col2\" class=\"data row9 col2\" >13/84</td>\n",
       "      <td id=\"T_c5e8f_row9_col3\" class=\"data row9 col3\" >0.001550</td>\n",
       "      <td id=\"T_c5e8f_row9_col4\" class=\"data row9 col4\" >0.005315</td>\n",
       "      <td id=\"T_c5e8f_row9_col5\" class=\"data row9 col5\" >2.954866</td>\n",
       "      <td id=\"T_c5e8f_row9_col6\" class=\"data row9 col6\" >19.116250</td>\n",
       "      <td id=\"T_c5e8f_row9_col7\" class=\"data row9 col7\" >F3;THBS1;GSN;ITGB3;ITGA2;CRIP2;F2RL2;THBD;F8;SERPINA1;CFH;TIMP3;FYN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row10\" class=\"row_heading level0 row10\" >10</th>\n",
       "      <td id=\"T_c5e8f_row10_col0\" class=\"data row10 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row10_col1\" class=\"data row10 col1\" >HALLMARK_COMPLEMENT</td>\n",
       "      <td id=\"T_c5e8f_row10_col2\" class=\"data row10 col2\" >14/141</td>\n",
       "      <td id=\"T_c5e8f_row10_col3\" class=\"data row10 col3\" >0.046824</td>\n",
       "      <td id=\"T_c5e8f_row10_col4\" class=\"data row10 col4\" >0.102162</td>\n",
       "      <td id=\"T_c5e8f_row10_col5\" class=\"data row10 col5\" >1.774132</td>\n",
       "      <td id=\"T_c5e8f_row10_col6\" class=\"data row10 col6\" >5.431244</td>\n",
       "      <td id=\"T_c5e8f_row10_col7\" class=\"data row10 col7\" >F3;KYNU;PLAUR;PRSS3;PCSK9;C4BPB;GATA3;CXCL1;CTSL;CDA;F8;SERPINA1;CFH;FYN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row11\" class=\"row_heading level0 row11\" >11</th>\n",
       "      <td id=\"T_c5e8f_row11_col0\" class=\"data row11 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row11_col1\" class=\"data row11 col1\" >HALLMARK_DNA_REPAIR</td>\n",
       "      <td id=\"T_c5e8f_row11_col2\" class=\"data row11 col2\" >3/145</td>\n",
       "      <td id=\"T_c5e8f_row11_col3\" class=\"data row11 col3\" >0.993973</td>\n",
       "      <td id=\"T_c5e8f_row11_col4\" class=\"data row11 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row11_col5\" class=\"data row11 col5\" >0.377589</td>\n",
       "      <td id=\"T_c5e8f_row11_col6\" class=\"data row11 col6\" >0.002283</td>\n",
       "      <td id=\"T_c5e8f_row11_col7\" class=\"data row11 col7\" >UPF3B;BCAP31;CDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row12\" class=\"row_heading level0 row12\" >12</th>\n",
       "      <td id=\"T_c5e8f_row12_col0\" class=\"data row12 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row12_col1\" class=\"data row12 col1\" >HALLMARK_E2F_TARGETS</td>\n",
       "      <td id=\"T_c5e8f_row12_col2\" class=\"data row12 col2\" >5/199</td>\n",
       "      <td id=\"T_c5e8f_row12_col3\" class=\"data row12 col3\" >0.994177</td>\n",
       "      <td id=\"T_c5e8f_row12_col4\" class=\"data row12 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row12_col5\" class=\"data row12 col5\" >0.433987</td>\n",
       "      <td id=\"T_c5e8f_row12_col6\" class=\"data row12 col6\" >0.002534</td>\n",
       "      <td id=\"T_c5e8f_row12_col7\" class=\"data row12 col7\" >HMGB3;CDC20;MYBL2;CDCA8;PRPS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row13\" class=\"row_heading level0 row13\" >13</th>\n",
       "      <td id=\"T_c5e8f_row13_col0\" class=\"data row13 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row13_col1\" class=\"data row13 col1\" >HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION</td>\n",
       "      <td id=\"T_c5e8f_row13_col2\" class=\"data row13 col2\" >32/157</td>\n",
       "      <td id=\"T_c5e8f_row13_col3\" class=\"data row13 col3\" >0.000000</td>\n",
       "      <td id=\"T_c5e8f_row13_col4\" class=\"data row13 col4\" >0.000000</td>\n",
       "      <td id=\"T_c5e8f_row13_col5\" class=\"data row13 col5\" >4.132167</td>\n",
       "      <td id=\"T_c5e8f_row13_col6\" class=\"data row13 col6\" >85.615963</td>\n",
       "      <td id=\"T_c5e8f_row13_col7\" class=\"data row13 col7\" >ABI3BP;COL12A1;OXTR;CXCL1;PDGFRB;FZD8;ITGB3;SERPINE2;MEST;MATN2;SAT1;ACTA2;TNC;FLNA;TIMP3;THBS2;MYLK;COL6A2;NNMT;COL1A2;GJA1;INHBA;PMEPA1;DKK1;THBS1;VCAM1;PLAUR;FSTL1;ITGA2;WNT5A;ADAM12;CDH11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row14\" class=\"row_heading level0 row14\" >14</th>\n",
       "      <td id=\"T_c5e8f_row14_col0\" class=\"data row14 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row14_col1\" class=\"data row14 col1\" >HALLMARK_ESTROGEN_RESPONSE_EARLY</td>\n",
       "      <td id=\"T_c5e8f_row14_col2\" class=\"data row14 col2\" >30/162</td>\n",
       "      <td id=\"T_c5e8f_row14_col3\" class=\"data row14 col3\" >0.000000</td>\n",
       "      <td id=\"T_c5e8f_row14_col4\" class=\"data row14 col4\" >0.000001</td>\n",
       "      <td id=\"T_c5e8f_row14_col5\" class=\"data row14 col5\" >3.661751</td>\n",
       "      <td id=\"T_c5e8f_row14_col6\" class=\"data row14 col6\" >62.946727</td>\n",
       "      <td id=\"T_c5e8f_row14_col7\" class=\"data row14 col7\" >ELF3;ESRP2;UGCG;FHL2;SCNN1A;SYT12;CELSR1;CALCR;RAB17;TIAM1;RETREG1;SLC1A1;DHRS3;CLDN7;RPS6KA2;TPD52L1;CALB2;KRT18;PTGES;PAPSS2;CLIC3;KRT8;MUC1;KAZN;INHBB;GJA1;KRT19;SLC2A1;DHCR7;MAPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row15\" class=\"row_heading level0 row15\" >15</th>\n",
       "      <td id=\"T_c5e8f_row15_col0\" class=\"data row15 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row15_col1\" class=\"data row15 col1\" >HALLMARK_ESTROGEN_RESPONSE_LATE</td>\n",
       "      <td id=\"T_c5e8f_row15_col2\" class=\"data row15 col2\" >23/155</td>\n",
       "      <td id=\"T_c5e8f_row15_col3\" class=\"data row15 col3\" >0.000059</td>\n",
       "      <td id=\"T_c5e8f_row15_col4\" class=\"data row15 col4\" >0.000428</td>\n",
       "      <td id=\"T_c5e8f_row15_col5\" class=\"data row15 col5\" >2.796739</td>\n",
       "      <td id=\"T_c5e8f_row15_col6\" class=\"data row15 col6\" >27.255656</td>\n",
       "      <td id=\"T_c5e8f_row15_col7\" class=\"data row15 col7\" >SCNN1A;SERPINA3;CKB;CALCR;TIAM1;MEST;CDC20;RPS6KA2;TPD52L1;TSPAN13;PTGES;PAPSS2;LLGL2;CLIC3;GPER1;CDH1;TRIM29;KRT19;SERPINA1;AGR2;DHCR7;PERP;MAPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row16\" class=\"row_heading level0 row16\" >16</th>\n",
       "      <td id=\"T_c5e8f_row16_col0\" class=\"data row16 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row16_col1\" class=\"data row16 col1\" >HALLMARK_FATTY_ACID_METABOLISM</td>\n",
       "      <td id=\"T_c5e8f_row16_col2\" class=\"data row16 col2\" >7/128</td>\n",
       "      <td id=\"T_c5e8f_row16_col3\" class=\"data row16 col3\" >0.664444</td>\n",
       "      <td id=\"T_c5e8f_row16_col4\" class=\"data row16 col4\" >0.885926</td>\n",
       "      <td id=\"T_c5e8f_row16_col5\" class=\"data row16 col5\" >0.955188</td>\n",
       "      <td id=\"T_c5e8f_row16_col6\" class=\"data row16 col6\" >0.390485</td>\n",
       "      <td id=\"T_c5e8f_row16_col7\" class=\"data row16 col7\" >IDH3G;VNN1;TP53INP2;CYP1A1;DHCR24;MGLL;NSDHL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row17\" class=\"row_heading level0 row17\" >17</th>\n",
       "      <td id=\"T_c5e8f_row17_col0\" class=\"data row17 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row17_col1\" class=\"data row17 col1\" >HALLMARK_G2M_CHECKPOINT</td>\n",
       "      <td id=\"T_c5e8f_row17_col2\" class=\"data row17 col2\" >5/196</td>\n",
       "      <td id=\"T_c5e8f_row17_col3\" class=\"data row17 col3\" >0.993343</td>\n",
       "      <td id=\"T_c5e8f_row17_col4\" class=\"data row17 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row17_col5\" class=\"data row17 col5\" >0.440890</td>\n",
       "      <td id=\"T_c5e8f_row17_col6\" class=\"data row17 col6\" >0.002945</td>\n",
       "      <td id=\"T_c5e8f_row17_col7\" class=\"data row17 col7\" >H2BC12;CDC20;SAP30;HMGB3;MYBL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row18\" class=\"row_heading level0 row18\" >18</th>\n",
       "      <td id=\"T_c5e8f_row18_col0\" class=\"data row18 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row18_col1\" class=\"data row18 col1\" >HALLMARK_GLYCOLYSIS</td>\n",
       "      <td id=\"T_c5e8f_row18_col2\" class=\"data row18 col2\" >17/176</td>\n",
       "      <td id=\"T_c5e8f_row18_col3\" class=\"data row18 col3\" >0.038519</td>\n",
       "      <td id=\"T_c5e8f_row18_col4\" class=\"data row18 col4\" >0.088043</td>\n",
       "      <td id=\"T_c5e8f_row18_col5\" class=\"data row18 col5\" >1.713621</td>\n",
       "      <td id=\"T_c5e8f_row18_col6\" class=\"data row18 col6\" >5.580600</td>\n",
       "      <td id=\"T_c5e8f_row18_col7\" class=\"data row18 col7\" >BIK;ARTN;NDST3;ANGPTL4;TGFA;SAP30;ELF3;CHST4;G6PD;CYB5A;PC;FUT8;PRPS1;EGLN3;VLDLR;NSDHL;PPFIA4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row19\" class=\"row_heading level0 row19\" >19</th>\n",
       "      <td id=\"T_c5e8f_row19_col0\" class=\"data row19 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row19_col1\" class=\"data row19 col1\" >HALLMARK_HEDGEHOG_SIGNALING</td>\n",
       "      <td id=\"T_c5e8f_row19_col2\" class=\"data row19 col2\" >8/27</td>\n",
       "      <td id=\"T_c5e8f_row19_col3\" class=\"data row19 col3\" >0.000139</td>\n",
       "      <td id=\"T_c5e8f_row19_col4\" class=\"data row19 col4\" >0.000834</td>\n",
       "      <td id=\"T_c5e8f_row19_col5\" class=\"data row19 col5\" >6.807732</td>\n",
       "      <td id=\"T_c5e8f_row19_col6\" class=\"data row19 col6\" >60.458321</td>\n",
       "      <td id=\"T_c5e8f_row19_col7\" class=\"data row19 col7\" >ETS2;NRCAM;NRP2;CELSR1;ADGRG1;VLDLR;AMOT;SHH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row20\" class=\"row_heading level0 row20\" >20</th>\n",
       "      <td id=\"T_c5e8f_row20_col0\" class=\"data row20 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row20_col1\" class=\"data row20 col1\" >HALLMARK_HEME_METABOLISM</td>\n",
       "      <td id=\"T_c5e8f_row20_col2\" class=\"data row20 col2\" >11/164</td>\n",
       "      <td id=\"T_c5e8f_row20_col3\" class=\"data row20 col3\" >0.409142</td>\n",
       "      <td id=\"T_c5e8f_row20_col4\" class=\"data row20 col4\" >0.595116</td>\n",
       "      <td id=\"T_c5e8f_row20_col5\" class=\"data row20 col5\" >1.162043</td>\n",
       "      <td id=\"T_c5e8f_row20_col6\" class=\"data row20 col6\" >1.038509</td>\n",
       "      <td id=\"T_c5e8f_row20_col7\" class=\"data row20 col7\" >MPP1;TNS1;PDZK1IP1;SLC7A11;UCP2;PC;SELENBP1;RAP1GAP;FN3K;ABCG2;SLC2A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row21\" class=\"row_heading level0 row21\" >21</th>\n",
       "      <td id=\"T_c5e8f_row21_col0\" class=\"data row21 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row21_col1\" class=\"data row21 col1\" >HALLMARK_HYPOXIA</td>\n",
       "      <td id=\"T_c5e8f_row21_col2\" class=\"data row21 col2\" >20/169</td>\n",
       "      <td id=\"T_c5e8f_row21_col3\" class=\"data row21 col3\" >0.003125</td>\n",
       "      <td id=\"T_c5e8f_row21_col4\" class=\"data row21 col4\" >0.008823</td>\n",
       "      <td id=\"T_c5e8f_row21_col5\" class=\"data row21 col5\" >2.151336</td>\n",
       "      <td id=\"T_c5e8f_row21_col6\" class=\"data row21 col6\" >12.409633</td>\n",
       "      <td id=\"T_c5e8f_row21_col7\" class=\"data row21 col7\" >NDRG1;PGF;ALDOC;SRPX;F3;ADM;HS3ST1;SAP30;PLAC8;EDN2;PPFIA4;FBP1;TMEM45A;ANGPTL4;SELENBP1;ACKR3;SLC2A1;PLAUR;MT1E;VLDLR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row22\" class=\"row_heading level0 row22\" >22</th>\n",
       "      <td id=\"T_c5e8f_row22_col0\" class=\"data row22 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row22_col1\" class=\"data row22 col1\" >HALLMARK_IL2_STAT5_SIGNALING</td>\n",
       "      <td id=\"T_c5e8f_row22_col2\" class=\"data row22 col2\" >14/149</td>\n",
       "      <td id=\"T_c5e8f_row22_col3\" class=\"data row22 col3\" >0.068054</td>\n",
       "      <td id=\"T_c5e8f_row22_col4\" class=\"data row22 col4\" >0.127863</td>\n",
       "      <td id=\"T_c5e8f_row22_col5\" class=\"data row22 col5\" >1.668331</td>\n",
       "      <td id=\"T_c5e8f_row22_col6\" class=\"data row22 col6\" >4.483559</td>\n",
       "      <td id=\"T_c5e8f_row22_col7\" class=\"data row22 col7\" >DHRS3;TNFSF10;NDRG1;GBP4;BATF3;TNFRSF1B;F2RL2;MUC1;PRNP;SPRY4;IRF6;EMP1;MAP6;TIAM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row23\" class=\"row_heading level0 row23\" >23</th>\n",
       "      <td id=\"T_c5e8f_row23_col0\" class=\"data row23 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row23_col1\" class=\"data row23 col1\" >HALLMARK_IL6_JAK_STAT3_SIGNALING</td>\n",
       "      <td id=\"T_c5e8f_row23_col2\" class=\"data row23 col2\" >4/58</td>\n",
       "      <td id=\"T_c5e8f_row23_col3\" class=\"data row23 col3\" >0.470391</td>\n",
       "      <td id=\"T_c5e8f_row23_col4\" class=\"data row23 col4\" >0.664082</td>\n",
       "      <td id=\"T_c5e8f_row23_col5\" class=\"data row23 col5\" >1.279741</td>\n",
       "      <td id=\"T_c5e8f_row23_col6\" class=\"data row23 col6\" >0.965168</td>\n",
       "      <td id=\"T_c5e8f_row23_col7\" class=\"data row23 col7\" >ITGB3;TNFRSF1B;CXCL1;CD14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row24\" class=\"row_heading level0 row24\" >24</th>\n",
       "      <td id=\"T_c5e8f_row24_col0\" class=\"data row24 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row24_col1\" class=\"data row24 col1\" >HALLMARK_INFLAMMATORY_RESPONSE</td>\n",
       "      <td id=\"T_c5e8f_row24_col2\" class=\"data row24 col2\" >22/135</td>\n",
       "      <td id=\"T_c5e8f_row24_col3\" class=\"data row24 col3\" >0.000019</td>\n",
       "      <td id=\"T_c5e8f_row24_col4\" class=\"data row24 col4\" >0.000186</td>\n",
       "      <td id=\"T_c5e8f_row24_col5\" class=\"data row24 col5\" >3.126780</td>\n",
       "      <td id=\"T_c5e8f_row24_col6\" class=\"data row24 col6\" >33.928454</td>\n",
       "      <td id=\"T_c5e8f_row24_col7\" class=\"data row24 col7\" >ICOSLG;LAMP3;CX3CL1;SLC4A4;RNF144B;EDN1;F3;TNFSF10;NAMPT;ADM;ITGB3;ICAM1;IL18;BST2;TNFRSF1B;CMKLR1;TNFSF15;INHBA;CD14;ACVR2A;PLAUR;HAS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row25\" class=\"row_heading level0 row25\" >25</th>\n",
       "      <td id=\"T_c5e8f_row25_col0\" class=\"data row25 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row25_col1\" class=\"data row25 col1\" >HALLMARK_INTERFERON_ALPHA_RESPONSE</td>\n",
       "      <td id=\"T_c5e8f_row25_col2\" class=\"data row25 col2\" >14/93</td>\n",
       "      <td id=\"T_c5e8f_row25_col3\" class=\"data row25 col3\" >0.001369</td>\n",
       "      <td id=\"T_c5e8f_row25_col4\" class=\"data row25 col4\" >0.005239</td>\n",
       "      <td id=\"T_c5e8f_row25_col5\" class=\"data row25 col5\" >2.856094</td>\n",
       "      <td id=\"T_c5e8f_row25_col6\" class=\"data row25 col6\" >18.833073</td>\n",
       "      <td id=\"T_c5e8f_row25_col7\" class=\"data row25 col7\" >OASL;DDX60;HERC6;GBP4;LAMP3;BST2;RSAD2;IFI44;ISG15;SAMD9;IFIH1;IFIT2;IFI27;IFIT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row26\" class=\"row_heading level0 row26\" >26</th>\n",
       "      <td id=\"T_c5e8f_row26_col0\" class=\"data row26 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row26_col1\" class=\"data row26 col1\" >HALLMARK_INTERFERON_GAMMA_RESPONSE</td>\n",
       "      <td id=\"T_c5e8f_row26_col2\" class=\"data row26 col2\" >24/166</td>\n",
       "      <td id=\"T_c5e8f_row26_col3\" class=\"data row26 col3\" >0.000062</td>\n",
       "      <td id=\"T_c5e8f_row26_col4\" class=\"data row26 col4\" >0.000428</td>\n",
       "      <td id=\"T_c5e8f_row26_col5\" class=\"data row26 col5\" >2.712373</td>\n",
       "      <td id=\"T_c5e8f_row26_col6\" class=\"data row26 col6\" >26.260184</td>\n",
       "      <td id=\"T_c5e8f_row26_col7\" class=\"data row26 col7\" >OASL;DDX60;GBP4;ISG15;IFI27;CFH;TNFSF10;NAMPT;IFIH1;OAS2;HLA-A;ICAM1;BST2;RSAD2;IFI44;CMKLR1;IFIT2;HERC6;VCAM1;SLAMF7;ST8SIA4;TNFAIP2;IFIT1;IFIT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row27\" class=\"row_heading level0 row27\" >27</th>\n",
       "      <td id=\"T_c5e8f_row27_col0\" class=\"data row27 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row27_col1\" class=\"data row27 col1\" >HALLMARK_KRAS_SIGNALING_DN</td>\n",
       "      <td id=\"T_c5e8f_row27_col2\" class=\"data row27 col2\" >9/80</td>\n",
       "      <td id=\"T_c5e8f_row27_col3\" class=\"data row27 col3\" >0.052350</td>\n",
       "      <td id=\"T_c5e8f_row27_col4\" class=\"data row27 col4\" >0.109251</td>\n",
       "      <td id=\"T_c5e8f_row27_col5\" class=\"data row27 col5\" >2.069164</td>\n",
       "      <td id=\"T_c5e8f_row27_col6\" class=\"data row27 col6\" >6.103645</td>\n",
       "      <td id=\"T_c5e8f_row27_col7\" class=\"data row27 col7\" >TENM2;NR4A2;RSAD2;TGFB2;C5;SLC16A7;EDN2;TCF7L1;EDN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row28\" class=\"row_heading level0 row28\" >28</th>\n",
       "      <td id=\"T_c5e8f_row28_col0\" class=\"data row28 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row28_col1\" class=\"data row28 col1\" >HALLMARK_KRAS_SIGNALING_UP</td>\n",
       "      <td id=\"T_c5e8f_row28_col2\" class=\"data row28 col2\" >23/132</td>\n",
       "      <td id=\"T_c5e8f_row28_col3\" class=\"data row28 col3\" >0.000004</td>\n",
       "      <td id=\"T_c5e8f_row28_col4\" class=\"data row28 col4\" >0.000064</td>\n",
       "      <td id=\"T_c5e8f_row28_col5\" class=\"data row28 col5\" >3.390333</td>\n",
       "      <td id=\"T_c5e8f_row28_col6\" class=\"data row28 col6\" >42.145458</td>\n",
       "      <td id=\"T_c5e8f_row28_col7\" class=\"data row28 col7\" >KCNN4;ST6GAL1;ADAM8;SERPINA3;NGF;EMP1;SATB1;CFH;LCP1;TSPAN13;CAB39L;ANGPTL4;PLEK2;TNFRSF1B;CMKLR1;ITGB2;INHBA;PLAUR;GALNT3;GPRC5B;ITGA2;GFPT2;ETV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row29\" class=\"row_heading level0 row29\" >29</th>\n",
       "      <td id=\"T_c5e8f_row29_col0\" class=\"data row29 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row29_col1\" class=\"data row29 col1\" >HALLMARK_MITOTIC_SPINDLE</td>\n",
       "      <td id=\"T_c5e8f_row29_col2\" class=\"data row29 col2\" >7/196</td>\n",
       "      <td id=\"T_c5e8f_row29_col3\" class=\"data row29 col3\" >0.956918</td>\n",
       "      <td id=\"T_c5e8f_row29_col4\" class=\"data row29 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row29_col5\" class=\"data row29 col5\" >0.609141</td>\n",
       "      <td id=\"T_c5e8f_row29_col6\" class=\"data row29 col6\" >0.026825</td>\n",
       "      <td id=\"T_c5e8f_row29_col7\" class=\"data row29 col7\" >PCNT;GSN;MYH10;FSCN1;FLNA;TIAM1;SORBS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row30\" class=\"row_heading level0 row30\" >30</th>\n",
       "      <td id=\"T_c5e8f_row30_col0\" class=\"data row30 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row30_col1\" class=\"data row30 col1\" >HALLMARK_MTORC1_SIGNALING</td>\n",
       "      <td id=\"T_c5e8f_row30_col2\" class=\"data row30 col2\" >14/193</td>\n",
       "      <td id=\"T_c5e8f_row30_col3\" class=\"data row30 col3\" >0.281757</td>\n",
       "      <td id=\"T_c5e8f_row30_col4\" class=\"data row30 col4\" >0.422635</td>\n",
       "      <td id=\"T_c5e8f_row30_col5\" class=\"data row30 col5\" >1.255001</td>\n",
       "      <td id=\"T_c5e8f_row30_col6\" class=\"data row30 col6\" >1.589724</td>\n",
       "      <td id=\"T_c5e8f_row30_col7\" class=\"data row30 col7\" >PHGDH;NAMPT;FADS2;SLC7A11;G6PD;ITGB2;EGLN3;DHCR24;INSIG1;VLDLR;TRIB3;SCD;DHCR7;SLC2A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row31\" class=\"row_heading level0 row31\" >31</th>\n",
       "      <td id=\"T_c5e8f_row31_col0\" class=\"data row31 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row31_col1\" class=\"data row31 col1\" >HALLMARK_MYC_TARGETS_V1</td>\n",
       "      <td id=\"T_c5e8f_row31_col2\" class=\"data row31 col2\" >2/196</td>\n",
       "      <td id=\"T_c5e8f_row31_col3\" class=\"data row31 col3\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row31_col4\" class=\"data row31 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row31_col5\" class=\"data row31 col5\" >0.196548</td>\n",
       "      <td id=\"T_c5e8f_row31_col6\" class=\"data row31 col6\" >0.000012</td>\n",
       "      <td id=\"T_c5e8f_row31_col7\" class=\"data row31 col7\" >CDC20;VBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row32\" class=\"row_heading level0 row32\" >32</th>\n",
       "      <td id=\"T_c5e8f_row32_col0\" class=\"data row32 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row32_col1\" class=\"data row32 col1\" >HALLMARK_MYOGENESIS</td>\n",
       "      <td id=\"T_c5e8f_row32_col2\" class=\"data row32 col2\" >17/125</td>\n",
       "      <td id=\"T_c5e8f_row32_col3\" class=\"data row32 col3\" >0.001419</td>\n",
       "      <td id=\"T_c5e8f_row32_col4\" class=\"data row32 col4\" >0.005239</td>\n",
       "      <td id=\"T_c5e8f_row32_col5\" class=\"data row32 col5\" >2.529276</td>\n",
       "      <td id=\"T_c5e8f_row32_col6\" class=\"data row32 col6\" >16.586601</td>\n",
       "      <td id=\"T_c5e8f_row32_col7\" class=\"data row32 col7\" >FHL1;EPHB3;GSN;MYLK;COL6A2;PC;SPDEF;ADAM12;ERBB3;VIPR1;FST;TPD52L1;PRNP;CKB;WWTR1;PPFIA4;SCD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row33\" class=\"row_heading level0 row33\" >33</th>\n",
       "      <td id=\"T_c5e8f_row33_col0\" class=\"data row33 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row33_col1\" class=\"data row33 col1\" >HALLMARK_NOTCH_SIGNALING</td>\n",
       "      <td id=\"T_c5e8f_row33_col2\" class=\"data row33 col2\" >4/29</td>\n",
       "      <td id=\"T_c5e8f_row33_col3\" class=\"data row33 col3\" >0.095627</td>\n",
       "      <td id=\"T_c5e8f_row33_col4\" class=\"data row33 col4\" >0.170003</td>\n",
       "      <td id=\"T_c5e8f_row33_col5\" class=\"data row33 col5\" >2.741363</td>\n",
       "      <td id=\"T_c5e8f_row33_col6\" class=\"data row33 col6\" >6.434814</td>\n",
       "      <td id=\"T_c5e8f_row33_col7\" class=\"data row33 col7\" >SAP30;FZD1;WNT5A;FZD7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row34\" class=\"row_heading level0 row34\" >34</th>\n",
       "      <td id=\"T_c5e8f_row34_col0\" class=\"data row34 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row34_col1\" class=\"data row34 col1\" >HALLMARK_OXIDATIVE_PHOSPHORYLATION</td>\n",
       "      <td id=\"T_c5e8f_row34_col2\" class=\"data row34 col2\" >5/200</td>\n",
       "      <td id=\"T_c5e8f_row34_col3\" class=\"data row34 col3\" >0.994432</td>\n",
       "      <td id=\"T_c5e8f_row34_col4\" class=\"data row34 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row34_col5\" class=\"data row34 col5\" >0.431733</td>\n",
       "      <td id=\"T_c5e8f_row34_col6\" class=\"data row34 col6\" >0.002410</td>\n",
       "      <td id=\"T_c5e8f_row34_col7\" class=\"data row34 col7\" >IDH3G;CYB5A;PDK4;SLC25A5;PHYH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row35\" class=\"row_heading level0 row35\" >35</th>\n",
       "      <td id=\"T_c5e8f_row35_col0\" class=\"data row35 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row35_col1\" class=\"data row35 col1\" >HALLMARK_P53_PATHWAY</td>\n",
       "      <td id=\"T_c5e8f_row35_col2\" class=\"data row35 col2\" >13/177</td>\n",
       "      <td id=\"T_c5e8f_row35_col3\" class=\"data row35 col3\" >0.276709</td>\n",
       "      <td id=\"T_c5e8f_row35_col4\" class=\"data row35 col4\" >0.422635</td>\n",
       "      <td id=\"T_c5e8f_row35_col5\" class=\"data row35 col5\" >1.274940</td>\n",
       "      <td id=\"T_c5e8f_row35_col6\" class=\"data row35 col6\" >1.638029</td>\n",
       "      <td id=\"T_c5e8f_row35_col7\" class=\"data row35 col7\" >NDRG1;TGFA;SAT1;SLC7A11;PLK2;TPD52L1;INHBB;PRMT2;TSC22D1;TRIB3;PERP;H2AJ;IRAK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row36\" class=\"row_heading level0 row36\" >36</th>\n",
       "      <td id=\"T_c5e8f_row36_col0\" class=\"data row36 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row36_col1\" class=\"data row36 col1\" >HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "      <td id=\"T_c5e8f_row36_col2\" class=\"data row36 col2\" >3/15</td>\n",
       "      <td id=\"T_c5e8f_row36_col3\" class=\"data row36 col3\" >0.058475</td>\n",
       "      <td id=\"T_c5e8f_row36_col4\" class=\"data row36 col4\" >0.116950</td>\n",
       "      <td id=\"T_c5e8f_row36_col5\" class=\"data row36 col5\" >4.348766</td>\n",
       "      <td id=\"T_c5e8f_row36_col6\" class=\"data row36 col6\" >12.346818</td>\n",
       "      <td id=\"T_c5e8f_row36_col7\" class=\"data row36 col7\" >FOXA2;SRPRB;FOXO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row37\" class=\"row_heading level0 row37\" >37</th>\n",
       "      <td id=\"T_c5e8f_row37_col0\" class=\"data row37 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row37_col1\" class=\"data row37 col1\" >HALLMARK_PEROXISOME</td>\n",
       "      <td id=\"T_c5e8f_row37_col2\" class=\"data row37 col2\" >2/84</td>\n",
       "      <td id=\"T_c5e8f_row37_col3\" class=\"data row37 col3\" >0.966759</td>\n",
       "      <td id=\"T_c5e8f_row37_col4\" class=\"data row37 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row37_col5\" class=\"data row37 col5\" >0.467498</td>\n",
       "      <td id=\"T_c5e8f_row37_col6\" class=\"data row37 col6\" >0.015804</td>\n",
       "      <td id=\"T_c5e8f_row37_col7\" class=\"data row37 col7\" >DHRS3;DHCR24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row38\" class=\"row_heading level0 row38\" >38</th>\n",
       "      <td id=\"T_c5e8f_row38_col0\" class=\"data row38 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row38_col1\" class=\"data row38 col1\" >HALLMARK_PI3K_AKT_MTOR_SIGNALING</td>\n",
       "      <td id=\"T_c5e8f_row38_col2\" class=\"data row38 col2\" >5/89</td>\n",
       "      <td id=\"T_c5e8f_row38_col3\" class=\"data row38 col3\" >0.633543</td>\n",
       "      <td id=\"T_c5e8f_row38_col4\" class=\"data row38 col4\" >0.868860</td>\n",
       "      <td id=\"T_c5e8f_row38_col5\" class=\"data row38 col5\" >1.007666</td>\n",
       "      <td id=\"T_c5e8f_row38_col6\" class=\"data row38 col6\" >0.459926</td>\n",
       "      <td id=\"T_c5e8f_row38_col7\" class=\"data row38 col7\" >TIAM1;CAB39L;NGF;TRIB3;SLC2A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row39\" class=\"row_heading level0 row39\" >39</th>\n",
       "      <td id=\"T_c5e8f_row39_col0\" class=\"data row39 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row39_col1\" class=\"data row39 col1\" >HALLMARK_PROTEIN_SECRETION</td>\n",
       "      <td id=\"T_c5e8f_row39_col2\" class=\"data row39 col2\" >2/95</td>\n",
       "      <td id=\"T_c5e8f_row39_col3\" class=\"data row39 col3\" >0.981514</td>\n",
       "      <td id=\"T_c5e8f_row39_col4\" class=\"data row39 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row39_col5\" class=\"data row39 col5\" >0.412141</td>\n",
       "      <td id=\"T_c5e8f_row39_col6\" class=\"data row39 col6\" >0.007690</td>\n",
       "      <td id=\"T_c5e8f_row39_col7\" class=\"data row39 col7\" >KRT18;BNIP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row40\" class=\"row_heading level0 row40\" >40</th>\n",
       "      <td id=\"T_c5e8f_row40_col0\" class=\"data row40 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row40_col1\" class=\"data row40 col1\" >HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY</td>\n",
       "      <td id=\"T_c5e8f_row40_col2\" class=\"data row40 col2\" >2/44</td>\n",
       "      <td id=\"T_c5e8f_row40_col3\" class=\"data row40 col3\" >0.755365</td>\n",
       "      <td id=\"T_c5e8f_row40_col4\" class=\"data row40 col4\" >0.972599</td>\n",
       "      <td id=\"T_c5e8f_row40_col5\" class=\"data row40 col5\" >0.910354</td>\n",
       "      <td id=\"T_c5e8f_row40_col6\" class=\"data row40 col6\" >0.255404</td>\n",
       "      <td id=\"T_c5e8f_row40_col7\" class=\"data row40 col7\" >PRNP;G6PD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row41\" class=\"row_heading level0 row41\" >41</th>\n",
       "      <td id=\"T_c5e8f_row41_col0\" class=\"data row41 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row41_col1\" class=\"data row41 col1\" >HALLMARK_TGF_BETA_SIGNALING</td>\n",
       "      <td id=\"T_c5e8f_row41_col2\" class=\"data row41 col2\" >5/51</td>\n",
       "      <td id=\"T_c5e8f_row41_col3\" class=\"data row41 col3\" >0.194927</td>\n",
       "      <td id=\"T_c5e8f_row41_col4\" class=\"data row41 col4\" >0.311883</td>\n",
       "      <td id=\"T_c5e8f_row41_col5\" class=\"data row41 col5\" >1.836613</td>\n",
       "      <td id=\"T_c5e8f_row41_col6\" class=\"data row41 col6\" >3.003102</td>\n",
       "      <td id=\"T_c5e8f_row41_col7\" class=\"data row41 col7\" >THBS1;LTBP2;CDH1;WWTR1;PMEPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row42\" class=\"row_heading level0 row42\" >42</th>\n",
       "      <td id=\"T_c5e8f_row42_col0\" class=\"data row42 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row42_col1\" class=\"data row42 col1\" >HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "      <td id=\"T_c5e8f_row42_col2\" class=\"data row42 col2\" >23/180</td>\n",
       "      <td id=\"T_c5e8f_row42_col3\" class=\"data row42 col3\" >0.000556</td>\n",
       "      <td id=\"T_c5e8f_row42_col4\" class=\"data row42 col4\" >0.002966</td>\n",
       "      <td id=\"T_c5e8f_row42_col5\" class=\"data row42 col5\" >2.348164</td>\n",
       "      <td id=\"T_c5e8f_row42_col6\" class=\"data row42 col6\" >17.598221</td>\n",
       "      <td id=\"T_c5e8f_row42_col7\" class=\"data row42 col7\" >ICOSLG;NR4A2;CXCL1;EDN1;F3;NAMPT;SAT1;TNC;IFIH1;PLPP3;ICAM1;IL18;TSC22D1;ACKR3;IFIT2;INHBA;PMEPA1;ETS2;KYNU;PLAUR;PLK2;GFPT2;TNFAIP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row43\" class=\"row_heading level0 row43\" >43</th>\n",
       "      <td id=\"T_c5e8f_row43_col0\" class=\"data row43 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row43_col1\" class=\"data row43 col1\" >HALLMARK_UNFOLDED_PROTEIN_RESPONSE</td>\n",
       "      <td id=\"T_c5e8f_row43_col2\" class=\"data row43 col2\" >2/109</td>\n",
       "      <td id=\"T_c5e8f_row43_col3\" class=\"data row43 col3\" >0.991370</td>\n",
       "      <td id=\"T_c5e8f_row43_col4\" class=\"data row43 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row43_col5\" class=\"data row43 col5\" >0.358072</td>\n",
       "      <td id=\"T_c5e8f_row43_col6\" class=\"data row43 col6\" >0.003103</td>\n",
       "      <td id=\"T_c5e8f_row43_col7\" class=\"data row43 col7\" >IFIT1;SRPRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row44\" class=\"row_heading level0 row44\" >44</th>\n",
       "      <td id=\"T_c5e8f_row44_col0\" class=\"data row44 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row44_col1\" class=\"data row44 col1\" >HALLMARK_UV_RESPONSE_DN</td>\n",
       "      <td id=\"T_c5e8f_row44_col2\" class=\"data row44 col2\" >16/133</td>\n",
       "      <td id=\"T_c5e8f_row44_col3\" class=\"data row44 col3\" >0.006614</td>\n",
       "      <td id=\"T_c5e8f_row44_col4\" class=\"data row44 col4\" >0.017638</td>\n",
       "      <td id=\"T_c5e8f_row44_col5\" class=\"data row44 col5\" >2.197803</td>\n",
       "      <td id=\"T_c5e8f_row44_col6\" class=\"data row44 col6\" >11.029745</td>\n",
       "      <td id=\"T_c5e8f_row44_col7\" class=\"data row44 col7\" >F3;ACVR2A;ITGB3;GJA1;EFEMP1;MT1E;FHL2;COL1A2;VLDLR;INSIG1;MGLL;PLPP3;LTBP1;PDGFRB;HAS2;FYN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row45\" class=\"row_heading level0 row45\" >45</th>\n",
       "      <td id=\"T_c5e8f_row45_col0\" class=\"data row45 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row45_col1\" class=\"data row45 col1\" >HALLMARK_UV_RESPONSE_UP</td>\n",
       "      <td id=\"T_c5e8f_row45_col2\" class=\"data row45 col2\" >4/132</td>\n",
       "      <td id=\"T_c5e8f_row45_col3\" class=\"data row45 col3\" >0.962605</td>\n",
       "      <td id=\"T_c5e8f_row45_col4\" class=\"data row45 col4\" >0.999940</td>\n",
       "      <td id=\"T_c5e8f_row45_col5\" class=\"data row45 col5\" >0.539615</td>\n",
       "      <td id=\"T_c5e8f_row45_col6\" class=\"data row45 col6\" >0.020566</td>\n",
       "      <td id=\"T_c5e8f_row45_col7\" class=\"data row45 col7\" >C4BPB;CYP1A1;HTR7;ICAM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row46\" class=\"row_heading level0 row46\" >46</th>\n",
       "      <td id=\"T_c5e8f_row46_col0\" class=\"data row46 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row46_col1\" class=\"data row46 col1\" >HALLMARK_WNT_BETA_CATENIN_SIGNALING</td>\n",
       "      <td id=\"T_c5e8f_row46_col2\" class=\"data row46 col2\" >4/33</td>\n",
       "      <td id=\"T_c5e8f_row46_col3\" class=\"data row46 col3\" >0.136896</td>\n",
       "      <td id=\"T_c5e8f_row46_col4\" class=\"data row46 col4\" >0.227843</td>\n",
       "      <td id=\"T_c5e8f_row46_col5\" class=\"data row46 col5\" >2.368910</td>\n",
       "      <td id=\"T_c5e8f_row46_col6\" class=\"data row46 col6\" >4.710649</td>\n",
       "      <td id=\"T_c5e8f_row46_col7\" class=\"data row46 col7\" >LEF1;FZD1;DKK1;FZD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c5e8f_level0_row47\" class=\"row_heading level0 row47\" >47</th>\n",
       "      <td id=\"T_c5e8f_row47_col0\" class=\"data row47 col0\" >MSigDB Hallmark</td>\n",
       "      <td id=\"T_c5e8f_row47_col1\" class=\"data row47 col1\" >HALLMARK_XENOBIOTIC_METABOLISM</td>\n",
       "      <td id=\"T_c5e8f_row47_col2\" class=\"data row47 col2\" >19/146</td>\n",
       "      <td id=\"T_c5e8f_row47_col3\" class=\"data row47 col3\" >0.001309</td>\n",
       "      <td id=\"T_c5e8f_row47_col4\" class=\"data row47 col4\" >0.005239</td>\n",
       "      <td id=\"T_c5e8f_row47_col5\" class=\"data row47 col5\" >2.400693</td>\n",
       "      <td id=\"T_c5e8f_row47_col6\" class=\"data row47 col6\" >15.936955</td>\n",
       "      <td id=\"T_c5e8f_row47_col7\" class=\"data row47 col7\" >CDA;ITIH4;AOX1;PDK4;CYP1A1;RAP1GAP;PTGES;PAPSS2;FBP1;MAN1A1;ATOH8;PC;VTN;ETS2;KYNU;TGFB2;VNN1;CYB5A;SPINT2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x799489a40710>"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ora.res2d.style\n",
    "\n",
    "#There is a detailed explanation of each column in the Markdown cell below\n",
    "#Pay particular attention to the confusing columns: Odds Ratio and the Combined Score\n",
    "#These are less self-explanatory, but the Enrichr webpage also provides an explanation that I elaborate on below with a figure to help visualize"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b449a345-b817-42cb-a3bb-aeb1e2870ecd",
   "metadata": {},
   "source": [
    "So what we have in our Enrichr object results table are the following:\n",
    "\n",
    "<ul>\n",
    "    <li><code>Gene_set</code> - this is the database that we're using for our analysis. If we provided more than one database for our analysis, we would have additional gene set values corresponding to the other databases.</li>\n",
    "    <li><code>Term</code> - this is the functional groupings of the different genes.</li>\n",
    "    <li><code>Overlap</code> - this is the quantification of the number of genes contained in our significant hits (or other specified threshold) that overlap with the genes contained within the different terms</li>\n",
    "    <li><code>P-value</code> - this is the p-value resulting from the hypergeometric test</li>\n",
    "    <li><code>Adjusted P-value</code> - this is the q-value resulting from the Benjamini-Hochberg correction, which is one way to correct for multiple hypothesis testing</li>\n",
    "    <li><code>Odds Ratio</code> - the odds ratio can be thought of as a quantification of the enrichment of your significant hits, so a greater Odds Ratio value indicates a greater enrichment of genes contained in that gene set among your hits. The odds ratio is calculated as follows:</li>\n",
    "</ul>\n",
    "\n",
    "```\n",
    "Odds Ratio = (1 * a * d) / (whichever is greater of the two: (1 * b * c) or 1)\n",
    "```\n",
    "\n",
    "<ul>\n",
    "    <p>Where the variables are the numbers of genes broken down in the figure below:</p>\n",
    "</ul>\n",
    "\n",
    "<img src=\"./images/15_1_fig_1.jpg\" style=\"height: 250px; margin: auto;\"/>\n",
    "\n",
    "<ul>\n",
    "    <p>So then if we see a reduced enrichment for example, we would expect to see a smaller odds ratio.</p>\n",
    "</ul>\n",
    "\n",
    "<img src=\"./images/15_1_fig_2.jpg\" style=\"height: 250px; margin: auto;\"/>\n",
    "\n",
    "<ul>\n",
    "    <li><code>Combined Score</code> - a calculation that takes into account the <code>P-value</code> (not the <code>Adjusted P-value</code>) and the <code>Odds Ratio</code></li>\n",
    "</ul>\n",
    "\n",
    "<ul>\n",
    "    <p>The combined score is calculated as:</p>\n",
    "</ul>\n",
    "\n",
    "```\n",
    "Combined Score = -log(P-value) * (Odds Ratio)\n",
    "```\n",
    "\n",
    "<ul>\n",
    "    <p>The combined score is a compromise between the p-value and the odds ratio, which has been shown by the Ma'ayan lab to be better for ranking the results of an over-representation analysis.</p>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eff39425-a5ac-4cf8-8a9d-caf86536b141",
   "metadata": {},
   "source": [
    "<h2>Save the results of your over-representation analysis</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7114d680-c345-48a9-b5c9-33c1bd6b9134",
   "metadata": {},
   "source": [
    "Let's go ahead and save the results table that we got from our analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "41a26d48-7c49-40ef-8db0-209d8b590d5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We can save the results of our functional analysis in a separate folder but continue to dump everything into Week 10\n",
    "try:\n",
    "    os.mkdir('../Week_10/functional-analysis')\n",
    "except FileExistsError:\n",
    "    pass\n",
    "\n",
    "ora.res2d.sort_values(by='Adjusted P-value').to_csv('../Week_10/functional-analysis/over-representation-analysis.csv',\\\n",
    "                                                    index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf810fb7-7384-4b50-aa9b-cecc6e2b7911",
   "metadata": {},
   "source": [
    "<h2>Visualize results of over-representation analysis</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f991b1f-44c3-4a3c-af01-67b1c0dab606",
   "metadata": {},
   "source": [
    "We can then take our results and plot them a few different ways to visualize the results of our functional analysis. If you ran your over-representation analysis using multiple databases, you also have the option to plot the results from the databases together."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bf79191-6f6a-4fea-ab5c-1b97a74a03c4",
   "metadata": {},
   "source": [
    "<h3>Plot a dot plot</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8b280d-645d-47f1-96a4-4dbe605f04f8",
   "metadata": {},
   "source": [
    "We can plot our results on a dot plot (categorical scatter plot) that allow us to visualize the different categories (<code>Term</code>) together on a single plot and arrange them based on the output result scores, such as combined score, adjusted p-value, etc. The dot plot allows us to visualize multiple result scores on a single plot depending on what parameters we provide it.\n",
    "\n",
    "To do this, we'll make use of GSEApy's <code>gp.dotplot()</code> function. <a href=\"https://gseapy.readthedocs.io/en/latest/run.html#gseapy.plot.dotplot\" rel=\"noopener noreferrer\"><u>Documentation for <code>gp.dotplot()</code> can be found here.</u></a> We'll provide it with the results of our analysis and specify how we want it to generate the plot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "305949f5-fa96-40e6-a99c-931c3fc967a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cd9fce4d-268a-408a-8cf1-c9ee4ab6ba32",
   "metadata": {},
   "source": [
    "<h2>Let's break down the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eafb617-8e6e-44fb-abb2-1bb331148a6d",
   "metadata": {},
   "source": [
    "<code>gp.dotplot()</code>\n",
    "\n",
    "We're calling up the function to generate dot plots from the GSEApy package. The dot plot will visualize the data contained in three columns of your DataFrame: % Genes in set (derived from <code>Overlap</code>), and two others that you can select through the <code>column</code> and <code>x</code> parameters.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>df=ora.res2d</code>\n",
    "\n",
    "We provide it with a DataFrame containing our results. In this case, we're providing it with the DataFrame containing the results of our most recent query.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>column='Adjusted P-value'</code>\n",
    "\n",
    "We are specifying how we want to fill the color for each dot. The default viridis color map will be used to color in the dots according to their adjusted p-value.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>x='Combined Score'</code>\n",
    "\n",
    "This is another set of data that we want to visualize in the same dot plot. This parameter affects where along the X-axis the dots are plotted with the location determined by the results column you specify. In this case, we will plot the dots along the X-axis based on their <code>Combined Score</code>.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>y='Term'</code>\n",
    "\n",
    "This sets our Y-axis, and here we're specifying that we want to set our Y-axis as a categorical Y axis corresponding to each group of genes.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>top_term=20</code>\n",
    "\n",
    "This argument sets the number of enriched terms that we will plot, and it will pull the top terms according to the values specified by your <code>column</code> parameter. So in our case, this will pull the top 10 most significant terms to generate a dot plot.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>cutoff=0.05</code>\n",
    "\n",
    "This argument sets what the cutoff significance threshold is for plotting. By setting this to parameter to <code>0.05</code>, we specify that our significantly enriched terms will be those that have an <code>Adjusted P-value</code> less than 0.05.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>size=15</code>\n",
    "\n",
    "This sets the size of our dots, and we can adjust this to better display our dots on our plots.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>figsize=(3,10)</code>\n",
    "\n",
    "This parameter sets the size of our output figure, and we can supply it with a tuple corresponding to the width and the height of our figure.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>title = 'MSigDB Hallmark'</code>\n",
    "\n",
    "We can also add a title to our plots through this parameter.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>xticklabels_rot=0</code>\n",
    "\n",
    "Here, we can adjust the rotation of our X-axis tick labels. This is helpful to adjust when plotting categorical values along the X-axis (such as comparing multiple databases at the same time) to prevent the labels from overlapping one another.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>show_ring=False</code>\n",
    "\n",
    "If we want, we can also plot a ring around our dots corresponding to the largest percent value in our plot to better compare size changes between dots. By setting this to <code>False</code>, we'll hide the rings and just plot the dots.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>marker='o'</code>\n",
    "\n",
    "This parameter is the same as what we've seen with matplotlib, and <code>gp.dotplot()</code> follows the same syntax, where we can specify that we want to plot dots by passing the argument <code>'o'</code>.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>ofname=None</code>\n",
    "\n",
    "If we want to instead save our output image generated by this function to a file, we can provide the name of the file as a string. This functions similarly to some of the visualization functions we've encountered in R, where we provided the file name as an argument to get an output file. \n",
    "\n",
    "To save the output visualization, we can then provide it with the file path and name.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cf063b6-92bb-4aa0-869e-c82d71da65e9",
   "metadata": {},
   "source": [
    "<h2>Plot a bar plot</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "633a1b41-de48-4a55-b680-6ec01a9a0f82",
   "metadata": {},
   "source": [
    "Another way of visualizing our results is to generate a bar plot, which is a more focused graph, where we simply plot one of the result scores. We'll make use of GSEApy's <code>gp.barplot()</code> function. <a href=\"https://gseapy.readthedocs.io/en/latest/run.html#gseapy.plot.barplot\" rel=\"noopener noreferrer\"><u>Documentation for <code>gp.barplot()</code> can be found here.</u></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a994c9f-bf69-47d1-8d4b-7d0f1dcd6c31",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c0fde4c8-9e43-4d5e-b5aa-9e47ed2f1ddd",
   "metadata": {},
   "source": [
    "The setup is very similar to the dot plot and in some ways a little simpler. The only difference here is that we can provide the function with a <code>color</code> string or dictionary that sets the fill color of the bars. If we have multiple databases being compared, we can provide it with a dictionary containing key:value pairs corresponding to the database name and the color, respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87944e67-767c-4690-92d2-a76443752cb1",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Gene Set Enrichment Analysis (GSEA)</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e18f98a-ecad-4d0b-931d-4cb050a91b03",
   "metadata": {},
   "source": [
    "Another common functional analysis approach is to perform a <a href=\"https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html\" rel=\"noopener noreferrer\"><u>Gene Set Enrichment Analysis (GSEA)</u></a>. A key difference between the over-representation analysis that we performed above and GSEA is that we will take our counts matrix to perform this analysis. So we are not setting any arbitrary cutoffs for significance or log2 fold change. We provide the analysis with our expression matrix, the database that we want it to use for analysis, and metadata contained within a .cls file.\n",
    "\n",
    "Our dataset will get ordered based on the normalized count matrix. Then the function will perform a walk down the list of genes to determine if there is an enrichment of genes within a gene set above what it would normally expect to encounter based on random chance, where random chance is determined by reshuffling (permutation) the order of the genes and repeating the walk up to a specified number of permutations. The result is that we identify which gene sets are enriched in either our control samples or our TAZ KO samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3773ffc3-9ce5-432a-800d-34184a324b97",
   "metadata": {},
   "source": [
    "<h2>Prepare for GSEA</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0b3ff79-7c54-4893-9659-84f404f00741",
   "metadata": {},
   "source": [
    "To perform our analysis, we'll first generate the \"metadata\" that will be parsed from a .cls file. The <code>gp.parser.gsea_cls_parser()</code> function will pull out the conditions and how they should be assigned to each column of your normalized counts matrix. We can generate this ourselves directly in this notebook. \n",
    "\n",
    "First, we can create variables that specify what our two sample groups are:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94e0006a-5332-41d0-8a61-7773cbf8ee7e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "708bc9f4-19f8-4f51-8d7c-cb4f43a32e90",
   "metadata": {},
   "source": [
    "Then we can put them together into a list that will be used to assign the groups to each column of our normalized count matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3d393d7-6e4f-4ef6-8ab5-d566cb3d8273",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e9b91dc9-5341-409a-b090-3c97dc43d22d",
   "metadata": {},
   "source": [
    "Let's take a look at the variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9475e2b-b390-4056-8b98-d2b36231cb81",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c0b2913b-8866-4eae-ba1f-1938015cfe5c",
   "metadata": {},
   "source": [
    "So this is much the same as what we did when we provided DESeq2 with a conditions matrix."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8faa7b0c-9ea8-4145-973a-c0461e7e54a8",
   "metadata": {},
   "source": [
    "<h2>Perform GSEA</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acab5e2c-f891-420b-a753-119200da823c",
   "metadata": {},
   "source": [
    "Now, we can set up and then run GSEA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24e60ae0-a39a-4af5-b0d0-d08e6bc28ddb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "21898e0d-fe53-4e5f-9518-fed20163a54c",
   "metadata": {},
   "source": [
    "<h2>Let's break down the code:</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10d83a8c-aad3-4a9c-a57d-42584fb92301",
   "metadata": {},
   "source": [
    "<code>gs_res</code>\n",
    "\n",
    "This will be the variable to hold our GSEA results.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>&equals;</code>\n",
    "\n",
    "Our assignment variable.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>gp.GSEA()</code>\n",
    "\n",
    "This is the function to set up our GSEA run. This set up allows us to first provide the information for our analysis before running the analysis. If we wanted to just run the analysis directly, we can instead use the <code>gp.gsea()</code> function. This will become more apparent when we specify how we want our groups to be arranged.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>data=norm_counts</code>\n",
    "\n",
    "This is our normalized counts matrix that we'll pass to <code>gp.GSEA()</code>. We don't need to provide it with an ordered list of genes since it will perform the ranking according to the argument we set for the <code>method</code> parameter. \n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>gene_sets=hallmark_db</code>\n",
    "\n",
    "Like with before, we indicate which database we want it to use for functional analysis. Here, we provide it with the <code>dict</code> object containing the gene sets for the MSigDB Hallmark database that we pulled in earlier.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>classes=class_vector</code>\n",
    "\n",
    "This argument assigns the sample conditions (phenotypes) to our normalized counts matrix based on the information contained in our .cls file. So by making sure that the order of our class vector matches our normalized counts matrix columns, our samples will be assigned to their correct \"class\" or \"phenotype\".\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>method='signal_to_noise'</code>\n",
    "\n",
    "This parameter sets the ranking algorithm that will be used. This method calculates the difference of means scaled by the standard deviation, so that a large difference in gene expression classifies that gene as a stronger \"class marker\" for the condition. So if something is highly upregulated in our TAZ KO samples compared to control, then it will be highly ranked in our ranked list. As we then progress towards the end of the ranked list, the genes will be highly downregulatd in TAZ KO samples compared to control, so those genes will be a stronger \"class marker\" for our control condition.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>permutation_type='phenotype'</code>\n",
    "\n",
    "Recall from Dr. Ingolia's lecture that GSEA compares the enrichment relative to what would be expected by random chance. In order to determine \"what would be expected by random chance\", GSEA will randomly shuffle the data and repeat the analysis many times. The repeated reshuffling and \"walk\" down the list creates a reference for the level of enrichment that would be expected based on random chance. There are two methods to \"shuffle\" the data prior to analysis. \n",
    "\n",
    "The one discussed in lecture is a reshuffling of the sample condition prior to ranking and then repeating the walk down the list of genes. With the setup that we have indicated here with the <code>'phenotype'</code> argument, we randomly assign a condition (phenotype) to each sample, then the genes are ranked according to our specified method. Then the walk down the list is performed. This type of permutation ensures that the gene sets are constant between permutations, and by shuffling the assigned phenotypes for each of our samples, it creates a reference point for the level of enrichment when there is no differential expression in our data (a random ranking of genes). This permutation method requires that we have at least 7 samples for each condition, which we do. The stringency of this method allows us to use a cutoff FDR (q-value) of 0.25 rather than 0.05.\n",
    "\n",
    "Another permutation method, which is less stringent, is to hold the gene rankings constant (assume the differential expression is true) and then instead reshuffle what genes are contained in the gene sets. This method can be used by passing the <code>'gene_set'</code> argument. So the reference point in this case is not a random ranking of genes but rather comparing the enrichment level to a hypothetical randomly generated gene set. This can be used if the sample numbers are too small for the 'phenotype' permutation, and it requires a FDR (q-value) cutoff of 0.05.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>permutation_num=1000</code>\n",
    "\n",
    "This is the number of times we want to reshuffle and repeat the analysis to determine a reference enrichment level for \"random chance\".\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>outdir=None</code>\n",
    "\n",
    "If we wanted to save the output into a file, we can specify the file name here or separately save using <code>pd.DataFrame.to_csv()</code>.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>seed=135</code>\n",
    "\n",
    "This seed is for the random permutations, and by setting a \"seed\" value, we fix the randomization, allowing our results to be reproducible between runs. So if you provided someone else with the same settings, they'll be able to get the same results if they use the same \"seed\" value. However, if they run with a different seed value, their randomization will be different, so their reference enrichment level will be different as well. So they may have slightly different results.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>gs_res.pheno_pos = tazko</code>\n",
    "\n",
    "This sets which condition (phenotype) positive enrichment values are correlated with. Our setup will set positive values to be indicative of enrichment in TAZ KO samples.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>gs_res.pheno_neg = ctrl</code>\n",
    "\n",
    "This sets which condition (phenotpe) negative enrichment values are correlated with. Based on our setup, negative values would indicate enrichment in our control samples.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>\n",
    "\n",
    "<code>gs_res.run()</code>\n",
    "\n",
    "After we've set up how we want to run our GSEA, we can then run the analysis with this line.\n",
    "\n",
    "<hr style=\"border: 1px solid; border-color: #AAAAAA;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f15ce64e-23ec-4c39-b22e-76f5a402400f",
   "metadata": {},
   "source": [
    "Let's take a look at the results table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b21bdd2c-c084-4bd7-8b0d-8b67ec5ad7e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3b7c714e-2e9f-4fff-8081-91c7f64c2f39",
   "metadata": {},
   "source": [
    "Depending on what seed you chose, your results might differ slightly, but we can see enrichment for a number of pathways in our TAZ KO samples. \n",
    "\n",
    "Now let's save the results of our GSEA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2118f87-b7eb-4eff-9a6e-e8bd1199aa30",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f0e1f798-81ab-40a6-ac30-c4251b2aa715",
   "metadata": {},
   "source": [
    "<h2>Generate a GSEA plot</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f37769c-702c-416d-b4e5-50b2d2b41ebc",
   "metadata": {},
   "source": [
    "We can then visualize our results with a GSEA plot using the function <code>gp_obj.plot()</code>.\n",
    "\n",
    "This function will generate a GSEA plot for our indicated gene set term/category, and the object is a <code>matplotlib.Figure</code> object, which we can modify using the same methods we have been using for our other figures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96efcea9-528d-43cc-a6ad-ff2fc3e67451",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b2286857-fc5f-47b8-93d6-6078199f11a7",
   "metadata": {},
   "source": [
    "If we want to plot multiple GSEA plots (say of our top 5 most significant hits), we can make use of a for loop:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05d50237-f2a1-40d5-b83a-4a176c4a9a09",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1c87db7a-2d4c-4d9a-8aa0-a8930bbe939d",
   "metadata": {},
   "source": [
    "Another way of looking at multiple GSEA terms is to plot them all together onto a single GSEA plot. We'll use the same function as before but instead of providing it with a single term, we can provide a list of terms:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fd8d5fc-66b7-452a-a943-e9186a437967",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "da39c635-11ce-4a8e-86d4-cac3e3cbaaab",
   "metadata": {},
   "source": [
    "<h2>Generate heatmap of leading edge genes</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e478528-73fc-4f99-9471-6b75179ca15a",
   "metadata": {},
   "source": [
    "The leading edge of a GSEA plot are the genes that have contributed to the \"peak\" enrichment score of a term, and we can pull them out of our results matrix to generate a heatmap of their relative expression levels between our two conditions. To do this, we'll first need to pull out each of the genes identified as a leading edge gene for a term of interest:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6450c86b-c4f9-421e-8f2a-0f53769b3494",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "312a517d-cbbf-4e6a-a5a9-3404d8e2ccf5",
   "metadata": {},
   "source": [
    "This line of code pulls out the string containing the gene names of the leading edge genes for our top GSEA result, and since the gene names are all a single value in the data matrix separated by a semicolon, we can split them apart from each other with the <code>str.split()</code> function, where <code>str</code> is replaced with the string we want to split up into a list.\n",
    "\n",
    "Let's take a look at our leading edge gene names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8a46d7f-17e4-4583-8586-6625094d73ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7a10b87a-e532-4d21-9313-17953221e07e",
   "metadata": {},
   "source": [
    "We can pass this list to an organized counts matrix generated by <code>gp.GSEA()</code> and contained within the output <code>gs_res</code> object. So we use this list to pull out just the normalized count data for the genes contributing to the leading edge while passing it to <code>gp.heatmap()</code> for visualization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d13270ef-04f6-42c7-b211-0bddcb1a6f12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cc7a8194-b01c-49b8-bbc9-7c6377737b1c",
   "metadata": {},
   "source": [
    "Like with our GSEA plots, if we want to plot our top 5 terms, we can make use of a for loop:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0953b7ad-87bd-46aa-9fbe-b5b56f535527",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bfaa1b3c-0a08-4200-9c61-fd5797dd070b",
   "metadata": {},
   "source": [
    "<h2>Generate a dot plot from GSEA results</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ebf50a3-0ddf-4c9b-9c8d-aec690f952d5",
   "metadata": {},
   "source": [
    "Similar to our over-representation analysis, we can also generate a dot plot from the results of our GSEA run. We'll again make use of the <code>gp.dotplot</code> function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97a43e88-bbe0-42a9-877f-50535b7f819b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fb6dce4d-0fdb-4af3-9055-da6172fbe2d0",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Determine functional group enrichment in TAZ direct targets</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c45d6289-9f3e-42b1-9ad8-2f450de956a9",
   "metadata": {},
   "source": [
    "With our subsetted data, we can perform an over-representation analysis or GSEA to determine which functional groups TAZ may be directly or indirectly regulating. We can continue to use our gene sets from MSigDB Hallmark collection for this functional analysis as well.\n",
    "\n",
    "We can define what our \"interesting\" genes are that we want to determine what gene sets or functional groups are enriched:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef97bac3-9f6a-42aa-8f3e-7567cb95a3ac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6f57b771-18ae-4f4d-9f3a-83ecfc7c38ad",
   "metadata": {},
   "source": [
    "Let's run the over-representation analysis using the GSEApy package's <code>gp.enrich()</code> function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b72fbee1-079d-4548-9a9d-746b7ce648c6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3d8c7630-7ca8-4396-b533-5f13821c065a",
   "metadata": {},
   "source": [
    "We can sort our data by the adjusted p-value just to reorganize our data where the most significant over-representation is at the top of the DataFrame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c191a42-ce82-4691-8976-dda95e55e7cb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "428d9323-faaa-48e5-b1e7-f8a095effa41",
   "metadata": {},
   "source": [
    "Now let's take a look at the output of our analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2354e5e-c914-454d-b7e8-8d708beb804c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "02c630e3-32ae-4beb-9887-ca04845b91ae",
   "metadata": {},
   "source": [
    "<h2>Generate a dot plot of over-represented gene sets</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35287ece-4a55-400e-a708-194d955498d9",
   "metadata": {},
   "source": [
    "We can make use of the <code>gp.dotplot()</code> function to generate a dot plot of our over-representation analysis results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98694859-5af1-4196-a42f-40075784907f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2db01d4b-6745-4896-8bfe-36e264fd5f21",
   "metadata": {},
   "source": [
    "What adjustments can we make to our code to instead look at the set of genes that are indirectly regulated by TAZ?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e988a38-4481-4a9e-8dd1-cc6d028a2ac4",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Perform GSEA on TAZ direct targets</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ede53fe-207c-4eb5-82ba-c7955ac1b05e",
   "metadata": {},
   "source": [
    "Like with over-representation analysis, we can also perform GSEA on our overlapping genes to get an idea of what functional groups may be direct targets of TAZ by taking a look at the difference in enrichment between our TAZ KO and control samples. \n",
    "\n",
    "We can continue with our normalized counts matrix from earlier in the notebook to pass to <code>GSEApy</code>'s <code>gp.GSEA()</code> function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6afde5f4-d355-4fb3-87b2-47480e4d5475",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5a8f7730-1dd0-42d8-a314-bc8ab7067b94",
   "metadata": {},
   "source": [
    "We can then find the full set of overlaps with our ChIP-seq dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "367207e1-ed1c-4227-a67f-aff1fba111b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0d7fac2c-d829-4547-a7d6-36301a3c02bf",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise: Pull out leading edge genes to visualize counts</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c6063ad-3986-49ff-806b-73d2d2fe7f74",
   "metadata": {},
   "source": [
    "For this exercise, select a significantly enriched functional group based on our GSEA analysis, and pull out the leading edge genes. Once you have the genes, then use that information to generate a violin plot to visualize the distribution of the normalized counts for the set of leading edge genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e901f66f-05ec-422b-b335-b17be54fa6e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2043ca32-55a3-45d3-8843-e71002a2ae55",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Exercise: Run GSEA on TAZ indirect targets</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08250597-c6f3-4cb6-9e5e-ba71ca7e39cc",
   "metadata": {},
   "source": [
    "For this last exercise, practice running GSEA on our RNA-seq genes that were found to not overlap with our ChIP-seq genes, and after, we can take a look at what the result might tell us about the indirect targets of TAZ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dee73b5c-50cb-4cf0-9397-e8516dadd225",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
