{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0a8fbfb1-1d98-4c7a-a6c0-b63cfbb12cf0",
   "metadata": {},
   "source": [
    "<h1 style=\"font-size: 40px; margin-bottom: 0px;\">Week 15 Assignment</h1>\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab512b95-234d-4a2a-94af-1b92728369ab",
   "metadata": {},
   "source": [
    "<strong><u>Due: Saturday, December 13th at 11:59PM</u></strong>\n",
    "\n",
    "<strong><u>Total points: 5 pts</u></strong>\n",
    "\n",
    "<strong><u>To turn in your assignment:</u></strong>\n",
    "<ol>\n",
    "    <li>Complete assignment and make sure you have all the outputs requested for the notebook. <mark><strong>Switch between R and Python kernels as needed to complete each question.</strong></mark></li>\n",
    "    <li>Go to your server's file directory and locate the <code>week_15_homework</code> directory containing this notebook</li>\n",
    "    <li>Right click and select <strong>Download as an Archive</strong> to download the <code>week_15_homework</code> directory as a ZIP folder. Do not mix up with <strong>Download Current Folder as an Archive</strong></li>\n",
    "    <li>Upload your ZIP folder to your individual feedback Google Drive folder under <code>MCB201B</code></li>\n",
    "</ol>\n",
    "\n",
    "<strong><u>How this assignment will be graded:</u></strong>\n",
    "\n",
    "You will be graded on whether or not you have <u>all requested outputs</u> and if your output is correct. You will also be graded on if your code is expected to be able to run without errors. There can be more than one correct way to answer a question. If you find that you prefer to separate your answer for a single question into multiple code cells, feel free to do so.\n",
    "\n",
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "236c1dec-415f-4c6d-864a-003e995577a3",
   "metadata": {},
   "source": [
    "<h1>Question 1</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c718b983-fc6e-4789-9401-99f65efb11bc",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90a6f30a-9421-49ca-a5c7-a0750060a1cf",
   "metadata": {},
   "source": [
    "Run a differential expression analysis on the full class transcriptome dataset, and export the results table as a <code>.csv</code> file to be included in your homework submission. Export shrunken logfold change results and normalized counts as needed for the other parts of this homework.\n",
    "\n",
    "Perform hierarchical clustering analysis on the top 500 most variably expressed genes for the full class transcriptome dataset. Output the results of your hierarchical clustering analysis to a PDF file to be included in your homework submission."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a071b937-396d-4932-9ca4-777231a1cd23",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "library(ggplot2)\n",
    "library(pheatmap)\n",
    "library(viridis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0f0b77-9b09-4512-bdab-59063dc9af10",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "26f78243-5bc6-48be-86ce-6a0aef30f946",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3e4e694-197c-45bd-93e5-b814102abddb",
   "metadata": {},
   "source": [
    "<h1>Question 2</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b30dad88-d5ff-48d6-a93d-27d4e09fc6ce",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f198f6f-3382-439f-9f6a-c11cc35aba1f",
   "metadata": {},
   "source": [
    "Generate an MA plot (either using R or Python) to visualize the results of your differential expression analysis. Export into a PDF file to be included as part of this homework submission."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "407d096f-72b0-45d2-a77e-d82c07a033ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "18e73448-2ac4-4d4f-88d6-a0754d8479b0",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9054ec7a-ebbe-40ba-8008-2a4e432d80ba",
   "metadata": {},
   "source": [
    "<h1>Question 3</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba085da4-a232-4545-ac98-338127e0c687",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c55cc739-704a-4af6-a05e-e285d373c5b6",
   "metadata": {},
   "source": [
    "Run an over-representation analysis for the Hallmark gene sets on the full class results for the following \"interesting\" hits: genes that are at least 2-fold upregulated in TAZ KO cells compared to controls and are significant (p-adj < 0.05). Generate a dot-plot of the top 5 functional groups based on their significance (Adjusted P-value < 0.05) with the Combined Score values plotted along the x-axis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed5ef704-22ba-4869-a608-15382c3cda54",
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install gseapy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8385c2b8-bf53-44dc-8055-37bcd7ad9e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats as stats\n",
    "import gseapy as gp\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8245fbc-9953-4c36-92e9-96795ed1a082",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fd163986-c91f-41ec-b13e-4510d0c1cbde",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d70e829-6800-4aea-8dcc-aaaca9c7ec39",
   "metadata": {},
   "source": [
    "<h1>Question 4</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2caa87c9-7e14-4bf5-beab-4471a3bae978",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11c33358-8ff6-4e04-8609-b6332780eac4",
   "metadata": {},
   "source": [
    "Perform GSEA on the full class transcriptome dataset with the Hallmark gene sets, and generate a dot-plot of the top 5 functional groups enriched in TAZ KO samples based on their NES."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "318ab043-4f0b-4fc8-b5e6-62e43e14846a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9205e3ca-d764-487b-8601-71c527f0ad7f",
   "metadata": {},
   "source": [
    "Pull out the leading edge genes for the top functional group enriched in TAZ KO samples, and generate a heatmap to compare the normalized counts of leading edge genes for that functional group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5798b52c-444d-43df-a71a-13783e619316",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "102fe800-20f2-46dc-8a32-d56f0e6d7e30",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f6f9049-7c52-442f-8a60-7d56b3f44319",
   "metadata": {},
   "source": [
    "<h1>Question 5</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa4c7d81-2690-418a-a185-78585c946c5d",
   "metadata": {},
   "source": [
    "1pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80984e81-81f1-45e2-958a-b3c6d5e65246",
   "metadata": {},
   "source": [
    "Perform GSEA on direct targets of TAZ using the full class transcriptome dataset with Hallmark gene sets, and pull out the leading edge genes for the top functional group enriched in TAZ KO samples based on their NES. Generate a box-and-whisker plot to visualize the distribution of normalized counts for the leading edge genes for the top functional group. If there are more than 10 leading edge genes, plot up to 10 leading edge genes of your choice from the top functional group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b11dcf70-a213-4e9f-8cd8-e41e32c7b371",
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install BioVenn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e54a5a9c-7283-4f23-9fc0-d9382848f55c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import BioVenn as bv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b8ee596-cd0a-4a48-945e-854ee5eff93f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7a5e10b9-be9c-4014-8d02-d09a4af9ba55",
   "metadata": {},
   "source": [
    "<hr style=\"margin-left: 0px; border: 0.25px solid; border-color: #000000; width: 98%;\"></hr>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
